// @buildonspark/spark-sdk/dist/tests/test-utils.js
import { Et as WalletConfig, On as SparkError, Yt as Network, Zt as getNetwork, a as WalletConfigService, d as UnsafeStatelessSparkSigner, f as KeyDerivationType, n as BitcoinFaucet, r as DepositService, u as DefaultSparkSigner, zt as getP2TRAddressFromPublicKey } from "../spark-wallet-xLdRrfsG.js";
import { r as ConnectionManagerNodeJS, t as SparkWalletNodeJS } from "../spark-wallet.node-Dl-d7Hf1.js";
import "../empty-CbimPmjm.js";
import "../sdk-types-Dn_8aH9E.js";
import "../index.node.js";
import { secp256k1 } from "@noble/curves/secp256k1";
import { sha256 } from "@noble/hashes/sha2";
import { Address, OutScript, Transaction } from "@scure/btc-signer";

//#region src/tests/utils/spark-testing-wallet.ts
var SparkWalletTesting = class extends SparkWalletNodeJS {
	async setupBackgroundStream() {}
	async *optimizeLeaves(multiplicity = void 0) {}
	async proxyParentSetupBackgroundStream() {
		return super.setupBackgroundStream();
	}
	getSigner() {
		return this.config.signer;
	}
	async queryPendingTransfers() {
		return await this.transferService.queryPendingTransfers();
	}
	async verifyPendingTransfer(transfer) {
		return await this.transferService.verifyPendingTransfer(transfer);
	}
};

//#endregion
//#region src/tests/test-utils.ts
function getTestWalletConfig() {
	return getTestWalletConfigWithIdentityKey(secp256k1.utils.randomPrivateKey());
}
function getTestWalletConfigWithIdentityKey(identityPrivateKey) {
	return {
		...WalletConfig.LOCAL,
		identityPrivateKey
	};
}
async function createDeposit(wallet, leafId, faucet, amountSats = 100000n) {
	const faucetCoin = await faucet.fund();
	const configService = new WalletConfigService({ network: "LOCAL" }, wallet.getSigner());
	const depositService = new DepositService(configService, new ConnectionManagerNodeJS(configService));
	const pubKey = await wallet.getSigner().getPublicKeyFromDerivation({
		type: KeyDerivationType.LEAF,
		path: leafId
	});
	const depositResp = await depositService.generateDepositAddress({
		signingPubkey: pubKey,
		leafId
	});
	if (!depositResp.depositAddress) throw new SparkError("Deposit address not found", {
		signingPubkey: pubKey,
		leafId
	});
	const depositTx = new Transaction();
	depositTx.addInput(faucetCoin.outpoint);
	const addr = Address(getNetwork(Network.LOCAL)).decode(depositResp.depositAddress.address);
	const script = OutScript.encode(addr);
	depositTx.addOutput({
		script,
		amount: amountSats
	});
	let treeResp;
	treeResp = await depositService.createTreeRoot({
		keyDerivation: {
			type: KeyDerivationType.LEAF,
			path: leafId
		},
		verifyingKey: depositResp.depositAddress.verifyingKey,
		depositTx,
		vout: 0
	});
	const signedDepositTx = await faucet.signFaucetCoin(depositTx, faucetCoin.txout, faucetCoin.key);
	await faucet.broadcastTx(signedDepositTx.hex);
	const randomKey = secp256k1.utils.randomPrivateKey();
	const randomAddress = getP2TRAddressFromPublicKey(secp256k1.getPublicKey(randomKey), Network.LOCAL);
	await faucet.generateToAddress(3, randomAddress);
	await new Promise((resolve) => setTimeout(resolve, 100));
	return treeResp.nodes[0];
}
async function createNewTree(wallet, leafId, faucet, amountSats = 100000n) {
	return await createDeposit(wallet, leafId, faucet, amountSats);
}
const signerTypes = [{
	name: "DefaultSparkSigner",
	Signer: DefaultSparkSigner,
	createTree: createNewTree
}, {
	name: "UnsafeStatelessSparkSigner",
	Signer: UnsafeStatelessSparkSigner,
	createTree: createNewTree
}];
const walletTypes = [...signerTypes];
function createDeterministicKeys(seed) {
	const privateKey = sha256(new TextEncoder().encode(seed));
	return {
		privateKey,
		publicKey: secp256k1.getPublicKey(privateKey, true)
	};
}

//#endregion
export { BitcoinFaucet, SparkWalletTesting, createDeterministicKeys, createNewTree, getTestWalletConfig, getTestWalletConfigWithIdentityKey, signerTypes, walletTypes };
