// @utxos/api-contracts/dist/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  ApiErrorSchema: () => ApiErrorSchema,
  ChainSchema: () => ChainSchema,
  CreateOutputRequestSchema: () => CreateOutputRequestSchema,
  CreateSponsorshipRequestSchema: () => CreateSponsorshipRequestSchema,
  CreateTokenRequestSchema: () => CreateTokenRequestSchema,
  CreateWalletRequestSchema: () => CreateWalletRequestSchema,
  FreezeAddressRequestSchema: () => FreezeAddressRequestSchema,
  GetFrozenAddressesResponseSchema: () => GetFrozenAddressesResponseSchema,
  GetPendingUtxosResponseSchema: () => GetPendingUtxosResponseSchema,
  GetPoliciesResponseSchema: () => GetPoliciesResponseSchema,
  GetRefreshTxHashResponseSchema: () => GetRefreshTxHashResponseSchema,
  GetSponsorshipResponseSchema: () => GetSponsorshipResponseSchema,
  GetTransactionsResponseSchema: () => GetTransactionsResponseSchema,
  GetWalletResponseSchema: () => GetWalletResponseSchema,
  GetWalletsResponseSchema: () => GetWalletsResponseSchema,
  LogTransactionRequestSchema: () => LogTransactionRequestSchema,
  MultiChainWalletInfoSchema: () => MultiChainWalletInfoSchema,
  NetworkSchema: () => NetworkSchema,
  PaginationInfoSchema: () => PaginationInfoSchema,
  ProjectResponseSchema: () => ProjectResponseSchema,
  RefreshTxHashRequestSchema: () => RefreshTxHashRequestSchema,
  RefreshTxHashResponseSchema: () => RefreshTxHashResponseSchema,
  SponsorshipConfigSchema: () => SponsorshipConfigSchema,
  SponsorshipOutputSchema: () => SponsorshipOutputSchema,
  TokenizationFrozenAddressSchema: () => TokenizationFrozenAddressSchema,
  TokenizationPolicySchema: () => TokenizationPolicySchema,
  TokenizationTransactionSchema: () => TokenizationTransactionSchema,
  TransactionTypeSchema: () => TransactionTypeSchema,
  TxParserRequestSchema: () => TxParserRequestSchema,
  TxParserResponseSchema: () => TxParserResponseSchema,
  UnfreezeAddressRequestSchema: () => UnfreezeAddressRequestSchema,
  UpdateOutputRequestSchema: () => UpdateOutputRequestSchema,
  isValidMultiChainWalletInfo: () => isValidMultiChainWalletInfo,
  isValidProjectResponse: () => isValidProjectResponse,
  validateCreateWalletRequest: () => validateCreateWalletRequest,
  validateFrozenAddress: () => validateFrozenAddress,
  validateGetPoliciesResponse: () => validateGetPoliciesResponse,
  validateGetTransactionsResponse: () => validateGetTransactionsResponse,
  validateGetWalletsResponse: () => validateGetWalletsResponse,
  validateMultiChainWalletInfo: () => validateMultiChainWalletInfo,
  validateProjectResponse: () => validateProjectResponse,
  validateSponsorshipConfig: () => validateSponsorshipConfig,
  validateSponsorshipOutput: () => validateSponsorshipOutput,
  validateTokenizationPolicy: () => validateTokenizationPolicy,
  validateTransaction: () => validateTransaction,
  validateTxParserRequest: () => validateTxParserRequest,
  validateTxParserResponse: () => validateTxParserResponse,
  z: () => import_zod6.z
});
module.exports = __toCommonJS(index_exports);

// src/schemas/common.ts
var import_zod = require("zod");
var PaginationInfoSchema = import_zod.z.object({
  page: import_zod.z.number(),
  limit: import_zod.z.number(),
  total: import_zod.z.number(),
  totalPages: import_zod.z.number()
});
var NetworkSchema = import_zod.z.enum(["mainnet", "testnet"]);
var ChainSchema = import_zod.z.enum(["spark", "cardano"]);
var ApiErrorSchema = import_zod.z.object({
  error: import_zod.z.string(),
  message: import_zod.z.string().optional()
});

// src/schemas/project.ts
var import_zod2 = require("zod");
var ProjectResponseSchema = import_zod2.z.object({
  id: import_zod2.z.string().uuid(),
  name: import_zod2.z.string(),
  publicKey: import_zod2.z.string().nullable(),
  apiKey: import_zod2.z.string().optional(),
  credits: import_zod2.z.number().optional(),
  branding: import_zod2.z.object({
    logoUrl: import_zod2.z.string().optional(),
    primaryColor: import_zod2.z.string().optional()
  }).optional(),
  oauthGoogle: import_zod2.z.object({
    clientId: import_zod2.z.string().optional()
  }).optional(),
  oauthDiscord: import_zod2.z.object({
    clientId: import_zod2.z.string().optional()
  }).optional(),
  oauthTwitter: import_zod2.z.object({
    clientId: import_zod2.z.string().optional()
  }).optional(),
  oauthApple: import_zod2.z.object({
    clientId: import_zod2.z.string().optional()
  }).optional()
});
function validateProjectResponse(data) {
  return ProjectResponseSchema.parse(data);
}
function isValidProjectResponse(data) {
  return ProjectResponseSchema.safeParse(data).success;
}

// src/schemas/project-wallet.ts
var import_zod3 = require("zod");
var CardanoChainInfoSchema = import_zod3.z.object({
  pubKeyHash: import_zod3.z.string(),
  stakeCredentialHash: import_zod3.z.string().nullable()
});
var SparkChainInfoSchema = import_zod3.z.object({
  mainnetPublicKey: import_zod3.z.string(),
  regtestPublicKey: import_zod3.z.string()
});
var MultiChainWalletInfoSchema = import_zod3.z.object({
  id: import_zod3.z.string().uuid(),
  projectId: import_zod3.z.string().uuid(),
  tags: import_zod3.z.array(import_zod3.z.string()),
  key: import_zod3.z.string(),
  // Encrypted mnemonic
  chains: import_zod3.z.object({
    cardano: CardanoChainInfoSchema.optional(),
    spark: SparkChainInfoSchema.optional()
  })
});
var CreateWalletRequestSchema = import_zod3.z.object({
  id: import_zod3.z.string().uuid(),
  projectId: import_zod3.z.string().uuid(),
  tags: import_zod3.z.array(import_zod3.z.string()).optional().default([]),
  key: import_zod3.z.string(),
  chains: import_zod3.z.object({
    cardano: CardanoChainInfoSchema.optional(),
    spark: SparkChainInfoSchema.optional()
  })
});
var GetWalletResponseSchema = MultiChainWalletInfoSchema;
var GetWalletsResponseSchema = import_zod3.z.array(MultiChainWalletInfoSchema);
function validateMultiChainWalletInfo(data) {
  return MultiChainWalletInfoSchema.parse(data);
}
function isValidMultiChainWalletInfo(data) {
  return MultiChainWalletInfoSchema.safeParse(data).success;
}
function validateCreateWalletRequest(data) {
  return CreateWalletRequestSchema.parse(data);
}
function validateGetWalletsResponse(data) {
  return GetWalletsResponseSchema.parse(data);
}

// src/schemas/sponsorship.ts
var import_zod4 = require("zod");
var SponsorshipConfigSchema = import_zod4.z.object({
  id: import_zod4.z.string().uuid(),
  projectId: import_zod4.z.string().uuid(),
  projectWalletId: import_zod4.z.string().uuid(),
  numUtxosTriggerPrepare: import_zod4.z.number().int().positive(),
  numUtxosPrepare: import_zod4.z.number().int().positive(),
  utxoAmount: import_zod4.z.number().int().positive(),
  sponsorshipInfo: import_zod4.z.string().optional()
});
var GetSponsorshipResponseSchema = SponsorshipConfigSchema;
var CreateSponsorshipRequestSchema = import_zod4.z.object({
  projectId: import_zod4.z.string().uuid(),
  projectWalletId: import_zod4.z.string().uuid(),
  numUtxosTriggerPrepare: import_zod4.z.number().int().positive(),
  numUtxosPrepare: import_zod4.z.number().int().positive(),
  utxoAmount: import_zod4.z.number().int().positive(),
  sponsorshipInfo: import_zod4.z.string().optional()
});
var RefreshTxHashRequestSchema = import_zod4.z.object({
  txHash: import_zod4.z.string()
});
var RefreshTxHashResponseSchema = SponsorshipConfigSchema.extend({
  refreshTxHash: import_zod4.z.string().optional()
});
var GetRefreshTxHashResponseSchema = import_zod4.z.object({
  refreshTxHash: import_zod4.z.string().nullable()
});
var SponsorshipOutputSchema = import_zod4.z.object({
  projectWalletId: import_zod4.z.string().uuid(),
  txHash: import_zod4.z.string(),
  outputIndex: import_zod4.z.number().int().nonnegative(),
  createdAt: import_zod4.z.string().or(import_zod4.z.date()),
  isPending: import_zod4.z.boolean(),
  isSpent: import_zod4.z.boolean()
});
var GetPendingUtxosResponseSchema = import_zod4.z.array(SponsorshipOutputSchema);
var UpdateOutputRequestSchema = import_zod4.z.object({
  isPending: import_zod4.z.boolean().optional(),
  isSpent: import_zod4.z.boolean().optional()
});
var CreateOutputRequestSchema = import_zod4.z.object({
  txHash: import_zod4.z.string(),
  outputIndex: import_zod4.z.number().int().nonnegative(),
  projectWalletId: import_zod4.z.string().uuid(),
  isPending: import_zod4.z.boolean().optional().default(false),
  isSpent: import_zod4.z.boolean().optional().default(false)
});
var TxParserRequestSchema = import_zod4.z.object({
  txHex: import_zod4.z.string(),
  address: import_zod4.z.string(),
  utxos: import_zod4.z.string(),
  // Stringified JSON
  sponsorUtxo: import_zod4.z.string(),
  // Stringified JSON
  network: import_zod4.z.enum(["mainnet", "preprod", "preview"])
});
var TxParserResponseSchema = import_zod4.z.object({
  rebuiltTxHex: import_zod4.z.string()
});
function validateSponsorshipConfig(data) {
  return SponsorshipConfigSchema.parse(data);
}
function validateSponsorshipOutput(data) {
  return SponsorshipOutputSchema.parse(data);
}
function validateTxParserRequest(data) {
  return TxParserRequestSchema.parse(data);
}
function validateTxParserResponse(data) {
  return TxParserResponseSchema.parse(data);
}

// src/schemas/tokenization.ts
var import_zod5 = require("zod");
var TokenizationPolicySchema = import_zod5.z.object({
  tokenId: import_zod5.z.string(),
  projectId: import_zod5.z.string().uuid(),
  walletId: import_zod5.z.string().uuid(),
  // Note: API returns walletId, DB has projectWalletId
  chain: import_zod5.z.string(),
  network: import_zod5.z.string(),
  isActive: import_zod5.z.boolean(),
  createdAt: import_zod5.z.string()
});
var CreateTokenRequestSchema = import_zod5.z.object({
  tokenId: import_zod5.z.string(),
  projectId: import_zod5.z.string().uuid(),
  walletId: import_zod5.z.string().uuid(),
  chain: import_zod5.z.string(),
  network: import_zod5.z.string()
});
var GetPoliciesResponseSchema = import_zod5.z.object({
  tokens: import_zod5.z.array(TokenizationPolicySchema),
  pagination: PaginationInfoSchema
});
var TokenizationFrozenAddressSchema = import_zod5.z.object({
  id: import_zod5.z.string().uuid(),
  tokenId: import_zod5.z.string(),
  projectId: import_zod5.z.string().uuid(),
  projectWalletId: import_zod5.z.string().uuid(),
  publicKeyHash: import_zod5.z.string(),
  stakeKeyHash: import_zod5.z.string().nullable().optional(),
  chain: import_zod5.z.string(),
  network: import_zod5.z.string(),
  isFrozen: import_zod5.z.boolean(),
  freezeReason: import_zod5.z.string().nullable().optional(),
  frozenAt: import_zod5.z.string().nullable().optional(),
  unfrozenAt: import_zod5.z.string().nullable().optional(),
  createdAt: import_zod5.z.string()
});
var FreezeAddressRequestSchema = import_zod5.z.object({
  tokenId: import_zod5.z.string(),
  projectId: import_zod5.z.string().uuid(),
  projectWalletId: import_zod5.z.string().uuid(),
  chain: import_zod5.z.string(),
  network: import_zod5.z.string(),
  publicKeyHash: import_zod5.z.string(),
  stakeKeyHash: import_zod5.z.string().optional(),
  freezeReason: import_zod5.z.string().optional()
});
var UnfreezeAddressRequestSchema = import_zod5.z.object({
  tokenId: import_zod5.z.string(),
  publicKeyHash: import_zod5.z.string(),
  projectId: import_zod5.z.string().uuid(),
  projectWalletId: import_zod5.z.string().uuid(),
  stakeKeyHash: import_zod5.z.string().optional()
});
var GetFrozenAddressesResponseSchema = import_zod5.z.object({
  frozenAddresses: import_zod5.z.array(TokenizationFrozenAddressSchema),
  pagination: PaginationInfoSchema
});
var TransactionTypeSchema = import_zod5.z.enum([
  "create",
  "mint",
  "burn",
  "transfer",
  "freeze",
  "unfreeze"
]);
var TokenizationTransactionSchema = import_zod5.z.object({
  txId: import_zod5.z.string(),
  tokenId: import_zod5.z.string(),
  projectId: import_zod5.z.string().uuid(),
  projectWalletId: import_zod5.z.string().uuid(),
  type: TransactionTypeSchema,
  chain: import_zod5.z.string(),
  network: import_zod5.z.string(),
  amount: import_zod5.z.string().nullable().optional(),
  fromAddress: import_zod5.z.string().nullable().optional(),
  toAddress: import_zod5.z.string().nullable().optional(),
  status: import_zod5.z.string(),
  createdAt: import_zod5.z.string()
});
var LogTransactionRequestSchema = import_zod5.z.object({
  txId: import_zod5.z.string(),
  tokenId: import_zod5.z.string(),
  projectId: import_zod5.z.string().uuid(),
  projectWalletId: import_zod5.z.string().uuid(),
  type: TransactionTypeSchema,
  chain: import_zod5.z.string(),
  network: import_zod5.z.string(),
  amount: import_zod5.z.string().optional(),
  fromAddress: import_zod5.z.string().optional(),
  toAddress: import_zod5.z.string().optional(),
  status: import_zod5.z.string()
});
var GetTransactionsResponseSchema = import_zod5.z.object({
  transactions: import_zod5.z.array(TokenizationTransactionSchema),
  pagination: PaginationInfoSchema
});
function validateTokenizationPolicy(data) {
  return TokenizationPolicySchema.parse(data);
}
function validateGetPoliciesResponse(data) {
  return GetPoliciesResponseSchema.parse(data);
}
function validateFrozenAddress(data) {
  return TokenizationFrozenAddressSchema.parse(data);
}
function validateTransaction(data) {
  return TokenizationTransactionSchema.parse(data);
}
function validateGetTransactionsResponse(data) {
  return GetTransactionsResponseSchema.parse(data);
}

// src/index.ts
var import_zod6 = require("zod");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ApiErrorSchema,
  ChainSchema,
  CreateOutputRequestSchema,
  CreateSponsorshipRequestSchema,
  CreateTokenRequestSchema,
  CreateWalletRequestSchema,
  FreezeAddressRequestSchema,
  GetFrozenAddressesResponseSchema,
  GetPendingUtxosResponseSchema,
  GetPoliciesResponseSchema,
  GetRefreshTxHashResponseSchema,
  GetSponsorshipResponseSchema,
  GetTransactionsResponseSchema,
  GetWalletResponseSchema,
  GetWalletsResponseSchema,
  LogTransactionRequestSchema,
  MultiChainWalletInfoSchema,
  NetworkSchema,
  PaginationInfoSchema,
  ProjectResponseSchema,
  RefreshTxHashRequestSchema,
  RefreshTxHashResponseSchema,
  SponsorshipConfigSchema,
  SponsorshipOutputSchema,
  TokenizationFrozenAddressSchema,
  TokenizationPolicySchema,
  TokenizationTransactionSchema,
  TransactionTypeSchema,
  TxParserRequestSchema,
  TxParserResponseSchema,
  UnfreezeAddressRequestSchema,
  UpdateOutputRequestSchema,
  isValidMultiChainWalletInfo,
  isValidProjectResponse,
  validateCreateWalletRequest,
  validateFrozenAddress,
  validateGetPoliciesResponse,
  validateGetTransactionsResponse,
  validateGetWalletsResponse,
  validateMultiChainWalletInfo,
  validateProjectResponse,
  validateSponsorshipConfig,
  validateSponsorshipOutput,
  validateTokenizationPolicy,
  validateTransaction,
  validateTxParserRequest,
  validateTxParserResponse,
  z
});
