// @utxos/api-contracts/dist/index.js
// src/schemas/common.ts
import { z } from "zod";
var PaginationInfoSchema = z.object({
  page: z.number(),
  limit: z.number(),
  total: z.number(),
  totalPages: z.number()
});
var NetworkSchema = z.enum(["mainnet", "testnet"]);
var ChainSchema = z.enum(["spark", "cardano"]);
var ApiErrorSchema = z.object({
  error: z.string(),
  message: z.string().optional()
});

// src/schemas/project.ts
import { z as z2 } from "zod";
var ProjectResponseSchema = z2.object({
  id: z2.string().uuid(),
  name: z2.string(),
  publicKey: z2.string().nullable(),
  apiKey: z2.string().optional(),
  credits: z2.number().optional(),
  branding: z2.object({
    logoUrl: z2.string().optional(),
    primaryColor: z2.string().optional()
  }).optional(),
  oauthGoogle: z2.object({
    clientId: z2.string().optional()
  }).optional(),
  oauthDiscord: z2.object({
    clientId: z2.string().optional()
  }).optional(),
  oauthTwitter: z2.object({
    clientId: z2.string().optional()
  }).optional(),
  oauthApple: z2.object({
    clientId: z2.string().optional()
  }).optional()
});
function validateProjectResponse(data) {
  return ProjectResponseSchema.parse(data);
}
function isValidProjectResponse(data) {
  return ProjectResponseSchema.safeParse(data).success;
}

// src/schemas/project-wallet.ts
import { z as z3 } from "zod";
var CardanoChainInfoSchema = z3.object({
  pubKeyHash: z3.string(),
  stakeCredentialHash: z3.string().nullable()
});
var SparkChainInfoSchema = z3.object({
  mainnetPublicKey: z3.string(),
  regtestPublicKey: z3.string()
});
var MultiChainWalletInfoSchema = z3.object({
  id: z3.string().uuid(),
  projectId: z3.string().uuid(),
  tags: z3.array(z3.string()),
  key: z3.string(),
  // Encrypted mnemonic
  chains: z3.object({
    cardano: CardanoChainInfoSchema.optional(),
    spark: SparkChainInfoSchema.optional()
  })
});
var CreateWalletRequestSchema = z3.object({
  id: z3.string().uuid(),
  projectId: z3.string().uuid(),
  tags: z3.array(z3.string()).optional().default([]),
  key: z3.string(),
  chains: z3.object({
    cardano: CardanoChainInfoSchema.optional(),
    spark: SparkChainInfoSchema.optional()
  })
});
var GetWalletResponseSchema = MultiChainWalletInfoSchema;
var GetWalletsResponseSchema = z3.array(MultiChainWalletInfoSchema);
function validateMultiChainWalletInfo(data) {
  return MultiChainWalletInfoSchema.parse(data);
}
function isValidMultiChainWalletInfo(data) {
  return MultiChainWalletInfoSchema.safeParse(data).success;
}
function validateCreateWalletRequest(data) {
  return CreateWalletRequestSchema.parse(data);
}
function validateGetWalletsResponse(data) {
  return GetWalletsResponseSchema.parse(data);
}

// src/schemas/sponsorship.ts
import { z as z4 } from "zod";
var SponsorshipConfigSchema = z4.object({
  id: z4.string().uuid(),
  projectId: z4.string().uuid(),
  projectWalletId: z4.string().uuid(),
  numUtxosTriggerPrepare: z4.number().int().positive(),
  numUtxosPrepare: z4.number().int().positive(),
  utxoAmount: z4.number().int().positive(),
  sponsorshipInfo: z4.string().optional()
});
var GetSponsorshipResponseSchema = SponsorshipConfigSchema;
var CreateSponsorshipRequestSchema = z4.object({
  projectId: z4.string().uuid(),
  projectWalletId: z4.string().uuid(),
  numUtxosTriggerPrepare: z4.number().int().positive(),
  numUtxosPrepare: z4.number().int().positive(),
  utxoAmount: z4.number().int().positive(),
  sponsorshipInfo: z4.string().optional()
});
var RefreshTxHashRequestSchema = z4.object({
  txHash: z4.string()
});
var RefreshTxHashResponseSchema = SponsorshipConfigSchema.extend({
  refreshTxHash: z4.string().optional()
});
var GetRefreshTxHashResponseSchema = z4.object({
  refreshTxHash: z4.string().nullable()
});
var SponsorshipOutputSchema = z4.object({
  projectWalletId: z4.string().uuid(),
  txHash: z4.string(),
  outputIndex: z4.number().int().nonnegative(),
  createdAt: z4.string().or(z4.date()),
  isPending: z4.boolean(),
  isSpent: z4.boolean()
});
var GetPendingUtxosResponseSchema = z4.array(SponsorshipOutputSchema);
var UpdateOutputRequestSchema = z4.object({
  isPending: z4.boolean().optional(),
  isSpent: z4.boolean().optional()
});
var CreateOutputRequestSchema = z4.object({
  txHash: z4.string(),
  outputIndex: z4.number().int().nonnegative(),
  projectWalletId: z4.string().uuid(),
  isPending: z4.boolean().optional().default(false),
  isSpent: z4.boolean().optional().default(false)
});
var TxParserRequestSchema = z4.object({
  txHex: z4.string(),
  address: z4.string(),
  utxos: z4.string(),
  // Stringified JSON
  sponsorUtxo: z4.string(),
  // Stringified JSON
  network: z4.enum(["mainnet", "preprod", "preview"])
});
var TxParserResponseSchema = z4.object({
  rebuiltTxHex: z4.string()
});
function validateSponsorshipConfig(data) {
  return SponsorshipConfigSchema.parse(data);
}
function validateSponsorshipOutput(data) {
  return SponsorshipOutputSchema.parse(data);
}
function validateTxParserRequest(data) {
  return TxParserRequestSchema.parse(data);
}
function validateTxParserResponse(data) {
  return TxParserResponseSchema.parse(data);
}

// src/schemas/tokenization.ts
import { z as z5 } from "zod";
var TokenizationPolicySchema = z5.object({
  tokenId: z5.string(),
  projectId: z5.string().uuid(),
  walletId: z5.string().uuid(),
  // Note: API returns walletId, DB has projectWalletId
  chain: z5.string(),
  network: z5.string(),
  isActive: z5.boolean(),
  createdAt: z5.string()
});
var CreateTokenRequestSchema = z5.object({
  tokenId: z5.string(),
  projectId: z5.string().uuid(),
  walletId: z5.string().uuid(),
  chain: z5.string(),
  network: z5.string()
});
var GetPoliciesResponseSchema = z5.object({
  tokens: z5.array(TokenizationPolicySchema),
  pagination: PaginationInfoSchema
});
var TokenizationFrozenAddressSchema = z5.object({
  id: z5.string().uuid(),
  tokenId: z5.string(),
  projectId: z5.string().uuid(),
  projectWalletId: z5.string().uuid(),
  publicKeyHash: z5.string(),
  stakeKeyHash: z5.string().nullable().optional(),
  chain: z5.string(),
  network: z5.string(),
  isFrozen: z5.boolean(),
  freezeReason: z5.string().nullable().optional(),
  frozenAt: z5.string().nullable().optional(),
  unfrozenAt: z5.string().nullable().optional(),
  createdAt: z5.string()
});
var FreezeAddressRequestSchema = z5.object({
  tokenId: z5.string(),
  projectId: z5.string().uuid(),
  projectWalletId: z5.string().uuid(),
  chain: z5.string(),
  network: z5.string(),
  publicKeyHash: z5.string(),
  stakeKeyHash: z5.string().optional(),
  freezeReason: z5.string().optional()
});
var UnfreezeAddressRequestSchema = z5.object({
  tokenId: z5.string(),
  publicKeyHash: z5.string(),
  projectId: z5.string().uuid(),
  projectWalletId: z5.string().uuid(),
  stakeKeyHash: z5.string().optional()
});
var GetFrozenAddressesResponseSchema = z5.object({
  frozenAddresses: z5.array(TokenizationFrozenAddressSchema),
  pagination: PaginationInfoSchema
});
var TransactionTypeSchema = z5.enum([
  "create",
  "mint",
  "burn",
  "transfer",
  "freeze",
  "unfreeze"
]);
var TokenizationTransactionSchema = z5.object({
  txId: z5.string(),
  tokenId: z5.string(),
  projectId: z5.string().uuid(),
  projectWalletId: z5.string().uuid(),
  type: TransactionTypeSchema,
  chain: z5.string(),
  network: z5.string(),
  amount: z5.string().nullable().optional(),
  fromAddress: z5.string().nullable().optional(),
  toAddress: z5.string().nullable().optional(),
  status: z5.string(),
  createdAt: z5.string()
});
var LogTransactionRequestSchema = z5.object({
  txId: z5.string(),
  tokenId: z5.string(),
  projectId: z5.string().uuid(),
  projectWalletId: z5.string().uuid(),
  type: TransactionTypeSchema,
  chain: z5.string(),
  network: z5.string(),
  amount: z5.string().optional(),
  fromAddress: z5.string().optional(),
  toAddress: z5.string().optional(),
  status: z5.string()
});
var GetTransactionsResponseSchema = z5.object({
  transactions: z5.array(TokenizationTransactionSchema),
  pagination: PaginationInfoSchema
});
function validateTokenizationPolicy(data) {
  return TokenizationPolicySchema.parse(data);
}
function validateGetPoliciesResponse(data) {
  return GetPoliciesResponseSchema.parse(data);
}
function validateFrozenAddress(data) {
  return TokenizationFrozenAddressSchema.parse(data);
}
function validateTransaction(data) {
  return TokenizationTransactionSchema.parse(data);
}
function validateGetTransactionsResponse(data) {
  return GetTransactionsResponseSchema.parse(data);
}

// src/index.ts
import { z as z6 } from "zod";
export {
  ApiErrorSchema,
  ChainSchema,
  CreateOutputRequestSchema,
  CreateSponsorshipRequestSchema,
  CreateTokenRequestSchema,
  CreateWalletRequestSchema,
  FreezeAddressRequestSchema,
  GetFrozenAddressesResponseSchema,
  GetPendingUtxosResponseSchema,
  GetPoliciesResponseSchema,
  GetRefreshTxHashResponseSchema,
  GetSponsorshipResponseSchema,
  GetTransactionsResponseSchema,
  GetWalletResponseSchema,
  GetWalletsResponseSchema,
  LogTransactionRequestSchema,
  MultiChainWalletInfoSchema,
  NetworkSchema,
  PaginationInfoSchema,
  ProjectResponseSchema,
  RefreshTxHashRequestSchema,
  RefreshTxHashResponseSchema,
  SponsorshipConfigSchema,
  SponsorshipOutputSchema,
  TokenizationFrozenAddressSchema,
  TokenizationPolicySchema,
  TokenizationTransactionSchema,
  TransactionTypeSchema,
  TxParserRequestSchema,
  TxParserResponseSchema,
  UnfreezeAddressRequestSchema,
  UpdateOutputRequestSchema,
  isValidMultiChainWalletInfo,
  isValidProjectResponse,
  validateCreateWalletRequest,
  validateFrozenAddress,
  validateGetPoliciesResponse,
  validateGetTransactionsResponse,
  validateGetWalletsResponse,
  validateMultiChainWalletInfo,
  validateProjectResponse,
  validateSponsorshipConfig,
  validateSponsorshipOutput,
  validateTokenizationPolicy,
  validateTransaction,
  validateTxParserRequest,
  validateTxParserResponse,
  z6 as z
};
