// @lightsparkdev/core/dist/index.js
import {
  ConfigKeys,
  DefaultCrypto,
  KeyOrAlias,
  LightsparkAuthException_default,
  LightsparkSigningException_default,
  Logger,
  LoggingLevel,
  NodeKeyCache_default,
  RSASigningKey,
  Requester_default,
  Secp256k1SigningKey,
  ServerEnvironment_default,
  SigningKey,
  SigningKeyType,
  StubAuthProvider,
  apiDomainForEnvironment,
  logger
} from "./chunk-2QHVLQNP.js";
import {
  CurrencyUnit,
  LightsparkException_default,
  abbrCurrencyUnit,
  b64decode,
  b64encode,
  bytesToHex,
  clamp,
  convertCurrencyAmount,
  convertCurrencyAmountValue,
  countryCodesToCurrencyCodes,
  createSha256Hash,
  defaultCurrencyCode,
  deleteLocalStorageItem,
  ensureArray,
  errorToJSON,
  formatCurrencyStr,
  formatInviteAmount,
  getCurrencyAmount,
  getCurrentLocale,
  getErrorMsg,
  getLocalStorageBoolean,
  getLocalStorageConfigItem,
  hexToBytes,
  isBare,
  isBrowser,
  isCurrencyAmountInputObj,
  isCurrencyAmountObj,
  isCurrencyAmountPreferenceObj,
  isCurrencyMap,
  isDeprecatedCurrencyAmountObj,
  isError,
  isErrorMsg,
  isErrorWithMessage,
  isNode,
  isNumber,
  isObject,
  isReactNative,
  isRecord,
  isSDKCurrencyAmount,
  isTest,
  isType,
  isUint8Array,
  isUmaCurrencyAmount,
  linearInterpolate,
  localeToCurrencyCode,
  localeToCurrencySymbol,
  lsidToUUID,
  mapCurrencyAmount,
  notNullUndefined,
  pollUntil,
  round,
  separateCurrencyStrParts,
  setLocalStorageBoolean,
  sleep,
  urlsafe_b64decode,
  zipcodeToState,
  zipcodeToStateCode
} from "./chunk-L47HFANC.js";

// src/requester/DefaultRequester.ts
var DefaultRequester = class extends Requester_default {
  async initWsClient(baseUrl, authProvider) {
    if (!this.resolveWsClient) {
      return this.wsClient;
    }
    if (isBare) {
      return null;
    }
    let websocketImpl;
    if (isNode && typeof WebSocket === "undefined") {
      const wsModule = await import("ws");
      websocketImpl = wsModule.default;
    }
    let websocketProtocol = "wss";
    if (baseUrl.startsWith("http://")) {
      websocketProtocol = "ws";
    }
    const graphqlWsModule = await import("graphql-ws");
    const { createClient } = graphqlWsModule;
    const wsClient = createClient({
      url: `${websocketProtocol}://${this.stripProtocol(this.baseUrl)}/${this.schemaEndpoint}`,
      connectionParams: () => authProvider.addWsConnectionParams({}),
      webSocketImpl: websocketImpl
    });
    if (this.resolveWsClient) {
      this.resolveWsClient(wsClient);
      this.resolveWsClient = null;
    }
    return wsClient;
  }
};
export {
  ConfigKeys,
  CurrencyUnit,
  DefaultCrypto,
  KeyOrAlias,
  LightsparkAuthException_default as LightsparkAuthException,
  LightsparkException_default as LightsparkException,
  LightsparkSigningException_default as LightsparkSigningException,
  Logger,
  LoggingLevel,
  NodeKeyCache_default as NodeKeyCache,
  RSASigningKey,
  DefaultRequester as Requester,
  Secp256k1SigningKey,
  ServerEnvironment_default as ServerEnvironment,
  SigningKey,
  SigningKeyType,
  StubAuthProvider,
  abbrCurrencyUnit,
  apiDomainForEnvironment,
  b64decode,
  b64encode,
  bytesToHex,
  clamp,
  convertCurrencyAmount,
  convertCurrencyAmountValue,
  countryCodesToCurrencyCodes,
  createSha256Hash,
  defaultCurrencyCode,
  deleteLocalStorageItem,
  ensureArray,
  errorToJSON,
  formatCurrencyStr,
  formatInviteAmount,
  getCurrencyAmount,
  getCurrentLocale,
  getErrorMsg,
  getLocalStorageBoolean,
  getLocalStorageConfigItem,
  hexToBytes,
  isBare,
  isBrowser,
  isCurrencyAmountInputObj,
  isCurrencyAmountObj,
  isCurrencyAmountPreferenceObj,
  isCurrencyMap,
  isDeprecatedCurrencyAmountObj,
  isError,
  isErrorMsg,
  isErrorWithMessage,
  isNode,
  isNumber,
  isObject,
  isReactNative,
  isRecord,
  isSDKCurrencyAmount,
  isTest,
  isType,
  isUint8Array,
  isUmaCurrencyAmount,
  linearInterpolate,
  localeToCurrencyCode,
  localeToCurrencySymbol,
  logger,
  lsidToUUID,
  mapCurrencyAmount,
  notNullUndefined,
  pollUntil,
  round,
  separateCurrencyStrParts,
  setLocalStorageBoolean,
  sleep,
  urlsafe_b64decode,
  zipcodeToState,
  zipcodeToStateCode
};
