// @buildonspark/spark-sdk/dist/proto/spark.js
import {
  AdaptorPublicKeyPackage,
  Address,
  AddressNode,
  AddressRequestNode,
  BitcoinTransactionOutput,
  CancelTransferRequest,
  CancelTransferResponse,
  ClaimLeafKeyTweak,
  ClaimLeafKeyTweak_PubkeySharesTweakEntry,
  ClaimTransferSignRefundsRequest,
  ClaimTransferSignRefundsResponse,
  ClaimTransferTweakKeysRequest,
  ConnectedEvent,
  CooperativeExitRequest,
  CooperativeExitResponse,
  CounterLeafSwapRequest,
  CounterLeafSwapResponse,
  CreateTreeRequest,
  CreateTreeResponse,
  CreationNode,
  CreationResponseNode,
  DepositAddressProof,
  DepositAddressProof_AddressSignaturesEntry,
  DepositAddressQueryResult,
  DepositEvent,
  Direction,
  ExitSingleNodeTreeSigningResult,
  ExitSingleNodeTreesRequest,
  ExitSingleNodeTreesResponse,
  ExitingTree,
  ExtendLeafRequest,
  ExtendLeafResponse,
  ExtendLeafSigningResult,
  FinalizeNodeSignaturesRequest,
  FinalizeNodeSignaturesResponse,
  FinalizeTransferRequest,
  FinalizeTransferResponse,
  FinalizeTransferWithTransferPackageRequest,
  FreezeTokensPayload,
  FreezeTokensRequest,
  FreezeTokensResponse,
  GenerateDepositAddressRequest,
  GenerateDepositAddressResponse,
  GenerateStaticDepositAddressRequest,
  GenerateStaticDepositAddressResponse,
  GetSigningCommitmentsRequest,
  GetSigningCommitmentsResponse,
  GetSigningOperatorListResponse,
  GetSigningOperatorListResponse_SigningOperatorsEntry,
  GetUtxosForAddressRequest,
  GetUtxosForAddressResponse,
  InitiatePreimageSwapRequest,
  InitiatePreimageSwapRequest_Reason,
  InitiatePreimageSwapResponse,
  InitiateStaticDepositUtxoRefundRequest,
  InitiateStaticDepositUtxoRefundResponse,
  InitiateSwapPrimaryTransferRequest,
  InitiateSwapPrimaryTransferResponse,
  InitiateUtxoSwapRequest,
  InitiateUtxoSwapResponse,
  InvoiceAmount,
  InvoiceAmountProof,
  InvoiceResponse,
  InvoiceStatus,
  LeafRefundTxSigningJob,
  LeafRefundTxSigningResult,
  Network,
  NodeOutput,
  NodeSignatureShares,
  NodeSignatures,
  OperatorSpecificOwnerSignature,
  OperatorSpecificTokenTransactionSignablePayload,
  Order,
  OutPoint,
  PageRequest,
  PageResponse,
  PreimageRequestRole,
  PreimageRequestStatus,
  PreimageRequestWithTransfer,
  PrepareTreeAddressRequest,
  PrepareTreeAddressResponse,
  ProvidePreimageRequest,
  ProvidePreimageResponse,
  QueryBalanceRequest,
  QueryBalanceResponse,
  QueryBalanceResponse_NodeBalancesEntry,
  QueryHtlcRequest,
  QueryHtlcResponse,
  QueryNodesByValueRequest,
  QueryNodesByValueResponse,
  QueryNodesByValueResponse_NodesEntry,
  QueryNodesDistributionRequest,
  QueryNodesDistributionResponse,
  QueryNodesDistributionResponse_NodeDistributionEntry,
  QueryNodesRequest,
  QueryNodesResponse,
  QueryNodesResponse_NodesEntry,
  QuerySparkInvoicesRequest,
  QuerySparkInvoicesResponse,
  QueryStaticDepositAddressesRequest,
  QueryStaticDepositAddressesResponse,
  QueryTransfersResponse,
  QueryUnusedDepositAddressesRequest,
  QueryUnusedDepositAddressesResponse,
  QueryUserSignedRefundsRequest,
  QueryUserSignedRefundsResponse,
  QueryWalletSettingRequest,
  QueryWalletSettingResponse,
  RefreshTimelockRequest,
  RefreshTimelockResponse,
  RefreshTimelockSigningResult,
  RenewLeafRequest,
  RenewLeafResponse,
  RenewNodeTimelockResult,
  RenewNodeTimelockSigningJob,
  RenewNodeZeroTimelockResult,
  RenewNodeZeroTimelockSigningJob,
  RenewRefundTimelockResult,
  RenewRefundTimelockSigningJob,
  RequestedSigningCommitments,
  RequestedSigningCommitments_SigningNonceCommitmentsEntry,
  RevocationSecretWithIndex,
  SatsPayment,
  SatsTransfer,
  SecretProof,
  SecretShare,
  SendLeafKeyTweak,
  SendLeafKeyTweak_PubkeySharesTweakEntry,
  SendLeafKeyTweaks,
  SignatureWithIndex,
  SigningCommitments,
  SigningCommitments_SigningCommitmentsEntry,
  SigningJob,
  SigningKeyshare,
  SigningKeyshare_PublicSharesEntry,
  SigningOperatorInfo,
  SigningResult,
  SigningResult_PublicKeysEntry,
  SigningResult_SignatureSharesEntry,
  SigningResult_SigningNonceCommitmentsEntry,
  SparkAddress,
  SparkInvoiceFields,
  SparkServiceDefinition,
  StartDepositTreeCreationRequest,
  StartDepositTreeCreationResponse,
  StartTransferRequest,
  StartTransferResponse,
  StartTreeCreationRequest,
  StartTreeCreationResponse,
  StartUserSignedTransferRequest,
  StorePreimageShareRequest,
  SubscribeToEventsRequest,
  SubscribeToEventsResponse,
  TokenCreateInput,
  TokenMintInput,
  TokenOutput,
  TokenOutputToSpend,
  TokenTransaction,
  TokenTransfer,
  TokenTransferInput,
  TokensPayment,
  Transfer,
  TransferEvent,
  TransferFilter,
  TransferLeaf,
  TransferPackage,
  TransferPackage_KeyTweakPackageEntry,
  TransferStatus,
  TransferType,
  TreeNode,
  TreeNodeIds,
  TreeNodeStatus,
  UTXO,
  UpdateWalletSettingRequest,
  UpdateWalletSettingResponse,
  UserSignedRefund,
  UserSignedTxSigningJob,
  UtxoSwapRequestType,
  WalletSetting,
  directionFromJSON,
  directionToJSON,
  initiatePreimageSwapRequest_ReasonFromJSON,
  initiatePreimageSwapRequest_ReasonToJSON,
  invoiceStatusFromJSON,
  invoiceStatusToJSON,
  networkFromJSON,
  networkToJSON,
  orderFromJSON,
  orderToJSON,
  preimageRequestRoleFromJSON,
  preimageRequestRoleToJSON,
  preimageRequestStatusFromJSON,
  preimageRequestStatusToJSON,
  protobufPackage,
  transferStatusFromJSON,
  transferStatusToJSON,
  transferTypeFromJSON,
  transferTypeToJSON,
  treeNodeStatusFromJSON,
  treeNodeStatusToJSON,
  utxoSwapRequestTypeFromJSON,
  utxoSwapRequestTypeToJSON
} from "../chunk-JE3MXMPW.js";
import "../chunk-MVRQ5US7.js";
export {
  AdaptorPublicKeyPackage,
  Address,
  AddressNode,
  AddressRequestNode,
  BitcoinTransactionOutput,
  CancelTransferRequest,
  CancelTransferResponse,
  ClaimLeafKeyTweak,
  ClaimLeafKeyTweak_PubkeySharesTweakEntry,
  ClaimTransferSignRefundsRequest,
  ClaimTransferSignRefundsResponse,
  ClaimTransferTweakKeysRequest,
  ConnectedEvent,
  CooperativeExitRequest,
  CooperativeExitResponse,
  CounterLeafSwapRequest,
  CounterLeafSwapResponse,
  CreateTreeRequest,
  CreateTreeResponse,
  CreationNode,
  CreationResponseNode,
  DepositAddressProof,
  DepositAddressProof_AddressSignaturesEntry,
  DepositAddressQueryResult,
  DepositEvent,
  Direction,
  ExitSingleNodeTreeSigningResult,
  ExitSingleNodeTreesRequest,
  ExitSingleNodeTreesResponse,
  ExitingTree,
  ExtendLeafRequest,
  ExtendLeafResponse,
  ExtendLeafSigningResult,
  FinalizeNodeSignaturesRequest,
  FinalizeNodeSignaturesResponse,
  FinalizeTransferRequest,
  FinalizeTransferResponse,
  FinalizeTransferWithTransferPackageRequest,
  FreezeTokensPayload,
  FreezeTokensRequest,
  FreezeTokensResponse,
  GenerateDepositAddressRequest,
  GenerateDepositAddressResponse,
  GenerateStaticDepositAddressRequest,
  GenerateStaticDepositAddressResponse,
  GetSigningCommitmentsRequest,
  GetSigningCommitmentsResponse,
  GetSigningOperatorListResponse,
  GetSigningOperatorListResponse_SigningOperatorsEntry,
  GetUtxosForAddressRequest,
  GetUtxosForAddressResponse,
  InitiatePreimageSwapRequest,
  InitiatePreimageSwapRequest_Reason,
  InitiatePreimageSwapResponse,
  InitiateStaticDepositUtxoRefundRequest,
  InitiateStaticDepositUtxoRefundResponse,
  InitiateSwapPrimaryTransferRequest,
  InitiateSwapPrimaryTransferResponse,
  InitiateUtxoSwapRequest,
  InitiateUtxoSwapResponse,
  InvoiceAmount,
  InvoiceAmountProof,
  InvoiceResponse,
  InvoiceStatus,
  LeafRefundTxSigningJob,
  LeafRefundTxSigningResult,
  Network,
  NodeOutput,
  NodeSignatureShares,
  NodeSignatures,
  OperatorSpecificOwnerSignature,
  OperatorSpecificTokenTransactionSignablePayload,
  Order,
  OutPoint,
  PageRequest,
  PageResponse,
  PreimageRequestRole,
  PreimageRequestStatus,
  PreimageRequestWithTransfer,
  PrepareTreeAddressRequest,
  PrepareTreeAddressResponse,
  ProvidePreimageRequest,
  ProvidePreimageResponse,
  QueryBalanceRequest,
  QueryBalanceResponse,
  QueryBalanceResponse_NodeBalancesEntry,
  QueryHtlcRequest,
  QueryHtlcResponse,
  QueryNodesByValueRequest,
  QueryNodesByValueResponse,
  QueryNodesByValueResponse_NodesEntry,
  QueryNodesDistributionRequest,
  QueryNodesDistributionResponse,
  QueryNodesDistributionResponse_NodeDistributionEntry,
  QueryNodesRequest,
  QueryNodesResponse,
  QueryNodesResponse_NodesEntry,
  QuerySparkInvoicesRequest,
  QuerySparkInvoicesResponse,
  QueryStaticDepositAddressesRequest,
  QueryStaticDepositAddressesResponse,
  QueryTransfersResponse,
  QueryUnusedDepositAddressesRequest,
  QueryUnusedDepositAddressesResponse,
  QueryUserSignedRefundsRequest,
  QueryUserSignedRefundsResponse,
  QueryWalletSettingRequest,
  QueryWalletSettingResponse,
  RefreshTimelockRequest,
  RefreshTimelockResponse,
  RefreshTimelockSigningResult,
  RenewLeafRequest,
  RenewLeafResponse,
  RenewNodeTimelockResult,
  RenewNodeTimelockSigningJob,
  RenewNodeZeroTimelockResult,
  RenewNodeZeroTimelockSigningJob,
  RenewRefundTimelockResult,
  RenewRefundTimelockSigningJob,
  RequestedSigningCommitments,
  RequestedSigningCommitments_SigningNonceCommitmentsEntry,
  RevocationSecretWithIndex,
  SatsPayment,
  SatsTransfer,
  SecretProof,
  SecretShare,
  SendLeafKeyTweak,
  SendLeafKeyTweak_PubkeySharesTweakEntry,
  SendLeafKeyTweaks,
  SignatureWithIndex,
  SigningCommitments,
  SigningCommitments_SigningCommitmentsEntry,
  SigningJob,
  SigningKeyshare,
  SigningKeyshare_PublicSharesEntry,
  SigningOperatorInfo,
  SigningResult,
  SigningResult_PublicKeysEntry,
  SigningResult_SignatureSharesEntry,
  SigningResult_SigningNonceCommitmentsEntry,
  SparkAddress,
  SparkInvoiceFields,
  SparkServiceDefinition,
  StartDepositTreeCreationRequest,
  StartDepositTreeCreationResponse,
  StartTransferRequest,
  StartTransferResponse,
  StartTreeCreationRequest,
  StartTreeCreationResponse,
  StartUserSignedTransferRequest,
  StorePreimageShareRequest,
  SubscribeToEventsRequest,
  SubscribeToEventsResponse,
  TokenCreateInput,
  TokenMintInput,
  TokenOutput,
  TokenOutputToSpend,
  TokenTransaction,
  TokenTransfer,
  TokenTransferInput,
  TokensPayment,
  Transfer,
  TransferEvent,
  TransferFilter,
  TransferLeaf,
  TransferPackage,
  TransferPackage_KeyTweakPackageEntry,
  TransferStatus,
  TransferType,
  TreeNode,
  TreeNodeIds,
  TreeNodeStatus,
  UTXO,
  UpdateWalletSettingRequest,
  UpdateWalletSettingResponse,
  UserSignedRefund,
  UserSignedTxSigningJob,
  UtxoSwapRequestType,
  WalletSetting,
  directionFromJSON,
  directionToJSON,
  initiatePreimageSwapRequest_ReasonFromJSON,
  initiatePreimageSwapRequest_ReasonToJSON,
  invoiceStatusFromJSON,
  invoiceStatusToJSON,
  networkFromJSON,
  networkToJSON,
  orderFromJSON,
  orderToJSON,
  preimageRequestRoleFromJSON,
  preimageRequestRoleToJSON,
  preimageRequestStatusFromJSON,
  preimageRequestStatusToJSON,
  protobufPackage,
  transferStatusFromJSON,
  transferStatusToJSON,
  transferTypeFromJSON,
  transferTypeToJSON,
  treeNodeStatusFromJSON,
  treeNodeStatusToJSON,
  utxoSwapRequestTypeFromJSON,
  utxoSwapRequestTypeToJSON
};
