// @buildonspark/spark-sdk/dist/proto/spark_token.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/proto/spark_token.ts
var spark_token_exports = {};
__export(spark_token_exports, {
  BroadcastTransactionRequest: () => BroadcastTransactionRequest,
  BroadcastTransactionResponse: () => BroadcastTransactionResponse,
  CommitProgress: () => CommitProgress,
  CommitStatus: () => CommitStatus,
  CommitTransactionRequest: () => CommitTransactionRequest,
  CommitTransactionResponse: () => CommitTransactionResponse,
  FinalTokenOutput: () => FinalTokenOutput,
  FinalTokenTransaction: () => FinalTokenTransaction,
  FreezeTokensPayload: () => FreezeTokensPayload,
  FreezeTokensRequest: () => FreezeTokensRequest,
  FreezeTokensResponse: () => FreezeTokensResponse,
  InputTtxoSignaturesPerOperator: () => InputTtxoSignaturesPerOperator,
  InvoiceAttachment: () => InvoiceAttachment,
  OutputWithPreviousTransactionData: () => OutputWithPreviousTransactionData,
  PartialTokenOutput: () => PartialTokenOutput,
  PartialTokenTransaction: () => PartialTokenTransaction,
  QueryTokenMetadataRequest: () => QueryTokenMetadataRequest,
  QueryTokenMetadataResponse: () => QueryTokenMetadataResponse,
  QueryTokenOutputsRequest: () => QueryTokenOutputsRequest,
  QueryTokenOutputsResponse: () => QueryTokenOutputsResponse,
  QueryTokenTransactionsRequest: () => QueryTokenTransactionsRequest,
  QueryTokenTransactionsResponse: () => QueryTokenTransactionsResponse,
  SignatureWithIndex: () => SignatureWithIndex,
  SparkTokenServiceDefinition: () => SparkTokenServiceDefinition,
  SpentTokenOutputMetadata: () => SpentTokenOutputMetadata,
  StartTransactionRequest: () => StartTransactionRequest,
  StartTransactionResponse: () => StartTransactionResponse,
  TokenCreateInput: () => TokenCreateInput,
  TokenMetadata: () => TokenMetadata,
  TokenMintInput: () => TokenMintInput,
  TokenOutput: () => TokenOutput,
  TokenOutputToSpend: () => TokenOutputToSpend,
  TokenTransaction: () => TokenTransaction,
  TokenTransactionConfirmationMetadata: () => TokenTransactionConfirmationMetadata,
  TokenTransactionMetadata: () => TokenTransactionMetadata,
  TokenTransactionStatus: () => TokenTransactionStatus,
  TokenTransactionType: () => TokenTransactionType,
  TokenTransactionWithStatus: () => TokenTransactionWithStatus,
  TokenTransferInput: () => TokenTransferInput,
  commitStatusFromJSON: () => commitStatusFromJSON,
  commitStatusToJSON: () => commitStatusToJSON,
  protobufPackage: () => protobufPackage,
  tokenTransactionStatusFromJSON: () => tokenTransactionStatusFromJSON,
  tokenTransactionStatusToJSON: () => tokenTransactionStatusToJSON,
  tokenTransactionTypeFromJSON: () => tokenTransactionTypeFromJSON,
  tokenTransactionTypeToJSON: () => tokenTransactionTypeToJSON
});
module.exports = __toCommonJS(spark_token_exports);

// buffer.js
var import_buffer = require("buffer");
if (typeof globalThis.Buffer === "undefined") {
  globalThis.Buffer = import_buffer.Buffer;
}
if (typeof window !== "undefined") {
  if (typeof window.global === "undefined") {
    window.global = window;
  }
  if (typeof window.globalThis === "undefined") {
    window.globalThis = window;
  }
}

// src/proto/spark_token.ts
var import_wire5 = require("@bufbuild/protobuf/wire");

// src/proto/google/protobuf/timestamp.ts
var import_wire = require("@bufbuild/protobuf/wire");
function createBaseTimestamp() {
  return { seconds: 0, nanos: 0 };
}
var Timestamp = {
  encode(message, writer = new import_wire.BinaryWriter()) {
    if (message.seconds !== 0) {
      writer.uint32(8).int64(message.seconds);
    }
    if (message.nanos !== 0) {
      writer.uint32(16).int32(message.nanos);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire.BinaryReader ? input : new import_wire.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseTimestamp();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 8) {
            break;
          }
          message.seconds = longToNumber(reader.int64());
          continue;
        }
        case 2: {
          if (tag !== 16) {
            break;
          }
          message.nanos = reader.int32();
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      seconds: isSet(object.seconds) ? globalThis.Number(object.seconds) : 0,
      nanos: isSet(object.nanos) ? globalThis.Number(object.nanos) : 0
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.seconds !== 0) {
      obj.seconds = Math.round(message.seconds);
    }
    if (message.nanos !== 0) {
      obj.nanos = Math.round(message.nanos);
    }
    return obj;
  },
  create(base) {
    return Timestamp.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseTimestamp();
    message.seconds = object.seconds ?? 0;
    message.nanos = object.nanos ?? 0;
    return message;
  }
};
function longToNumber(int64) {
  const num = globalThis.Number(int64.toString());
  if (num > globalThis.Number.MAX_SAFE_INTEGER) {
    throw new globalThis.Error("Value is larger than Number.MAX_SAFE_INTEGER");
  }
  if (num < globalThis.Number.MIN_SAFE_INTEGER) {
    throw new globalThis.Error("Value is smaller than Number.MIN_SAFE_INTEGER");
  }
  return num;
}
function isSet(value) {
  return value !== null && value !== void 0;
}

// src/proto/spark.ts
var import_wire4 = require("@bufbuild/protobuf/wire");

// src/proto/common.ts
var import_wire2 = require("@bufbuild/protobuf/wire");

// src/proto/google/protobuf/empty.ts
var import_wire3 = require("@bufbuild/protobuf/wire");

// src/proto/spark.ts
function networkFromJSON(object) {
  switch (object) {
    case 0:
    case "UNSPECIFIED":
      return 0 /* UNSPECIFIED */;
    case 1:
    case "MAINNET":
      return 1 /* MAINNET */;
    case 2:
    case "REGTEST":
      return 2 /* REGTEST */;
    case 3:
    case "TESTNET":
      return 3 /* TESTNET */;
    case 4:
    case "SIGNET":
      return 4 /* SIGNET */;
    case -1:
    case "UNRECOGNIZED":
    default:
      return -1 /* UNRECOGNIZED */;
  }
}
function networkToJSON(object) {
  switch (object) {
    case 0 /* UNSPECIFIED */:
      return "UNSPECIFIED";
    case 1 /* MAINNET */:
      return "MAINNET";
    case 2 /* REGTEST */:
      return "REGTEST";
    case 3 /* TESTNET */:
      return "TESTNET";
    case 4 /* SIGNET */:
      return "SIGNET";
    case -1 /* UNRECOGNIZED */:
    default:
      return "UNRECOGNIZED";
  }
}
function directionFromJSON(object) {
  switch (object) {
    case 0:
    case "NEXT":
      return 0 /* NEXT */;
    case 1:
    case "PREVIOUS":
      return 1 /* PREVIOUS */;
    case -1:
    case "UNRECOGNIZED":
    default:
      return -1 /* UNRECOGNIZED */;
  }
}
function directionToJSON(object) {
  switch (object) {
    case 0 /* NEXT */:
      return "NEXT";
    case 1 /* PREVIOUS */:
      return "PREVIOUS";
    case -1 /* UNRECOGNIZED */:
    default:
      return "UNRECOGNIZED";
  }
}
function orderFromJSON(object) {
  switch (object) {
    case 0:
    case "DESCENDING":
      return 0 /* DESCENDING */;
    case 1:
    case "ASCENDING":
      return 1 /* ASCENDING */;
    case -1:
    case "UNRECOGNIZED":
    default:
      return -1 /* UNRECOGNIZED */;
  }
}
function orderToJSON(object) {
  switch (object) {
    case 0 /* DESCENDING */:
      return "DESCENDING";
    case 1 /* ASCENDING */:
      return "ASCENDING";
    case -1 /* UNRECOGNIZED */:
    default:
      return "UNRECOGNIZED";
  }
}
function createBasePageRequest() {
  return { pageSize: 0, cursor: "", direction: 0 };
}
var PageRequest = {
  encode(message, writer = new import_wire4.BinaryWriter()) {
    if (message.pageSize !== 0) {
      writer.uint32(8).int32(message.pageSize);
    }
    if (message.cursor !== "") {
      writer.uint32(18).string(message.cursor);
    }
    if (message.direction !== 0) {
      writer.uint32(24).int32(message.direction);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire4.BinaryReader ? input : new import_wire4.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBasePageRequest();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 8) {
            break;
          }
          message.pageSize = reader.int32();
          continue;
        }
        case 2: {
          if (tag !== 18) {
            break;
          }
          message.cursor = reader.string();
          continue;
        }
        case 3: {
          if (tag !== 24) {
            break;
          }
          message.direction = reader.int32();
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      pageSize: isSet2(object.pageSize) ? globalThis.Number(object.pageSize) : 0,
      cursor: isSet2(object.cursor) ? globalThis.String(object.cursor) : "",
      direction: isSet2(object.direction) ? directionFromJSON(object.direction) : 0
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.pageSize !== 0) {
      obj.pageSize = Math.round(message.pageSize);
    }
    if (message.cursor !== "") {
      obj.cursor = message.cursor;
    }
    if (message.direction !== 0) {
      obj.direction = directionToJSON(message.direction);
    }
    return obj;
  },
  create(base) {
    return PageRequest.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBasePageRequest();
    message.pageSize = object.pageSize ?? 0;
    message.cursor = object.cursor ?? "";
    message.direction = object.direction ?? 0;
    return message;
  }
};
function createBasePageResponse() {
  return { hasNextPage: false, hasPreviousPage: false, nextCursor: "", previousCursor: "" };
}
var PageResponse = {
  encode(message, writer = new import_wire4.BinaryWriter()) {
    if (message.hasNextPage !== false) {
      writer.uint32(8).bool(message.hasNextPage);
    }
    if (message.hasPreviousPage !== false) {
      writer.uint32(16).bool(message.hasPreviousPage);
    }
    if (message.nextCursor !== "") {
      writer.uint32(26).string(message.nextCursor);
    }
    if (message.previousCursor !== "") {
      writer.uint32(34).string(message.previousCursor);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire4.BinaryReader ? input : new import_wire4.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBasePageResponse();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 8) {
            break;
          }
          message.hasNextPage = reader.bool();
          continue;
        }
        case 2: {
          if (tag !== 16) {
            break;
          }
          message.hasPreviousPage = reader.bool();
          continue;
        }
        case 3: {
          if (tag !== 26) {
            break;
          }
          message.nextCursor = reader.string();
          continue;
        }
        case 4: {
          if (tag !== 34) {
            break;
          }
          message.previousCursor = reader.string();
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      hasNextPage: isSet2(object.hasNextPage) ? globalThis.Boolean(object.hasNextPage) : false,
      hasPreviousPage: isSet2(object.hasPreviousPage) ? globalThis.Boolean(object.hasPreviousPage) : false,
      nextCursor: isSet2(object.nextCursor) ? globalThis.String(object.nextCursor) : "",
      previousCursor: isSet2(object.previousCursor) ? globalThis.String(object.previousCursor) : ""
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.hasNextPage !== false) {
      obj.hasNextPage = message.hasNextPage;
    }
    if (message.hasPreviousPage !== false) {
      obj.hasPreviousPage = message.hasPreviousPage;
    }
    if (message.nextCursor !== "") {
      obj.nextCursor = message.nextCursor;
    }
    if (message.previousCursor !== "") {
      obj.previousCursor = message.previousCursor;
    }
    return obj;
  },
  create(base) {
    return PageResponse.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBasePageResponse();
    message.hasNextPage = object.hasNextPage ?? false;
    message.hasPreviousPage = object.hasPreviousPage ?? false;
    message.nextCursor = object.nextCursor ?? "";
    message.previousCursor = object.previousCursor ?? "";
    return message;
  }
};
function createBaseSigningKeyshare() {
  return { ownerIdentifiers: [], threshold: 0, publicKey: new Uint8Array(0), publicShares: {}, updatedTime: void 0 };
}
var SigningKeyshare = {
  encode(message, writer = new import_wire4.BinaryWriter()) {
    for (const v of message.ownerIdentifiers) {
      writer.uint32(10).string(v);
    }
    if (message.threshold !== 0) {
      writer.uint32(16).uint32(message.threshold);
    }
    if (message.publicKey.length !== 0) {
      writer.uint32(26).bytes(message.publicKey);
    }
    Object.entries(message.publicShares).forEach(([key, value]) => {
      SigningKeyshare_PublicSharesEntry.encode({ key, value }, writer.uint32(34).fork()).join();
    });
    if (message.updatedTime !== void 0) {
      Timestamp.encode(toTimestamp(message.updatedTime), writer.uint32(42).fork()).join();
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire4.BinaryReader ? input : new import_wire4.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseSigningKeyshare();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 10) {
            break;
          }
          message.ownerIdentifiers.push(reader.string());
          continue;
        }
        case 2: {
          if (tag !== 16) {
            break;
          }
          message.threshold = reader.uint32();
          continue;
        }
        case 3: {
          if (tag !== 26) {
            break;
          }
          message.publicKey = reader.bytes();
          continue;
        }
        case 4: {
          if (tag !== 34) {
            break;
          }
          const entry4 = SigningKeyshare_PublicSharesEntry.decode(reader, reader.uint32());
          if (entry4.value !== void 0) {
            message.publicShares[entry4.key] = entry4.value;
          }
          continue;
        }
        case 5: {
          if (tag !== 42) {
            break;
          }
          message.updatedTime = fromTimestamp(Timestamp.decode(reader, reader.uint32()));
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      ownerIdentifiers: globalThis.Array.isArray(object?.ownerIdentifiers) ? object.ownerIdentifiers.map((e) => globalThis.String(e)) : [],
      threshold: isSet2(object.threshold) ? globalThis.Number(object.threshold) : 0,
      publicKey: isSet2(object.publicKey) ? bytesFromBase64(object.publicKey) : new Uint8Array(0),
      publicShares: isObject(object.publicShares) ? Object.entries(object.publicShares).reduce((acc, [key, value]) => {
        acc[key] = bytesFromBase64(value);
        return acc;
      }, {}) : {},
      updatedTime: isSet2(object.updatedTime) ? fromJsonTimestamp(object.updatedTime) : void 0
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.ownerIdentifiers?.length) {
      obj.ownerIdentifiers = message.ownerIdentifiers;
    }
    if (message.threshold !== 0) {
      obj.threshold = Math.round(message.threshold);
    }
    if (message.publicKey.length !== 0) {
      obj.publicKey = base64FromBytes(message.publicKey);
    }
    if (message.publicShares) {
      const entries = Object.entries(message.publicShares);
      if (entries.length > 0) {
        obj.publicShares = {};
        entries.forEach(([k, v]) => {
          obj.publicShares[k] = base64FromBytes(v);
        });
      }
    }
    if (message.updatedTime !== void 0) {
      obj.updatedTime = message.updatedTime.toISOString();
    }
    return obj;
  },
  create(base) {
    return SigningKeyshare.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseSigningKeyshare();
    message.ownerIdentifiers = object.ownerIdentifiers?.map((e) => e) || [];
    message.threshold = object.threshold ?? 0;
    message.publicKey = object.publicKey ?? new Uint8Array(0);
    message.publicShares = Object.entries(object.publicShares ?? {}).reduce(
      (acc, [key, value]) => {
        if (value !== void 0) {
          acc[key] = value;
        }
        return acc;
      },
      {}
    );
    message.updatedTime = object.updatedTime ?? void 0;
    return message;
  }
};
function createBaseSigningKeyshare_PublicSharesEntry() {
  return { key: "", value: new Uint8Array(0) };
}
var SigningKeyshare_PublicSharesEntry = {
  encode(message, writer = new import_wire4.BinaryWriter()) {
    if (message.key !== "") {
      writer.uint32(10).string(message.key);
    }
    if (message.value.length !== 0) {
      writer.uint32(18).bytes(message.value);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire4.BinaryReader ? input : new import_wire4.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseSigningKeyshare_PublicSharesEntry();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 10) {
            break;
          }
          message.key = reader.string();
          continue;
        }
        case 2: {
          if (tag !== 18) {
            break;
          }
          message.value = reader.bytes();
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      key: isSet2(object.key) ? globalThis.String(object.key) : "",
      value: isSet2(object.value) ? bytesFromBase64(object.value) : new Uint8Array(0)
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.key !== "") {
      obj.key = message.key;
    }
    if (message.value.length !== 0) {
      obj.value = base64FromBytes(message.value);
    }
    return obj;
  },
  create(base) {
    return SigningKeyshare_PublicSharesEntry.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseSigningKeyshare_PublicSharesEntry();
    message.key = object.key ?? "";
    message.value = object.value ?? new Uint8Array(0);
    return message;
  }
};
function bytesFromBase64(b64) {
  if (globalThis.Buffer) {
    return Uint8Array.from(globalThis.Buffer.from(b64, "base64"));
  } else {
    const bin = globalThis.atob(b64);
    const arr = new Uint8Array(bin.length);
    for (let i = 0; i < bin.length; ++i) {
      arr[i] = bin.charCodeAt(i);
    }
    return arr;
  }
}
function base64FromBytes(arr) {
  if (globalThis.Buffer) {
    return globalThis.Buffer.from(arr).toString("base64");
  } else {
    const bin = [];
    arr.forEach((byte) => {
      bin.push(globalThis.String.fromCharCode(byte));
    });
    return globalThis.btoa(bin.join(""));
  }
}
function toTimestamp(date) {
  const seconds = Math.trunc(date.getTime() / 1e3);
  const nanos = date.getTime() % 1e3 * 1e6;
  return { seconds, nanos };
}
function fromTimestamp(t) {
  let millis = (t.seconds || 0) * 1e3;
  millis += (t.nanos || 0) / 1e6;
  return new globalThis.Date(millis);
}
function fromJsonTimestamp(o) {
  if (o instanceof globalThis.Date) {
    return o;
  } else if (typeof o === "string") {
    return new globalThis.Date(o);
  } else {
    return fromTimestamp(Timestamp.fromJSON(o));
  }
}
function isObject(value) {
  return typeof value === "object" && value !== null;
}
function isSet2(value) {
  return value !== null && value !== void 0;
}

// src/proto/spark_token.ts
var protobufPackage = "spark_token";
var TokenTransactionType = /* @__PURE__ */ ((TokenTransactionType2) => {
  TokenTransactionType2[TokenTransactionType2["TOKEN_TRANSACTION_TYPE_UNSPECIFIED"] = 0] = "TOKEN_TRANSACTION_TYPE_UNSPECIFIED";
  TokenTransactionType2[TokenTransactionType2["TOKEN_TRANSACTION_TYPE_CREATE"] = 1] = "TOKEN_TRANSACTION_TYPE_CREATE";
  TokenTransactionType2[TokenTransactionType2["TOKEN_TRANSACTION_TYPE_MINT"] = 2] = "TOKEN_TRANSACTION_TYPE_MINT";
  TokenTransactionType2[TokenTransactionType2["TOKEN_TRANSACTION_TYPE_TRANSFER"] = 3] = "TOKEN_TRANSACTION_TYPE_TRANSFER";
  TokenTransactionType2[TokenTransactionType2["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
  return TokenTransactionType2;
})(TokenTransactionType || {});
function tokenTransactionTypeFromJSON(object) {
  switch (object) {
    case 0:
    case "TOKEN_TRANSACTION_TYPE_UNSPECIFIED":
      return 0 /* TOKEN_TRANSACTION_TYPE_UNSPECIFIED */;
    case 1:
    case "TOKEN_TRANSACTION_TYPE_CREATE":
      return 1 /* TOKEN_TRANSACTION_TYPE_CREATE */;
    case 2:
    case "TOKEN_TRANSACTION_TYPE_MINT":
      return 2 /* TOKEN_TRANSACTION_TYPE_MINT */;
    case 3:
    case "TOKEN_TRANSACTION_TYPE_TRANSFER":
      return 3 /* TOKEN_TRANSACTION_TYPE_TRANSFER */;
    case -1:
    case "UNRECOGNIZED":
    default:
      return -1 /* UNRECOGNIZED */;
  }
}
function tokenTransactionTypeToJSON(object) {
  switch (object) {
    case 0 /* TOKEN_TRANSACTION_TYPE_UNSPECIFIED */:
      return "TOKEN_TRANSACTION_TYPE_UNSPECIFIED";
    case 1 /* TOKEN_TRANSACTION_TYPE_CREATE */:
      return "TOKEN_TRANSACTION_TYPE_CREATE";
    case 2 /* TOKEN_TRANSACTION_TYPE_MINT */:
      return "TOKEN_TRANSACTION_TYPE_MINT";
    case 3 /* TOKEN_TRANSACTION_TYPE_TRANSFER */:
      return "TOKEN_TRANSACTION_TYPE_TRANSFER";
    case -1 /* UNRECOGNIZED */:
    default:
      return "UNRECOGNIZED";
  }
}
var CommitStatus = /* @__PURE__ */ ((CommitStatus2) => {
  CommitStatus2[CommitStatus2["COMMIT_UNSPECIFIED"] = 0] = "COMMIT_UNSPECIFIED";
  CommitStatus2[CommitStatus2["COMMIT_PROCESSING"] = 1] = "COMMIT_PROCESSING";
  CommitStatus2[CommitStatus2["COMMIT_FINALIZED"] = 2] = "COMMIT_FINALIZED";
  CommitStatus2[CommitStatus2["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
  return CommitStatus2;
})(CommitStatus || {});
function commitStatusFromJSON(object) {
  switch (object) {
    case 0:
    case "COMMIT_UNSPECIFIED":
      return 0 /* COMMIT_UNSPECIFIED */;
    case 1:
    case "COMMIT_PROCESSING":
      return 1 /* COMMIT_PROCESSING */;
    case 2:
    case "COMMIT_FINALIZED":
      return 2 /* COMMIT_FINALIZED */;
    case -1:
    case "UNRECOGNIZED":
    default:
      return -1 /* UNRECOGNIZED */;
  }
}
function commitStatusToJSON(object) {
  switch (object) {
    case 0 /* COMMIT_UNSPECIFIED */:
      return "COMMIT_UNSPECIFIED";
    case 1 /* COMMIT_PROCESSING */:
      return "COMMIT_PROCESSING";
    case 2 /* COMMIT_FINALIZED */:
      return "COMMIT_FINALIZED";
    case -1 /* UNRECOGNIZED */:
    default:
      return "UNRECOGNIZED";
  }
}
var TokenTransactionStatus = /* @__PURE__ */ ((TokenTransactionStatus2) => {
  TokenTransactionStatus2[TokenTransactionStatus2["TOKEN_TRANSACTION_STARTED"] = 0] = "TOKEN_TRANSACTION_STARTED";
  TokenTransactionStatus2[TokenTransactionStatus2["TOKEN_TRANSACTION_SIGNED"] = 1] = "TOKEN_TRANSACTION_SIGNED";
  TokenTransactionStatus2[TokenTransactionStatus2["TOKEN_TRANSACTION_REVEALED"] = 5] = "TOKEN_TRANSACTION_REVEALED";
  TokenTransactionStatus2[TokenTransactionStatus2["TOKEN_TRANSACTION_FINALIZED"] = 2] = "TOKEN_TRANSACTION_FINALIZED";
  TokenTransactionStatus2[TokenTransactionStatus2["TOKEN_TRANSACTION_STARTED_CANCELLED"] = 3] = "TOKEN_TRANSACTION_STARTED_CANCELLED";
  TokenTransactionStatus2[TokenTransactionStatus2["TOKEN_TRANSACTION_SIGNED_CANCELLED"] = 4] = "TOKEN_TRANSACTION_SIGNED_CANCELLED";
  TokenTransactionStatus2[TokenTransactionStatus2["TOKEN_TRANSACTION_UNKNOWN"] = 10] = "TOKEN_TRANSACTION_UNKNOWN";
  TokenTransactionStatus2[TokenTransactionStatus2["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
  return TokenTransactionStatus2;
})(TokenTransactionStatus || {});
function tokenTransactionStatusFromJSON(object) {
  switch (object) {
    case 0:
    case "TOKEN_TRANSACTION_STARTED":
      return 0 /* TOKEN_TRANSACTION_STARTED */;
    case 1:
    case "TOKEN_TRANSACTION_SIGNED":
      return 1 /* TOKEN_TRANSACTION_SIGNED */;
    case 5:
    case "TOKEN_TRANSACTION_REVEALED":
      return 5 /* TOKEN_TRANSACTION_REVEALED */;
    case 2:
    case "TOKEN_TRANSACTION_FINALIZED":
      return 2 /* TOKEN_TRANSACTION_FINALIZED */;
    case 3:
    case "TOKEN_TRANSACTION_STARTED_CANCELLED":
      return 3 /* TOKEN_TRANSACTION_STARTED_CANCELLED */;
    case 4:
    case "TOKEN_TRANSACTION_SIGNED_CANCELLED":
      return 4 /* TOKEN_TRANSACTION_SIGNED_CANCELLED */;
    case 10:
    case "TOKEN_TRANSACTION_UNKNOWN":
      return 10 /* TOKEN_TRANSACTION_UNKNOWN */;
    case -1:
    case "UNRECOGNIZED":
    default:
      return -1 /* UNRECOGNIZED */;
  }
}
function tokenTransactionStatusToJSON(object) {
  switch (object) {
    case 0 /* TOKEN_TRANSACTION_STARTED */:
      return "TOKEN_TRANSACTION_STARTED";
    case 1 /* TOKEN_TRANSACTION_SIGNED */:
      return "TOKEN_TRANSACTION_SIGNED";
    case 5 /* TOKEN_TRANSACTION_REVEALED */:
      return "TOKEN_TRANSACTION_REVEALED";
    case 2 /* TOKEN_TRANSACTION_FINALIZED */:
      return "TOKEN_TRANSACTION_FINALIZED";
    case 3 /* TOKEN_TRANSACTION_STARTED_CANCELLED */:
      return "TOKEN_TRANSACTION_STARTED_CANCELLED";
    case 4 /* TOKEN_TRANSACTION_SIGNED_CANCELLED */:
      return "TOKEN_TRANSACTION_SIGNED_CANCELLED";
    case 10 /* TOKEN_TRANSACTION_UNKNOWN */:
      return "TOKEN_TRANSACTION_UNKNOWN";
    case -1 /* UNRECOGNIZED */:
    default:
      return "UNRECOGNIZED";
  }
}
function createBaseTokenOutputToSpend() {
  return { prevTokenTransactionHash: new Uint8Array(0), prevTokenTransactionVout: 0 };
}
var TokenOutputToSpend = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    if (message.prevTokenTransactionHash.length !== 0) {
      writer.uint32(10).bytes(message.prevTokenTransactionHash);
    }
    if (message.prevTokenTransactionVout !== 0) {
      writer.uint32(16).uint32(message.prevTokenTransactionVout);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseTokenOutputToSpend();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 10) {
            break;
          }
          message.prevTokenTransactionHash = reader.bytes();
          continue;
        }
        case 2: {
          if (tag !== 16) {
            break;
          }
          message.prevTokenTransactionVout = reader.uint32();
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      prevTokenTransactionHash: isSet3(object.prevTokenTransactionHash) ? bytesFromBase642(object.prevTokenTransactionHash) : new Uint8Array(0),
      prevTokenTransactionVout: isSet3(object.prevTokenTransactionVout) ? globalThis.Number(object.prevTokenTransactionVout) : 0
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.prevTokenTransactionHash.length !== 0) {
      obj.prevTokenTransactionHash = base64FromBytes2(message.prevTokenTransactionHash);
    }
    if (message.prevTokenTransactionVout !== 0) {
      obj.prevTokenTransactionVout = Math.round(message.prevTokenTransactionVout);
    }
    return obj;
  },
  create(base) {
    return TokenOutputToSpend.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseTokenOutputToSpend();
    message.prevTokenTransactionHash = object.prevTokenTransactionHash ?? new Uint8Array(0);
    message.prevTokenTransactionVout = object.prevTokenTransactionVout ?? 0;
    return message;
  }
};
function createBaseTokenTransferInput() {
  return { outputsToSpend: [] };
}
var TokenTransferInput = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    for (const v of message.outputsToSpend) {
      TokenOutputToSpend.encode(v, writer.uint32(10).fork()).join();
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseTokenTransferInput();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 10) {
            break;
          }
          message.outputsToSpend.push(TokenOutputToSpend.decode(reader, reader.uint32()));
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      outputsToSpend: globalThis.Array.isArray(object?.outputsToSpend) ? object.outputsToSpend.map((e) => TokenOutputToSpend.fromJSON(e)) : []
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.outputsToSpend?.length) {
      obj.outputsToSpend = message.outputsToSpend.map((e) => TokenOutputToSpend.toJSON(e));
    }
    return obj;
  },
  create(base) {
    return TokenTransferInput.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseTokenTransferInput();
    message.outputsToSpend = object.outputsToSpend?.map((e) => TokenOutputToSpend.fromPartial(e)) || [];
    return message;
  }
};
function createBaseTokenMintInput() {
  return { issuerPublicKey: new Uint8Array(0), tokenIdentifier: void 0 };
}
var TokenMintInput = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    if (message.issuerPublicKey.length !== 0) {
      writer.uint32(10).bytes(message.issuerPublicKey);
    }
    if (message.tokenIdentifier !== void 0) {
      writer.uint32(18).bytes(message.tokenIdentifier);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseTokenMintInput();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 10) {
            break;
          }
          message.issuerPublicKey = reader.bytes();
          continue;
        }
        case 2: {
          if (tag !== 18) {
            break;
          }
          message.tokenIdentifier = reader.bytes();
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      issuerPublicKey: isSet3(object.issuerPublicKey) ? bytesFromBase642(object.issuerPublicKey) : new Uint8Array(0),
      tokenIdentifier: isSet3(object.tokenIdentifier) ? bytesFromBase642(object.tokenIdentifier) : void 0
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.issuerPublicKey.length !== 0) {
      obj.issuerPublicKey = base64FromBytes2(message.issuerPublicKey);
    }
    if (message.tokenIdentifier !== void 0) {
      obj.tokenIdentifier = base64FromBytes2(message.tokenIdentifier);
    }
    return obj;
  },
  create(base) {
    return TokenMintInput.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseTokenMintInput();
    message.issuerPublicKey = object.issuerPublicKey ?? new Uint8Array(0);
    message.tokenIdentifier = object.tokenIdentifier ?? void 0;
    return message;
  }
};
function createBaseTokenCreateInput() {
  return {
    issuerPublicKey: new Uint8Array(0),
    tokenName: "",
    tokenTicker: "",
    decimals: 0,
    maxSupply: new Uint8Array(0),
    isFreezable: false,
    creationEntityPublicKey: void 0
  };
}
var TokenCreateInput = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    if (message.issuerPublicKey.length !== 0) {
      writer.uint32(10).bytes(message.issuerPublicKey);
    }
    if (message.tokenName !== "") {
      writer.uint32(18).string(message.tokenName);
    }
    if (message.tokenTicker !== "") {
      writer.uint32(26).string(message.tokenTicker);
    }
    if (message.decimals !== 0) {
      writer.uint32(32).uint32(message.decimals);
    }
    if (message.maxSupply.length !== 0) {
      writer.uint32(42).bytes(message.maxSupply);
    }
    if (message.isFreezable !== false) {
      writer.uint32(48).bool(message.isFreezable);
    }
    if (message.creationEntityPublicKey !== void 0) {
      writer.uint32(58).bytes(message.creationEntityPublicKey);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseTokenCreateInput();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 10) {
            break;
          }
          message.issuerPublicKey = reader.bytes();
          continue;
        }
        case 2: {
          if (tag !== 18) {
            break;
          }
          message.tokenName = reader.string();
          continue;
        }
        case 3: {
          if (tag !== 26) {
            break;
          }
          message.tokenTicker = reader.string();
          continue;
        }
        case 4: {
          if (tag !== 32) {
            break;
          }
          message.decimals = reader.uint32();
          continue;
        }
        case 5: {
          if (tag !== 42) {
            break;
          }
          message.maxSupply = reader.bytes();
          continue;
        }
        case 6: {
          if (tag !== 48) {
            break;
          }
          message.isFreezable = reader.bool();
          continue;
        }
        case 7: {
          if (tag !== 58) {
            break;
          }
          message.creationEntityPublicKey = reader.bytes();
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      issuerPublicKey: isSet3(object.issuerPublicKey) ? bytesFromBase642(object.issuerPublicKey) : new Uint8Array(0),
      tokenName: isSet3(object.tokenName) ? globalThis.String(object.tokenName) : "",
      tokenTicker: isSet3(object.tokenTicker) ? globalThis.String(object.tokenTicker) : "",
      decimals: isSet3(object.decimals) ? globalThis.Number(object.decimals) : 0,
      maxSupply: isSet3(object.maxSupply) ? bytesFromBase642(object.maxSupply) : new Uint8Array(0),
      isFreezable: isSet3(object.isFreezable) ? globalThis.Boolean(object.isFreezable) : false,
      creationEntityPublicKey: isSet3(object.creationEntityPublicKey) ? bytesFromBase642(object.creationEntityPublicKey) : void 0
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.issuerPublicKey.length !== 0) {
      obj.issuerPublicKey = base64FromBytes2(message.issuerPublicKey);
    }
    if (message.tokenName !== "") {
      obj.tokenName = message.tokenName;
    }
    if (message.tokenTicker !== "") {
      obj.tokenTicker = message.tokenTicker;
    }
    if (message.decimals !== 0) {
      obj.decimals = Math.round(message.decimals);
    }
    if (message.maxSupply.length !== 0) {
      obj.maxSupply = base64FromBytes2(message.maxSupply);
    }
    if (message.isFreezable !== false) {
      obj.isFreezable = message.isFreezable;
    }
    if (message.creationEntityPublicKey !== void 0) {
      obj.creationEntityPublicKey = base64FromBytes2(message.creationEntityPublicKey);
    }
    return obj;
  },
  create(base) {
    return TokenCreateInput.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseTokenCreateInput();
    message.issuerPublicKey = object.issuerPublicKey ?? new Uint8Array(0);
    message.tokenName = object.tokenName ?? "";
    message.tokenTicker = object.tokenTicker ?? "";
    message.decimals = object.decimals ?? 0;
    message.maxSupply = object.maxSupply ?? new Uint8Array(0);
    message.isFreezable = object.isFreezable ?? false;
    message.creationEntityPublicKey = object.creationEntityPublicKey ?? void 0;
    return message;
  }
};
function createBaseTokenOutput() {
  return {
    id: void 0,
    ownerPublicKey: new Uint8Array(0),
    revocationCommitment: void 0,
    withdrawBondSats: void 0,
    withdrawRelativeBlockLocktime: void 0,
    tokenPublicKey: void 0,
    tokenIdentifier: void 0,
    tokenAmount: new Uint8Array(0)
  };
}
var TokenOutput = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    if (message.id !== void 0) {
      writer.uint32(10).string(message.id);
    }
    if (message.ownerPublicKey.length !== 0) {
      writer.uint32(18).bytes(message.ownerPublicKey);
    }
    if (message.revocationCommitment !== void 0) {
      writer.uint32(26).bytes(message.revocationCommitment);
    }
    if (message.withdrawBondSats !== void 0) {
      writer.uint32(32).uint64(message.withdrawBondSats);
    }
    if (message.withdrawRelativeBlockLocktime !== void 0) {
      writer.uint32(40).uint64(message.withdrawRelativeBlockLocktime);
    }
    if (message.tokenPublicKey !== void 0) {
      writer.uint32(50).bytes(message.tokenPublicKey);
    }
    if (message.tokenIdentifier !== void 0) {
      writer.uint32(66).bytes(message.tokenIdentifier);
    }
    if (message.tokenAmount.length !== 0) {
      writer.uint32(58).bytes(message.tokenAmount);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseTokenOutput();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 10) {
            break;
          }
          message.id = reader.string();
          continue;
        }
        case 2: {
          if (tag !== 18) {
            break;
          }
          message.ownerPublicKey = reader.bytes();
          continue;
        }
        case 3: {
          if (tag !== 26) {
            break;
          }
          message.revocationCommitment = reader.bytes();
          continue;
        }
        case 4: {
          if (tag !== 32) {
            break;
          }
          message.withdrawBondSats = longToNumber2(reader.uint64());
          continue;
        }
        case 5: {
          if (tag !== 40) {
            break;
          }
          message.withdrawRelativeBlockLocktime = longToNumber2(reader.uint64());
          continue;
        }
        case 6: {
          if (tag !== 50) {
            break;
          }
          message.tokenPublicKey = reader.bytes();
          continue;
        }
        case 8: {
          if (tag !== 66) {
            break;
          }
          message.tokenIdentifier = reader.bytes();
          continue;
        }
        case 7: {
          if (tag !== 58) {
            break;
          }
          message.tokenAmount = reader.bytes();
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      id: isSet3(object.id) ? globalThis.String(object.id) : void 0,
      ownerPublicKey: isSet3(object.ownerPublicKey) ? bytesFromBase642(object.ownerPublicKey) : new Uint8Array(0),
      revocationCommitment: isSet3(object.revocationCommitment) ? bytesFromBase642(object.revocationCommitment) : void 0,
      withdrawBondSats: isSet3(object.withdrawBondSats) ? globalThis.Number(object.withdrawBondSats) : void 0,
      withdrawRelativeBlockLocktime: isSet3(object.withdrawRelativeBlockLocktime) ? globalThis.Number(object.withdrawRelativeBlockLocktime) : void 0,
      tokenPublicKey: isSet3(object.tokenPublicKey) ? bytesFromBase642(object.tokenPublicKey) : void 0,
      tokenIdentifier: isSet3(object.tokenIdentifier) ? bytesFromBase642(object.tokenIdentifier) : void 0,
      tokenAmount: isSet3(object.tokenAmount) ? bytesFromBase642(object.tokenAmount) : new Uint8Array(0)
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.id !== void 0) {
      obj.id = message.id;
    }
    if (message.ownerPublicKey.length !== 0) {
      obj.ownerPublicKey = base64FromBytes2(message.ownerPublicKey);
    }
    if (message.revocationCommitment !== void 0) {
      obj.revocationCommitment = base64FromBytes2(message.revocationCommitment);
    }
    if (message.withdrawBondSats !== void 0) {
      obj.withdrawBondSats = Math.round(message.withdrawBondSats);
    }
    if (message.withdrawRelativeBlockLocktime !== void 0) {
      obj.withdrawRelativeBlockLocktime = Math.round(message.withdrawRelativeBlockLocktime);
    }
    if (message.tokenPublicKey !== void 0) {
      obj.tokenPublicKey = base64FromBytes2(message.tokenPublicKey);
    }
    if (message.tokenIdentifier !== void 0) {
      obj.tokenIdentifier = base64FromBytes2(message.tokenIdentifier);
    }
    if (message.tokenAmount.length !== 0) {
      obj.tokenAmount = base64FromBytes2(message.tokenAmount);
    }
    return obj;
  },
  create(base) {
    return TokenOutput.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseTokenOutput();
    message.id = object.id ?? void 0;
    message.ownerPublicKey = object.ownerPublicKey ?? new Uint8Array(0);
    message.revocationCommitment = object.revocationCommitment ?? void 0;
    message.withdrawBondSats = object.withdrawBondSats ?? void 0;
    message.withdrawRelativeBlockLocktime = object.withdrawRelativeBlockLocktime ?? void 0;
    message.tokenPublicKey = object.tokenPublicKey ?? void 0;
    message.tokenIdentifier = object.tokenIdentifier ?? void 0;
    message.tokenAmount = object.tokenAmount ?? new Uint8Array(0);
    return message;
  }
};
function createBasePartialTokenOutput() {
  return {
    ownerPublicKey: new Uint8Array(0),
    withdrawBondSats: 0,
    withdrawRelativeBlockLocktime: 0,
    tokenIdentifier: new Uint8Array(0),
    tokenAmount: new Uint8Array(0)
  };
}
var PartialTokenOutput = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    if (message.ownerPublicKey.length !== 0) {
      writer.uint32(10).bytes(message.ownerPublicKey);
    }
    if (message.withdrawBondSats !== 0) {
      writer.uint32(16).uint64(message.withdrawBondSats);
    }
    if (message.withdrawRelativeBlockLocktime !== 0) {
      writer.uint32(24).uint64(message.withdrawRelativeBlockLocktime);
    }
    if (message.tokenIdentifier.length !== 0) {
      writer.uint32(34).bytes(message.tokenIdentifier);
    }
    if (message.tokenAmount.length !== 0) {
      writer.uint32(42).bytes(message.tokenAmount);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBasePartialTokenOutput();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 10) {
            break;
          }
          message.ownerPublicKey = reader.bytes();
          continue;
        }
        case 2: {
          if (tag !== 16) {
            break;
          }
          message.withdrawBondSats = longToNumber2(reader.uint64());
          continue;
        }
        case 3: {
          if (tag !== 24) {
            break;
          }
          message.withdrawRelativeBlockLocktime = longToNumber2(reader.uint64());
          continue;
        }
        case 4: {
          if (tag !== 34) {
            break;
          }
          message.tokenIdentifier = reader.bytes();
          continue;
        }
        case 5: {
          if (tag !== 42) {
            break;
          }
          message.tokenAmount = reader.bytes();
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      ownerPublicKey: isSet3(object.ownerPublicKey) ? bytesFromBase642(object.ownerPublicKey) : new Uint8Array(0),
      withdrawBondSats: isSet3(object.withdrawBondSats) ? globalThis.Number(object.withdrawBondSats) : 0,
      withdrawRelativeBlockLocktime: isSet3(object.withdrawRelativeBlockLocktime) ? globalThis.Number(object.withdrawRelativeBlockLocktime) : 0,
      tokenIdentifier: isSet3(object.tokenIdentifier) ? bytesFromBase642(object.tokenIdentifier) : new Uint8Array(0),
      tokenAmount: isSet3(object.tokenAmount) ? bytesFromBase642(object.tokenAmount) : new Uint8Array(0)
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.ownerPublicKey.length !== 0) {
      obj.ownerPublicKey = base64FromBytes2(message.ownerPublicKey);
    }
    if (message.withdrawBondSats !== 0) {
      obj.withdrawBondSats = Math.round(message.withdrawBondSats);
    }
    if (message.withdrawRelativeBlockLocktime !== 0) {
      obj.withdrawRelativeBlockLocktime = Math.round(message.withdrawRelativeBlockLocktime);
    }
    if (message.tokenIdentifier.length !== 0) {
      obj.tokenIdentifier = base64FromBytes2(message.tokenIdentifier);
    }
    if (message.tokenAmount.length !== 0) {
      obj.tokenAmount = base64FromBytes2(message.tokenAmount);
    }
    return obj;
  },
  create(base) {
    return PartialTokenOutput.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBasePartialTokenOutput();
    message.ownerPublicKey = object.ownerPublicKey ?? new Uint8Array(0);
    message.withdrawBondSats = object.withdrawBondSats ?? 0;
    message.withdrawRelativeBlockLocktime = object.withdrawRelativeBlockLocktime ?? 0;
    message.tokenIdentifier = object.tokenIdentifier ?? new Uint8Array(0);
    message.tokenAmount = object.tokenAmount ?? new Uint8Array(0);
    return message;
  }
};
function createBaseFinalTokenOutput() {
  return { partialTokenOutput: void 0, revocationCommitment: new Uint8Array(0) };
}
var FinalTokenOutput = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    if (message.partialTokenOutput !== void 0) {
      PartialTokenOutput.encode(message.partialTokenOutput, writer.uint32(10).fork()).join();
    }
    if (message.revocationCommitment.length !== 0) {
      writer.uint32(18).bytes(message.revocationCommitment);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseFinalTokenOutput();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 10) {
            break;
          }
          message.partialTokenOutput = PartialTokenOutput.decode(reader, reader.uint32());
          continue;
        }
        case 2: {
          if (tag !== 18) {
            break;
          }
          message.revocationCommitment = reader.bytes();
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      partialTokenOutput: isSet3(object.partialTokenOutput) ? PartialTokenOutput.fromJSON(object.partialTokenOutput) : void 0,
      revocationCommitment: isSet3(object.revocationCommitment) ? bytesFromBase642(object.revocationCommitment) : new Uint8Array(0)
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.partialTokenOutput !== void 0) {
      obj.partialTokenOutput = PartialTokenOutput.toJSON(message.partialTokenOutput);
    }
    if (message.revocationCommitment.length !== 0) {
      obj.revocationCommitment = base64FromBytes2(message.revocationCommitment);
    }
    return obj;
  },
  create(base) {
    return FinalTokenOutput.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseFinalTokenOutput();
    message.partialTokenOutput = object.partialTokenOutput !== void 0 && object.partialTokenOutput !== null ? PartialTokenOutput.fromPartial(object.partialTokenOutput) : void 0;
    message.revocationCommitment = object.revocationCommitment ?? new Uint8Array(0);
    return message;
  }
};
function createBaseTokenTransaction() {
  return {
    version: 0,
    tokenInputs: void 0,
    tokenOutputs: [],
    sparkOperatorIdentityPublicKeys: [],
    expiryTime: void 0,
    network: 0,
    clientCreatedTimestamp: void 0,
    invoiceAttachments: [],
    validityDurationSeconds: void 0
  };
}
var TokenTransaction = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    if (message.version !== 0) {
      writer.uint32(8).uint32(message.version);
    }
    switch (message.tokenInputs?.$case) {
      case "mintInput":
        TokenMintInput.encode(message.tokenInputs.mintInput, writer.uint32(18).fork()).join();
        break;
      case "transferInput":
        TokenTransferInput.encode(message.tokenInputs.transferInput, writer.uint32(26).fork()).join();
        break;
      case "createInput":
        TokenCreateInput.encode(message.tokenInputs.createInput, writer.uint32(66).fork()).join();
        break;
    }
    for (const v of message.tokenOutputs) {
      TokenOutput.encode(v, writer.uint32(34).fork()).join();
    }
    for (const v of message.sparkOperatorIdentityPublicKeys) {
      writer.uint32(42).bytes(v);
    }
    if (message.expiryTime !== void 0) {
      Timestamp.encode(toTimestamp2(message.expiryTime), writer.uint32(50).fork()).join();
    }
    if (message.network !== 0) {
      writer.uint32(56).int32(message.network);
    }
    if (message.clientCreatedTimestamp !== void 0) {
      Timestamp.encode(toTimestamp2(message.clientCreatedTimestamp), writer.uint32(74).fork()).join();
    }
    for (const v of message.invoiceAttachments) {
      InvoiceAttachment.encode(v, writer.uint32(82).fork()).join();
    }
    if (message.validityDurationSeconds !== void 0) {
      writer.uint32(88).uint64(message.validityDurationSeconds);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseTokenTransaction();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 8) {
            break;
          }
          message.version = reader.uint32();
          continue;
        }
        case 2: {
          if (tag !== 18) {
            break;
          }
          message.tokenInputs = { $case: "mintInput", mintInput: TokenMintInput.decode(reader, reader.uint32()) };
          continue;
        }
        case 3: {
          if (tag !== 26) {
            break;
          }
          message.tokenInputs = {
            $case: "transferInput",
            transferInput: TokenTransferInput.decode(reader, reader.uint32())
          };
          continue;
        }
        case 8: {
          if (tag !== 66) {
            break;
          }
          message.tokenInputs = { $case: "createInput", createInput: TokenCreateInput.decode(reader, reader.uint32()) };
          continue;
        }
        case 4: {
          if (tag !== 34) {
            break;
          }
          message.tokenOutputs.push(TokenOutput.decode(reader, reader.uint32()));
          continue;
        }
        case 5: {
          if (tag !== 42) {
            break;
          }
          message.sparkOperatorIdentityPublicKeys.push(reader.bytes());
          continue;
        }
        case 6: {
          if (tag !== 50) {
            break;
          }
          message.expiryTime = fromTimestamp2(Timestamp.decode(reader, reader.uint32()));
          continue;
        }
        case 7: {
          if (tag !== 56) {
            break;
          }
          message.network = reader.int32();
          continue;
        }
        case 9: {
          if (tag !== 74) {
            break;
          }
          message.clientCreatedTimestamp = fromTimestamp2(Timestamp.decode(reader, reader.uint32()));
          continue;
        }
        case 10: {
          if (tag !== 82) {
            break;
          }
          message.invoiceAttachments.push(InvoiceAttachment.decode(reader, reader.uint32()));
          continue;
        }
        case 11: {
          if (tag !== 88) {
            break;
          }
          message.validityDurationSeconds = longToNumber2(reader.uint64());
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      version: isSet3(object.version) ? globalThis.Number(object.version) : 0,
      tokenInputs: isSet3(object.mintInput) ? { $case: "mintInput", mintInput: TokenMintInput.fromJSON(object.mintInput) } : isSet3(object.transferInput) ? { $case: "transferInput", transferInput: TokenTransferInput.fromJSON(object.transferInput) } : isSet3(object.createInput) ? { $case: "createInput", createInput: TokenCreateInput.fromJSON(object.createInput) } : void 0,
      tokenOutputs: globalThis.Array.isArray(object?.tokenOutputs) ? object.tokenOutputs.map((e) => TokenOutput.fromJSON(e)) : [],
      sparkOperatorIdentityPublicKeys: globalThis.Array.isArray(object?.sparkOperatorIdentityPublicKeys) ? object.sparkOperatorIdentityPublicKeys.map((e) => bytesFromBase642(e)) : [],
      expiryTime: isSet3(object.expiryTime) ? fromJsonTimestamp2(object.expiryTime) : void 0,
      network: isSet3(object.network) ? networkFromJSON(object.network) : 0,
      clientCreatedTimestamp: isSet3(object.clientCreatedTimestamp) ? fromJsonTimestamp2(object.clientCreatedTimestamp) : void 0,
      invoiceAttachments: globalThis.Array.isArray(object?.invoiceAttachments) ? object.invoiceAttachments.map((e) => InvoiceAttachment.fromJSON(e)) : [],
      validityDurationSeconds: isSet3(object.validityDurationSeconds) ? globalThis.Number(object.validityDurationSeconds) : void 0
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.version !== 0) {
      obj.version = Math.round(message.version);
    }
    if (message.tokenInputs?.$case === "mintInput") {
      obj.mintInput = TokenMintInput.toJSON(message.tokenInputs.mintInput);
    } else if (message.tokenInputs?.$case === "transferInput") {
      obj.transferInput = TokenTransferInput.toJSON(message.tokenInputs.transferInput);
    } else if (message.tokenInputs?.$case === "createInput") {
      obj.createInput = TokenCreateInput.toJSON(message.tokenInputs.createInput);
    }
    if (message.tokenOutputs?.length) {
      obj.tokenOutputs = message.tokenOutputs.map((e) => TokenOutput.toJSON(e));
    }
    if (message.sparkOperatorIdentityPublicKeys?.length) {
      obj.sparkOperatorIdentityPublicKeys = message.sparkOperatorIdentityPublicKeys.map((e) => base64FromBytes2(e));
    }
    if (message.expiryTime !== void 0) {
      obj.expiryTime = message.expiryTime.toISOString();
    }
    if (message.network !== 0) {
      obj.network = networkToJSON(message.network);
    }
    if (message.clientCreatedTimestamp !== void 0) {
      obj.clientCreatedTimestamp = message.clientCreatedTimestamp.toISOString();
    }
    if (message.invoiceAttachments?.length) {
      obj.invoiceAttachments = message.invoiceAttachments.map((e) => InvoiceAttachment.toJSON(e));
    }
    if (message.validityDurationSeconds !== void 0) {
      obj.validityDurationSeconds = Math.round(message.validityDurationSeconds);
    }
    return obj;
  },
  create(base) {
    return TokenTransaction.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseTokenTransaction();
    message.version = object.version ?? 0;
    switch (object.tokenInputs?.$case) {
      case "mintInput": {
        if (object.tokenInputs?.mintInput !== void 0 && object.tokenInputs?.mintInput !== null) {
          message.tokenInputs = {
            $case: "mintInput",
            mintInput: TokenMintInput.fromPartial(object.tokenInputs.mintInput)
          };
        }
        break;
      }
      case "transferInput": {
        if (object.tokenInputs?.transferInput !== void 0 && object.tokenInputs?.transferInput !== null) {
          message.tokenInputs = {
            $case: "transferInput",
            transferInput: TokenTransferInput.fromPartial(object.tokenInputs.transferInput)
          };
        }
        break;
      }
      case "createInput": {
        if (object.tokenInputs?.createInput !== void 0 && object.tokenInputs?.createInput !== null) {
          message.tokenInputs = {
            $case: "createInput",
            createInput: TokenCreateInput.fromPartial(object.tokenInputs.createInput)
          };
        }
        break;
      }
    }
    message.tokenOutputs = object.tokenOutputs?.map((e) => TokenOutput.fromPartial(e)) || [];
    message.sparkOperatorIdentityPublicKeys = object.sparkOperatorIdentityPublicKeys?.map((e) => e) || [];
    message.expiryTime = object.expiryTime ?? void 0;
    message.network = object.network ?? 0;
    message.clientCreatedTimestamp = object.clientCreatedTimestamp ?? void 0;
    message.invoiceAttachments = object.invoiceAttachments?.map((e) => InvoiceAttachment.fromPartial(e)) || [];
    message.validityDurationSeconds = object.validityDurationSeconds ?? void 0;
    return message;
  }
};
function createBaseTokenTransactionMetadata() {
  return {
    sparkOperatorIdentityPublicKeys: [],
    network: 0,
    clientCreatedTimestamp: void 0,
    validityDurationSeconds: 0,
    invoiceAttachments: []
  };
}
var TokenTransactionMetadata = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    for (const v of message.sparkOperatorIdentityPublicKeys) {
      writer.uint32(18).bytes(v);
    }
    if (message.network !== 0) {
      writer.uint32(24).int32(message.network);
    }
    if (message.clientCreatedTimestamp !== void 0) {
      Timestamp.encode(toTimestamp2(message.clientCreatedTimestamp), writer.uint32(34).fork()).join();
    }
    if (message.validityDurationSeconds !== 0) {
      writer.uint32(40).uint64(message.validityDurationSeconds);
    }
    for (const v of message.invoiceAttachments) {
      InvoiceAttachment.encode(v, writer.uint32(50).fork()).join();
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseTokenTransactionMetadata();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 2: {
          if (tag !== 18) {
            break;
          }
          message.sparkOperatorIdentityPublicKeys.push(reader.bytes());
          continue;
        }
        case 3: {
          if (tag !== 24) {
            break;
          }
          message.network = reader.int32();
          continue;
        }
        case 4: {
          if (tag !== 34) {
            break;
          }
          message.clientCreatedTimestamp = fromTimestamp2(Timestamp.decode(reader, reader.uint32()));
          continue;
        }
        case 5: {
          if (tag !== 40) {
            break;
          }
          message.validityDurationSeconds = longToNumber2(reader.uint64());
          continue;
        }
        case 6: {
          if (tag !== 50) {
            break;
          }
          message.invoiceAttachments.push(InvoiceAttachment.decode(reader, reader.uint32()));
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      sparkOperatorIdentityPublicKeys: globalThis.Array.isArray(object?.sparkOperatorIdentityPublicKeys) ? object.sparkOperatorIdentityPublicKeys.map((e) => bytesFromBase642(e)) : [],
      network: isSet3(object.network) ? networkFromJSON(object.network) : 0,
      clientCreatedTimestamp: isSet3(object.clientCreatedTimestamp) ? fromJsonTimestamp2(object.clientCreatedTimestamp) : void 0,
      validityDurationSeconds: isSet3(object.validityDurationSeconds) ? globalThis.Number(object.validityDurationSeconds) : 0,
      invoiceAttachments: globalThis.Array.isArray(object?.invoiceAttachments) ? object.invoiceAttachments.map((e) => InvoiceAttachment.fromJSON(e)) : []
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.sparkOperatorIdentityPublicKeys?.length) {
      obj.sparkOperatorIdentityPublicKeys = message.sparkOperatorIdentityPublicKeys.map((e) => base64FromBytes2(e));
    }
    if (message.network !== 0) {
      obj.network = networkToJSON(message.network);
    }
    if (message.clientCreatedTimestamp !== void 0) {
      obj.clientCreatedTimestamp = message.clientCreatedTimestamp.toISOString();
    }
    if (message.validityDurationSeconds !== 0) {
      obj.validityDurationSeconds = Math.round(message.validityDurationSeconds);
    }
    if (message.invoiceAttachments?.length) {
      obj.invoiceAttachments = message.invoiceAttachments.map((e) => InvoiceAttachment.toJSON(e));
    }
    return obj;
  },
  create(base) {
    return TokenTransactionMetadata.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseTokenTransactionMetadata();
    message.sparkOperatorIdentityPublicKeys = object.sparkOperatorIdentityPublicKeys?.map((e) => e) || [];
    message.network = object.network ?? 0;
    message.clientCreatedTimestamp = object.clientCreatedTimestamp ?? void 0;
    message.validityDurationSeconds = object.validityDurationSeconds ?? 0;
    message.invoiceAttachments = object.invoiceAttachments?.map((e) => InvoiceAttachment.fromPartial(e)) || [];
    return message;
  }
};
function createBasePartialTokenTransaction() {
  return { version: 0, tokenTransactionMetadata: void 0, tokenInputs: void 0, partialTokenOutputs: [] };
}
var PartialTokenTransaction = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    if (message.version !== 0) {
      writer.uint32(8).uint32(message.version);
    }
    if (message.tokenTransactionMetadata !== void 0) {
      TokenTransactionMetadata.encode(message.tokenTransactionMetadata, writer.uint32(18).fork()).join();
    }
    switch (message.tokenInputs?.$case) {
      case "mintInput":
        TokenMintInput.encode(message.tokenInputs.mintInput, writer.uint32(26).fork()).join();
        break;
      case "transferInput":
        TokenTransferInput.encode(message.tokenInputs.transferInput, writer.uint32(34).fork()).join();
        break;
      case "createInput":
        TokenCreateInput.encode(message.tokenInputs.createInput, writer.uint32(42).fork()).join();
        break;
    }
    for (const v of message.partialTokenOutputs) {
      PartialTokenOutput.encode(v, writer.uint32(50).fork()).join();
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBasePartialTokenTransaction();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 8) {
            break;
          }
          message.version = reader.uint32();
          continue;
        }
        case 2: {
          if (tag !== 18) {
            break;
          }
          message.tokenTransactionMetadata = TokenTransactionMetadata.decode(reader, reader.uint32());
          continue;
        }
        case 3: {
          if (tag !== 26) {
            break;
          }
          message.tokenInputs = { $case: "mintInput", mintInput: TokenMintInput.decode(reader, reader.uint32()) };
          continue;
        }
        case 4: {
          if (tag !== 34) {
            break;
          }
          message.tokenInputs = {
            $case: "transferInput",
            transferInput: TokenTransferInput.decode(reader, reader.uint32())
          };
          continue;
        }
        case 5: {
          if (tag !== 42) {
            break;
          }
          message.tokenInputs = { $case: "createInput", createInput: TokenCreateInput.decode(reader, reader.uint32()) };
          continue;
        }
        case 6: {
          if (tag !== 50) {
            break;
          }
          message.partialTokenOutputs.push(PartialTokenOutput.decode(reader, reader.uint32()));
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      version: isSet3(object.version) ? globalThis.Number(object.version) : 0,
      tokenTransactionMetadata: isSet3(object.tokenTransactionMetadata) ? TokenTransactionMetadata.fromJSON(object.tokenTransactionMetadata) : void 0,
      tokenInputs: isSet3(object.mintInput) ? { $case: "mintInput", mintInput: TokenMintInput.fromJSON(object.mintInput) } : isSet3(object.transferInput) ? { $case: "transferInput", transferInput: TokenTransferInput.fromJSON(object.transferInput) } : isSet3(object.createInput) ? { $case: "createInput", createInput: TokenCreateInput.fromJSON(object.createInput) } : void 0,
      partialTokenOutputs: globalThis.Array.isArray(object?.partialTokenOutputs) ? object.partialTokenOutputs.map((e) => PartialTokenOutput.fromJSON(e)) : []
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.version !== 0) {
      obj.version = Math.round(message.version);
    }
    if (message.tokenTransactionMetadata !== void 0) {
      obj.tokenTransactionMetadata = TokenTransactionMetadata.toJSON(message.tokenTransactionMetadata);
    }
    if (message.tokenInputs?.$case === "mintInput") {
      obj.mintInput = TokenMintInput.toJSON(message.tokenInputs.mintInput);
    } else if (message.tokenInputs?.$case === "transferInput") {
      obj.transferInput = TokenTransferInput.toJSON(message.tokenInputs.transferInput);
    } else if (message.tokenInputs?.$case === "createInput") {
      obj.createInput = TokenCreateInput.toJSON(message.tokenInputs.createInput);
    }
    if (message.partialTokenOutputs?.length) {
      obj.partialTokenOutputs = message.partialTokenOutputs.map((e) => PartialTokenOutput.toJSON(e));
    }
    return obj;
  },
  create(base) {
    return PartialTokenTransaction.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBasePartialTokenTransaction();
    message.version = object.version ?? 0;
    message.tokenTransactionMetadata = object.tokenTransactionMetadata !== void 0 && object.tokenTransactionMetadata !== null ? TokenTransactionMetadata.fromPartial(object.tokenTransactionMetadata) : void 0;
    switch (object.tokenInputs?.$case) {
      case "mintInput": {
        if (object.tokenInputs?.mintInput !== void 0 && object.tokenInputs?.mintInput !== null) {
          message.tokenInputs = {
            $case: "mintInput",
            mintInput: TokenMintInput.fromPartial(object.tokenInputs.mintInput)
          };
        }
        break;
      }
      case "transferInput": {
        if (object.tokenInputs?.transferInput !== void 0 && object.tokenInputs?.transferInput !== null) {
          message.tokenInputs = {
            $case: "transferInput",
            transferInput: TokenTransferInput.fromPartial(object.tokenInputs.transferInput)
          };
        }
        break;
      }
      case "createInput": {
        if (object.tokenInputs?.createInput !== void 0 && object.tokenInputs?.createInput !== null) {
          message.tokenInputs = {
            $case: "createInput",
            createInput: TokenCreateInput.fromPartial(object.tokenInputs.createInput)
          };
        }
        break;
      }
    }
    message.partialTokenOutputs = object.partialTokenOutputs?.map((e) => PartialTokenOutput.fromPartial(e)) || [];
    return message;
  }
};
function createBaseFinalTokenTransaction() {
  return { version: 0, tokenTransactionMetadata: void 0, tokenInputs: void 0, finalTokenOutputs: [] };
}
var FinalTokenTransaction = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    if (message.version !== 0) {
      writer.uint32(8).uint32(message.version);
    }
    if (message.tokenTransactionMetadata !== void 0) {
      TokenTransactionMetadata.encode(message.tokenTransactionMetadata, writer.uint32(18).fork()).join();
    }
    switch (message.tokenInputs?.$case) {
      case "mintInput":
        TokenMintInput.encode(message.tokenInputs.mintInput, writer.uint32(26).fork()).join();
        break;
      case "transferInput":
        TokenTransferInput.encode(message.tokenInputs.transferInput, writer.uint32(34).fork()).join();
        break;
      case "createInput":
        TokenCreateInput.encode(message.tokenInputs.createInput, writer.uint32(42).fork()).join();
        break;
    }
    for (const v of message.finalTokenOutputs) {
      FinalTokenOutput.encode(v, writer.uint32(50).fork()).join();
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseFinalTokenTransaction();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 8) {
            break;
          }
          message.version = reader.uint32();
          continue;
        }
        case 2: {
          if (tag !== 18) {
            break;
          }
          message.tokenTransactionMetadata = TokenTransactionMetadata.decode(reader, reader.uint32());
          continue;
        }
        case 3: {
          if (tag !== 26) {
            break;
          }
          message.tokenInputs = { $case: "mintInput", mintInput: TokenMintInput.decode(reader, reader.uint32()) };
          continue;
        }
        case 4: {
          if (tag !== 34) {
            break;
          }
          message.tokenInputs = {
            $case: "transferInput",
            transferInput: TokenTransferInput.decode(reader, reader.uint32())
          };
          continue;
        }
        case 5: {
          if (tag !== 42) {
            break;
          }
          message.tokenInputs = { $case: "createInput", createInput: TokenCreateInput.decode(reader, reader.uint32()) };
          continue;
        }
        case 6: {
          if (tag !== 50) {
            break;
          }
          message.finalTokenOutputs.push(FinalTokenOutput.decode(reader, reader.uint32()));
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      version: isSet3(object.version) ? globalThis.Number(object.version) : 0,
      tokenTransactionMetadata: isSet3(object.tokenTransactionMetadata) ? TokenTransactionMetadata.fromJSON(object.tokenTransactionMetadata) : void 0,
      tokenInputs: isSet3(object.mintInput) ? { $case: "mintInput", mintInput: TokenMintInput.fromJSON(object.mintInput) } : isSet3(object.transferInput) ? { $case: "transferInput", transferInput: TokenTransferInput.fromJSON(object.transferInput) } : isSet3(object.createInput) ? { $case: "createInput", createInput: TokenCreateInput.fromJSON(object.createInput) } : void 0,
      finalTokenOutputs: globalThis.Array.isArray(object?.finalTokenOutputs) ? object.finalTokenOutputs.map((e) => FinalTokenOutput.fromJSON(e)) : []
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.version !== 0) {
      obj.version = Math.round(message.version);
    }
    if (message.tokenTransactionMetadata !== void 0) {
      obj.tokenTransactionMetadata = TokenTransactionMetadata.toJSON(message.tokenTransactionMetadata);
    }
    if (message.tokenInputs?.$case === "mintInput") {
      obj.mintInput = TokenMintInput.toJSON(message.tokenInputs.mintInput);
    } else if (message.tokenInputs?.$case === "transferInput") {
      obj.transferInput = TokenTransferInput.toJSON(message.tokenInputs.transferInput);
    } else if (message.tokenInputs?.$case === "createInput") {
      obj.createInput = TokenCreateInput.toJSON(message.tokenInputs.createInput);
    }
    if (message.finalTokenOutputs?.length) {
      obj.finalTokenOutputs = message.finalTokenOutputs.map((e) => FinalTokenOutput.toJSON(e));
    }
    return obj;
  },
  create(base) {
    return FinalTokenTransaction.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseFinalTokenTransaction();
    message.version = object.version ?? 0;
    message.tokenTransactionMetadata = object.tokenTransactionMetadata !== void 0 && object.tokenTransactionMetadata !== null ? TokenTransactionMetadata.fromPartial(object.tokenTransactionMetadata) : void 0;
    switch (object.tokenInputs?.$case) {
      case "mintInput": {
        if (object.tokenInputs?.mintInput !== void 0 && object.tokenInputs?.mintInput !== null) {
          message.tokenInputs = {
            $case: "mintInput",
            mintInput: TokenMintInput.fromPartial(object.tokenInputs.mintInput)
          };
        }
        break;
      }
      case "transferInput": {
        if (object.tokenInputs?.transferInput !== void 0 && object.tokenInputs?.transferInput !== null) {
          message.tokenInputs = {
            $case: "transferInput",
            transferInput: TokenTransferInput.fromPartial(object.tokenInputs.transferInput)
          };
        }
        break;
      }
      case "createInput": {
        if (object.tokenInputs?.createInput !== void 0 && object.tokenInputs?.createInput !== null) {
          message.tokenInputs = {
            $case: "createInput",
            createInput: TokenCreateInput.fromPartial(object.tokenInputs.createInput)
          };
        }
        break;
      }
    }
    message.finalTokenOutputs = object.finalTokenOutputs?.map((e) => FinalTokenOutput.fromPartial(e)) || [];
    return message;
  }
};
function createBaseInvoiceAttachment() {
  return { sparkInvoice: "" };
}
var InvoiceAttachment = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    if (message.sparkInvoice !== "") {
      writer.uint32(10).string(message.sparkInvoice);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseInvoiceAttachment();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 10) {
            break;
          }
          message.sparkInvoice = reader.string();
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return { sparkInvoice: isSet3(object.sparkInvoice) ? globalThis.String(object.sparkInvoice) : "" };
  },
  toJSON(message) {
    const obj = {};
    if (message.sparkInvoice !== "") {
      obj.sparkInvoice = message.sparkInvoice;
    }
    return obj;
  },
  create(base) {
    return InvoiceAttachment.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseInvoiceAttachment();
    message.sparkInvoice = object.sparkInvoice ?? "";
    return message;
  }
};
function createBaseSignatureWithIndex() {
  return { signature: new Uint8Array(0), inputIndex: 0 };
}
var SignatureWithIndex = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    if (message.signature.length !== 0) {
      writer.uint32(10).bytes(message.signature);
    }
    if (message.inputIndex !== 0) {
      writer.uint32(16).uint32(message.inputIndex);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseSignatureWithIndex();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 10) {
            break;
          }
          message.signature = reader.bytes();
          continue;
        }
        case 2: {
          if (tag !== 16) {
            break;
          }
          message.inputIndex = reader.uint32();
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      signature: isSet3(object.signature) ? bytesFromBase642(object.signature) : new Uint8Array(0),
      inputIndex: isSet3(object.inputIndex) ? globalThis.Number(object.inputIndex) : 0
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.signature.length !== 0) {
      obj.signature = base64FromBytes2(message.signature);
    }
    if (message.inputIndex !== 0) {
      obj.inputIndex = Math.round(message.inputIndex);
    }
    return obj;
  },
  create(base) {
    return SignatureWithIndex.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseSignatureWithIndex();
    message.signature = object.signature ?? new Uint8Array(0);
    message.inputIndex = object.inputIndex ?? 0;
    return message;
  }
};
function createBaseInputTtxoSignaturesPerOperator() {
  return { ttxoSignatures: [], operatorIdentityPublicKey: new Uint8Array(0) };
}
var InputTtxoSignaturesPerOperator = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    for (const v of message.ttxoSignatures) {
      SignatureWithIndex.encode(v, writer.uint32(10).fork()).join();
    }
    if (message.operatorIdentityPublicKey.length !== 0) {
      writer.uint32(18).bytes(message.operatorIdentityPublicKey);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseInputTtxoSignaturesPerOperator();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 10) {
            break;
          }
          message.ttxoSignatures.push(SignatureWithIndex.decode(reader, reader.uint32()));
          continue;
        }
        case 2: {
          if (tag !== 18) {
            break;
          }
          message.operatorIdentityPublicKey = reader.bytes();
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      ttxoSignatures: globalThis.Array.isArray(object?.ttxoSignatures) ? object.ttxoSignatures.map((e) => SignatureWithIndex.fromJSON(e)) : [],
      operatorIdentityPublicKey: isSet3(object.operatorIdentityPublicKey) ? bytesFromBase642(object.operatorIdentityPublicKey) : new Uint8Array(0)
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.ttxoSignatures?.length) {
      obj.ttxoSignatures = message.ttxoSignatures.map((e) => SignatureWithIndex.toJSON(e));
    }
    if (message.operatorIdentityPublicKey.length !== 0) {
      obj.operatorIdentityPublicKey = base64FromBytes2(message.operatorIdentityPublicKey);
    }
    return obj;
  },
  create(base) {
    return InputTtxoSignaturesPerOperator.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseInputTtxoSignaturesPerOperator();
    message.ttxoSignatures = object.ttxoSignatures?.map((e) => SignatureWithIndex.fromPartial(e)) || [];
    message.operatorIdentityPublicKey = object.operatorIdentityPublicKey ?? new Uint8Array(0);
    return message;
  }
};
function createBaseStartTransactionRequest() {
  return {
    identityPublicKey: new Uint8Array(0),
    partialTokenTransaction: void 0,
    partialTokenTransactionOwnerSignatures: [],
    validityDurationSeconds: 0
  };
}
var StartTransactionRequest = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    if (message.identityPublicKey.length !== 0) {
      writer.uint32(10).bytes(message.identityPublicKey);
    }
    if (message.partialTokenTransaction !== void 0) {
      TokenTransaction.encode(message.partialTokenTransaction, writer.uint32(18).fork()).join();
    }
    for (const v of message.partialTokenTransactionOwnerSignatures) {
      SignatureWithIndex.encode(v, writer.uint32(26).fork()).join();
    }
    if (message.validityDurationSeconds !== 0) {
      writer.uint32(32).uint64(message.validityDurationSeconds);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseStartTransactionRequest();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 10) {
            break;
          }
          message.identityPublicKey = reader.bytes();
          continue;
        }
        case 2: {
          if (tag !== 18) {
            break;
          }
          message.partialTokenTransaction = TokenTransaction.decode(reader, reader.uint32());
          continue;
        }
        case 3: {
          if (tag !== 26) {
            break;
          }
          message.partialTokenTransactionOwnerSignatures.push(SignatureWithIndex.decode(reader, reader.uint32()));
          continue;
        }
        case 4: {
          if (tag !== 32) {
            break;
          }
          message.validityDurationSeconds = longToNumber2(reader.uint64());
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      identityPublicKey: isSet3(object.identityPublicKey) ? bytesFromBase642(object.identityPublicKey) : new Uint8Array(0),
      partialTokenTransaction: isSet3(object.partialTokenTransaction) ? TokenTransaction.fromJSON(object.partialTokenTransaction) : void 0,
      partialTokenTransactionOwnerSignatures: globalThis.Array.isArray(object?.partialTokenTransactionOwnerSignatures) ? object.partialTokenTransactionOwnerSignatures.map((e) => SignatureWithIndex.fromJSON(e)) : [],
      validityDurationSeconds: isSet3(object.validityDurationSeconds) ? globalThis.Number(object.validityDurationSeconds) : 0
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.identityPublicKey.length !== 0) {
      obj.identityPublicKey = base64FromBytes2(message.identityPublicKey);
    }
    if (message.partialTokenTransaction !== void 0) {
      obj.partialTokenTransaction = TokenTransaction.toJSON(message.partialTokenTransaction);
    }
    if (message.partialTokenTransactionOwnerSignatures?.length) {
      obj.partialTokenTransactionOwnerSignatures = message.partialTokenTransactionOwnerSignatures.map(
        (e) => SignatureWithIndex.toJSON(e)
      );
    }
    if (message.validityDurationSeconds !== 0) {
      obj.validityDurationSeconds = Math.round(message.validityDurationSeconds);
    }
    return obj;
  },
  create(base) {
    return StartTransactionRequest.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseStartTransactionRequest();
    message.identityPublicKey = object.identityPublicKey ?? new Uint8Array(0);
    message.partialTokenTransaction = object.partialTokenTransaction !== void 0 && object.partialTokenTransaction !== null ? TokenTransaction.fromPartial(object.partialTokenTransaction) : void 0;
    message.partialTokenTransactionOwnerSignatures = object.partialTokenTransactionOwnerSignatures?.map((e) => SignatureWithIndex.fromPartial(e)) || [];
    message.validityDurationSeconds = object.validityDurationSeconds ?? 0;
    return message;
  }
};
function createBaseStartTransactionResponse() {
  return { finalTokenTransaction: void 0, keyshareInfo: void 0 };
}
var StartTransactionResponse = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    if (message.finalTokenTransaction !== void 0) {
      TokenTransaction.encode(message.finalTokenTransaction, writer.uint32(10).fork()).join();
    }
    if (message.keyshareInfo !== void 0) {
      SigningKeyshare.encode(message.keyshareInfo, writer.uint32(18).fork()).join();
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseStartTransactionResponse();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 10) {
            break;
          }
          message.finalTokenTransaction = TokenTransaction.decode(reader, reader.uint32());
          continue;
        }
        case 2: {
          if (tag !== 18) {
            break;
          }
          message.keyshareInfo = SigningKeyshare.decode(reader, reader.uint32());
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      finalTokenTransaction: isSet3(object.finalTokenTransaction) ? TokenTransaction.fromJSON(object.finalTokenTransaction) : void 0,
      keyshareInfo: isSet3(object.keyshareInfo) ? SigningKeyshare.fromJSON(object.keyshareInfo) : void 0
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.finalTokenTransaction !== void 0) {
      obj.finalTokenTransaction = TokenTransaction.toJSON(message.finalTokenTransaction);
    }
    if (message.keyshareInfo !== void 0) {
      obj.keyshareInfo = SigningKeyshare.toJSON(message.keyshareInfo);
    }
    return obj;
  },
  create(base) {
    return StartTransactionResponse.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseStartTransactionResponse();
    message.finalTokenTransaction = object.finalTokenTransaction !== void 0 && object.finalTokenTransaction !== null ? TokenTransaction.fromPartial(object.finalTokenTransaction) : void 0;
    message.keyshareInfo = object.keyshareInfo !== void 0 && object.keyshareInfo !== null ? SigningKeyshare.fromPartial(object.keyshareInfo) : void 0;
    return message;
  }
};
function createBaseCommitTransactionRequest() {
  return {
    finalTokenTransaction: void 0,
    finalTokenTransactionHash: new Uint8Array(0),
    inputTtxoSignaturesPerOperator: [],
    ownerIdentityPublicKey: new Uint8Array(0)
  };
}
var CommitTransactionRequest = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    if (message.finalTokenTransaction !== void 0) {
      TokenTransaction.encode(message.finalTokenTransaction, writer.uint32(10).fork()).join();
    }
    if (message.finalTokenTransactionHash.length !== 0) {
      writer.uint32(18).bytes(message.finalTokenTransactionHash);
    }
    for (const v of message.inputTtxoSignaturesPerOperator) {
      InputTtxoSignaturesPerOperator.encode(v, writer.uint32(26).fork()).join();
    }
    if (message.ownerIdentityPublicKey.length !== 0) {
      writer.uint32(34).bytes(message.ownerIdentityPublicKey);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseCommitTransactionRequest();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 10) {
            break;
          }
          message.finalTokenTransaction = TokenTransaction.decode(reader, reader.uint32());
          continue;
        }
        case 2: {
          if (tag !== 18) {
            break;
          }
          message.finalTokenTransactionHash = reader.bytes();
          continue;
        }
        case 3: {
          if (tag !== 26) {
            break;
          }
          message.inputTtxoSignaturesPerOperator.push(InputTtxoSignaturesPerOperator.decode(reader, reader.uint32()));
          continue;
        }
        case 4: {
          if (tag !== 34) {
            break;
          }
          message.ownerIdentityPublicKey = reader.bytes();
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      finalTokenTransaction: isSet3(object.finalTokenTransaction) ? TokenTransaction.fromJSON(object.finalTokenTransaction) : void 0,
      finalTokenTransactionHash: isSet3(object.finalTokenTransactionHash) ? bytesFromBase642(object.finalTokenTransactionHash) : new Uint8Array(0),
      inputTtxoSignaturesPerOperator: globalThis.Array.isArray(object?.inputTtxoSignaturesPerOperator) ? object.inputTtxoSignaturesPerOperator.map((e) => InputTtxoSignaturesPerOperator.fromJSON(e)) : [],
      ownerIdentityPublicKey: isSet3(object.ownerIdentityPublicKey) ? bytesFromBase642(object.ownerIdentityPublicKey) : new Uint8Array(0)
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.finalTokenTransaction !== void 0) {
      obj.finalTokenTransaction = TokenTransaction.toJSON(message.finalTokenTransaction);
    }
    if (message.finalTokenTransactionHash.length !== 0) {
      obj.finalTokenTransactionHash = base64FromBytes2(message.finalTokenTransactionHash);
    }
    if (message.inputTtxoSignaturesPerOperator?.length) {
      obj.inputTtxoSignaturesPerOperator = message.inputTtxoSignaturesPerOperator.map(
        (e) => InputTtxoSignaturesPerOperator.toJSON(e)
      );
    }
    if (message.ownerIdentityPublicKey.length !== 0) {
      obj.ownerIdentityPublicKey = base64FromBytes2(message.ownerIdentityPublicKey);
    }
    return obj;
  },
  create(base) {
    return CommitTransactionRequest.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseCommitTransactionRequest();
    message.finalTokenTransaction = object.finalTokenTransaction !== void 0 && object.finalTokenTransaction !== null ? TokenTransaction.fromPartial(object.finalTokenTransaction) : void 0;
    message.finalTokenTransactionHash = object.finalTokenTransactionHash ?? new Uint8Array(0);
    message.inputTtxoSignaturesPerOperator = object.inputTtxoSignaturesPerOperator?.map((e) => InputTtxoSignaturesPerOperator.fromPartial(e)) || [];
    message.ownerIdentityPublicKey = object.ownerIdentityPublicKey ?? new Uint8Array(0);
    return message;
  }
};
function createBaseCommitProgress() {
  return { committedOperatorPublicKeys: [], uncommittedOperatorPublicKeys: [] };
}
var CommitProgress = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    for (const v of message.committedOperatorPublicKeys) {
      writer.uint32(10).bytes(v);
    }
    for (const v of message.uncommittedOperatorPublicKeys) {
      writer.uint32(18).bytes(v);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseCommitProgress();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 10) {
            break;
          }
          message.committedOperatorPublicKeys.push(reader.bytes());
          continue;
        }
        case 2: {
          if (tag !== 18) {
            break;
          }
          message.uncommittedOperatorPublicKeys.push(reader.bytes());
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      committedOperatorPublicKeys: globalThis.Array.isArray(object?.committedOperatorPublicKeys) ? object.committedOperatorPublicKeys.map((e) => bytesFromBase642(e)) : [],
      uncommittedOperatorPublicKeys: globalThis.Array.isArray(object?.uncommittedOperatorPublicKeys) ? object.uncommittedOperatorPublicKeys.map((e) => bytesFromBase642(e)) : []
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.committedOperatorPublicKeys?.length) {
      obj.committedOperatorPublicKeys = message.committedOperatorPublicKeys.map((e) => base64FromBytes2(e));
    }
    if (message.uncommittedOperatorPublicKeys?.length) {
      obj.uncommittedOperatorPublicKeys = message.uncommittedOperatorPublicKeys.map((e) => base64FromBytes2(e));
    }
    return obj;
  },
  create(base) {
    return CommitProgress.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseCommitProgress();
    message.committedOperatorPublicKeys = object.committedOperatorPublicKeys?.map((e) => e) || [];
    message.uncommittedOperatorPublicKeys = object.uncommittedOperatorPublicKeys?.map((e) => e) || [];
    return message;
  }
};
function createBaseCommitTransactionResponse() {
  return { commitStatus: 0, commitProgress: void 0 };
}
var CommitTransactionResponse = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    if (message.commitStatus !== 0) {
      writer.uint32(8).int32(message.commitStatus);
    }
    if (message.commitProgress !== void 0) {
      CommitProgress.encode(message.commitProgress, writer.uint32(18).fork()).join();
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseCommitTransactionResponse();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 8) {
            break;
          }
          message.commitStatus = reader.int32();
          continue;
        }
        case 2: {
          if (tag !== 18) {
            break;
          }
          message.commitProgress = CommitProgress.decode(reader, reader.uint32());
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      commitStatus: isSet3(object.commitStatus) ? commitStatusFromJSON(object.commitStatus) : 0,
      commitProgress: isSet3(object.commitProgress) ? CommitProgress.fromJSON(object.commitProgress) : void 0
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.commitStatus !== 0) {
      obj.commitStatus = commitStatusToJSON(message.commitStatus);
    }
    if (message.commitProgress !== void 0) {
      obj.commitProgress = CommitProgress.toJSON(message.commitProgress);
    }
    return obj;
  },
  create(base) {
    return CommitTransactionResponse.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseCommitTransactionResponse();
    message.commitStatus = object.commitStatus ?? 0;
    message.commitProgress = object.commitProgress !== void 0 && object.commitProgress !== null ? CommitProgress.fromPartial(object.commitProgress) : void 0;
    return message;
  }
};
function createBaseBroadcastTransactionRequest() {
  return {
    identityPublicKey: new Uint8Array(0),
    partialTokenTransaction: void 0,
    tokenTransactionOwnerSignatures: []
  };
}
var BroadcastTransactionRequest = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    if (message.identityPublicKey.length !== 0) {
      writer.uint32(10).bytes(message.identityPublicKey);
    }
    if (message.partialTokenTransaction !== void 0) {
      PartialTokenTransaction.encode(message.partialTokenTransaction, writer.uint32(18).fork()).join();
    }
    for (const v of message.tokenTransactionOwnerSignatures) {
      SignatureWithIndex.encode(v, writer.uint32(26).fork()).join();
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseBroadcastTransactionRequest();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 10) {
            break;
          }
          message.identityPublicKey = reader.bytes();
          continue;
        }
        case 2: {
          if (tag !== 18) {
            break;
          }
          message.partialTokenTransaction = PartialTokenTransaction.decode(reader, reader.uint32());
          continue;
        }
        case 3: {
          if (tag !== 26) {
            break;
          }
          message.tokenTransactionOwnerSignatures.push(SignatureWithIndex.decode(reader, reader.uint32()));
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      identityPublicKey: isSet3(object.identityPublicKey) ? bytesFromBase642(object.identityPublicKey) : new Uint8Array(0),
      partialTokenTransaction: isSet3(object.partialTokenTransaction) ? PartialTokenTransaction.fromJSON(object.partialTokenTransaction) : void 0,
      tokenTransactionOwnerSignatures: globalThis.Array.isArray(object?.tokenTransactionOwnerSignatures) ? object.tokenTransactionOwnerSignatures.map((e) => SignatureWithIndex.fromJSON(e)) : []
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.identityPublicKey.length !== 0) {
      obj.identityPublicKey = base64FromBytes2(message.identityPublicKey);
    }
    if (message.partialTokenTransaction !== void 0) {
      obj.partialTokenTransaction = PartialTokenTransaction.toJSON(message.partialTokenTransaction);
    }
    if (message.tokenTransactionOwnerSignatures?.length) {
      obj.tokenTransactionOwnerSignatures = message.tokenTransactionOwnerSignatures.map(
        (e) => SignatureWithIndex.toJSON(e)
      );
    }
    return obj;
  },
  create(base) {
    return BroadcastTransactionRequest.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseBroadcastTransactionRequest();
    message.identityPublicKey = object.identityPublicKey ?? new Uint8Array(0);
    message.partialTokenTransaction = object.partialTokenTransaction !== void 0 && object.partialTokenTransaction !== null ? PartialTokenTransaction.fromPartial(object.partialTokenTransaction) : void 0;
    message.tokenTransactionOwnerSignatures = object.tokenTransactionOwnerSignatures?.map((e) => SignatureWithIndex.fromPartial(e)) || [];
    return message;
  }
};
function createBaseBroadcastTransactionResponse() {
  return { finalTokenTransaction: void 0, commitStatus: 0, commitProgress: void 0 };
}
var BroadcastTransactionResponse = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    if (message.finalTokenTransaction !== void 0) {
      FinalTokenTransaction.encode(message.finalTokenTransaction, writer.uint32(10).fork()).join();
    }
    if (message.commitStatus !== 0) {
      writer.uint32(16).int32(message.commitStatus);
    }
    if (message.commitProgress !== void 0) {
      CommitProgress.encode(message.commitProgress, writer.uint32(26).fork()).join();
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseBroadcastTransactionResponse();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 10) {
            break;
          }
          message.finalTokenTransaction = FinalTokenTransaction.decode(reader, reader.uint32());
          continue;
        }
        case 2: {
          if (tag !== 16) {
            break;
          }
          message.commitStatus = reader.int32();
          continue;
        }
        case 3: {
          if (tag !== 26) {
            break;
          }
          message.commitProgress = CommitProgress.decode(reader, reader.uint32());
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      finalTokenTransaction: isSet3(object.finalTokenTransaction) ? FinalTokenTransaction.fromJSON(object.finalTokenTransaction) : void 0,
      commitStatus: isSet3(object.commitStatus) ? commitStatusFromJSON(object.commitStatus) : 0,
      commitProgress: isSet3(object.commitProgress) ? CommitProgress.fromJSON(object.commitProgress) : void 0
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.finalTokenTransaction !== void 0) {
      obj.finalTokenTransaction = FinalTokenTransaction.toJSON(message.finalTokenTransaction);
    }
    if (message.commitStatus !== 0) {
      obj.commitStatus = commitStatusToJSON(message.commitStatus);
    }
    if (message.commitProgress !== void 0) {
      obj.commitProgress = CommitProgress.toJSON(message.commitProgress);
    }
    return obj;
  },
  create(base) {
    return BroadcastTransactionResponse.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseBroadcastTransactionResponse();
    message.finalTokenTransaction = object.finalTokenTransaction !== void 0 && object.finalTokenTransaction !== null ? FinalTokenTransaction.fromPartial(object.finalTokenTransaction) : void 0;
    message.commitStatus = object.commitStatus ?? 0;
    message.commitProgress = object.commitProgress !== void 0 && object.commitProgress !== null ? CommitProgress.fromPartial(object.commitProgress) : void 0;
    return message;
  }
};
function createBaseQueryTokenMetadataRequest() {
  return { tokenIdentifiers: [], issuerPublicKeys: [] };
}
var QueryTokenMetadataRequest = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    for (const v of message.tokenIdentifiers) {
      writer.uint32(10).bytes(v);
    }
    for (const v of message.issuerPublicKeys) {
      writer.uint32(18).bytes(v);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseQueryTokenMetadataRequest();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 10) {
            break;
          }
          message.tokenIdentifiers.push(reader.bytes());
          continue;
        }
        case 2: {
          if (tag !== 18) {
            break;
          }
          message.issuerPublicKeys.push(reader.bytes());
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      tokenIdentifiers: globalThis.Array.isArray(object?.tokenIdentifiers) ? object.tokenIdentifiers.map((e) => bytesFromBase642(e)) : [],
      issuerPublicKeys: globalThis.Array.isArray(object?.issuerPublicKeys) ? object.issuerPublicKeys.map((e) => bytesFromBase642(e)) : []
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.tokenIdentifiers?.length) {
      obj.tokenIdentifiers = message.tokenIdentifiers.map((e) => base64FromBytes2(e));
    }
    if (message.issuerPublicKeys?.length) {
      obj.issuerPublicKeys = message.issuerPublicKeys.map((e) => base64FromBytes2(e));
    }
    return obj;
  },
  create(base) {
    return QueryTokenMetadataRequest.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseQueryTokenMetadataRequest();
    message.tokenIdentifiers = object.tokenIdentifiers?.map((e) => e) || [];
    message.issuerPublicKeys = object.issuerPublicKeys?.map((e) => e) || [];
    return message;
  }
};
function createBaseTokenMetadata() {
  return {
    issuerPublicKey: new Uint8Array(0),
    tokenName: "",
    tokenTicker: "",
    decimals: 0,
    maxSupply: new Uint8Array(0),
    isFreezable: false,
    creationEntityPublicKey: void 0,
    tokenIdentifier: new Uint8Array(0)
  };
}
var TokenMetadata = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    if (message.issuerPublicKey.length !== 0) {
      writer.uint32(10).bytes(message.issuerPublicKey);
    }
    if (message.tokenName !== "") {
      writer.uint32(18).string(message.tokenName);
    }
    if (message.tokenTicker !== "") {
      writer.uint32(26).string(message.tokenTicker);
    }
    if (message.decimals !== 0) {
      writer.uint32(32).uint32(message.decimals);
    }
    if (message.maxSupply.length !== 0) {
      writer.uint32(42).bytes(message.maxSupply);
    }
    if (message.isFreezable !== false) {
      writer.uint32(48).bool(message.isFreezable);
    }
    if (message.creationEntityPublicKey !== void 0) {
      writer.uint32(58).bytes(message.creationEntityPublicKey);
    }
    if (message.tokenIdentifier.length !== 0) {
      writer.uint32(66).bytes(message.tokenIdentifier);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseTokenMetadata();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 10) {
            break;
          }
          message.issuerPublicKey = reader.bytes();
          continue;
        }
        case 2: {
          if (tag !== 18) {
            break;
          }
          message.tokenName = reader.string();
          continue;
        }
        case 3: {
          if (tag !== 26) {
            break;
          }
          message.tokenTicker = reader.string();
          continue;
        }
        case 4: {
          if (tag !== 32) {
            break;
          }
          message.decimals = reader.uint32();
          continue;
        }
        case 5: {
          if (tag !== 42) {
            break;
          }
          message.maxSupply = reader.bytes();
          continue;
        }
        case 6: {
          if (tag !== 48) {
            break;
          }
          message.isFreezable = reader.bool();
          continue;
        }
        case 7: {
          if (tag !== 58) {
            break;
          }
          message.creationEntityPublicKey = reader.bytes();
          continue;
        }
        case 8: {
          if (tag !== 66) {
            break;
          }
          message.tokenIdentifier = reader.bytes();
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      issuerPublicKey: isSet3(object.issuerPublicKey) ? bytesFromBase642(object.issuerPublicKey) : new Uint8Array(0),
      tokenName: isSet3(object.tokenName) ? globalThis.String(object.tokenName) : "",
      tokenTicker: isSet3(object.tokenTicker) ? globalThis.String(object.tokenTicker) : "",
      decimals: isSet3(object.decimals) ? globalThis.Number(object.decimals) : 0,
      maxSupply: isSet3(object.maxSupply) ? bytesFromBase642(object.maxSupply) : new Uint8Array(0),
      isFreezable: isSet3(object.isFreezable) ? globalThis.Boolean(object.isFreezable) : false,
      creationEntityPublicKey: isSet3(object.creationEntityPublicKey) ? bytesFromBase642(object.creationEntityPublicKey) : void 0,
      tokenIdentifier: isSet3(object.tokenIdentifier) ? bytesFromBase642(object.tokenIdentifier) : new Uint8Array(0)
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.issuerPublicKey.length !== 0) {
      obj.issuerPublicKey = base64FromBytes2(message.issuerPublicKey);
    }
    if (message.tokenName !== "") {
      obj.tokenName = message.tokenName;
    }
    if (message.tokenTicker !== "") {
      obj.tokenTicker = message.tokenTicker;
    }
    if (message.decimals !== 0) {
      obj.decimals = Math.round(message.decimals);
    }
    if (message.maxSupply.length !== 0) {
      obj.maxSupply = base64FromBytes2(message.maxSupply);
    }
    if (message.isFreezable !== false) {
      obj.isFreezable = message.isFreezable;
    }
    if (message.creationEntityPublicKey !== void 0) {
      obj.creationEntityPublicKey = base64FromBytes2(message.creationEntityPublicKey);
    }
    if (message.tokenIdentifier.length !== 0) {
      obj.tokenIdentifier = base64FromBytes2(message.tokenIdentifier);
    }
    return obj;
  },
  create(base) {
    return TokenMetadata.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseTokenMetadata();
    message.issuerPublicKey = object.issuerPublicKey ?? new Uint8Array(0);
    message.tokenName = object.tokenName ?? "";
    message.tokenTicker = object.tokenTicker ?? "";
    message.decimals = object.decimals ?? 0;
    message.maxSupply = object.maxSupply ?? new Uint8Array(0);
    message.isFreezable = object.isFreezable ?? false;
    message.creationEntityPublicKey = object.creationEntityPublicKey ?? void 0;
    message.tokenIdentifier = object.tokenIdentifier ?? new Uint8Array(0);
    return message;
  }
};
function createBaseQueryTokenMetadataResponse() {
  return { tokenMetadata: [] };
}
var QueryTokenMetadataResponse = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    for (const v of message.tokenMetadata) {
      TokenMetadata.encode(v, writer.uint32(10).fork()).join();
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseQueryTokenMetadataResponse();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 10) {
            break;
          }
          message.tokenMetadata.push(TokenMetadata.decode(reader, reader.uint32()));
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      tokenMetadata: globalThis.Array.isArray(object?.tokenMetadata) ? object.tokenMetadata.map((e) => TokenMetadata.fromJSON(e)) : []
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.tokenMetadata?.length) {
      obj.tokenMetadata = message.tokenMetadata.map((e) => TokenMetadata.toJSON(e));
    }
    return obj;
  },
  create(base) {
    return QueryTokenMetadataResponse.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseQueryTokenMetadataResponse();
    message.tokenMetadata = object.tokenMetadata?.map((e) => TokenMetadata.fromPartial(e)) || [];
    return message;
  }
};
function createBaseQueryTokenOutputsRequest() {
  return { ownerPublicKeys: [], issuerPublicKeys: [], tokenIdentifiers: [], network: 0, pageRequest: void 0 };
}
var QueryTokenOutputsRequest = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    for (const v of message.ownerPublicKeys) {
      writer.uint32(10).bytes(v);
    }
    for (const v of message.issuerPublicKeys) {
      writer.uint32(18).bytes(v);
    }
    for (const v of message.tokenIdentifiers) {
      writer.uint32(34).bytes(v);
    }
    if (message.network !== 0) {
      writer.uint32(24).int32(message.network);
    }
    if (message.pageRequest !== void 0) {
      PageRequest.encode(message.pageRequest, writer.uint32(42).fork()).join();
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseQueryTokenOutputsRequest();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 10) {
            break;
          }
          message.ownerPublicKeys.push(reader.bytes());
          continue;
        }
        case 2: {
          if (tag !== 18) {
            break;
          }
          message.issuerPublicKeys.push(reader.bytes());
          continue;
        }
        case 4: {
          if (tag !== 34) {
            break;
          }
          message.tokenIdentifiers.push(reader.bytes());
          continue;
        }
        case 3: {
          if (tag !== 24) {
            break;
          }
          message.network = reader.int32();
          continue;
        }
        case 5: {
          if (tag !== 42) {
            break;
          }
          message.pageRequest = PageRequest.decode(reader, reader.uint32());
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      ownerPublicKeys: globalThis.Array.isArray(object?.ownerPublicKeys) ? object.ownerPublicKeys.map((e) => bytesFromBase642(e)) : [],
      issuerPublicKeys: globalThis.Array.isArray(object?.issuerPublicKeys) ? object.issuerPublicKeys.map((e) => bytesFromBase642(e)) : [],
      tokenIdentifiers: globalThis.Array.isArray(object?.tokenIdentifiers) ? object.tokenIdentifiers.map((e) => bytesFromBase642(e)) : [],
      network: isSet3(object.network) ? networkFromJSON(object.network) : 0,
      pageRequest: isSet3(object.pageRequest) ? PageRequest.fromJSON(object.pageRequest) : void 0
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.ownerPublicKeys?.length) {
      obj.ownerPublicKeys = message.ownerPublicKeys.map((e) => base64FromBytes2(e));
    }
    if (message.issuerPublicKeys?.length) {
      obj.issuerPublicKeys = message.issuerPublicKeys.map((e) => base64FromBytes2(e));
    }
    if (message.tokenIdentifiers?.length) {
      obj.tokenIdentifiers = message.tokenIdentifiers.map((e) => base64FromBytes2(e));
    }
    if (message.network !== 0) {
      obj.network = networkToJSON(message.network);
    }
    if (message.pageRequest !== void 0) {
      obj.pageRequest = PageRequest.toJSON(message.pageRequest);
    }
    return obj;
  },
  create(base) {
    return QueryTokenOutputsRequest.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseQueryTokenOutputsRequest();
    message.ownerPublicKeys = object.ownerPublicKeys?.map((e) => e) || [];
    message.issuerPublicKeys = object.issuerPublicKeys?.map((e) => e) || [];
    message.tokenIdentifiers = object.tokenIdentifiers?.map((e) => e) || [];
    message.network = object.network ?? 0;
    message.pageRequest = object.pageRequest !== void 0 && object.pageRequest !== null ? PageRequest.fromPartial(object.pageRequest) : void 0;
    return message;
  }
};
function createBaseQueryTokenTransactionsRequest() {
  return {
    outputIds: [],
    ownerPublicKeys: [],
    issuerPublicKeys: [],
    tokenIdentifiers: [],
    tokenTransactionHashes: [],
    order: 0,
    limit: 0,
    offset: 0
  };
}
var QueryTokenTransactionsRequest = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    for (const v of message.outputIds) {
      writer.uint32(10).string(v);
    }
    for (const v of message.ownerPublicKeys) {
      writer.uint32(18).bytes(v);
    }
    for (const v of message.issuerPublicKeys) {
      writer.uint32(26).bytes(v);
    }
    for (const v of message.tokenIdentifiers) {
      writer.uint32(58).bytes(v);
    }
    for (const v of message.tokenTransactionHashes) {
      writer.uint32(34).bytes(v);
    }
    if (message.order !== 0) {
      writer.uint32(64).int32(message.order);
    }
    if (message.limit !== 0) {
      writer.uint32(40).int64(message.limit);
    }
    if (message.offset !== 0) {
      writer.uint32(48).int64(message.offset);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseQueryTokenTransactionsRequest();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 10) {
            break;
          }
          message.outputIds.push(reader.string());
          continue;
        }
        case 2: {
          if (tag !== 18) {
            break;
          }
          message.ownerPublicKeys.push(reader.bytes());
          continue;
        }
        case 3: {
          if (tag !== 26) {
            break;
          }
          message.issuerPublicKeys.push(reader.bytes());
          continue;
        }
        case 7: {
          if (tag !== 58) {
            break;
          }
          message.tokenIdentifiers.push(reader.bytes());
          continue;
        }
        case 4: {
          if (tag !== 34) {
            break;
          }
          message.tokenTransactionHashes.push(reader.bytes());
          continue;
        }
        case 8: {
          if (tag !== 64) {
            break;
          }
          message.order = reader.int32();
          continue;
        }
        case 5: {
          if (tag !== 40) {
            break;
          }
          message.limit = longToNumber2(reader.int64());
          continue;
        }
        case 6: {
          if (tag !== 48) {
            break;
          }
          message.offset = longToNumber2(reader.int64());
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      outputIds: globalThis.Array.isArray(object?.outputIds) ? object.outputIds.map((e) => globalThis.String(e)) : [],
      ownerPublicKeys: globalThis.Array.isArray(object?.ownerPublicKeys) ? object.ownerPublicKeys.map((e) => bytesFromBase642(e)) : [],
      issuerPublicKeys: globalThis.Array.isArray(object?.issuerPublicKeys) ? object.issuerPublicKeys.map((e) => bytesFromBase642(e)) : [],
      tokenIdentifiers: globalThis.Array.isArray(object?.tokenIdentifiers) ? object.tokenIdentifiers.map((e) => bytesFromBase642(e)) : [],
      tokenTransactionHashes: globalThis.Array.isArray(object?.tokenTransactionHashes) ? object.tokenTransactionHashes.map((e) => bytesFromBase642(e)) : [],
      order: isSet3(object.order) ? orderFromJSON(object.order) : 0,
      limit: isSet3(object.limit) ? globalThis.Number(object.limit) : 0,
      offset: isSet3(object.offset) ? globalThis.Number(object.offset) : 0
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.outputIds?.length) {
      obj.outputIds = message.outputIds;
    }
    if (message.ownerPublicKeys?.length) {
      obj.ownerPublicKeys = message.ownerPublicKeys.map((e) => base64FromBytes2(e));
    }
    if (message.issuerPublicKeys?.length) {
      obj.issuerPublicKeys = message.issuerPublicKeys.map((e) => base64FromBytes2(e));
    }
    if (message.tokenIdentifiers?.length) {
      obj.tokenIdentifiers = message.tokenIdentifiers.map((e) => base64FromBytes2(e));
    }
    if (message.tokenTransactionHashes?.length) {
      obj.tokenTransactionHashes = message.tokenTransactionHashes.map((e) => base64FromBytes2(e));
    }
    if (message.order !== 0) {
      obj.order = orderToJSON(message.order);
    }
    if (message.limit !== 0) {
      obj.limit = Math.round(message.limit);
    }
    if (message.offset !== 0) {
      obj.offset = Math.round(message.offset);
    }
    return obj;
  },
  create(base) {
    return QueryTokenTransactionsRequest.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseQueryTokenTransactionsRequest();
    message.outputIds = object.outputIds?.map((e) => e) || [];
    message.ownerPublicKeys = object.ownerPublicKeys?.map((e) => e) || [];
    message.issuerPublicKeys = object.issuerPublicKeys?.map((e) => e) || [];
    message.tokenIdentifiers = object.tokenIdentifiers?.map((e) => e) || [];
    message.tokenTransactionHashes = object.tokenTransactionHashes?.map((e) => e) || [];
    message.order = object.order ?? 0;
    message.limit = object.limit ?? 0;
    message.offset = object.offset ?? 0;
    return message;
  }
};
function createBaseQueryTokenTransactionsResponse() {
  return { tokenTransactionsWithStatus: [], offset: 0 };
}
var QueryTokenTransactionsResponse = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    for (const v of message.tokenTransactionsWithStatus) {
      TokenTransactionWithStatus.encode(v, writer.uint32(10).fork()).join();
    }
    if (message.offset !== 0) {
      writer.uint32(16).int64(message.offset);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseQueryTokenTransactionsResponse();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 10) {
            break;
          }
          message.tokenTransactionsWithStatus.push(TokenTransactionWithStatus.decode(reader, reader.uint32()));
          continue;
        }
        case 2: {
          if (tag !== 16) {
            break;
          }
          message.offset = longToNumber2(reader.int64());
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      tokenTransactionsWithStatus: globalThis.Array.isArray(object?.tokenTransactionsWithStatus) ? object.tokenTransactionsWithStatus.map((e) => TokenTransactionWithStatus.fromJSON(e)) : [],
      offset: isSet3(object.offset) ? globalThis.Number(object.offset) : 0
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.tokenTransactionsWithStatus?.length) {
      obj.tokenTransactionsWithStatus = message.tokenTransactionsWithStatus.map(
        (e) => TokenTransactionWithStatus.toJSON(e)
      );
    }
    if (message.offset !== 0) {
      obj.offset = Math.round(message.offset);
    }
    return obj;
  },
  create(base) {
    return QueryTokenTransactionsResponse.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseQueryTokenTransactionsResponse();
    message.tokenTransactionsWithStatus = object.tokenTransactionsWithStatus?.map((e) => TokenTransactionWithStatus.fromPartial(e)) || [];
    message.offset = object.offset ?? 0;
    return message;
  }
};
function createBaseOutputWithPreviousTransactionData() {
  return { output: void 0, previousTransactionHash: new Uint8Array(0), previousTransactionVout: 0 };
}
var OutputWithPreviousTransactionData = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    if (message.output !== void 0) {
      TokenOutput.encode(message.output, writer.uint32(10).fork()).join();
    }
    if (message.previousTransactionHash.length !== 0) {
      writer.uint32(18).bytes(message.previousTransactionHash);
    }
    if (message.previousTransactionVout !== 0) {
      writer.uint32(24).uint32(message.previousTransactionVout);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseOutputWithPreviousTransactionData();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 10) {
            break;
          }
          message.output = TokenOutput.decode(reader, reader.uint32());
          continue;
        }
        case 2: {
          if (tag !== 18) {
            break;
          }
          message.previousTransactionHash = reader.bytes();
          continue;
        }
        case 3: {
          if (tag !== 24) {
            break;
          }
          message.previousTransactionVout = reader.uint32();
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      output: isSet3(object.output) ? TokenOutput.fromJSON(object.output) : void 0,
      previousTransactionHash: isSet3(object.previousTransactionHash) ? bytesFromBase642(object.previousTransactionHash) : new Uint8Array(0),
      previousTransactionVout: isSet3(object.previousTransactionVout) ? globalThis.Number(object.previousTransactionVout) : 0
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.output !== void 0) {
      obj.output = TokenOutput.toJSON(message.output);
    }
    if (message.previousTransactionHash.length !== 0) {
      obj.previousTransactionHash = base64FromBytes2(message.previousTransactionHash);
    }
    if (message.previousTransactionVout !== 0) {
      obj.previousTransactionVout = Math.round(message.previousTransactionVout);
    }
    return obj;
  },
  create(base) {
    return OutputWithPreviousTransactionData.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseOutputWithPreviousTransactionData();
    message.output = object.output !== void 0 && object.output !== null ? TokenOutput.fromPartial(object.output) : void 0;
    message.previousTransactionHash = object.previousTransactionHash ?? new Uint8Array(0);
    message.previousTransactionVout = object.previousTransactionVout ?? 0;
    return message;
  }
};
function createBaseQueryTokenOutputsResponse() {
  return { outputsWithPreviousTransactionData: [], pageResponse: void 0 };
}
var QueryTokenOutputsResponse = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    for (const v of message.outputsWithPreviousTransactionData) {
      OutputWithPreviousTransactionData.encode(v, writer.uint32(10).fork()).join();
    }
    if (message.pageResponse !== void 0) {
      PageResponse.encode(message.pageResponse, writer.uint32(18).fork()).join();
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseQueryTokenOutputsResponse();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 10) {
            break;
          }
          message.outputsWithPreviousTransactionData.push(
            OutputWithPreviousTransactionData.decode(reader, reader.uint32())
          );
          continue;
        }
        case 2: {
          if (tag !== 18) {
            break;
          }
          message.pageResponse = PageResponse.decode(reader, reader.uint32());
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      outputsWithPreviousTransactionData: globalThis.Array.isArray(object?.outputsWithPreviousTransactionData) ? object.outputsWithPreviousTransactionData.map((e) => OutputWithPreviousTransactionData.fromJSON(e)) : [],
      pageResponse: isSet3(object.pageResponse) ? PageResponse.fromJSON(object.pageResponse) : void 0
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.outputsWithPreviousTransactionData?.length) {
      obj.outputsWithPreviousTransactionData = message.outputsWithPreviousTransactionData.map(
        (e) => OutputWithPreviousTransactionData.toJSON(e)
      );
    }
    if (message.pageResponse !== void 0) {
      obj.pageResponse = PageResponse.toJSON(message.pageResponse);
    }
    return obj;
  },
  create(base) {
    return QueryTokenOutputsResponse.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseQueryTokenOutputsResponse();
    message.outputsWithPreviousTransactionData = object.outputsWithPreviousTransactionData?.map((e) => OutputWithPreviousTransactionData.fromPartial(e)) || [];
    message.pageResponse = object.pageResponse !== void 0 && object.pageResponse !== null ? PageResponse.fromPartial(object.pageResponse) : void 0;
    return message;
  }
};
function createBaseSpentTokenOutputMetadata() {
  return { outputId: "", revocationSecret: new Uint8Array(0) };
}
var SpentTokenOutputMetadata = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    if (message.outputId !== "") {
      writer.uint32(10).string(message.outputId);
    }
    if (message.revocationSecret.length !== 0) {
      writer.uint32(18).bytes(message.revocationSecret);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseSpentTokenOutputMetadata();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 10) {
            break;
          }
          message.outputId = reader.string();
          continue;
        }
        case 2: {
          if (tag !== 18) {
            break;
          }
          message.revocationSecret = reader.bytes();
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      outputId: isSet3(object.outputId) ? globalThis.String(object.outputId) : "",
      revocationSecret: isSet3(object.revocationSecret) ? bytesFromBase642(object.revocationSecret) : new Uint8Array(0)
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.outputId !== "") {
      obj.outputId = message.outputId;
    }
    if (message.revocationSecret.length !== 0) {
      obj.revocationSecret = base64FromBytes2(message.revocationSecret);
    }
    return obj;
  },
  create(base) {
    return SpentTokenOutputMetadata.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseSpentTokenOutputMetadata();
    message.outputId = object.outputId ?? "";
    message.revocationSecret = object.revocationSecret ?? new Uint8Array(0);
    return message;
  }
};
function createBaseTokenTransactionConfirmationMetadata() {
  return { spentTokenOutputsMetadata: [] };
}
var TokenTransactionConfirmationMetadata = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    for (const v of message.spentTokenOutputsMetadata) {
      SpentTokenOutputMetadata.encode(v, writer.uint32(10).fork()).join();
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseTokenTransactionConfirmationMetadata();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 10) {
            break;
          }
          message.spentTokenOutputsMetadata.push(SpentTokenOutputMetadata.decode(reader, reader.uint32()));
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      spentTokenOutputsMetadata: globalThis.Array.isArray(object?.spentTokenOutputsMetadata) ? object.spentTokenOutputsMetadata.map((e) => SpentTokenOutputMetadata.fromJSON(e)) : []
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.spentTokenOutputsMetadata?.length) {
      obj.spentTokenOutputsMetadata = message.spentTokenOutputsMetadata.map((e) => SpentTokenOutputMetadata.toJSON(e));
    }
    return obj;
  },
  create(base) {
    return TokenTransactionConfirmationMetadata.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseTokenTransactionConfirmationMetadata();
    message.spentTokenOutputsMetadata = object.spentTokenOutputsMetadata?.map((e) => SpentTokenOutputMetadata.fromPartial(e)) || [];
    return message;
  }
};
function createBaseTokenTransactionWithStatus() {
  return {
    tokenTransaction: void 0,
    status: 0,
    confirmationMetadata: void 0,
    tokenTransactionHash: new Uint8Array(0)
  };
}
var TokenTransactionWithStatus = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    if (message.tokenTransaction !== void 0) {
      TokenTransaction.encode(message.tokenTransaction, writer.uint32(10).fork()).join();
    }
    if (message.status !== 0) {
      writer.uint32(16).int32(message.status);
    }
    if (message.confirmationMetadata !== void 0) {
      TokenTransactionConfirmationMetadata.encode(message.confirmationMetadata, writer.uint32(26).fork()).join();
    }
    if (message.tokenTransactionHash.length !== 0) {
      writer.uint32(34).bytes(message.tokenTransactionHash);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseTokenTransactionWithStatus();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 10) {
            break;
          }
          message.tokenTransaction = TokenTransaction.decode(reader, reader.uint32());
          continue;
        }
        case 2: {
          if (tag !== 16) {
            break;
          }
          message.status = reader.int32();
          continue;
        }
        case 3: {
          if (tag !== 26) {
            break;
          }
          message.confirmationMetadata = TokenTransactionConfirmationMetadata.decode(reader, reader.uint32());
          continue;
        }
        case 4: {
          if (tag !== 34) {
            break;
          }
          message.tokenTransactionHash = reader.bytes();
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      tokenTransaction: isSet3(object.tokenTransaction) ? TokenTransaction.fromJSON(object.tokenTransaction) : void 0,
      status: isSet3(object.status) ? tokenTransactionStatusFromJSON(object.status) : 0,
      confirmationMetadata: isSet3(object.confirmationMetadata) ? TokenTransactionConfirmationMetadata.fromJSON(object.confirmationMetadata) : void 0,
      tokenTransactionHash: isSet3(object.tokenTransactionHash) ? bytesFromBase642(object.tokenTransactionHash) : new Uint8Array(0)
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.tokenTransaction !== void 0) {
      obj.tokenTransaction = TokenTransaction.toJSON(message.tokenTransaction);
    }
    if (message.status !== 0) {
      obj.status = tokenTransactionStatusToJSON(message.status);
    }
    if (message.confirmationMetadata !== void 0) {
      obj.confirmationMetadata = TokenTransactionConfirmationMetadata.toJSON(message.confirmationMetadata);
    }
    if (message.tokenTransactionHash.length !== 0) {
      obj.tokenTransactionHash = base64FromBytes2(message.tokenTransactionHash);
    }
    return obj;
  },
  create(base) {
    return TokenTransactionWithStatus.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseTokenTransactionWithStatus();
    message.tokenTransaction = object.tokenTransaction !== void 0 && object.tokenTransaction !== null ? TokenTransaction.fromPartial(object.tokenTransaction) : void 0;
    message.status = object.status ?? 0;
    message.confirmationMetadata = object.confirmationMetadata !== void 0 && object.confirmationMetadata !== null ? TokenTransactionConfirmationMetadata.fromPartial(object.confirmationMetadata) : void 0;
    message.tokenTransactionHash = object.tokenTransactionHash ?? new Uint8Array(0);
    return message;
  }
};
function createBaseFreezeTokensPayload() {
  return {
    version: 0,
    ownerPublicKey: new Uint8Array(0),
    tokenPublicKey: void 0,
    tokenIdentifier: void 0,
    issuerProvidedTimestamp: 0,
    operatorIdentityPublicKey: new Uint8Array(0),
    shouldUnfreeze: false
  };
}
var FreezeTokensPayload = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    if (message.version !== 0) {
      writer.uint32(8).uint32(message.version);
    }
    if (message.ownerPublicKey.length !== 0) {
      writer.uint32(18).bytes(message.ownerPublicKey);
    }
    if (message.tokenPublicKey !== void 0) {
      writer.uint32(26).bytes(message.tokenPublicKey);
    }
    if (message.tokenIdentifier !== void 0) {
      writer.uint32(34).bytes(message.tokenIdentifier);
    }
    if (message.issuerProvidedTimestamp !== 0) {
      writer.uint32(40).uint64(message.issuerProvidedTimestamp);
    }
    if (message.operatorIdentityPublicKey.length !== 0) {
      writer.uint32(50).bytes(message.operatorIdentityPublicKey);
    }
    if (message.shouldUnfreeze !== false) {
      writer.uint32(56).bool(message.shouldUnfreeze);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseFreezeTokensPayload();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 8) {
            break;
          }
          message.version = reader.uint32();
          continue;
        }
        case 2: {
          if (tag !== 18) {
            break;
          }
          message.ownerPublicKey = reader.bytes();
          continue;
        }
        case 3: {
          if (tag !== 26) {
            break;
          }
          message.tokenPublicKey = reader.bytes();
          continue;
        }
        case 4: {
          if (tag !== 34) {
            break;
          }
          message.tokenIdentifier = reader.bytes();
          continue;
        }
        case 5: {
          if (tag !== 40) {
            break;
          }
          message.issuerProvidedTimestamp = longToNumber2(reader.uint64());
          continue;
        }
        case 6: {
          if (tag !== 50) {
            break;
          }
          message.operatorIdentityPublicKey = reader.bytes();
          continue;
        }
        case 7: {
          if (tag !== 56) {
            break;
          }
          message.shouldUnfreeze = reader.bool();
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      version: isSet3(object.version) ? globalThis.Number(object.version) : 0,
      ownerPublicKey: isSet3(object.ownerPublicKey) ? bytesFromBase642(object.ownerPublicKey) : new Uint8Array(0),
      tokenPublicKey: isSet3(object.tokenPublicKey) ? bytesFromBase642(object.tokenPublicKey) : void 0,
      tokenIdentifier: isSet3(object.tokenIdentifier) ? bytesFromBase642(object.tokenIdentifier) : void 0,
      issuerProvidedTimestamp: isSet3(object.issuerProvidedTimestamp) ? globalThis.Number(object.issuerProvidedTimestamp) : 0,
      operatorIdentityPublicKey: isSet3(object.operatorIdentityPublicKey) ? bytesFromBase642(object.operatorIdentityPublicKey) : new Uint8Array(0),
      shouldUnfreeze: isSet3(object.shouldUnfreeze) ? globalThis.Boolean(object.shouldUnfreeze) : false
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.version !== 0) {
      obj.version = Math.round(message.version);
    }
    if (message.ownerPublicKey.length !== 0) {
      obj.ownerPublicKey = base64FromBytes2(message.ownerPublicKey);
    }
    if (message.tokenPublicKey !== void 0) {
      obj.tokenPublicKey = base64FromBytes2(message.tokenPublicKey);
    }
    if (message.tokenIdentifier !== void 0) {
      obj.tokenIdentifier = base64FromBytes2(message.tokenIdentifier);
    }
    if (message.issuerProvidedTimestamp !== 0) {
      obj.issuerProvidedTimestamp = Math.round(message.issuerProvidedTimestamp);
    }
    if (message.operatorIdentityPublicKey.length !== 0) {
      obj.operatorIdentityPublicKey = base64FromBytes2(message.operatorIdentityPublicKey);
    }
    if (message.shouldUnfreeze !== false) {
      obj.shouldUnfreeze = message.shouldUnfreeze;
    }
    return obj;
  },
  create(base) {
    return FreezeTokensPayload.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseFreezeTokensPayload();
    message.version = object.version ?? 0;
    message.ownerPublicKey = object.ownerPublicKey ?? new Uint8Array(0);
    message.tokenPublicKey = object.tokenPublicKey ?? void 0;
    message.tokenIdentifier = object.tokenIdentifier ?? void 0;
    message.issuerProvidedTimestamp = object.issuerProvidedTimestamp ?? 0;
    message.operatorIdentityPublicKey = object.operatorIdentityPublicKey ?? new Uint8Array(0);
    message.shouldUnfreeze = object.shouldUnfreeze ?? false;
    return message;
  }
};
function createBaseFreezeTokensRequest() {
  return { freezeTokensPayload: void 0, issuerSignature: new Uint8Array(0) };
}
var FreezeTokensRequest = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    if (message.freezeTokensPayload !== void 0) {
      FreezeTokensPayload.encode(message.freezeTokensPayload, writer.uint32(10).fork()).join();
    }
    if (message.issuerSignature.length !== 0) {
      writer.uint32(18).bytes(message.issuerSignature);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseFreezeTokensRequest();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 10) {
            break;
          }
          message.freezeTokensPayload = FreezeTokensPayload.decode(reader, reader.uint32());
          continue;
        }
        case 2: {
          if (tag !== 18) {
            break;
          }
          message.issuerSignature = reader.bytes();
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      freezeTokensPayload: isSet3(object.freezeTokensPayload) ? FreezeTokensPayload.fromJSON(object.freezeTokensPayload) : void 0,
      issuerSignature: isSet3(object.issuerSignature) ? bytesFromBase642(object.issuerSignature) : new Uint8Array(0)
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.freezeTokensPayload !== void 0) {
      obj.freezeTokensPayload = FreezeTokensPayload.toJSON(message.freezeTokensPayload);
    }
    if (message.issuerSignature.length !== 0) {
      obj.issuerSignature = base64FromBytes2(message.issuerSignature);
    }
    return obj;
  },
  create(base) {
    return FreezeTokensRequest.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseFreezeTokensRequest();
    message.freezeTokensPayload = object.freezeTokensPayload !== void 0 && object.freezeTokensPayload !== null ? FreezeTokensPayload.fromPartial(object.freezeTokensPayload) : void 0;
    message.issuerSignature = object.issuerSignature ?? new Uint8Array(0);
    return message;
  }
};
function createBaseFreezeTokensResponse() {
  return { impactedOutputIds: [], impactedTokenAmount: new Uint8Array(0) };
}
var FreezeTokensResponse = {
  encode(message, writer = new import_wire5.BinaryWriter()) {
    for (const v of message.impactedOutputIds) {
      writer.uint32(10).string(v);
    }
    if (message.impactedTokenAmount.length !== 0) {
      writer.uint32(18).bytes(message.impactedTokenAmount);
    }
    return writer;
  },
  decode(input, length) {
    const reader = input instanceof import_wire5.BinaryReader ? input : new import_wire5.BinaryReader(input);
    const end = length === void 0 ? reader.len : reader.pos + length;
    const message = createBaseFreezeTokensResponse();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1: {
          if (tag !== 10) {
            break;
          }
          message.impactedOutputIds.push(reader.string());
          continue;
        }
        case 2: {
          if (tag !== 18) {
            break;
          }
          message.impactedTokenAmount = reader.bytes();
          continue;
        }
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skip(tag & 7);
    }
    return message;
  },
  fromJSON(object) {
    return {
      impactedOutputIds: globalThis.Array.isArray(object?.impactedOutputIds) ? object.impactedOutputIds.map((e) => globalThis.String(e)) : [],
      impactedTokenAmount: isSet3(object.impactedTokenAmount) ? bytesFromBase642(object.impactedTokenAmount) : new Uint8Array(0)
    };
  },
  toJSON(message) {
    const obj = {};
    if (message.impactedOutputIds?.length) {
      obj.impactedOutputIds = message.impactedOutputIds;
    }
    if (message.impactedTokenAmount.length !== 0) {
      obj.impactedTokenAmount = base64FromBytes2(message.impactedTokenAmount);
    }
    return obj;
  },
  create(base) {
    return FreezeTokensResponse.fromPartial(base ?? {});
  },
  fromPartial(object) {
    const message = createBaseFreezeTokensResponse();
    message.impactedOutputIds = object.impactedOutputIds?.map((e) => e) || [];
    message.impactedTokenAmount = object.impactedTokenAmount ?? new Uint8Array(0);
    return message;
  }
};
var SparkTokenServiceDefinition = {
  name: "SparkTokenService",
  fullName: "spark_token.SparkTokenService",
  methods: {
    /**
     * Start process to create final token transaction with all inputs required
     * from user and SOs (including revocation secret commitment)
     */
    start_transaction: {
      name: "start_transaction",
      requestType: StartTransactionRequest,
      requestStream: false,
      responseType: StartTransactionResponse,
      responseStream: false,
      options: {}
    },
    /**
     * Complete the transaction and commit it with all SOs. This will be
     * coordinated by one SO.
     */
    commit_transaction: {
      name: "commit_transaction",
      requestType: CommitTransactionRequest,
      requestStream: false,
      responseType: CommitTransactionResponse,
      responseStream: false,
      options: {}
    },
    query_token_metadata: {
      name: "query_token_metadata",
      requestType: QueryTokenMetadataRequest,
      requestStream: false,
      responseType: QueryTokenMetadataResponse,
      responseStream: false,
      options: {}
    },
    query_token_transactions: {
      name: "query_token_transactions",
      requestType: QueryTokenTransactionsRequest,
      requestStream: false,
      responseType: QueryTokenTransactionsResponse,
      responseStream: false,
      options: {}
    },
    query_token_outputs: {
      name: "query_token_outputs",
      requestType: QueryTokenOutputsRequest,
      requestStream: false,
      responseType: QueryTokenOutputsResponse,
      responseStream: false,
      options: {}
    },
    freeze_tokens: {
      name: "freeze_tokens",
      requestType: FreezeTokensRequest,
      requestStream: false,
      responseType: FreezeTokensResponse,
      responseStream: false,
      options: {}
    },
    /** Replaces start_transaction and commit_transaction in single phase transaction flow. */
    broadcast_transaction: {
      name: "broadcast_transaction",
      requestType: BroadcastTransactionRequest,
      requestStream: false,
      responseType: BroadcastTransactionResponse,
      responseStream: false,
      options: {}
    }
  }
};
function bytesFromBase642(b64) {
  if (globalThis.Buffer) {
    return Uint8Array.from(globalThis.Buffer.from(b64, "base64"));
  } else {
    const bin = globalThis.atob(b64);
    const arr = new Uint8Array(bin.length);
    for (let i = 0; i < bin.length; ++i) {
      arr[i] = bin.charCodeAt(i);
    }
    return arr;
  }
}
function base64FromBytes2(arr) {
  if (globalThis.Buffer) {
    return globalThis.Buffer.from(arr).toString("base64");
  } else {
    const bin = [];
    arr.forEach((byte) => {
      bin.push(globalThis.String.fromCharCode(byte));
    });
    return globalThis.btoa(bin.join(""));
  }
}
function toTimestamp2(date) {
  const seconds = Math.trunc(date.getTime() / 1e3);
  const nanos = date.getTime() % 1e3 * 1e6;
  return { seconds, nanos };
}
function fromTimestamp2(t) {
  let millis = (t.seconds || 0) * 1e3;
  millis += (t.nanos || 0) / 1e6;
  return new globalThis.Date(millis);
}
function fromJsonTimestamp2(o) {
  if (o instanceof globalThis.Date) {
    return o;
  } else if (typeof o === "string") {
    return new globalThis.Date(o);
  } else {
    return fromTimestamp2(Timestamp.fromJSON(o));
  }
}
function longToNumber2(int64) {
  const num = globalThis.Number(int64.toString());
  if (num > globalThis.Number.MAX_SAFE_INTEGER) {
    throw new globalThis.Error("Value is larger than Number.MAX_SAFE_INTEGER");
  }
  if (num < globalThis.Number.MIN_SAFE_INTEGER) {
    throw new globalThis.Error("Value is smaller than Number.MIN_SAFE_INTEGER");
  }
  return num;
}
function isSet3(value) {
  return value !== null && value !== void 0;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  BroadcastTransactionRequest,
  BroadcastTransactionResponse,
  CommitProgress,
  CommitStatus,
  CommitTransactionRequest,
  CommitTransactionResponse,
  FinalTokenOutput,
  FinalTokenTransaction,
  FreezeTokensPayload,
  FreezeTokensRequest,
  FreezeTokensResponse,
  InputTtxoSignaturesPerOperator,
  InvoiceAttachment,
  OutputWithPreviousTransactionData,
  PartialTokenOutput,
  PartialTokenTransaction,
  QueryTokenMetadataRequest,
  QueryTokenMetadataResponse,
  QueryTokenOutputsRequest,
  QueryTokenOutputsResponse,
  QueryTokenTransactionsRequest,
  QueryTokenTransactionsResponse,
  SignatureWithIndex,
  SparkTokenServiceDefinition,
  SpentTokenOutputMetadata,
  StartTransactionRequest,
  StartTransactionResponse,
  TokenCreateInput,
  TokenMetadata,
  TokenMintInput,
  TokenOutput,
  TokenOutputToSpend,
  TokenTransaction,
  TokenTransactionConfirmationMetadata,
  TokenTransactionMetadata,
  TokenTransactionStatus,
  TokenTransactionType,
  TokenTransactionWithStatus,
  TokenTransferInput,
  commitStatusFromJSON,
  commitStatusToJSON,
  protobufPackage,
  tokenTransactionStatusFromJSON,
  tokenTransactionStatusToJSON,
  tokenTransactionTypeFromJSON,
  tokenTransactionTypeToJSON
});
