// @buildonspark/spark-sdk/dist/chunk-UYEB2VPG.js
// src/graphql/objects/BitcoinNetwork.ts
var BitcoinNetwork = /* @__PURE__ */ ((BitcoinNetwork2) => {
  BitcoinNetwork2["FUTURE_VALUE"] = "FUTURE_VALUE";
  BitcoinNetwork2["MAINNET"] = "MAINNET";
  BitcoinNetwork2["REGTEST"] = "REGTEST";
  BitcoinNetwork2["SIGNET"] = "SIGNET";
  BitcoinNetwork2["TESTNET"] = "TESTNET";
  return BitcoinNetwork2;
})(BitcoinNetwork || {});
var BitcoinNetwork_default = BitcoinNetwork;

// src/graphql/objects/ClaimStaticDeposit.ts
import { isObject } from "@lightsparkdev/core";

// src/graphql/objects/ClaimStaticDepositStatus.ts
var ClaimStaticDepositStatus = /* @__PURE__ */ ((ClaimStaticDepositStatus2) => {
  ClaimStaticDepositStatus2["FUTURE_VALUE"] = "FUTURE_VALUE";
  ClaimStaticDepositStatus2["CREATED"] = "CREATED";
  ClaimStaticDepositStatus2["TRANSFER_CREATED"] = "TRANSFER_CREATED";
  ClaimStaticDepositStatus2["TRANSFER_CREATION_FAILED"] = "TRANSFER_CREATION_FAILED";
  ClaimStaticDepositStatus2["REFUND_SIGNING_COMMITMENTS_QUERYING_FAILED"] = "REFUND_SIGNING_COMMITMENTS_QUERYING_FAILED";
  ClaimStaticDepositStatus2["REFUND_SIGNING_FAILED"] = "REFUND_SIGNING_FAILED";
  ClaimStaticDepositStatus2["UTXO_SWAPPING_FAILED"] = "UTXO_SWAPPING_FAILED";
  ClaimStaticDepositStatus2["TRANSFER_COMPLETED"] = "TRANSFER_COMPLETED";
  ClaimStaticDepositStatus2["SPEND_TX_CREATED"] = "SPEND_TX_CREATED";
  ClaimStaticDepositStatus2["SPEND_TX_BROADCAST"] = "SPEND_TX_BROADCAST";
  return ClaimStaticDepositStatus2;
})(ClaimStaticDepositStatus || {});
var ClaimStaticDepositStatus_default = ClaimStaticDepositStatus;

// src/graphql/objects/CurrencyUnit.ts
var CurrencyUnit = /* @__PURE__ */ ((CurrencyUnit2) => {
  CurrencyUnit2["FUTURE_VALUE"] = "FUTURE_VALUE";
  CurrencyUnit2["BITCOIN"] = "BITCOIN";
  CurrencyUnit2["SATOSHI"] = "SATOSHI";
  CurrencyUnit2["MILLISATOSHI"] = "MILLISATOSHI";
  CurrencyUnit2["USD"] = "USD";
  CurrencyUnit2["MXN"] = "MXN";
  CurrencyUnit2["PHP"] = "PHP";
  CurrencyUnit2["EUR"] = "EUR";
  CurrencyUnit2["NANOBITCOIN"] = "NANOBITCOIN";
  CurrencyUnit2["MICROBITCOIN"] = "MICROBITCOIN";
  CurrencyUnit2["MILLIBITCOIN"] = "MILLIBITCOIN";
  return CurrencyUnit2;
})(CurrencyUnit || {});
var CurrencyUnit_default = CurrencyUnit;

// src/graphql/objects/CurrencyAmount.ts
var CurrencyAmountFromJson = (obj) => {
  return {
    originalValue: obj["currency_amount_original_value"],
    originalUnit: CurrencyUnit_default[obj["currency_amount_original_unit"]] ?? CurrencyUnit_default.FUTURE_VALUE
  };
};
var CurrencyAmountToJson = (obj) => {
  return {
    currency_amount_original_value: obj.originalValue,
    currency_amount_original_unit: obj.originalUnit
  };
};

// src/graphql/objects/ClaimStaticDeposit.ts
var ClaimStaticDepositFromJson = (obj) => {
  return {
    id: obj["claim_static_deposit_id"],
    createdAt: obj["claim_static_deposit_created_at"],
    updatedAt: obj["claim_static_deposit_updated_at"],
    network: BitcoinNetwork_default[obj["claim_static_deposit_network"]] ?? BitcoinNetwork_default.FUTURE_VALUE,
    creditAmount: CurrencyAmountFromJson(
      obj["claim_static_deposit_credit_amount"]
    ),
    maxFee: CurrencyAmountFromJson(obj["claim_static_deposit_max_fee"]),
    status: ClaimStaticDepositStatus_default[obj["claim_static_deposit_status"]] ?? ClaimStaticDepositStatus_default.FUTURE_VALUE,
    transactionId: obj["claim_static_deposit_transaction_id"],
    outputIndex: obj["claim_static_deposit_output_index"],
    bitcoinNetwork: BitcoinNetwork_default[obj["claim_static_deposit_bitcoin_network"]] ?? BitcoinNetwork_default.FUTURE_VALUE,
    typename: "ClaimStaticDeposit",
    transferSparkId: obj["claim_static_deposit_transfer_spark_id"]
  };
};
var FRAGMENT = `
fragment ClaimStaticDepositFragment on ClaimStaticDeposit {
    __typename
    claim_static_deposit_id: id
    claim_static_deposit_created_at: created_at
    claim_static_deposit_updated_at: updated_at
    claim_static_deposit_network: network
    claim_static_deposit_credit_amount: credit_amount {
        __typename
        currency_amount_original_value: original_value
        currency_amount_original_unit: original_unit
    }
    claim_static_deposit_max_fee: max_fee {
        __typename
        currency_amount_original_value: original_value
        currency_amount_original_unit: original_unit
    }
    claim_static_deposit_status: status
    claim_static_deposit_transaction_id: transaction_id
    claim_static_deposit_output_index: output_index
    claim_static_deposit_bitcoin_network: bitcoin_network
    claim_static_deposit_transfer_spark_id: transfer_spark_id
}`;
var getClaimStaticDepositQuery = (id) => {
  return {
    queryPayload: `
query GetClaimStaticDeposit($id: ID!) {
    entity(id: $id) {
        ... on ClaimStaticDeposit {
            ...ClaimStaticDepositFragment
        }
    }
}

${FRAGMENT}    
`,
    variables: { id },
    constructObject: (data) => isObject(data) && "entity" in data && isObject(data.entity) ? ClaimStaticDepositFromJson(data.entity) : null
  };
};

// src/graphql/objects/ClaimStaticDepositRequestType.ts
var ClaimStaticDepositRequestType = /* @__PURE__ */ ((ClaimStaticDepositRequestType2) => {
  ClaimStaticDepositRequestType2["FUTURE_VALUE"] = "FUTURE_VALUE";
  ClaimStaticDepositRequestType2["FIXED_AMOUNT"] = "FIXED_AMOUNT";
  ClaimStaticDepositRequestType2["MAX_FEE"] = "MAX_FEE";
  return ClaimStaticDepositRequestType2;
})(ClaimStaticDepositRequestType || {});
var ClaimStaticDepositRequestType_default = ClaimStaticDepositRequestType;

// src/graphql/objects/CoopExitFeeQuote.ts
import { isObject as isObject2 } from "@lightsparkdev/core";
var CoopExitFeeQuoteFromJson = (obj) => {
  return {
    id: obj["coop_exit_fee_quote_id"],
    createdAt: obj["coop_exit_fee_quote_created_at"],
    updatedAt: obj["coop_exit_fee_quote_updated_at"],
    network: BitcoinNetwork_default[obj["coop_exit_fee_quote_network"]] ?? BitcoinNetwork_default.FUTURE_VALUE,
    totalAmount: CurrencyAmountFromJson(
      obj["coop_exit_fee_quote_total_amount"]
    ),
    userFeeFast: CurrencyAmountFromJson(
      obj["coop_exit_fee_quote_user_fee_fast"]
    ),
    userFeeMedium: CurrencyAmountFromJson(
      obj["coop_exit_fee_quote_user_fee_medium"]
    ),
    userFeeSlow: CurrencyAmountFromJson(
      obj["coop_exit_fee_quote_user_fee_slow"]
    ),
    l1BroadcastFeeFast: CurrencyAmountFromJson(
      obj["coop_exit_fee_quote_l1_broadcast_fee_fast"]
    ),
    l1BroadcastFeeMedium: CurrencyAmountFromJson(
      obj["coop_exit_fee_quote_l1_broadcast_fee_medium"]
    ),
    l1BroadcastFeeSlow: CurrencyAmountFromJson(
      obj["coop_exit_fee_quote_l1_broadcast_fee_slow"]
    ),
    expiresAt: obj["coop_exit_fee_quote_expires_at"],
    typename: "CoopExitFeeQuote"
  };
};
var FRAGMENT2 = `
fragment CoopExitFeeQuoteFragment on CoopExitFeeQuote {
    __typename
    coop_exit_fee_quote_id: id
    coop_exit_fee_quote_created_at: created_at
    coop_exit_fee_quote_updated_at: updated_at
    coop_exit_fee_quote_network: network
    coop_exit_fee_quote_total_amount: total_amount {
        __typename
        currency_amount_original_value: original_value
        currency_amount_original_unit: original_unit
    }
    coop_exit_fee_quote_user_fee_fast: user_fee_fast {
        __typename
        currency_amount_original_value: original_value
        currency_amount_original_unit: original_unit
    }
    coop_exit_fee_quote_user_fee_medium: user_fee_medium {
        __typename
        currency_amount_original_value: original_value
        currency_amount_original_unit: original_unit
    }
    coop_exit_fee_quote_user_fee_slow: user_fee_slow {
        __typename
        currency_amount_original_value: original_value
        currency_amount_original_unit: original_unit
    }
    coop_exit_fee_quote_l1_broadcast_fee_fast: l1_broadcast_fee_fast {
        __typename
        currency_amount_original_value: original_value
        currency_amount_original_unit: original_unit
    }
    coop_exit_fee_quote_l1_broadcast_fee_medium: l1_broadcast_fee_medium {
        __typename
        currency_amount_original_value: original_value
        currency_amount_original_unit: original_unit
    }
    coop_exit_fee_quote_l1_broadcast_fee_slow: l1_broadcast_fee_slow {
        __typename
        currency_amount_original_value: original_value
        currency_amount_original_unit: original_unit
    }
    coop_exit_fee_quote_expires_at: expires_at
}`;
var getCoopExitFeeQuoteQuery = (id) => {
  return {
    queryPayload: `
query GetCoopExitFeeQuote($id: ID!) {
    entity(id: $id) {
        ... on CoopExitFeeQuote {
            ...CoopExitFeeQuoteFragment
        }
    }
}

${FRAGMENT2}    
`,
    variables: { id },
    constructObject: (data) => isObject2(data) && "entity" in data && isObject2(data.entity) ? CoopExitFeeQuoteFromJson(data.entity) : null
  };
};

// src/graphql/objects/ExitSpeed.ts
var ExitSpeed = /* @__PURE__ */ ((ExitSpeed2) => {
  ExitSpeed2["FUTURE_VALUE"] = "FUTURE_VALUE";
  ExitSpeed2["FAST"] = "FAST";
  ExitSpeed2["MEDIUM"] = "MEDIUM";
  ExitSpeed2["SLOW"] = "SLOW";
  return ExitSpeed2;
})(ExitSpeed || {});
var ExitSpeed_default = ExitSpeed;

// src/graphql/objects/LeavesSwapRequest.ts
import { isObject as isObject3 } from "@lightsparkdev/core";

// src/graphql/objects/SparkLeavesSwapRequestStatus.ts
var SparkLeavesSwapRequestStatus = /* @__PURE__ */ ((SparkLeavesSwapRequestStatus2) => {
  SparkLeavesSwapRequestStatus2["FUTURE_VALUE"] = "FUTURE_VALUE";
  SparkLeavesSwapRequestStatus2["CREATED"] = "CREATED";
  SparkLeavesSwapRequestStatus2["INITIATED"] = "INITIATED";
  SparkLeavesSwapRequestStatus2["LEAVES_LOCKED"] = "LEAVES_LOCKED";
  SparkLeavesSwapRequestStatus2["REFUND_TX_ADAPTOR_SIGNED"] = "REFUND_TX_ADAPTOR_SIGNED";
  SparkLeavesSwapRequestStatus2["INBOUND_TRANSFER_CLAIMED"] = "INBOUND_TRANSFER_CLAIMED";
  SparkLeavesSwapRequestStatus2["SUCCEEDED"] = "SUCCEEDED";
  SparkLeavesSwapRequestStatus2["EXPIRED"] = "EXPIRED";
  SparkLeavesSwapRequestStatus2["FAILED"] = "FAILED";
  return SparkLeavesSwapRequestStatus2;
})(SparkLeavesSwapRequestStatus || {});
var SparkLeavesSwapRequestStatus_default = SparkLeavesSwapRequestStatus;

// src/graphql/objects/SwapLeaf.ts
var SwapLeafFromJson = (obj) => {
  return {
    leafId: obj["swap_leaf_leaf_id"],
    rawUnsignedRefundTransaction: obj["swap_leaf_raw_unsigned_refund_transaction"],
    adaptorSignedSignature: obj["swap_leaf_adaptor_signed_signature"],
    directRawUnsignedRefundTransaction: obj["swap_leaf_direct_raw_unsigned_refund_transaction"],
    directAdaptorSignedSignature: obj["swap_leaf_direct_adaptor_signed_signature"],
    directFromCpfpRawUnsignedRefundTransaction: obj["swap_leaf_direct_from_cpfp_raw_unsigned_refund_transaction"],
    directFromCpfpAdaptorSignedSignature: obj["swap_leaf_direct_from_cpfp_adaptor_signed_signature"]
  };
};

// src/auto-bind.ts
function getAllProperties(object) {
  const properties = /* @__PURE__ */ new Set();
  let current = object;
  while (current && current !== Object.prototype) {
    for (const key of Reflect.ownKeys(current)) {
      properties.add([current, key]);
    }
    current = Reflect.getPrototypeOf(current);
  }
  return properties;
}
function autoBind(self, { include, exclude } = {}) {
  const filter = (key) => {
    const keyStr = typeof key === "string" ? key : key.toString();
    const match = (pattern) => typeof pattern === "string" ? keyStr === pattern : pattern.test(keyStr);
    if (include) {
      return include.some(match);
    }
    if (exclude) {
      return !exclude.some(match);
    }
    return true;
  };
  for (const [proto, key] of getAllProperties(self.constructor.prototype)) {
    if (key === "constructor" || !filter(key)) {
      continue;
    }
    const descriptor = Reflect.getOwnPropertyDescriptor(proto, key);
    if (descriptor && typeof descriptor.value === "function") {
      self[key] = self[key].bind(self);
    }
  }
  return self;
}

// src/graphql/objects/Leaf.ts
var LeafFromJson = (obj) => {
  return {
    amount: CurrencyAmountFromJson(obj["leaf_amount"]),
    sparkNodeId: obj["leaf_spark_node_id"]
  };
};

// src/graphql/objects/PageInfo.ts
var PageInfoFromJson = (obj) => {
  return {
    hasNextPage: obj["page_info_has_next_page"],
    hasPreviousPage: obj["page_info_has_previous_page"],
    startCursor: obj["page_info_start_cursor"],
    endCursor: obj["page_info_end_cursor"]
  };
};

// src/graphql/objects/SparkTransferToLeavesConnection.ts
var SparkTransferToLeavesConnectionFromJson = (obj) => {
  return {
    count: obj["spark_transfer_to_leaves_connection_count"],
    pageInfo: PageInfoFromJson(
      obj["spark_transfer_to_leaves_connection_page_info"]
    ),
    entities: obj["spark_transfer_to_leaves_connection_entities"].map(
      (e) => LeafFromJson(e)
    ),
    typename: "SparkTransferToLeavesConnection"
  };
};

// src/graphql/objects/Transfer.ts
var Transfer = class {
  constructor(totalAmount, sparkId, userRequestId) {
    this.totalAmount = totalAmount;
    this.sparkId = sparkId;
    this.userRequestId = userRequestId;
    autoBind(this);
  }
  async getLeaves(client, first = void 0, after = void 0) {
    return await client.executeRawQuery({
      queryPayload: ` 
query FetchSparkTransferToLeavesConnection($entity_id: ID!, $first: Int, $after: String) {
    entity(id: $entity_id) {
        ... on Transfer {
            leaves(, first: $first, after: $after) {
                __typename
                spark_transfer_to_leaves_connection_count: count
                spark_transfer_to_leaves_connection_page_info: page_info {
                    __typename
                    page_info_has_next_page: has_next_page
                    page_info_has_previous_page: has_previous_page
                    page_info_start_cursor: start_cursor
                    page_info_end_cursor: end_cursor
                }
                spark_transfer_to_leaves_connection_entities: entities {
                    __typename
                    leaf_amount: amount {
                        __typename
                        currency_amount_original_value: original_value
                        currency_amount_original_unit: original_unit
                    }
                    leaf_spark_node_id: spark_node_id
                }
            }
        }
    }
}
`,
      variables: { entity_id: this.sparkId, first, after },
      constructObject: (json) => {
        const connection = json["entity"]["leaves"];
        return SparkTransferToLeavesConnectionFromJson(connection);
      }
    });
  }
  toJson() {
    return {
      transfer_total_amount: CurrencyAmountToJson(this.totalAmount),
      transfer_spark_id: this.sparkId,
      transfer_user_request: { id: this.userRequestId }
    };
  }
};
var TransferFromJson = (obj) => {
  return new Transfer(
    CurrencyAmountFromJson(obj["transfer_total_amount"]),
    obj["transfer_spark_id"],
    obj["transfer_user_request"]?.id ?? void 0
  );
};
var FRAGMENT3 = `
fragment TransferFragment on Transfer {
    __typename
    transfer_total_amount: total_amount {
        __typename
        currency_amount_original_value: original_value
        currency_amount_original_unit: original_unit
    }
    transfer_spark_id: spark_id
    transfer_user_request: user_request {
        id
    }
}`;
var Transfer_default = Transfer;

// src/graphql/objects/LeavesSwapRequest.ts
var LeavesSwapRequestFromJson = (obj) => {
  return {
    id: obj["leaves_swap_request_id"],
    createdAt: obj["leaves_swap_request_created_at"],
    updatedAt: obj["leaves_swap_request_updated_at"],
    network: BitcoinNetwork_default[obj["leaves_swap_request_network"]] ?? BitcoinNetwork_default.FUTURE_VALUE,
    status: SparkLeavesSwapRequestStatus_default[obj["leaves_swap_request_status"]] ?? SparkLeavesSwapRequestStatus_default.FUTURE_VALUE,
    totalAmount: CurrencyAmountFromJson(
      obj["leaves_swap_request_total_amount"]
    ),
    targetAmount: CurrencyAmountFromJson(
      obj["leaves_swap_request_target_amount"]
    ),
    fee: CurrencyAmountFromJson(obj["leaves_swap_request_fee"]),
    inboundTransfer: TransferFromJson(
      obj["leaves_swap_request_inbound_transfer"]
    ),
    swapLeaves: obj["leaves_swap_request_swap_leaves"].map(
      (e) => SwapLeafFromJson(e)
    ),
    typename: "LeavesSwapRequest",
    outboundTransfer: !!obj["leaves_swap_request_outbound_transfer"] ? TransferFromJson(obj["leaves_swap_request_outbound_transfer"]) : void 0,
    expiresAt: obj["leaves_swap_request_expires_at"]
  };
};
var FRAGMENT4 = `
fragment LeavesSwapRequestFragment on LeavesSwapRequest {
    __typename
    leaves_swap_request_id: id
    leaves_swap_request_created_at: created_at
    leaves_swap_request_updated_at: updated_at
    leaves_swap_request_network: network
    leaves_swap_request_status: status
    leaves_swap_request_total_amount: total_amount {
        __typename
        currency_amount_original_value: original_value
        currency_amount_original_unit: original_unit
    }
    leaves_swap_request_target_amount: target_amount {
        __typename
        currency_amount_original_value: original_value
        currency_amount_original_unit: original_unit
    }
    leaves_swap_request_fee: fee {
        __typename
        currency_amount_original_value: original_value
        currency_amount_original_unit: original_unit
    }
    leaves_swap_request_inbound_transfer: inbound_transfer {
        __typename
        transfer_total_amount: total_amount {
            __typename
            currency_amount_original_value: original_value
            currency_amount_original_unit: original_unit
        }
        transfer_spark_id: spark_id
        transfer_user_request: user_request {
            id
        }
    }
    leaves_swap_request_outbound_transfer: outbound_transfer {
        __typename
        transfer_total_amount: total_amount {
            __typename
            currency_amount_original_value: original_value
            currency_amount_original_unit: original_unit
        }
        transfer_spark_id: spark_id
        transfer_user_request: user_request {
            id
        }
    }
    leaves_swap_request_expires_at: expires_at
    leaves_swap_request_swap_leaves: swap_leaves {
        __typename
        swap_leaf_leaf_id: leaf_id
        swap_leaf_raw_unsigned_refund_transaction: raw_unsigned_refund_transaction
        swap_leaf_adaptor_signed_signature: adaptor_signed_signature
        swap_leaf_direct_raw_unsigned_refund_transaction: direct_raw_unsigned_refund_transaction
        swap_leaf_direct_adaptor_signed_signature: direct_adaptor_signed_signature
        swap_leaf_direct_from_cpfp_raw_unsigned_refund_transaction: direct_from_cpfp_raw_unsigned_refund_transaction
        swap_leaf_direct_from_cpfp_adaptor_signed_signature: direct_from_cpfp_adaptor_signed_signature
    }
}`;
var getLeavesSwapRequestQuery = (id) => {
  return {
    queryPayload: `
query GetLeavesSwapRequest($id: ID!) {
    entity(id: $id) {
        ... on LeavesSwapRequest {
            ...LeavesSwapRequestFragment
        }
    }
}

${FRAGMENT4}    
`,
    variables: { id },
    constructObject: (data) => isObject3(data) && "entity" in data && isObject3(data.entity) ? LeavesSwapRequestFromJson(data.entity) : null
  };
};

// src/graphql/objects/LightningReceiveRequest.ts
import { isObject as isObject4 } from "@lightsparkdev/core";

// src/graphql/objects/Invoice.ts
var InvoiceFromJson = (obj) => {
  return {
    encodedInvoice: obj["invoice_encoded_invoice"],
    bitcoinNetwork: BitcoinNetwork_default[obj["invoice_bitcoin_network"]] ?? BitcoinNetwork_default.FUTURE_VALUE,
    paymentHash: obj["invoice_payment_hash"],
    amount: CurrencyAmountFromJson(obj["invoice_amount"]),
    createdAt: obj["invoice_created_at"],
    expiresAt: obj["invoice_expires_at"],
    memo: obj["invoice_memo"]
  };
};

// src/graphql/objects/LightningReceiveRequestStatus.ts
var LightningReceiveRequestStatus = /* @__PURE__ */ ((LightningReceiveRequestStatus2) => {
  LightningReceiveRequestStatus2["FUTURE_VALUE"] = "FUTURE_VALUE";
  LightningReceiveRequestStatus2["INVOICE_CREATED"] = "INVOICE_CREATED";
  LightningReceiveRequestStatus2["TRANSFER_CREATED"] = "TRANSFER_CREATED";
  LightningReceiveRequestStatus2["TRANSFER_CREATION_FAILED"] = "TRANSFER_CREATION_FAILED";
  LightningReceiveRequestStatus2["REFUND_SIGNING_COMMITMENTS_QUERYING_FAILED"] = "REFUND_SIGNING_COMMITMENTS_QUERYING_FAILED";
  LightningReceiveRequestStatus2["REFUND_SIGNING_FAILED"] = "REFUND_SIGNING_FAILED";
  LightningReceiveRequestStatus2["PAYMENT_PREIMAGE_RECOVERED"] = "PAYMENT_PREIMAGE_RECOVERED";
  LightningReceiveRequestStatus2["PAYMENT_PREIMAGE_RECOVERING_FAILED"] = "PAYMENT_PREIMAGE_RECOVERING_FAILED";
  LightningReceiveRequestStatus2["LIGHTNING_PAYMENT_RECEIVED"] = "LIGHTNING_PAYMENT_RECEIVED";
  LightningReceiveRequestStatus2["TRANSFER_FAILED"] = "TRANSFER_FAILED";
  LightningReceiveRequestStatus2["TRANSFER_COMPLETED"] = "TRANSFER_COMPLETED";
  return LightningReceiveRequestStatus2;
})(LightningReceiveRequestStatus || {});
var LightningReceiveRequestStatus_default = LightningReceiveRequestStatus;

// src/graphql/objects/LightningReceiveRequest.ts
var LightningReceiveRequestFromJson = (obj) => {
  return {
    id: obj["lightning_receive_request_id"],
    createdAt: obj["lightning_receive_request_created_at"],
    updatedAt: obj["lightning_receive_request_updated_at"],
    network: BitcoinNetwork_default[obj["lightning_receive_request_network"]] ?? BitcoinNetwork_default.FUTURE_VALUE,
    invoice: InvoiceFromJson(obj["lightning_receive_request_invoice"]),
    status: LightningReceiveRequestStatus_default[obj["lightning_receive_request_status"]] ?? LightningReceiveRequestStatus_default.FUTURE_VALUE,
    typename: "LightningReceiveRequest",
    transfer: !!obj["lightning_receive_request_transfer"] ? TransferFromJson(obj["lightning_receive_request_transfer"]) : void 0,
    paymentPreimage: obj["lightning_receive_request_payment_preimage"],
    receiverIdentityPublicKey: obj["lightning_receive_request_receiver_identity_public_key"]
  };
};
var FRAGMENT5 = `
fragment LightningReceiveRequestFragment on LightningReceiveRequest {
    __typename
    lightning_receive_request_id: id
    lightning_receive_request_created_at: created_at
    lightning_receive_request_updated_at: updated_at
    lightning_receive_request_network: network
    lightning_receive_request_invoice: invoice {
        __typename
        invoice_encoded_invoice: encoded_invoice
        invoice_bitcoin_network: bitcoin_network
        invoice_payment_hash: payment_hash
        invoice_amount: amount {
            __typename
            currency_amount_original_value: original_value
            currency_amount_original_unit: original_unit
        }
        invoice_created_at: created_at
        invoice_expires_at: expires_at
        invoice_memo: memo
    }
    lightning_receive_request_status: status
    lightning_receive_request_transfer: transfer {
        __typename
        transfer_total_amount: total_amount {
            __typename
            currency_amount_original_value: original_value
            currency_amount_original_unit: original_unit
        }
        transfer_spark_id: spark_id
        transfer_user_request: user_request {
            id
        }
    }
    lightning_receive_request_payment_preimage: payment_preimage
    lightning_receive_request_receiver_identity_public_key: receiver_identity_public_key
}`;
var getLightningReceiveRequestQuery = (id) => {
  return {
    queryPayload: `
query GetLightningReceiveRequest($id: ID!) {
    entity(id: $id) {
        ... on LightningReceiveRequest {
            ...LightningReceiveRequestFragment
        }
    }
}

${FRAGMENT5}    
`,
    variables: { id },
    constructObject: (data) => isObject4(data) && "entity" in data && isObject4(data.entity) ? LightningReceiveRequestFromJson(data.entity) : null
  };
};

// src/graphql/objects/LightningSendRequest.ts
import { isObject as isObject5 } from "@lightsparkdev/core";

// src/graphql/objects/LightningSendRequestStatus.ts
var LightningSendRequestStatus = /* @__PURE__ */ ((LightningSendRequestStatus2) => {
  LightningSendRequestStatus2["FUTURE_VALUE"] = "FUTURE_VALUE";
  LightningSendRequestStatus2["CREATED"] = "CREATED";
  LightningSendRequestStatus2["REQUEST_VALIDATED"] = "REQUEST_VALIDATED";
  LightningSendRequestStatus2["LIGHTNING_PAYMENT_INITIATED"] = "LIGHTNING_PAYMENT_INITIATED";
  LightningSendRequestStatus2["LIGHTNING_PAYMENT_FAILED"] = "LIGHTNING_PAYMENT_FAILED";
  LightningSendRequestStatus2["LIGHTNING_PAYMENT_SUCCEEDED"] = "LIGHTNING_PAYMENT_SUCCEEDED";
  LightningSendRequestStatus2["PREIMAGE_PROVIDED"] = "PREIMAGE_PROVIDED";
  LightningSendRequestStatus2["TRANSFER_COMPLETED"] = "TRANSFER_COMPLETED";
  return LightningSendRequestStatus2;
})(LightningSendRequestStatus || {});
var LightningSendRequestStatus_default = LightningSendRequestStatus;

// src/graphql/objects/LightningSendRequest.ts
var LightningSendRequestFromJson = (obj) => {
  return {
    id: obj["lightning_send_request_id"],
    createdAt: obj["lightning_send_request_created_at"],
    updatedAt: obj["lightning_send_request_updated_at"],
    network: BitcoinNetwork_default[obj["lightning_send_request_network"]] ?? BitcoinNetwork_default.FUTURE_VALUE,
    encodedInvoice: obj["lightning_send_request_encoded_invoice"],
    fee: CurrencyAmountFromJson(obj["lightning_send_request_fee"]),
    idempotencyKey: obj["lightning_send_request_idempotency_key"],
    status: LightningSendRequestStatus_default[obj["lightning_send_request_status"]] ?? LightningSendRequestStatus_default.FUTURE_VALUE,
    typename: "LightningSendRequest",
    transfer: !!obj["lightning_send_request_transfer"] ? TransferFromJson(obj["lightning_send_request_transfer"]) : void 0,
    paymentPreimage: obj["lightning_send_request_payment_preimage"]
  };
};
var FRAGMENT6 = `
fragment LightningSendRequestFragment on LightningSendRequest {
    __typename
    lightning_send_request_id: id
    lightning_send_request_created_at: created_at
    lightning_send_request_updated_at: updated_at
    lightning_send_request_network: network
    lightning_send_request_encoded_invoice: encoded_invoice
    lightning_send_request_fee: fee {
        __typename
        currency_amount_original_value: original_value
        currency_amount_original_unit: original_unit
    }
    lightning_send_request_idempotency_key: idempotency_key
    lightning_send_request_status: status
    lightning_send_request_transfer: transfer {
        __typename
        transfer_total_amount: total_amount {
            __typename
            currency_amount_original_value: original_value
            currency_amount_original_unit: original_unit
        }
        transfer_spark_id: spark_id
        transfer_user_request: user_request {
            id
        }
    }
    lightning_send_request_payment_preimage: payment_preimage
}`;
var getLightningSendRequestQuery = (id) => {
  return {
    queryPayload: `
query GetLightningSendRequest($id: ID!) {
    entity(id: $id) {
        ... on LightningSendRequest {
            ...LightningSendRequestFragment
        }
    }
}

${FRAGMENT6}    
`,
    variables: { id },
    constructObject: (data) => isObject5(data) && "entity" in data && isObject5(data.entity) ? LightningSendRequestFromJson(data.entity) : null
  };
};

// src/graphql/objects/SparkCoopExitRequestStatus.ts
var SparkCoopExitRequestStatus = /* @__PURE__ */ ((SparkCoopExitRequestStatus2) => {
  SparkCoopExitRequestStatus2["FUTURE_VALUE"] = "FUTURE_VALUE";
  SparkCoopExitRequestStatus2["INITIATED"] = "INITIATED";
  SparkCoopExitRequestStatus2["INBOUND_TRANSFER_CHECKED"] = "INBOUND_TRANSFER_CHECKED";
  SparkCoopExitRequestStatus2["TX_SIGNED"] = "TX_SIGNED";
  SparkCoopExitRequestStatus2["TX_BROADCASTED"] = "TX_BROADCASTED";
  SparkCoopExitRequestStatus2["WAITING_ON_TX_CONFIRMATIONS"] = "WAITING_ON_TX_CONFIRMATIONS";
  SparkCoopExitRequestStatus2["SUCCEEDED"] = "SUCCEEDED";
  SparkCoopExitRequestStatus2["EXPIRED"] = "EXPIRED";
  SparkCoopExitRequestStatus2["FAILED"] = "FAILED";
  return SparkCoopExitRequestStatus2;
})(SparkCoopExitRequestStatus || {});
var SparkCoopExitRequestStatus_default = SparkCoopExitRequestStatus;

// src/graphql/objects/SparkUserRequestStatus.ts
var SparkUserRequestStatus = /* @__PURE__ */ ((SparkUserRequestStatus2) => {
  SparkUserRequestStatus2["FUTURE_VALUE"] = "FUTURE_VALUE";
  SparkUserRequestStatus2["CREATED"] = "CREATED";
  SparkUserRequestStatus2["IN_PROGRESS"] = "IN_PROGRESS";
  SparkUserRequestStatus2["SUCCEEDED"] = "SUCCEEDED";
  SparkUserRequestStatus2["FAILED"] = "FAILED";
  SparkUserRequestStatus2["CANCELED"] = "CANCELED";
  SparkUserRequestStatus2["UNKNOWN"] = "UNKNOWN";
  return SparkUserRequestStatus2;
})(SparkUserRequestStatus || {});
var SparkUserRequestStatus_default = SparkUserRequestStatus;

// src/graphql/objects/SparkUserRequestType.ts
var SparkUserRequestType = /* @__PURE__ */ ((SparkUserRequestType2) => {
  SparkUserRequestType2["FUTURE_VALUE"] = "FUTURE_VALUE";
  SparkUserRequestType2["LIGHTNING_SEND"] = "LIGHTNING_SEND";
  SparkUserRequestType2["LIGHTNING_RECEIVE"] = "LIGHTNING_RECEIVE";
  SparkUserRequestType2["COOP_EXIT"] = "COOP_EXIT";
  SparkUserRequestType2["LEAVES_SWAP"] = "LEAVES_SWAP";
  SparkUserRequestType2["CLAIM_STATIC_DEPOSIT"] = "CLAIM_STATIC_DEPOSIT";
  return SparkUserRequestType2;
})(SparkUserRequestType || {});
var SparkUserRequestType_default = SparkUserRequestType;

// src/graphql/objects/SparkWalletUser.ts
import { isObject as isObject7 } from "@lightsparkdev/core";

// src/graphql/objects/UserRequest.ts
import { LightsparkException, isObject as isObject6 } from "@lightsparkdev/core";
var UserRequestFromJson = (obj) => {
  if (obj["__typename"] == "ClaimStaticDeposit") {
    return {
      id: obj["claim_static_deposit_id"],
      createdAt: obj["claim_static_deposit_created_at"],
      updatedAt: obj["claim_static_deposit_updated_at"],
      network: BitcoinNetwork_default[obj["claim_static_deposit_network"]] ?? BitcoinNetwork_default.FUTURE_VALUE,
      creditAmount: CurrencyAmountFromJson(
        obj["claim_static_deposit_credit_amount"]
      ),
      maxFee: CurrencyAmountFromJson(obj["claim_static_deposit_max_fee"]),
      status: ClaimStaticDepositStatus_default[obj["claim_static_deposit_status"]] ?? ClaimStaticDepositStatus_default.FUTURE_VALUE,
      transactionId: obj["claim_static_deposit_transaction_id"],
      outputIndex: obj["claim_static_deposit_output_index"],
      bitcoinNetwork: BitcoinNetwork_default[obj["claim_static_deposit_bitcoin_network"]] ?? BitcoinNetwork_default.FUTURE_VALUE,
      typename: "ClaimStaticDeposit",
      transferSparkId: obj["claim_static_deposit_transfer_spark_id"]
    };
  }
  if (obj["__typename"] == "CoopExitRequest") {
    return {
      id: obj["coop_exit_request_id"],
      createdAt: obj["coop_exit_request_created_at"],
      updatedAt: obj["coop_exit_request_updated_at"],
      network: BitcoinNetwork_default[obj["coop_exit_request_network"]] ?? BitcoinNetwork_default.FUTURE_VALUE,
      fee: CurrencyAmountFromJson(obj["coop_exit_request_fee"]),
      l1BroadcastFee: CurrencyAmountFromJson(
        obj["coop_exit_request_l1_broadcast_fee"]
      ),
      status: SparkCoopExitRequestStatus_default[obj["coop_exit_request_status"]] ?? SparkCoopExitRequestStatus_default.FUTURE_VALUE,
      expiresAt: obj["coop_exit_request_expires_at"],
      rawConnectorTransaction: obj["coop_exit_request_raw_connector_transaction"],
      rawCoopExitTransaction: obj["coop_exit_request_raw_coop_exit_transaction"],
      coopExitTxid: obj["coop_exit_request_coop_exit_txid"],
      typename: "CoopExitRequest",
      feeQuoteId: obj["coop_exit_request_fee_quote"]?.id ?? void 0,
      exitSpeed: !!obj["coop_exit_request_exit_speed"] ? ExitSpeed_default[obj["coop_exit_request_exit_speed"]] ?? ExitSpeed_default.FUTURE_VALUE : null,
      transfer: !!obj["coop_exit_request_transfer"] ? TransferFromJson(obj["coop_exit_request_transfer"]) : void 0
    };
  }
  if (obj["__typename"] == "LeavesSwapRequest") {
    return {
      id: obj["leaves_swap_request_id"],
      createdAt: obj["leaves_swap_request_created_at"],
      updatedAt: obj["leaves_swap_request_updated_at"],
      network: BitcoinNetwork_default[obj["leaves_swap_request_network"]] ?? BitcoinNetwork_default.FUTURE_VALUE,
      status: SparkLeavesSwapRequestStatus_default[obj["leaves_swap_request_status"]] ?? SparkLeavesSwapRequestStatus_default.FUTURE_VALUE,
      totalAmount: CurrencyAmountFromJson(
        obj["leaves_swap_request_total_amount"]
      ),
      targetAmount: CurrencyAmountFromJson(
        obj["leaves_swap_request_target_amount"]
      ),
      fee: CurrencyAmountFromJson(obj["leaves_swap_request_fee"]),
      inboundTransfer: TransferFromJson(
        obj["leaves_swap_request_inbound_transfer"]
      ),
      swapLeaves: obj["leaves_swap_request_swap_leaves"].map(
        (e) => SwapLeafFromJson(e)
      ),
      typename: "LeavesSwapRequest",
      outboundTransfer: !!obj["leaves_swap_request_outbound_transfer"] ? TransferFromJson(obj["leaves_swap_request_outbound_transfer"]) : void 0,
      expiresAt: obj["leaves_swap_request_expires_at"]
    };
  }
  if (obj["__typename"] == "LightningReceiveRequest") {
    return {
      id: obj["lightning_receive_request_id"],
      createdAt: obj["lightning_receive_request_created_at"],
      updatedAt: obj["lightning_receive_request_updated_at"],
      network: BitcoinNetwork_default[obj["lightning_receive_request_network"]] ?? BitcoinNetwork_default.FUTURE_VALUE,
      invoice: InvoiceFromJson(obj["lightning_receive_request_invoice"]),
      status: LightningReceiveRequestStatus_default[obj["lightning_receive_request_status"]] ?? LightningReceiveRequestStatus_default.FUTURE_VALUE,
      typename: "LightningReceiveRequest",
      transfer: !!obj["lightning_receive_request_transfer"] ? TransferFromJson(obj["lightning_receive_request_transfer"]) : void 0,
      paymentPreimage: obj["lightning_receive_request_payment_preimage"],
      receiverIdentityPublicKey: obj["lightning_receive_request_receiver_identity_public_key"]
    };
  }
  if (obj["__typename"] == "LightningSendRequest") {
    return {
      id: obj["lightning_send_request_id"],
      createdAt: obj["lightning_send_request_created_at"],
      updatedAt: obj["lightning_send_request_updated_at"],
      network: BitcoinNetwork_default[obj["lightning_send_request_network"]] ?? BitcoinNetwork_default.FUTURE_VALUE,
      encodedInvoice: obj["lightning_send_request_encoded_invoice"],
      fee: CurrencyAmountFromJson(obj["lightning_send_request_fee"]),
      idempotencyKey: obj["lightning_send_request_idempotency_key"],
      status: LightningSendRequestStatus_default[obj["lightning_send_request_status"]] ?? LightningSendRequestStatus_default.FUTURE_VALUE,
      typename: "LightningSendRequest",
      transfer: !!obj["lightning_send_request_transfer"] ? TransferFromJson(obj["lightning_send_request_transfer"]) : void 0,
      paymentPreimage: obj["lightning_send_request_payment_preimage"]
    };
  }
  throw new LightsparkException(
    "DeserializationError",
    `Couldn't find a concrete type for interface UserRequest corresponding to the typename=${obj["__typename"]}`
  );
};
var FRAGMENT7 = `
fragment UserRequestFragment on UserRequest {
    __typename
    ... on ClaimStaticDeposit {
        __typename
        claim_static_deposit_id: id
        claim_static_deposit_created_at: created_at
        claim_static_deposit_updated_at: updated_at
        claim_static_deposit_network: network
        claim_static_deposit_credit_amount: credit_amount {
            __typename
            currency_amount_original_value: original_value
            currency_amount_original_unit: original_unit
        }
        claim_static_deposit_max_fee: max_fee {
            __typename
            currency_amount_original_value: original_value
            currency_amount_original_unit: original_unit
        }
        claim_static_deposit_status: status
        claim_static_deposit_transaction_id: transaction_id
        claim_static_deposit_output_index: output_index
        claim_static_deposit_bitcoin_network: bitcoin_network
        claim_static_deposit_transfer_spark_id: transfer_spark_id
    }
    ... on CoopExitRequest {
        __typename
        coop_exit_request_id: id
        coop_exit_request_created_at: created_at
        coop_exit_request_updated_at: updated_at
        coop_exit_request_network: network
        coop_exit_request_fee: fee {
            __typename
            currency_amount_original_value: original_value
            currency_amount_original_unit: original_unit
        }
        coop_exit_request_l1_broadcast_fee: l1_broadcast_fee {
            __typename
            currency_amount_original_value: original_value
            currency_amount_original_unit: original_unit
        }
        coop_exit_request_fee_quote: fee_quote {
            id
        }
        coop_exit_request_exit_speed: exit_speed
        coop_exit_request_status: status
        coop_exit_request_expires_at: expires_at
        coop_exit_request_raw_connector_transaction: raw_connector_transaction
        coop_exit_request_raw_coop_exit_transaction: raw_coop_exit_transaction
        coop_exit_request_coop_exit_txid: coop_exit_txid
        coop_exit_request_transfer: transfer {
            __typename
            transfer_total_amount: total_amount {
                __typename
                currency_amount_original_value: original_value
                currency_amount_original_unit: original_unit
            }
            transfer_spark_id: spark_id
            transfer_user_request: user_request {
                id
            }
        }
    }
    ... on LeavesSwapRequest {
        __typename
        leaves_swap_request_id: id
        leaves_swap_request_created_at: created_at
        leaves_swap_request_updated_at: updated_at
        leaves_swap_request_network: network
        leaves_swap_request_status: status
        leaves_swap_request_total_amount: total_amount {
            __typename
            currency_amount_original_value: original_value
            currency_amount_original_unit: original_unit
        }
        leaves_swap_request_target_amount: target_amount {
            __typename
            currency_amount_original_value: original_value
            currency_amount_original_unit: original_unit
        }
        leaves_swap_request_fee: fee {
            __typename
            currency_amount_original_value: original_value
            currency_amount_original_unit: original_unit
        }
        leaves_swap_request_inbound_transfer: inbound_transfer {
            __typename
            transfer_total_amount: total_amount {
                __typename
                currency_amount_original_value: original_value
                currency_amount_original_unit: original_unit
            }
            transfer_spark_id: spark_id
            transfer_user_request: user_request {
                id
            }
        }
        leaves_swap_request_outbound_transfer: outbound_transfer {
            __typename
            transfer_total_amount: total_amount {
                __typename
                currency_amount_original_value: original_value
                currency_amount_original_unit: original_unit
            }
            transfer_spark_id: spark_id
            transfer_user_request: user_request {
                id
            }
        }
        leaves_swap_request_expires_at: expires_at
        leaves_swap_request_swap_leaves: swap_leaves {
            __typename
            swap_leaf_leaf_id: leaf_id
            swap_leaf_raw_unsigned_refund_transaction: raw_unsigned_refund_transaction
            swap_leaf_adaptor_signed_signature: adaptor_signed_signature
            swap_leaf_direct_raw_unsigned_refund_transaction: direct_raw_unsigned_refund_transaction
            swap_leaf_direct_adaptor_signed_signature: direct_adaptor_signed_signature
            swap_leaf_direct_from_cpfp_raw_unsigned_refund_transaction: direct_from_cpfp_raw_unsigned_refund_transaction
            swap_leaf_direct_from_cpfp_adaptor_signed_signature: direct_from_cpfp_adaptor_signed_signature
        }
    }
    ... on LightningReceiveRequest {
        __typename
        lightning_receive_request_id: id
        lightning_receive_request_created_at: created_at
        lightning_receive_request_updated_at: updated_at
        lightning_receive_request_network: network
        lightning_receive_request_invoice: invoice {
            __typename
            invoice_encoded_invoice: encoded_invoice
            invoice_bitcoin_network: bitcoin_network
            invoice_payment_hash: payment_hash
            invoice_amount: amount {
                __typename
                currency_amount_original_value: original_value
                currency_amount_original_unit: original_unit
            }
            invoice_created_at: created_at
            invoice_expires_at: expires_at
            invoice_memo: memo
        }
        lightning_receive_request_status: status
        lightning_receive_request_transfer: transfer {
            __typename
            transfer_total_amount: total_amount {
                __typename
                currency_amount_original_value: original_value
                currency_amount_original_unit: original_unit
            }
            transfer_spark_id: spark_id
            transfer_user_request: user_request {
                id
            }
        }
        lightning_receive_request_payment_preimage: payment_preimage
        lightning_receive_request_receiver_identity_public_key: receiver_identity_public_key
    }
    ... on LightningSendRequest {
        __typename
        lightning_send_request_id: id
        lightning_send_request_created_at: created_at
        lightning_send_request_updated_at: updated_at
        lightning_send_request_network: network
        lightning_send_request_encoded_invoice: encoded_invoice
        lightning_send_request_fee: fee {
            __typename
            currency_amount_original_value: original_value
            currency_amount_original_unit: original_unit
        }
        lightning_send_request_idempotency_key: idempotency_key
        lightning_send_request_status: status
        lightning_send_request_transfer: transfer {
            __typename
            transfer_total_amount: total_amount {
                __typename
                currency_amount_original_value: original_value
                currency_amount_original_unit: original_unit
            }
            transfer_spark_id: spark_id
            transfer_user_request: user_request {
                id
            }
        }
        lightning_send_request_payment_preimage: payment_preimage
    }
}`;
var getUserRequestQuery = (id) => {
  return {
    queryPayload: `
query GetUserRequest($id: ID!) {
    entity(id: $id) {
        ... on UserRequest {
            ...UserRequestFragment
        }
    }
}

${FRAGMENT7}    
`,
    variables: { id },
    constructObject: (data) => isObject6(data) && "entity" in data && isObject6(data.entity) ? UserRequestFromJson(data.entity) : null
  };
};

// src/graphql/objects/SparkWalletUserToUserRequestsConnection.ts
var SparkWalletUserToUserRequestsConnectionFromJson = (obj) => {
  return {
    count: obj["spark_wallet_user_to_user_requests_connection_count"],
    pageInfo: PageInfoFromJson(
      obj["spark_wallet_user_to_user_requests_connection_page_info"]
    ),
    entities: obj["spark_wallet_user_to_user_requests_connection_entities"].map(
      (e) => UserRequestFromJson(e)
    ),
    typename: "SparkWalletUserToUserRequestsConnection"
  };
};

// src/graphql/objects/SparkWalletUser.ts
var SparkWalletUser = class {
  constructor(id, createdAt, updatedAt, identityPublicKey, typename) {
    this.id = id;
    this.createdAt = createdAt;
    this.updatedAt = updatedAt;
    this.identityPublicKey = identityPublicKey;
    this.typename = typename;
    autoBind(this);
  }
  async getUserRequests(client, first = void 0, after = void 0, types = void 0, statuses = void 0, networks = void 0) {
    return await client.executeRawQuery({
      queryPayload: ` 
query FetchSparkWalletUserToUserRequestsConnection($entity_id: ID!, $first: Int, $after: String, $types: [SparkUserRequestType!], $statuses: [SparkUserRequestStatus!], $networks: [BitcoinNetwork!]) {
    entity(id: $entity_id) {
        ... on SparkWalletUser {
            user_requests(, first: $first, after: $after, types: $types, statuses: $statuses, networks: $networks) {
                __typename
                spark_wallet_user_to_user_requests_connection_count: count
                spark_wallet_user_to_user_requests_connection_page_info: page_info {
                    __typename
                    page_info_has_next_page: has_next_page
                    page_info_has_previous_page: has_previous_page
                    page_info_start_cursor: start_cursor
                    page_info_end_cursor: end_cursor
                }
                spark_wallet_user_to_user_requests_connection_entities: entities {
                    __typename
                    ... on ClaimStaticDeposit {
                        __typename
                        claim_static_deposit_id: id
                        claim_static_deposit_created_at: created_at
                        claim_static_deposit_updated_at: updated_at
                        claim_static_deposit_network: network
                        claim_static_deposit_credit_amount: credit_amount {
                            __typename
                            currency_amount_original_value: original_value
                            currency_amount_original_unit: original_unit
                        }
                        claim_static_deposit_max_fee: max_fee {
                            __typename
                            currency_amount_original_value: original_value
                            currency_amount_original_unit: original_unit
                        }
                        claim_static_deposit_status: status
                        claim_static_deposit_transaction_id: transaction_id
                        claim_static_deposit_output_index: output_index
                        claim_static_deposit_bitcoin_network: bitcoin_network
                        claim_static_deposit_transfer_spark_id: transfer_spark_id
                    }
                    ... on CoopExitRequest {
                        __typename
                        coop_exit_request_id: id
                        coop_exit_request_created_at: created_at
                        coop_exit_request_updated_at: updated_at
                        coop_exit_request_network: network
                        coop_exit_request_fee: fee {
                            __typename
                            currency_amount_original_value: original_value
                            currency_amount_original_unit: original_unit
                        }
                        coop_exit_request_l1_broadcast_fee: l1_broadcast_fee {
                            __typename
                            currency_amount_original_value: original_value
                            currency_amount_original_unit: original_unit
                        }
                        coop_exit_request_fee_quote: fee_quote {
                            id
                        }
                        coop_exit_request_exit_speed: exit_speed
                        coop_exit_request_status: status
                        coop_exit_request_expires_at: expires_at
                        coop_exit_request_raw_connector_transaction: raw_connector_transaction
                        coop_exit_request_raw_coop_exit_transaction: raw_coop_exit_transaction
                        coop_exit_request_coop_exit_txid: coop_exit_txid
                        coop_exit_request_transfer: transfer {
                            __typename
                            transfer_total_amount: total_amount {
                                __typename
                                currency_amount_original_value: original_value
                                currency_amount_original_unit: original_unit
                            }
                            transfer_spark_id: spark_id
                            transfer_user_request: user_request {
                                id
                            }
                        }
                    }
                    ... on LeavesSwapRequest {
                        __typename
                        leaves_swap_request_id: id
                        leaves_swap_request_created_at: created_at
                        leaves_swap_request_updated_at: updated_at
                        leaves_swap_request_network: network
                        leaves_swap_request_status: status
                        leaves_swap_request_total_amount: total_amount {
                            __typename
                            currency_amount_original_value: original_value
                            currency_amount_original_unit: original_unit
                        }
                        leaves_swap_request_target_amount: target_amount {
                            __typename
                            currency_amount_original_value: original_value
                            currency_amount_original_unit: original_unit
                        }
                        leaves_swap_request_fee: fee {
                            __typename
                            currency_amount_original_value: original_value
                            currency_amount_original_unit: original_unit
                        }
                        leaves_swap_request_inbound_transfer: inbound_transfer {
                            __typename
                            transfer_total_amount: total_amount {
                                __typename
                                currency_amount_original_value: original_value
                                currency_amount_original_unit: original_unit
                            }
                            transfer_spark_id: spark_id
                            transfer_user_request: user_request {
                                id
                            }
                        }
                        leaves_swap_request_outbound_transfer: outbound_transfer {
                            __typename
                            transfer_total_amount: total_amount {
                                __typename
                                currency_amount_original_value: original_value
                                currency_amount_original_unit: original_unit
                            }
                            transfer_spark_id: spark_id
                            transfer_user_request: user_request {
                                id
                            }
                        }
                        leaves_swap_request_expires_at: expires_at
                        leaves_swap_request_swap_leaves: swap_leaves {
                            __typename
                            swap_leaf_leaf_id: leaf_id
                            swap_leaf_raw_unsigned_refund_transaction: raw_unsigned_refund_transaction
                            swap_leaf_adaptor_signed_signature: adaptor_signed_signature
                        }
                    }
                    ... on LightningReceiveRequest {
                        __typename
                        lightning_receive_request_id: id
                        lightning_receive_request_created_at: created_at
                        lightning_receive_request_updated_at: updated_at
                        lightning_receive_request_network: network
                        lightning_receive_request_invoice: invoice {
                            __typename
                            invoice_encoded_invoice: encoded_invoice
                            invoice_bitcoin_network: bitcoin_network
                            invoice_payment_hash: payment_hash
                            invoice_amount: amount {
                                __typename
                                currency_amount_original_value: original_value
                                currency_amount_original_unit: original_unit
                            }
                            invoice_created_at: created_at
                            invoice_expires_at: expires_at
                            invoice_memo: memo
                        }
                        lightning_receive_request_status: status
                        lightning_receive_request_transfer: transfer {
                            __typename
                            transfer_total_amount: total_amount {
                                __typename
                                currency_amount_original_value: original_value
                                currency_amount_original_unit: original_unit
                            }
                            transfer_spark_id: spark_id
                            transfer_user_request: user_request {
                                id
                            }
                        }
                        lightning_receive_request_payment_preimage: payment_preimage
                        lightning_receive_request_receiver_identity_public_key: receiver_identity_public_key
                    }
                    ... on LightningSendRequest {
                        __typename
                        lightning_send_request_id: id
                        lightning_send_request_created_at: created_at
                        lightning_send_request_updated_at: updated_at
                        lightning_send_request_network: network
                        lightning_send_request_encoded_invoice: encoded_invoice
                        lightning_send_request_fee: fee {
                            __typename
                            currency_amount_original_value: original_value
                            currency_amount_original_unit: original_unit
                        }
                        lightning_send_request_idempotency_key: idempotency_key
                        lightning_send_request_status: status
                        lightning_send_request_transfer: transfer {
                            __typename
                            transfer_total_amount: total_amount {
                                __typename
                                currency_amount_original_value: original_value
                                currency_amount_original_unit: original_unit
                            }
                            transfer_spark_id: spark_id
                            transfer_user_request: user_request {
                                id
                            }
                        }
                        lightning_send_request_payment_preimage: payment_preimage
                    }
                }
            }
        }
    }
}
`,
      variables: {
        entity_id: this.id,
        first,
        after,
        types,
        statuses,
        networks
      },
      constructObject: (json) => {
        const connection = json["entity"]["user_requests"];
        return SparkWalletUserToUserRequestsConnectionFromJson(connection);
      }
    });
  }
  static getSparkWalletUserQuery(id) {
    return {
      queryPayload: `
query GetSparkWalletUser($id: ID!) {
    entity(id: $id) {
        ... on SparkWalletUser {
            ...SparkWalletUserFragment
        }
    }
}

${FRAGMENT8}    
`,
      variables: { id },
      constructObject: (data) => isObject7(data) && "entity" in data && isObject7(data.entity) ? SparkWalletUserFromJson(data.entity) : null
    };
  }
  toJson() {
    return {
      __typename: "SparkWalletUser",
      spark_wallet_user_id: this.id,
      spark_wallet_user_created_at: this.createdAt,
      spark_wallet_user_updated_at: this.updatedAt,
      spark_wallet_user_identity_public_key: this.identityPublicKey
    };
  }
};
var SparkWalletUserFromJson = (obj) => {
  return new SparkWalletUser(
    obj["spark_wallet_user_id"],
    obj["spark_wallet_user_created_at"],
    obj["spark_wallet_user_updated_at"],
    obj["spark_wallet_user_identity_public_key"],
    "SparkWalletUser"
  );
};
var FRAGMENT8 = `
fragment SparkWalletUserFragment on SparkWalletUser {
    __typename
    spark_wallet_user_id: id
    spark_wallet_user_created_at: created_at
    spark_wallet_user_updated_at: updated_at
    spark_wallet_user_identity_public_key: identity_public_key
}`;
var SparkWalletUser_default = SparkWalletUser;

export {
  BitcoinNetwork_default,
  ClaimStaticDepositStatus_default,
  CurrencyUnit_default,
  CurrencyAmountFromJson,
  ClaimStaticDepositFromJson,
  getClaimStaticDepositQuery,
  ClaimStaticDepositRequestType_default,
  CoopExitFeeQuoteFromJson,
  FRAGMENT2 as FRAGMENT,
  getCoopExitFeeQuoteQuery,
  ExitSpeed_default,
  SparkLeavesSwapRequestStatus_default,
  TransferFromJson,
  FRAGMENT3 as FRAGMENT2,
  Transfer_default,
  LeavesSwapRequestFromJson,
  FRAGMENT4 as FRAGMENT3,
  getLeavesSwapRequestQuery,
  LightningReceiveRequestStatus_default,
  LightningReceiveRequestFromJson,
  FRAGMENT5 as FRAGMENT4,
  getLightningReceiveRequestQuery,
  LightningSendRequestStatus_default,
  LightningSendRequestFromJson,
  FRAGMENT6 as FRAGMENT5,
  getLightningSendRequestQuery,
  SparkCoopExitRequestStatus_default,
  SparkUserRequestStatus_default,
  SparkUserRequestType_default,
  FRAGMENT7 as FRAGMENT6,
  getUserRequestQuery,
  SparkWalletUserToUserRequestsConnectionFromJson,
  SparkWalletUser_default
};
