// @buildonspark/spark-sdk/dist/chunk-EU3I7GFB.js
import {
  Network,
  TransferStatus,
  TransferType
} from "./chunk-JE3MXMPW.js";

// src/types/sdk-types.ts
import { bytesToHex } from "@noble/curves/utils";
function mapSettingsProtoToWalletSettings(proto) {
  return {
    ownerIdentityPublicKey: bytesToHex(proto.ownerIdentityPublicKey),
    privateEnabled: proto.privateEnabled
  };
}
function mapTreeNodeToWalletLeaf(proto) {
  return {
    id: proto.id,
    treeId: proto.treeId,
    value: proto.value,
    parentNodeId: proto.parentNodeId,
    nodeTx: bytesToHex(proto.nodeTx),
    refundTx: bytesToHex(proto.refundTx),
    vout: proto.vout,
    verifyingPublicKey: bytesToHex(proto.verifyingPublicKey),
    ownerIdentityPublicKey: bytesToHex(proto.ownerIdentityPublicKey),
    signingKeyshare: proto.signingKeyshare,
    status: proto.status,
    network: Network[proto.network]
  };
}
var TransferDirection = /* @__PURE__ */ ((TransferDirection2) => {
  TransferDirection2["INCOMING"] = "INCOMING";
  TransferDirection2["OUTGOING"] = "OUTGOING";
  return TransferDirection2;
})(TransferDirection || {});
function mapTransferLeafToWalletTransferLeaf(proto) {
  return {
    leaf: proto.leaf ? mapTreeNodeToWalletLeaf(proto.leaf) : void 0,
    secretCipher: bytesToHex(proto.secretCipher),
    signature: bytesToHex(proto.signature),
    intermediateRefundTx: bytesToHex(proto.intermediateRefundTx)
  };
}
function mapTransferToWalletTransfer(proto, identityPublicKey, userRequest) {
  const receiverIdentityPublicKey = bytesToHex(proto.receiverIdentityPublicKey);
  const senderIdentityPublicKey = bytesToHex(proto.senderIdentityPublicKey);
  return {
    id: proto.id,
    senderIdentityPublicKey,
    receiverIdentityPublicKey,
    status: TransferStatus[proto.status],
    totalValue: proto.totalValue,
    expiryTime: proto.expiryTime ? new Date(proto.expiryTime) : void 0,
    leaves: proto.leaves.map(mapTransferLeafToWalletTransferLeaf),
    createdTime: proto.createdTime ? new Date(proto.createdTime) : void 0,
    updatedTime: proto.updatedTime ? new Date(proto.updatedTime) : void 0,
    type: TransferType[proto.type],
    transferDirection: receiverIdentityPublicKey === identityPublicKey ? "INCOMING" /* INCOMING */ : "OUTGOING" /* OUTGOING */,
    userRequest
  };
}

export {
  mapSettingsProtoToWalletSettings,
  mapTreeNodeToWalletLeaf,
  TransferDirection,
  mapTransferLeafToWalletTransferLeaf,
  mapTransferToWalletTransfer
};
