// @buildonspark/spark-sdk/dist/index.node.js
import {
  AuthenticationError,
  ConfigurationError,
  ConnectionManagerNodeJS,
  DEFAULT_FEE_SATS,
  DIRECT_HTLC_TIMELOCK_OFFSET,
  DIRECT_TIMELOCK_OFFSET,
  DefaultSparkSigner,
  HTLC_TIMELOCK_OFFSET,
  INITIAL_SEQUENCE,
  InternalValidationError,
  LOGGER_NAMES,
  Network,
  NetworkError,
  NetworkToProto,
  RPCError,
  SparkAuthenticationError,
  SparkError,
  SparkRequestError,
  SparkSdkError,
  SparkSdkLogger,
  SparkValidationError,
  SparkWalletEvent,
  SparkWalletNodeJS,
  TEST_UNILATERAL_DIRECT_SEQUENCE,
  TEST_UNILATERAL_SEQUENCE,
  TIME_LOCK_INTERVAL,
  TokenTransactionService,
  UnsafeStatelessSparkSigner,
  ValidationError,
  WalletConfig,
  addPrivateKeys,
  addPublicKeys,
  applyAdaptorToSignature,
  applyAdditiveTweakToPublicKey,
  assertBech32,
  bech32mDecode,
  bigIntToPrivateKey,
  checkIfSelectedOutputsAreAvailable,
  checkIfValidSequence,
  collectResponses,
  computeTaprootKeyNoScript,
  computerLagrangeCoefficients,
  constructFeeBumpTx,
  constructUnilateralExitFeeBumpPackages,
  constructUnilateralExitTxs,
  createConnectorRefundTxs,
  createCurrentTimelockRefundTxs,
  createDecrementedTimelockNodeTx,
  createDecrementedTimelockRefundTxs,
  createInitialTimelockNodeTx,
  createInitialTimelockRefundTxs,
  createRootNodeTx,
  createSigningCommitment,
  createSigningNonce,
  createTestUnilateralRefundTxs,
  createTestUnilateralTimelockNodeTx,
  createZeroTimelockNodeTx,
  decodeBech32mTokenIdentifier,
  decodeBytesToSigningCommitment,
  decodeBytesToSigningNonce,
  decodeSparkAddress,
  doesLeafNeedRefresh,
  doesTxnNeedRenewed,
  encodeBech32mTokenIdentifier,
  encodeSigningCommitmentToBytes,
  encodeSigningNonceToBytes,
  encodeSparkAddress,
  encodeSparkAddressWithSignature,
  evaluatePolynomial,
  fieldDiv,
  filterTokenBalanceForTokenIdentifier,
  generateAdaptorFromSignature,
  generatePolynomialForSecretSharing,
  generateSignatureFromExistingAdaptor,
  getCurrentTimelock,
  getEphemeralAnchorOutput,
  getLatestDepositTxId,
  getNetwork,
  getNetworkFromAddress,
  getNetworkFromBech32mTokenIdentifier,
  getNetworkFromSparkAddress,
  getNetworkFromString,
  getNextHTLCTransactionSequence,
  getNextTransactionSequence,
  getP2TRAddressFromPkScript,
  getP2TRAddressFromPublicKey,
  getP2TRScriptFromPublicKey,
  getP2WPKHAddressFromPublicKey,
  getRandomBigInt,
  getRandomSigningNonce,
  getSigHashFromTx,
  getSigningCommitmentFromNonce,
  getSparkAddressFromTaproot,
  getSparkFrost,
  getTransactionSequence,
  getTransferPackageSigningPayload,
  getTxEstimatedVbytesSizeByNumberOfInputsOutputs,
  getTxFromRawTxBytes,
  getTxFromRawTxHex,
  getTxId,
  getTxIdNoReverse,
  hash160,
  initializeTracerEnvNodeJS,
  isEphemeralAnchorOutput,
  isLegacySparkAddress,
  isSafeForNumber,
  isTxBroadcast,
  isValidPublicKey,
  isValidSparkAddress,
  isZeroTimelock,
  lastKeyWithTarget,
  maybeApplyFee,
  modInverse,
  proofOfPossessionMessageHashForDepositAddress,
  protoToNetwork,
  recoverSecret,
  splitSecret,
  splitSecretWithProofs,
  subtractPrivateKeys,
  subtractPublicKeys,
  sumAvailableTokens,
  sumOfPrivateKeys,
  toProtoTimestamp,
  validateOutboundAdaptorSignature,
  validateShare,
  validateSparkInvoiceFields,
  validateSparkInvoiceSignature
} from "./chunk-RU434ZAE.js";
import "./chunk-ZP6Z6DFX.js";
import "./chunk-EU3I7GFB.js";
import "./chunk-JE3MXMPW.js";
import "./chunk-UYEB2VPG.js";
import "./chunk-MVRQ5US7.js";
export {
  AuthenticationError,
  ConfigurationError,
  ConnectionManagerNodeJS as ConnectionManager,
  DEFAULT_FEE_SATS,
  DIRECT_HTLC_TIMELOCK_OFFSET,
  DIRECT_TIMELOCK_OFFSET,
  DefaultSparkSigner,
  HTLC_TIMELOCK_OFFSET,
  INITIAL_SEQUENCE,
  InternalValidationError,
  LOGGER_NAMES,
  Network,
  NetworkError,
  NetworkToProto,
  RPCError,
  SparkAuthenticationError,
  SparkError,
  SparkRequestError,
  SparkSdkError,
  SparkSdkLogger,
  SparkValidationError,
  SparkWalletNodeJS as SparkWallet,
  SparkWalletEvent,
  TEST_UNILATERAL_DIRECT_SEQUENCE,
  TEST_UNILATERAL_SEQUENCE,
  TIME_LOCK_INTERVAL,
  TokenTransactionService,
  UnsafeStatelessSparkSigner,
  ValidationError,
  WalletConfig,
  addPrivateKeys,
  addPublicKeys,
  applyAdaptorToSignature,
  applyAdditiveTweakToPublicKey,
  assertBech32,
  bech32mDecode,
  bigIntToPrivateKey,
  checkIfSelectedOutputsAreAvailable,
  checkIfValidSequence,
  collectResponses,
  computeTaprootKeyNoScript,
  computerLagrangeCoefficients,
  constructFeeBumpTx,
  constructUnilateralExitFeeBumpPackages,
  constructUnilateralExitTxs,
  createConnectorRefundTxs,
  createCurrentTimelockRefundTxs,
  createDecrementedTimelockNodeTx,
  createDecrementedTimelockRefundTxs,
  createInitialTimelockNodeTx,
  createInitialTimelockRefundTxs,
  createRootNodeTx,
  createSigningCommitment,
  createSigningNonce,
  createTestUnilateralRefundTxs,
  createTestUnilateralTimelockNodeTx,
  createZeroTimelockNodeTx,
  decodeBech32mTokenIdentifier,
  decodeBytesToSigningCommitment,
  decodeBytesToSigningNonce,
  decodeSparkAddress,
  doesLeafNeedRefresh,
  doesTxnNeedRenewed,
  encodeBech32mTokenIdentifier,
  encodeSigningCommitmentToBytes,
  encodeSigningNonceToBytes,
  encodeSparkAddress,
  encodeSparkAddressWithSignature,
  evaluatePolynomial,
  fieldDiv,
  filterTokenBalanceForTokenIdentifier,
  generateAdaptorFromSignature,
  generatePolynomialForSecretSharing,
  generateSignatureFromExistingAdaptor,
  getCurrentTimelock,
  getEphemeralAnchorOutput,
  getLatestDepositTxId,
  getNetwork,
  getNetworkFromAddress,
  getNetworkFromBech32mTokenIdentifier,
  getNetworkFromSparkAddress,
  getNetworkFromString,
  getNextHTLCTransactionSequence,
  getNextTransactionSequence,
  getP2TRAddressFromPkScript,
  getP2TRAddressFromPublicKey,
  getP2TRScriptFromPublicKey,
  getP2WPKHAddressFromPublicKey,
  getRandomBigInt,
  getRandomSigningNonce,
  getSigHashFromTx,
  getSigningCommitmentFromNonce,
  getSparkAddressFromTaproot,
  getSparkFrost,
  getTransactionSequence,
  getTransferPackageSigningPayload,
  getTxEstimatedVbytesSizeByNumberOfInputsOutputs,
  getTxFromRawTxBytes,
  getTxFromRawTxHex,
  getTxId,
  getTxIdNoReverse,
  hash160,
  initializeTracerEnvNodeJS as initializeTracerEnv,
  isEphemeralAnchorOutput,
  isLegacySparkAddress,
  isSafeForNumber,
  isTxBroadcast,
  isValidPublicKey,
  isValidSparkAddress,
  isZeroTimelock,
  lastKeyWithTarget,
  maybeApplyFee,
  modInverse,
  proofOfPossessionMessageHashForDepositAddress,
  protoToNetwork,
  recoverSecret,
  splitSecret,
  splitSecretWithProofs,
  subtractPrivateKeys,
  subtractPublicKeys,
  sumAvailableTokens,
  sumOfPrivateKeys,
  toProtoTimestamp,
  validateOutboundAdaptorSignature,
  validateShare,
  validateSparkInvoiceFields,
  validateSparkInvoiceSignature
};
