// @buildonspark/spark-sdk/dist/sdk-types-Dn_8aH9E.js
import { Network, TransferStatus, TransferType } from "./proto/spark.js";
import { bytesToHex } from "@noble/curves/utils";
import { LightsparkException, isObject } from "@lightsparkdev/core";

//#region src/graphql/objects/BitcoinNetwork.ts
/** This is an enum identifying a particular Bitcoin Network. **/
let BitcoinNetwork = /* @__PURE__ */ function(BitcoinNetwork) {
	/**
	* This is an enum value that represents values that could be added in the future.
	* Clients should support unknown values as more of them could be added without notice.
	*/
	BitcoinNetwork["FUTURE_VALUE"] = "FUTURE_VALUE";
	/** The production version of the Bitcoin Blockchain. **/
	BitcoinNetwork["MAINNET"] = "MAINNET";
	/** A test version of the Bitcoin Blockchain, maintained by Lightspark. **/
	BitcoinNetwork["REGTEST"] = "REGTEST";
	/** A test version of the Bitcoin Blockchain, maintained by a centralized organization. Not in use at Lightspark. **/
	BitcoinNetwork["SIGNET"] = "SIGNET";
	/** A test version of the Bitcoin Blockchain, publicly available. **/
	BitcoinNetwork["TESTNET"] = "TESTNET";
	return BitcoinNetwork;
}({});

//#endregion
//#region src/graphql/objects/ClaimStaticDepositStatus.ts
let ClaimStaticDepositStatus = /* @__PURE__ */ function(ClaimStaticDepositStatus) {
	/**
	* This is an enum value that represents values that could be added in the future.
	* Clients should support unknown values as more of them could be added without notice.
	*/
	ClaimStaticDepositStatus["FUTURE_VALUE"] = "FUTURE_VALUE";
	ClaimStaticDepositStatus["CREATED"] = "CREATED";
	ClaimStaticDepositStatus["TRANSFER_CREATED"] = "TRANSFER_CREATED";
	ClaimStaticDepositStatus["TRANSFER_CREATION_FAILED"] = "TRANSFER_CREATION_FAILED";
	ClaimStaticDepositStatus["REFUND_SIGNING_COMMITMENTS_QUERYING_FAILED"] = "REFUND_SIGNING_COMMITMENTS_QUERYING_FAILED";
	ClaimStaticDepositStatus["REFUND_SIGNING_FAILED"] = "REFUND_SIGNING_FAILED";
	ClaimStaticDepositStatus["UTXO_SWAPPING_FAILED"] = "UTXO_SWAPPING_FAILED";
	ClaimStaticDepositStatus["TRANSFER_COMPLETED"] = "TRANSFER_COMPLETED";
	ClaimStaticDepositStatus["SPEND_TX_CREATED"] = "SPEND_TX_CREATED";
	ClaimStaticDepositStatus["SPEND_TX_BROADCAST"] = "SPEND_TX_BROADCAST";
	return ClaimStaticDepositStatus;
}({});

//#endregion
//#region src/graphql/objects/CurrencyUnit.ts
/** This enum identifies the unit of currency associated with a CurrencyAmount. **/
let CurrencyUnit$1 = /* @__PURE__ */ function(CurrencyUnit) {
	/**
	* This is an enum value that represents values that could be added in the future.
	* Clients should support unknown values as more of them could be added without notice.
	*/
	CurrencyUnit["FUTURE_VALUE"] = "FUTURE_VALUE";
	/** Bitcoin is the cryptocurrency native to the Bitcoin network. It is used as the native medium for value transfer for the Lightning Network. **/
	CurrencyUnit["BITCOIN"] = "BITCOIN";
	/** 0.00000001 (10e-8) Bitcoin or one hundred millionth of a Bitcoin. This is the unit most commonly used in Lightning transactions. **/
	CurrencyUnit["SATOSHI"] = "SATOSHI";
	/** 0.001 Satoshi, or 10e-11 Bitcoin. We recommend using the Satoshi unit instead when possible. **/
	CurrencyUnit["MILLISATOSHI"] = "MILLISATOSHI";
	/** United States Dollar. **/
	CurrencyUnit["USD"] = "USD";
	/** Mexican Peso. **/
	CurrencyUnit["MXN"] = "MXN";
	/** Philippine Peso. **/
	CurrencyUnit["PHP"] = "PHP";
	/** Euro. **/
	CurrencyUnit["EUR"] = "EUR";
	/** 0.000000001 (10e-9) Bitcoin or a billionth of a Bitcoin. We recommend using the Satoshi unit instead when possible. **/
	CurrencyUnit["NANOBITCOIN"] = "NANOBITCOIN";
	/** 0.000001 (10e-6) Bitcoin or a millionth of a Bitcoin. We recommend using the Satoshi unit instead when possible. **/
	CurrencyUnit["MICROBITCOIN"] = "MICROBITCOIN";
	/** 0.001 (10e-3) Bitcoin or a thousandth of a Bitcoin. We recommend using the Satoshi unit instead when possible. **/
	CurrencyUnit["MILLIBITCOIN"] = "MILLIBITCOIN";
	return CurrencyUnit;
}({});

//#endregion
//#region src/graphql/objects/CurrencyAmount.ts
const CurrencyAmountFromJson = (obj) => {
	return {
		originalValue: obj["currency_amount_original_value"],
		originalUnit: CurrencyUnit$1[obj["currency_amount_original_unit"]] ?? CurrencyUnit$1.FUTURE_VALUE
	};
};
const CurrencyAmountToJson = (obj) => {
	return {
		currency_amount_original_value: obj.originalValue,
		currency_amount_original_unit: obj.originalUnit
	};
};

//#endregion
//#region src/graphql/objects/ClaimStaticDeposit.ts
const ClaimStaticDepositFromJson = (obj) => {
	return {
		id: obj["claim_static_deposit_id"],
		createdAt: obj["claim_static_deposit_created_at"],
		updatedAt: obj["claim_static_deposit_updated_at"],
		network: BitcoinNetwork[obj["claim_static_deposit_network"]] ?? BitcoinNetwork.FUTURE_VALUE,
		creditAmount: CurrencyAmountFromJson(obj["claim_static_deposit_credit_amount"]),
		maxFee: CurrencyAmountFromJson(obj["claim_static_deposit_max_fee"]),
		status: ClaimStaticDepositStatus[obj["claim_static_deposit_status"]] ?? ClaimStaticDepositStatus.FUTURE_VALUE,
		transactionId: obj["claim_static_deposit_transaction_id"],
		outputIndex: obj["claim_static_deposit_output_index"],
		bitcoinNetwork: BitcoinNetwork[obj["claim_static_deposit_bitcoin_network"]] ?? BitcoinNetwork.FUTURE_VALUE,
		typename: "ClaimStaticDeposit",
		transferSparkId: obj["claim_static_deposit_transfer_spark_id"]
	};
};
const FRAGMENT$7 = `
fragment ClaimStaticDepositFragment on ClaimStaticDeposit {
    __typename
    claim_static_deposit_id: id
    claim_static_deposit_created_at: created_at
    claim_static_deposit_updated_at: updated_at
    claim_static_deposit_network: network
    claim_static_deposit_credit_amount: credit_amount {
        __typename
        currency_amount_original_value: original_value
        currency_amount_original_unit: original_unit
    }
    claim_static_deposit_max_fee: max_fee {
        __typename
        currency_amount_original_value: original_value
        currency_amount_original_unit: original_unit
    }
    claim_static_deposit_status: status
    claim_static_deposit_transaction_id: transaction_id
    claim_static_deposit_output_index: output_index
    claim_static_deposit_bitcoin_network: bitcoin_network
    claim_static_deposit_transfer_spark_id: transfer_spark_id
}`;
const getClaimStaticDepositQuery = (id) => {
	return {
		queryPayload: `
query GetClaimStaticDeposit($id: ID!) {
    entity(id: $id) {
        ... on ClaimStaticDeposit {
            ...ClaimStaticDepositFragment
        }
    }
}

${FRAGMENT$7}    
`,
		variables: { id },
		constructObject: (data) => isObject(data) && "entity" in data && isObject(data.entity) ? ClaimStaticDepositFromJson(data.entity) : null
	};
};

//#endregion
//#region src/graphql/objects/ClaimStaticDepositRequestType.ts
let ClaimStaticDepositRequestType = /* @__PURE__ */ function(ClaimStaticDepositRequestType) {
	/**
	* This is an enum value that represents values that could be added in the future.
	* Clients should support unknown values as more of them could be added without notice.
	*/
	ClaimStaticDepositRequestType["FUTURE_VALUE"] = "FUTURE_VALUE";
	ClaimStaticDepositRequestType["FIXED_AMOUNT"] = "FIXED_AMOUNT";
	ClaimStaticDepositRequestType["MAX_FEE"] = "MAX_FEE";
	return ClaimStaticDepositRequestType;
}({});

//#endregion
//#region src/graphql/objects/CoopExitFeeQuote.ts
const CoopExitFeeQuoteFromJson = (obj) => {
	return {
		id: obj["coop_exit_fee_quote_id"],
		createdAt: obj["coop_exit_fee_quote_created_at"],
		updatedAt: obj["coop_exit_fee_quote_updated_at"],
		network: BitcoinNetwork[obj["coop_exit_fee_quote_network"]] ?? BitcoinNetwork.FUTURE_VALUE,
		totalAmount: CurrencyAmountFromJson(obj["coop_exit_fee_quote_total_amount"]),
		userFeeFast: CurrencyAmountFromJson(obj["coop_exit_fee_quote_user_fee_fast"]),
		userFeeMedium: CurrencyAmountFromJson(obj["coop_exit_fee_quote_user_fee_medium"]),
		userFeeSlow: CurrencyAmountFromJson(obj["coop_exit_fee_quote_user_fee_slow"]),
		l1BroadcastFeeFast: CurrencyAmountFromJson(obj["coop_exit_fee_quote_l1_broadcast_fee_fast"]),
		l1BroadcastFeeMedium: CurrencyAmountFromJson(obj["coop_exit_fee_quote_l1_broadcast_fee_medium"]),
		l1BroadcastFeeSlow: CurrencyAmountFromJson(obj["coop_exit_fee_quote_l1_broadcast_fee_slow"]),
		expiresAt: obj["coop_exit_fee_quote_expires_at"],
		typename: "CoopExitFeeQuote"
	};
};
const FRAGMENT$6 = `
fragment CoopExitFeeQuoteFragment on CoopExitFeeQuote {
    __typename
    coop_exit_fee_quote_id: id
    coop_exit_fee_quote_created_at: created_at
    coop_exit_fee_quote_updated_at: updated_at
    coop_exit_fee_quote_network: network
    coop_exit_fee_quote_total_amount: total_amount {
        __typename
        currency_amount_original_value: original_value
        currency_amount_original_unit: original_unit
    }
    coop_exit_fee_quote_user_fee_fast: user_fee_fast {
        __typename
        currency_amount_original_value: original_value
        currency_amount_original_unit: original_unit
    }
    coop_exit_fee_quote_user_fee_medium: user_fee_medium {
        __typename
        currency_amount_original_value: original_value
        currency_amount_original_unit: original_unit
    }
    coop_exit_fee_quote_user_fee_slow: user_fee_slow {
        __typename
        currency_amount_original_value: original_value
        currency_amount_original_unit: original_unit
    }
    coop_exit_fee_quote_l1_broadcast_fee_fast: l1_broadcast_fee_fast {
        __typename
        currency_amount_original_value: original_value
        currency_amount_original_unit: original_unit
    }
    coop_exit_fee_quote_l1_broadcast_fee_medium: l1_broadcast_fee_medium {
        __typename
        currency_amount_original_value: original_value
        currency_amount_original_unit: original_unit
    }
    coop_exit_fee_quote_l1_broadcast_fee_slow: l1_broadcast_fee_slow {
        __typename
        currency_amount_original_value: original_value
        currency_amount_original_unit: original_unit
    }
    coop_exit_fee_quote_expires_at: expires_at
}`;
const getCoopExitFeeQuoteQuery = (id) => {
	return {
		queryPayload: `
query GetCoopExitFeeQuote($id: ID!) {
    entity(id: $id) {
        ... on CoopExitFeeQuote {
            ...CoopExitFeeQuoteFragment
        }
    }
}

${FRAGMENT$6}    
`,
		variables: { id },
		constructObject: (data) => isObject(data) && "entity" in data && isObject(data.entity) ? CoopExitFeeQuoteFromJson(data.entity) : null
	};
};

//#endregion
//#region src/graphql/objects/ExitSpeed.ts
let ExitSpeed = /* @__PURE__ */ function(ExitSpeed) {
	/**
	* This is an enum value that represents values that could be added in the future.
	* Clients should support unknown values as more of them could be added without notice.
	*/
	ExitSpeed["FUTURE_VALUE"] = "FUTURE_VALUE";
	ExitSpeed["FAST"] = "FAST";
	ExitSpeed["MEDIUM"] = "MEDIUM";
	ExitSpeed["SLOW"] = "SLOW";
	return ExitSpeed;
}({});

//#endregion
//#region src/graphql/objects/SparkLeavesSwapRequestStatus.ts
let SparkLeavesSwapRequestStatus = /* @__PURE__ */ function(SparkLeavesSwapRequestStatus) {
	/**
	* This is an enum value that represents values that could be added in the future.
	* Clients should support unknown values as more of them could be added without notice.
	*/
	SparkLeavesSwapRequestStatus["FUTURE_VALUE"] = "FUTURE_VALUE";
	SparkLeavesSwapRequestStatus["CREATED"] = "CREATED";
	SparkLeavesSwapRequestStatus["INITIATED"] = "INITIATED";
	SparkLeavesSwapRequestStatus["LEAVES_LOCKED"] = "LEAVES_LOCKED";
	SparkLeavesSwapRequestStatus["REFUND_TX_ADAPTOR_SIGNED"] = "REFUND_TX_ADAPTOR_SIGNED";
	SparkLeavesSwapRequestStatus["INBOUND_TRANSFER_CLAIMED"] = "INBOUND_TRANSFER_CLAIMED";
	SparkLeavesSwapRequestStatus["SUCCEEDED"] = "SUCCEEDED";
	SparkLeavesSwapRequestStatus["EXPIRED"] = "EXPIRED";
	SparkLeavesSwapRequestStatus["FAILED"] = "FAILED";
	return SparkLeavesSwapRequestStatus;
}({});

//#endregion
//#region src/graphql/objects/SwapLeaf.ts
const SwapLeafFromJson = (obj) => {
	return {
		leafId: obj["swap_leaf_leaf_id"],
		rawUnsignedRefundTransaction: obj["swap_leaf_raw_unsigned_refund_transaction"],
		adaptorSignedSignature: obj["swap_leaf_adaptor_signed_signature"],
		directRawUnsignedRefundTransaction: obj["swap_leaf_direct_raw_unsigned_refund_transaction"],
		directAdaptorSignedSignature: obj["swap_leaf_direct_adaptor_signed_signature"],
		directFromCpfpRawUnsignedRefundTransaction: obj["swap_leaf_direct_from_cpfp_raw_unsigned_refund_transaction"],
		directFromCpfpAdaptorSignedSignature: obj["swap_leaf_direct_from_cpfp_adaptor_signed_signature"]
	};
};

//#endregion
//#region src/auto-bind.ts
function getAllProperties(object) {
	const properties = /* @__PURE__ */ new Set();
	let current = object;
	while (current && current !== Object.prototype) {
		for (const key of Reflect.ownKeys(current)) properties.add([current, key]);
		current = Reflect.getPrototypeOf(current);
	}
	return properties;
}
function autoBind(self, { include, exclude } = {}) {
	const filter = (key) => {
		const keyStr = typeof key === "string" ? key : key.toString();
		const match = (pattern) => typeof pattern === "string" ? keyStr === pattern : pattern.test(keyStr);
		if (include) return include.some(match);
		if (exclude) return !exclude.some(match);
		return true;
	};
	for (const [proto, key] of getAllProperties(self.constructor.prototype)) {
		if (key === "constructor" || !filter(key)) continue;
		const descriptor = Reflect.getOwnPropertyDescriptor(proto, key);
		if (descriptor && typeof descriptor.value === "function") self[key] = self[key].bind(self);
	}
	return self;
}

//#endregion
//#region src/graphql/objects/Leaf.ts
const LeafFromJson = (obj) => {
	return {
		amount: CurrencyAmountFromJson(obj["leaf_amount"]),
		sparkNodeId: obj["leaf_spark_node_id"]
	};
};

//#endregion
//#region src/graphql/objects/PageInfo.ts
const PageInfoFromJson = (obj) => {
	return {
		hasNextPage: obj["page_info_has_next_page"],
		hasPreviousPage: obj["page_info_has_previous_page"],
		startCursor: obj["page_info_start_cursor"],
		endCursor: obj["page_info_end_cursor"]
	};
};

//#endregion
//#region src/graphql/objects/SparkTransferToLeavesConnection.ts
const SparkTransferToLeavesConnectionFromJson = (obj) => {
	return {
		count: obj["spark_transfer_to_leaves_connection_count"],
		pageInfo: PageInfoFromJson(obj["spark_transfer_to_leaves_connection_page_info"]),
		entities: obj["spark_transfer_to_leaves_connection_entities"].map((e) => LeafFromJson(e)),
		typename: "SparkTransferToLeavesConnection"
	};
};

//#endregion
//#region src/graphql/objects/Transfer.ts
var Transfer = class {
	constructor(totalAmount, sparkId, userRequestId) {
		this.totalAmount = totalAmount;
		this.sparkId = sparkId;
		this.userRequestId = userRequestId;
		autoBind(this);
	}
	async getLeaves(client, first = void 0, after = void 0) {
		return await client.executeRawQuery({
			queryPayload: ` 
query FetchSparkTransferToLeavesConnection($entity_id: ID!, $first: Int, $after: String) {
    entity(id: $entity_id) {
        ... on Transfer {
            leaves(, first: $first, after: $after) {
                __typename
                spark_transfer_to_leaves_connection_count: count
                spark_transfer_to_leaves_connection_page_info: page_info {
                    __typename
                    page_info_has_next_page: has_next_page
                    page_info_has_previous_page: has_previous_page
                    page_info_start_cursor: start_cursor
                    page_info_end_cursor: end_cursor
                }
                spark_transfer_to_leaves_connection_entities: entities {
                    __typename
                    leaf_amount: amount {
                        __typename
                        currency_amount_original_value: original_value
                        currency_amount_original_unit: original_unit
                    }
                    leaf_spark_node_id: spark_node_id
                }
            }
        }
    }
}
`,
			variables: {
				entity_id: this.sparkId,
				first,
				after
			},
			constructObject: (json) => {
				const connection = json["entity"]["leaves"];
				return SparkTransferToLeavesConnectionFromJson(connection);
			}
		});
	}
	toJson() {
		return {
			transfer_total_amount: CurrencyAmountToJson(this.totalAmount),
			transfer_spark_id: this.sparkId,
			transfer_user_request: { id: this.userRequestId }
		};
	}
};
const TransferFromJson = (obj) => {
	return new Transfer(CurrencyAmountFromJson(obj["transfer_total_amount"]), obj["transfer_spark_id"], obj["transfer_user_request"]?.id ?? void 0);
};
const FRAGMENT$5 = `
fragment TransferFragment on Transfer {
    __typename
    transfer_total_amount: total_amount {
        __typename
        currency_amount_original_value: original_value
        currency_amount_original_unit: original_unit
    }
    transfer_spark_id: spark_id
    transfer_user_request: user_request {
        id
    }
}`;

//#endregion
//#region src/graphql/objects/LeavesSwapRequest.ts
const LeavesSwapRequestFromJson = (obj) => {
	return {
		id: obj["leaves_swap_request_id"],
		createdAt: obj["leaves_swap_request_created_at"],
		updatedAt: obj["leaves_swap_request_updated_at"],
		network: BitcoinNetwork[obj["leaves_swap_request_network"]] ?? BitcoinNetwork.FUTURE_VALUE,
		status: SparkLeavesSwapRequestStatus[obj["leaves_swap_request_status"]] ?? SparkLeavesSwapRequestStatus.FUTURE_VALUE,
		totalAmount: CurrencyAmountFromJson(obj["leaves_swap_request_total_amount"]),
		targetAmount: CurrencyAmountFromJson(obj["leaves_swap_request_target_amount"]),
		fee: CurrencyAmountFromJson(obj["leaves_swap_request_fee"]),
		inboundTransfer: TransferFromJson(obj["leaves_swap_request_inbound_transfer"]),
		swapLeaves: obj["leaves_swap_request_swap_leaves"].map((e) => SwapLeafFromJson(e)),
		typename: "LeavesSwapRequest",
		outboundTransfer: !!obj["leaves_swap_request_outbound_transfer"] ? TransferFromJson(obj["leaves_swap_request_outbound_transfer"]) : void 0,
		expiresAt: obj["leaves_swap_request_expires_at"]
	};
};
const FRAGMENT$4 = `
fragment LeavesSwapRequestFragment on LeavesSwapRequest {
    __typename
    leaves_swap_request_id: id
    leaves_swap_request_created_at: created_at
    leaves_swap_request_updated_at: updated_at
    leaves_swap_request_network: network
    leaves_swap_request_status: status
    leaves_swap_request_total_amount: total_amount {
        __typename
        currency_amount_original_value: original_value
        currency_amount_original_unit: original_unit
    }
    leaves_swap_request_target_amount: target_amount {
        __typename
        currency_amount_original_value: original_value
        currency_amount_original_unit: original_unit
    }
    leaves_swap_request_fee: fee {
        __typename
        currency_amount_original_value: original_value
        currency_amount_original_unit: original_unit
    }
    leaves_swap_request_inbound_transfer: inbound_transfer {
        __typename
        transfer_total_amount: total_amount {
            __typename
            currency_amount_original_value: original_value
            currency_amount_original_unit: original_unit
        }
        transfer_spark_id: spark_id
        transfer_user_request: user_request {
            id
        }
    }
    leaves_swap_request_outbound_transfer: outbound_transfer {
        __typename
        transfer_total_amount: total_amount {
            __typename
            currency_amount_original_value: original_value
            currency_amount_original_unit: original_unit
        }
        transfer_spark_id: spark_id
        transfer_user_request: user_request {
            id
        }
    }
    leaves_swap_request_expires_at: expires_at
    leaves_swap_request_swap_leaves: swap_leaves {
        __typename
        swap_leaf_leaf_id: leaf_id
        swap_leaf_raw_unsigned_refund_transaction: raw_unsigned_refund_transaction
        swap_leaf_adaptor_signed_signature: adaptor_signed_signature
        swap_leaf_direct_raw_unsigned_refund_transaction: direct_raw_unsigned_refund_transaction
        swap_leaf_direct_adaptor_signed_signature: direct_adaptor_signed_signature
        swap_leaf_direct_from_cpfp_raw_unsigned_refund_transaction: direct_from_cpfp_raw_unsigned_refund_transaction
        swap_leaf_direct_from_cpfp_adaptor_signed_signature: direct_from_cpfp_adaptor_signed_signature
    }
}`;
const getLeavesSwapRequestQuery = (id) => {
	return {
		queryPayload: `
query GetLeavesSwapRequest($id: ID!) {
    entity(id: $id) {
        ... on LeavesSwapRequest {
            ...LeavesSwapRequestFragment
        }
    }
}

${FRAGMENT$4}    
`,
		variables: { id },
		constructObject: (data) => isObject(data) && "entity" in data && isObject(data.entity) ? LeavesSwapRequestFromJson(data.entity) : null
	};
};

//#endregion
//#region src/graphql/objects/Invoice.ts
const InvoiceFromJson = (obj) => {
	return {
		encodedInvoice: obj["invoice_encoded_invoice"],
		bitcoinNetwork: BitcoinNetwork[obj["invoice_bitcoin_network"]] ?? BitcoinNetwork.FUTURE_VALUE,
		paymentHash: obj["invoice_payment_hash"],
		amount: CurrencyAmountFromJson(obj["invoice_amount"]),
		createdAt: obj["invoice_created_at"],
		expiresAt: obj["invoice_expires_at"],
		memo: obj["invoice_memo"]
	};
};

//#endregion
//#region src/graphql/objects/LightningReceiveRequestStatus.ts
let LightningReceiveRequestStatus = /* @__PURE__ */ function(LightningReceiveRequestStatus) {
	/**
	* This is an enum value that represents values that could be added in the future.
	* Clients should support unknown values as more of them could be added without notice.
	*/
	LightningReceiveRequestStatus["FUTURE_VALUE"] = "FUTURE_VALUE";
	LightningReceiveRequestStatus["INVOICE_CREATED"] = "INVOICE_CREATED";
	LightningReceiveRequestStatus["TRANSFER_CREATED"] = "TRANSFER_CREATED";
	LightningReceiveRequestStatus["TRANSFER_CREATION_FAILED"] = "TRANSFER_CREATION_FAILED";
	LightningReceiveRequestStatus["REFUND_SIGNING_COMMITMENTS_QUERYING_FAILED"] = "REFUND_SIGNING_COMMITMENTS_QUERYING_FAILED";
	LightningReceiveRequestStatus["REFUND_SIGNING_FAILED"] = "REFUND_SIGNING_FAILED";
	LightningReceiveRequestStatus["PAYMENT_PREIMAGE_RECOVERED"] = "PAYMENT_PREIMAGE_RECOVERED";
	LightningReceiveRequestStatus["PAYMENT_PREIMAGE_RECOVERING_FAILED"] = "PAYMENT_PREIMAGE_RECOVERING_FAILED";
	LightningReceiveRequestStatus["LIGHTNING_PAYMENT_RECEIVED"] = "LIGHTNING_PAYMENT_RECEIVED";
	LightningReceiveRequestStatus["TRANSFER_FAILED"] = "TRANSFER_FAILED";
	LightningReceiveRequestStatus["TRANSFER_COMPLETED"] = "TRANSFER_COMPLETED";
	return LightningReceiveRequestStatus;
}({});

//#endregion
//#region src/graphql/objects/LightningReceiveRequest.ts
const LightningReceiveRequestFromJson = (obj) => {
	return {
		id: obj["lightning_receive_request_id"],
		createdAt: obj["lightning_receive_request_created_at"],
		updatedAt: obj["lightning_receive_request_updated_at"],
		network: BitcoinNetwork[obj["lightning_receive_request_network"]] ?? BitcoinNetwork.FUTURE_VALUE,
		invoice: InvoiceFromJson(obj["lightning_receive_request_invoice"]),
		status: LightningReceiveRequestStatus[obj["lightning_receive_request_status"]] ?? LightningReceiveRequestStatus.FUTURE_VALUE,
		typename: "LightningReceiveRequest",
		transfer: !!obj["lightning_receive_request_transfer"] ? TransferFromJson(obj["lightning_receive_request_transfer"]) : void 0,
		paymentPreimage: obj["lightning_receive_request_payment_preimage"],
		receiverIdentityPublicKey: obj["lightning_receive_request_receiver_identity_public_key"],
		sparkInvoice: obj["lightning_receive_request_spark_invoice"]
	};
};
const FRAGMENT$3 = `
fragment LightningReceiveRequestFragment on LightningReceiveRequest {
    __typename
    lightning_receive_request_id: id
    lightning_receive_request_created_at: created_at
    lightning_receive_request_updated_at: updated_at
    lightning_receive_request_network: network
    lightning_receive_request_invoice: invoice {
        __typename
        invoice_encoded_invoice: encoded_invoice
        invoice_bitcoin_network: bitcoin_network
        invoice_payment_hash: payment_hash
        invoice_amount: amount {
            __typename
            currency_amount_original_value: original_value
            currency_amount_original_unit: original_unit
        }
        invoice_created_at: created_at
        invoice_expires_at: expires_at
        invoice_memo: memo
    }
    lightning_receive_request_status: status
    lightning_receive_request_transfer: transfer {
        __typename
        transfer_total_amount: total_amount {
            __typename
            currency_amount_original_value: original_value
            currency_amount_original_unit: original_unit
        }
        transfer_spark_id: spark_id
        transfer_user_request: user_request {
            id
        }
    }
    lightning_receive_request_payment_preimage: payment_preimage
    lightning_receive_request_receiver_identity_public_key: receiver_identity_public_key
    lightning_receive_request_spark_invoice: spark_invoice
}`;
const getLightningReceiveRequestQuery = (id) => {
	return {
		queryPayload: `
query GetLightningReceiveRequest($id: ID!) {
    entity(id: $id) {
        ... on LightningReceiveRequest {
            ...LightningReceiveRequestFragment
        }
    }
}

${FRAGMENT$3}    
`,
		variables: { id },
		constructObject: (data) => isObject(data) && "entity" in data && isObject(data.entity) ? LightningReceiveRequestFromJson(data.entity) : null
	};
};

//#endregion
//#region src/graphql/objects/LightningSendRequestStatus.ts
let LightningSendRequestStatus = /* @__PURE__ */ function(LightningSendRequestStatus) {
	/**
	* This is an enum value that represents values that could be added in the future.
	* Clients should support unknown values as more of them could be added without notice.
	*/
	LightningSendRequestStatus["FUTURE_VALUE"] = "FUTURE_VALUE";
	LightningSendRequestStatus["CREATED"] = "CREATED";
	LightningSendRequestStatus["USER_TRANSFER_VALIDATION_FAILED"] = "USER_TRANSFER_VALIDATION_FAILED";
	LightningSendRequestStatus["LIGHTNING_PAYMENT_INITIATED"] = "LIGHTNING_PAYMENT_INITIATED";
	LightningSendRequestStatus["LIGHTNING_PAYMENT_FAILED"] = "LIGHTNING_PAYMENT_FAILED";
	LightningSendRequestStatus["LIGHTNING_PAYMENT_SUCCEEDED"] = "LIGHTNING_PAYMENT_SUCCEEDED";
	LightningSendRequestStatus["PREIMAGE_PROVIDED"] = "PREIMAGE_PROVIDED";
	LightningSendRequestStatus["PREIMAGE_PROVIDING_FAILED"] = "PREIMAGE_PROVIDING_FAILED";
	LightningSendRequestStatus["TRANSFER_COMPLETED"] = "TRANSFER_COMPLETED";
	LightningSendRequestStatus["TRANSFER_FAILED"] = "TRANSFER_FAILED";
	LightningSendRequestStatus["PENDING_USER_SWAP_RETURN"] = "PENDING_USER_SWAP_RETURN";
	LightningSendRequestStatus["USER_SWAP_RETURNED"] = "USER_SWAP_RETURNED";
	LightningSendRequestStatus["USER_SWAP_RETURN_FAILED"] = "USER_SWAP_RETURN_FAILED";
	LightningSendRequestStatus["REQUEST_VALIDATED"] = "REQUEST_VALIDATED";
	return LightningSendRequestStatus;
}({});

//#endregion
//#region src/graphql/objects/LightningSendRequest.ts
const LightningSendRequestFromJson = (obj) => {
	return {
		id: obj["lightning_send_request_id"],
		createdAt: obj["lightning_send_request_created_at"],
		updatedAt: obj["lightning_send_request_updated_at"],
		network: BitcoinNetwork[obj["lightning_send_request_network"]] ?? BitcoinNetwork.FUTURE_VALUE,
		encodedInvoice: obj["lightning_send_request_encoded_invoice"],
		fee: CurrencyAmountFromJson(obj["lightning_send_request_fee"]),
		idempotencyKey: obj["lightning_send_request_idempotency_key"],
		status: LightningSendRequestStatus[obj["lightning_send_request_status"]] ?? LightningSendRequestStatus.FUTURE_VALUE,
		typename: "LightningSendRequest",
		transfer: !!obj["lightning_send_request_transfer"] ? TransferFromJson(obj["lightning_send_request_transfer"]) : void 0,
		paymentPreimage: obj["lightning_send_request_payment_preimage"]
	};
};
const FRAGMENT$2 = `
fragment LightningSendRequestFragment on LightningSendRequest {
    __typename
    lightning_send_request_id: id
    lightning_send_request_created_at: created_at
    lightning_send_request_updated_at: updated_at
    lightning_send_request_network: network
    lightning_send_request_encoded_invoice: encoded_invoice
    lightning_send_request_fee: fee {
        __typename
        currency_amount_original_value: original_value
        currency_amount_original_unit: original_unit
    }
    lightning_send_request_idempotency_key: idempotency_key
    lightning_send_request_status: status
    lightning_send_request_transfer: transfer {
        __typename
        transfer_total_amount: total_amount {
            __typename
            currency_amount_original_value: original_value
            currency_amount_original_unit: original_unit
        }
        transfer_spark_id: spark_id
        transfer_user_request: user_request {
            id
        }
    }
    lightning_send_request_payment_preimage: payment_preimage
}`;
const getLightningSendRequestQuery = (id) => {
	return {
		queryPayload: `
query GetLightningSendRequest($id: ID!) {
    entity(id: $id) {
        ... on LightningSendRequest {
            ...LightningSendRequestFragment
        }
    }
}

${FRAGMENT$2}    
`,
		variables: { id },
		constructObject: (data) => isObject(data) && "entity" in data && isObject(data.entity) ? LightningSendRequestFromJson(data.entity) : null
	};
};

//#endregion
//#region src/graphql/objects/SparkCoopExitRequestStatus.ts
let SparkCoopExitRequestStatus = /* @__PURE__ */ function(SparkCoopExitRequestStatus) {
	/**
	* This is an enum value that represents values that could be added in the future.
	* Clients should support unknown values as more of them could be added without notice.
	*/
	SparkCoopExitRequestStatus["FUTURE_VALUE"] = "FUTURE_VALUE";
	SparkCoopExitRequestStatus["INITIATED"] = "INITIATED";
	SparkCoopExitRequestStatus["INBOUND_TRANSFER_CHECKED"] = "INBOUND_TRANSFER_CHECKED";
	SparkCoopExitRequestStatus["TX_SIGNED"] = "TX_SIGNED";
	SparkCoopExitRequestStatus["TX_BROADCASTED"] = "TX_BROADCASTED";
	SparkCoopExitRequestStatus["WAITING_ON_TX_CONFIRMATIONS"] = "WAITING_ON_TX_CONFIRMATIONS";
	SparkCoopExitRequestStatus["SUCCEEDED"] = "SUCCEEDED";
	SparkCoopExitRequestStatus["EXPIRED"] = "EXPIRED";
	SparkCoopExitRequestStatus["FAILED"] = "FAILED";
	return SparkCoopExitRequestStatus;
}({});

//#endregion
//#region src/graphql/objects/SparkUserRequestStatus.ts
let SparkUserRequestStatus = /* @__PURE__ */ function(SparkUserRequestStatus) {
	/**
	* This is an enum value that represents values that could be added in the future.
	* Clients should support unknown values as more of them could be added without notice.
	*/
	SparkUserRequestStatus["FUTURE_VALUE"] = "FUTURE_VALUE";
	SparkUserRequestStatus["CREATED"] = "CREATED";
	SparkUserRequestStatus["IN_PROGRESS"] = "IN_PROGRESS";
	SparkUserRequestStatus["SUCCEEDED"] = "SUCCEEDED";
	SparkUserRequestStatus["FAILED"] = "FAILED";
	SparkUserRequestStatus["CANCELED"] = "CANCELED";
	SparkUserRequestStatus["UNKNOWN"] = "UNKNOWN";
	return SparkUserRequestStatus;
}({});

//#endregion
//#region src/graphql/objects/SparkUserRequestType.ts
let SparkUserRequestType = /* @__PURE__ */ function(SparkUserRequestType) {
	/**
	* This is an enum value that represents values that could be added in the future.
	* Clients should support unknown values as more of them could be added without notice.
	*/
	SparkUserRequestType["FUTURE_VALUE"] = "FUTURE_VALUE";
	SparkUserRequestType["LIGHTNING_SEND"] = "LIGHTNING_SEND";
	SparkUserRequestType["LIGHTNING_RECEIVE"] = "LIGHTNING_RECEIVE";
	SparkUserRequestType["COOP_EXIT"] = "COOP_EXIT";
	SparkUserRequestType["LEAVES_SWAP"] = "LEAVES_SWAP";
	SparkUserRequestType["CLAIM_STATIC_DEPOSIT"] = "CLAIM_STATIC_DEPOSIT";
	return SparkUserRequestType;
}({});

//#endregion
//#region src/graphql/objects/UserRequest.ts
const UserRequestFromJson = (obj) => {
	if (obj["__typename"] == "ClaimStaticDeposit") return {
		id: obj["claim_static_deposit_id"],
		createdAt: obj["claim_static_deposit_created_at"],
		updatedAt: obj["claim_static_deposit_updated_at"],
		network: BitcoinNetwork[obj["claim_static_deposit_network"]] ?? BitcoinNetwork.FUTURE_VALUE,
		creditAmount: CurrencyAmountFromJson(obj["claim_static_deposit_credit_amount"]),
		maxFee: CurrencyAmountFromJson(obj["claim_static_deposit_max_fee"]),
		status: ClaimStaticDepositStatus[obj["claim_static_deposit_status"]] ?? ClaimStaticDepositStatus.FUTURE_VALUE,
		transactionId: obj["claim_static_deposit_transaction_id"],
		outputIndex: obj["claim_static_deposit_output_index"],
		bitcoinNetwork: BitcoinNetwork[obj["claim_static_deposit_bitcoin_network"]] ?? BitcoinNetwork.FUTURE_VALUE,
		typename: "ClaimStaticDeposit",
		transferSparkId: obj["claim_static_deposit_transfer_spark_id"]
	};
	if (obj["__typename"] == "CoopExitRequest") return {
		id: obj["coop_exit_request_id"],
		createdAt: obj["coop_exit_request_created_at"],
		updatedAt: obj["coop_exit_request_updated_at"],
		network: BitcoinNetwork[obj["coop_exit_request_network"]] ?? BitcoinNetwork.FUTURE_VALUE,
		fee: CurrencyAmountFromJson(obj["coop_exit_request_fee"]),
		l1BroadcastFee: CurrencyAmountFromJson(obj["coop_exit_request_l1_broadcast_fee"]),
		status: SparkCoopExitRequestStatus[obj["coop_exit_request_status"]] ?? SparkCoopExitRequestStatus.FUTURE_VALUE,
		expiresAt: obj["coop_exit_request_expires_at"],
		rawConnectorTransaction: obj["coop_exit_request_raw_connector_transaction"],
		rawCoopExitTransaction: obj["coop_exit_request_raw_coop_exit_transaction"],
		coopExitTxid: obj["coop_exit_request_coop_exit_txid"],
		typename: "CoopExitRequest",
		feeQuoteId: obj["coop_exit_request_fee_quote"]?.id ?? void 0,
		exitSpeed: !!obj["coop_exit_request_exit_speed"] ? ExitSpeed[obj["coop_exit_request_exit_speed"]] ?? ExitSpeed.FUTURE_VALUE : null,
		transfer: !!obj["coop_exit_request_transfer"] ? TransferFromJson(obj["coop_exit_request_transfer"]) : void 0
	};
	if (obj["__typename"] == "LeavesSwapRequest") return {
		id: obj["leaves_swap_request_id"],
		createdAt: obj["leaves_swap_request_created_at"],
		updatedAt: obj["leaves_swap_request_updated_at"],
		network: BitcoinNetwork[obj["leaves_swap_request_network"]] ?? BitcoinNetwork.FUTURE_VALUE,
		status: SparkLeavesSwapRequestStatus[obj["leaves_swap_request_status"]] ?? SparkLeavesSwapRequestStatus.FUTURE_VALUE,
		totalAmount: CurrencyAmountFromJson(obj["leaves_swap_request_total_amount"]),
		targetAmount: CurrencyAmountFromJson(obj["leaves_swap_request_target_amount"]),
		fee: CurrencyAmountFromJson(obj["leaves_swap_request_fee"]),
		inboundTransfer: TransferFromJson(obj["leaves_swap_request_inbound_transfer"]),
		swapLeaves: obj["leaves_swap_request_swap_leaves"].map((e) => SwapLeafFromJson(e)),
		typename: "LeavesSwapRequest",
		outboundTransfer: !!obj["leaves_swap_request_outbound_transfer"] ? TransferFromJson(obj["leaves_swap_request_outbound_transfer"]) : void 0,
		expiresAt: obj["leaves_swap_request_expires_at"]
	};
	if (obj["__typename"] == "LightningReceiveRequest") return {
		id: obj["lightning_receive_request_id"],
		createdAt: obj["lightning_receive_request_created_at"],
		updatedAt: obj["lightning_receive_request_updated_at"],
		network: BitcoinNetwork[obj["lightning_receive_request_network"]] ?? BitcoinNetwork.FUTURE_VALUE,
		invoice: InvoiceFromJson(obj["lightning_receive_request_invoice"]),
		status: LightningReceiveRequestStatus[obj["lightning_receive_request_status"]] ?? LightningReceiveRequestStatus.FUTURE_VALUE,
		typename: "LightningReceiveRequest",
		transfer: !!obj["lightning_receive_request_transfer"] ? TransferFromJson(obj["lightning_receive_request_transfer"]) : void 0,
		paymentPreimage: obj["lightning_receive_request_payment_preimage"],
		receiverIdentityPublicKey: obj["lightning_receive_request_receiver_identity_public_key"]
	};
	if (obj["__typename"] == "LightningSendRequest") return {
		id: obj["lightning_send_request_id"],
		createdAt: obj["lightning_send_request_created_at"],
		updatedAt: obj["lightning_send_request_updated_at"],
		network: BitcoinNetwork[obj["lightning_send_request_network"]] ?? BitcoinNetwork.FUTURE_VALUE,
		encodedInvoice: obj["lightning_send_request_encoded_invoice"],
		fee: CurrencyAmountFromJson(obj["lightning_send_request_fee"]),
		idempotencyKey: obj["lightning_send_request_idempotency_key"],
		status: LightningSendRequestStatus[obj["lightning_send_request_status"]] ?? LightningSendRequestStatus.FUTURE_VALUE,
		typename: "LightningSendRequest",
		transfer: !!obj["lightning_send_request_transfer"] ? TransferFromJson(obj["lightning_send_request_transfer"]) : void 0,
		paymentPreimage: obj["lightning_send_request_payment_preimage"]
	};
	throw new LightsparkException("DeserializationError", `Couldn't find a concrete type for interface UserRequest corresponding to the typename=${obj["__typename"]}`);
};
const FRAGMENT$1 = `
fragment UserRequestFragment on UserRequest {
    __typename
    ... on ClaimStaticDeposit {
        __typename
        claim_static_deposit_id: id
        claim_static_deposit_created_at: created_at
        claim_static_deposit_updated_at: updated_at
        claim_static_deposit_network: network
        claim_static_deposit_credit_amount: credit_amount {
            __typename
            currency_amount_original_value: original_value
            currency_amount_original_unit: original_unit
        }
        claim_static_deposit_max_fee: max_fee {
            __typename
            currency_amount_original_value: original_value
            currency_amount_original_unit: original_unit
        }
        claim_static_deposit_status: status
        claim_static_deposit_transaction_id: transaction_id
        claim_static_deposit_output_index: output_index
        claim_static_deposit_bitcoin_network: bitcoin_network
        claim_static_deposit_transfer_spark_id: transfer_spark_id
    }
    ... on CoopExitRequest {
        __typename
        coop_exit_request_id: id
        coop_exit_request_created_at: created_at
        coop_exit_request_updated_at: updated_at
        coop_exit_request_network: network
        coop_exit_request_fee: fee {
            __typename
            currency_amount_original_value: original_value
            currency_amount_original_unit: original_unit
        }
        coop_exit_request_l1_broadcast_fee: l1_broadcast_fee {
            __typename
            currency_amount_original_value: original_value
            currency_amount_original_unit: original_unit
        }
        coop_exit_request_fee_quote: fee_quote {
            id
        }
        coop_exit_request_exit_speed: exit_speed
        coop_exit_request_status: status
        coop_exit_request_expires_at: expires_at
        coop_exit_request_raw_connector_transaction: raw_connector_transaction
        coop_exit_request_raw_coop_exit_transaction: raw_coop_exit_transaction
        coop_exit_request_coop_exit_txid: coop_exit_txid
        coop_exit_request_transfer: transfer {
            __typename
            transfer_total_amount: total_amount {
                __typename
                currency_amount_original_value: original_value
                currency_amount_original_unit: original_unit
            }
            transfer_spark_id: spark_id
            transfer_user_request: user_request {
                id
            }
        }
    }
    ... on LeavesSwapRequest {
        __typename
        leaves_swap_request_id: id
        leaves_swap_request_created_at: created_at
        leaves_swap_request_updated_at: updated_at
        leaves_swap_request_network: network
        leaves_swap_request_status: status
        leaves_swap_request_total_amount: total_amount {
            __typename
            currency_amount_original_value: original_value
            currency_amount_original_unit: original_unit
        }
        leaves_swap_request_target_amount: target_amount {
            __typename
            currency_amount_original_value: original_value
            currency_amount_original_unit: original_unit
        }
        leaves_swap_request_fee: fee {
            __typename
            currency_amount_original_value: original_value
            currency_amount_original_unit: original_unit
        }
        leaves_swap_request_inbound_transfer: inbound_transfer {
            __typename
            transfer_total_amount: total_amount {
                __typename
                currency_amount_original_value: original_value
                currency_amount_original_unit: original_unit
            }
            transfer_spark_id: spark_id
            transfer_user_request: user_request {
                id
            }
        }
        leaves_swap_request_outbound_transfer: outbound_transfer {
            __typename
            transfer_total_amount: total_amount {
                __typename
                currency_amount_original_value: original_value
                currency_amount_original_unit: original_unit
            }
            transfer_spark_id: spark_id
            transfer_user_request: user_request {
                id
            }
        }
        leaves_swap_request_expires_at: expires_at
        leaves_swap_request_swap_leaves: swap_leaves {
            __typename
            swap_leaf_leaf_id: leaf_id
            swap_leaf_raw_unsigned_refund_transaction: raw_unsigned_refund_transaction
            swap_leaf_adaptor_signed_signature: adaptor_signed_signature
            swap_leaf_direct_raw_unsigned_refund_transaction: direct_raw_unsigned_refund_transaction
            swap_leaf_direct_adaptor_signed_signature: direct_adaptor_signed_signature
            swap_leaf_direct_from_cpfp_raw_unsigned_refund_transaction: direct_from_cpfp_raw_unsigned_refund_transaction
            swap_leaf_direct_from_cpfp_adaptor_signed_signature: direct_from_cpfp_adaptor_signed_signature
        }
    }
    ... on LightningReceiveRequest {
        __typename
        lightning_receive_request_id: id
        lightning_receive_request_created_at: created_at
        lightning_receive_request_updated_at: updated_at
        lightning_receive_request_network: network
        lightning_receive_request_invoice: invoice {
            __typename
            invoice_encoded_invoice: encoded_invoice
            invoice_bitcoin_network: bitcoin_network
            invoice_payment_hash: payment_hash
            invoice_amount: amount {
                __typename
                currency_amount_original_value: original_value
                currency_amount_original_unit: original_unit
            }
            invoice_created_at: created_at
            invoice_expires_at: expires_at
            invoice_memo: memo
        }
        lightning_receive_request_status: status
        lightning_receive_request_transfer: transfer {
            __typename
            transfer_total_amount: total_amount {
                __typename
                currency_amount_original_value: original_value
                currency_amount_original_unit: original_unit
            }
            transfer_spark_id: spark_id
            transfer_user_request: user_request {
                id
            }
        }
        lightning_receive_request_payment_preimage: payment_preimage
        lightning_receive_request_receiver_identity_public_key: receiver_identity_public_key
    }
    ... on LightningSendRequest {
        __typename
        lightning_send_request_id: id
        lightning_send_request_created_at: created_at
        lightning_send_request_updated_at: updated_at
        lightning_send_request_network: network
        lightning_send_request_encoded_invoice: encoded_invoice
        lightning_send_request_fee: fee {
            __typename
            currency_amount_original_value: original_value
            currency_amount_original_unit: original_unit
        }
        lightning_send_request_idempotency_key: idempotency_key
        lightning_send_request_status: status
        lightning_send_request_transfer: transfer {
            __typename
            transfer_total_amount: total_amount {
                __typename
                currency_amount_original_value: original_value
                currency_amount_original_unit: original_unit
            }
            transfer_spark_id: spark_id
            transfer_user_request: user_request {
                id
            }
        }
        lightning_send_request_payment_preimage: payment_preimage
    }
}`;
const getUserRequestQuery = (id) => {
	return {
		queryPayload: `
query GetUserRequest($id: ID!) {
    entity(id: $id) {
        ... on UserRequest {
            ...UserRequestFragment
        }
    }
}

${FRAGMENT$1}    
`,
		variables: { id },
		constructObject: (data) => isObject(data) && "entity" in data && isObject(data.entity) ? UserRequestFromJson(data.entity) : null
	};
};

//#endregion
//#region src/graphql/objects/SparkWalletUserToUserRequestsConnection.ts
const SparkWalletUserToUserRequestsConnectionFromJson = (obj) => {
	return {
		count: obj["spark_wallet_user_to_user_requests_connection_count"],
		pageInfo: PageInfoFromJson(obj["spark_wallet_user_to_user_requests_connection_page_info"]),
		entities: obj["spark_wallet_user_to_user_requests_connection_entities"].map((e) => UserRequestFromJson(e)),
		typename: "SparkWalletUserToUserRequestsConnection"
	};
};

//#endregion
//#region src/graphql/objects/SparkWalletUser.ts
var SparkWalletUser = class {
	constructor(id, createdAt, updatedAt, identityPublicKey, typename) {
		this.id = id;
		this.createdAt = createdAt;
		this.updatedAt = updatedAt;
		this.identityPublicKey = identityPublicKey;
		this.typename = typename;
		autoBind(this);
	}
	async getUserRequests(client, first = void 0, after = void 0, types = void 0, statuses = void 0, networks = void 0) {
		return await client.executeRawQuery({
			queryPayload: ` 
query FetchSparkWalletUserToUserRequestsConnection($entity_id: ID!, $first: Int, $after: String, $types: [SparkUserRequestType!], $statuses: [SparkUserRequestStatus!], $networks: [BitcoinNetwork!]) {
    entity(id: $entity_id) {
        ... on SparkWalletUser {
            user_requests(, first: $first, after: $after, types: $types, statuses: $statuses, networks: $networks) {
                __typename
                spark_wallet_user_to_user_requests_connection_count: count
                spark_wallet_user_to_user_requests_connection_page_info: page_info {
                    __typename
                    page_info_has_next_page: has_next_page
                    page_info_has_previous_page: has_previous_page
                    page_info_start_cursor: start_cursor
                    page_info_end_cursor: end_cursor
                }
                spark_wallet_user_to_user_requests_connection_entities: entities {
                    __typename
                    ... on ClaimStaticDeposit {
                        __typename
                        claim_static_deposit_id: id
                        claim_static_deposit_created_at: created_at
                        claim_static_deposit_updated_at: updated_at
                        claim_static_deposit_network: network
                        claim_static_deposit_credit_amount: credit_amount {
                            __typename
                            currency_amount_original_value: original_value
                            currency_amount_original_unit: original_unit
                        }
                        claim_static_deposit_max_fee: max_fee {
                            __typename
                            currency_amount_original_value: original_value
                            currency_amount_original_unit: original_unit
                        }
                        claim_static_deposit_status: status
                        claim_static_deposit_transaction_id: transaction_id
                        claim_static_deposit_output_index: output_index
                        claim_static_deposit_bitcoin_network: bitcoin_network
                        claim_static_deposit_transfer_spark_id: transfer_spark_id
                    }
                    ... on CoopExitRequest {
                        __typename
                        coop_exit_request_id: id
                        coop_exit_request_created_at: created_at
                        coop_exit_request_updated_at: updated_at
                        coop_exit_request_network: network
                        coop_exit_request_fee: fee {
                            __typename
                            currency_amount_original_value: original_value
                            currency_amount_original_unit: original_unit
                        }
                        coop_exit_request_l1_broadcast_fee: l1_broadcast_fee {
                            __typename
                            currency_amount_original_value: original_value
                            currency_amount_original_unit: original_unit
                        }
                        coop_exit_request_fee_quote: fee_quote {
                            id
                        }
                        coop_exit_request_exit_speed: exit_speed
                        coop_exit_request_status: status
                        coop_exit_request_expires_at: expires_at
                        coop_exit_request_raw_connector_transaction: raw_connector_transaction
                        coop_exit_request_raw_coop_exit_transaction: raw_coop_exit_transaction
                        coop_exit_request_coop_exit_txid: coop_exit_txid
                        coop_exit_request_transfer: transfer {
                            __typename
                            transfer_total_amount: total_amount {
                                __typename
                                currency_amount_original_value: original_value
                                currency_amount_original_unit: original_unit
                            }
                            transfer_spark_id: spark_id
                            transfer_user_request: user_request {
                                id
                            }
                        }
                    }
                    ... on LeavesSwapRequest {
                        __typename
                        leaves_swap_request_id: id
                        leaves_swap_request_created_at: created_at
                        leaves_swap_request_updated_at: updated_at
                        leaves_swap_request_network: network
                        leaves_swap_request_status: status
                        leaves_swap_request_total_amount: total_amount {
                            __typename
                            currency_amount_original_value: original_value
                            currency_amount_original_unit: original_unit
                        }
                        leaves_swap_request_target_amount: target_amount {
                            __typename
                            currency_amount_original_value: original_value
                            currency_amount_original_unit: original_unit
                        }
                        leaves_swap_request_fee: fee {
                            __typename
                            currency_amount_original_value: original_value
                            currency_amount_original_unit: original_unit
                        }
                        leaves_swap_request_inbound_transfer: inbound_transfer {
                            __typename
                            transfer_total_amount: total_amount {
                                __typename
                                currency_amount_original_value: original_value
                                currency_amount_original_unit: original_unit
                            }
                            transfer_spark_id: spark_id
                            transfer_user_request: user_request {
                                id
                            }
                        }
                        leaves_swap_request_outbound_transfer: outbound_transfer {
                            __typename
                            transfer_total_amount: total_amount {
                                __typename
                                currency_amount_original_value: original_value
                                currency_amount_original_unit: original_unit
                            }
                            transfer_spark_id: spark_id
                            transfer_user_request: user_request {
                                id
                            }
                        }
                        leaves_swap_request_expires_at: expires_at
                        leaves_swap_request_swap_leaves: swap_leaves {
                            __typename
                            swap_leaf_leaf_id: leaf_id
                            swap_leaf_raw_unsigned_refund_transaction: raw_unsigned_refund_transaction
                            swap_leaf_adaptor_signed_signature: adaptor_signed_signature
                        }
                    }
                    ... on LightningReceiveRequest {
                        __typename
                        lightning_receive_request_id: id
                        lightning_receive_request_created_at: created_at
                        lightning_receive_request_updated_at: updated_at
                        lightning_receive_request_network: network
                        lightning_receive_request_invoice: invoice {
                            __typename
                            invoice_encoded_invoice: encoded_invoice
                            invoice_bitcoin_network: bitcoin_network
                            invoice_payment_hash: payment_hash
                            invoice_amount: amount {
                                __typename
                                currency_amount_original_value: original_value
                                currency_amount_original_unit: original_unit
                            }
                            invoice_created_at: created_at
                            invoice_expires_at: expires_at
                            invoice_memo: memo
                        }
                        lightning_receive_request_status: status
                        lightning_receive_request_transfer: transfer {
                            __typename
                            transfer_total_amount: total_amount {
                                __typename
                                currency_amount_original_value: original_value
                                currency_amount_original_unit: original_unit
                            }
                            transfer_spark_id: spark_id
                            transfer_user_request: user_request {
                                id
                            }
                        }
                        lightning_receive_request_payment_preimage: payment_preimage
                        lightning_receive_request_receiver_identity_public_key: receiver_identity_public_key
                    }
                    ... on LightningSendRequest {
                        __typename
                        lightning_send_request_id: id
                        lightning_send_request_created_at: created_at
                        lightning_send_request_updated_at: updated_at
                        lightning_send_request_network: network
                        lightning_send_request_encoded_invoice: encoded_invoice
                        lightning_send_request_fee: fee {
                            __typename
                            currency_amount_original_value: original_value
                            currency_amount_original_unit: original_unit
                        }
                        lightning_send_request_idempotency_key: idempotency_key
                        lightning_send_request_status: status
                        lightning_send_request_transfer: transfer {
                            __typename
                            transfer_total_amount: total_amount {
                                __typename
                                currency_amount_original_value: original_value
                                currency_amount_original_unit: original_unit
                            }
                            transfer_spark_id: spark_id
                            transfer_user_request: user_request {
                                id
                            }
                        }
                        lightning_send_request_payment_preimage: payment_preimage
                    }
                }
            }
        }
    }
}
`,
			variables: {
				entity_id: this.id,
				first,
				after,
				types,
				statuses,
				networks
			},
			constructObject: (json) => {
				const connection = json["entity"]["user_requests"];
				return SparkWalletUserToUserRequestsConnectionFromJson(connection);
			}
		});
	}
	static getSparkWalletUserQuery(id) {
		return {
			queryPayload: `
query GetSparkWalletUser($id: ID!) {
    entity(id: $id) {
        ... on SparkWalletUser {
            ...SparkWalletUserFragment
        }
    }
}

${FRAGMENT}    
`,
			variables: { id },
			constructObject: (data) => isObject(data) && "entity" in data && isObject(data.entity) ? SparkWalletUserFromJson(data.entity) : null
		};
	}
	toJson() {
		return {
			__typename: "SparkWalletUser",
			spark_wallet_user_id: this.id,
			spark_wallet_user_created_at: this.createdAt,
			spark_wallet_user_updated_at: this.updatedAt,
			spark_wallet_user_identity_public_key: this.identityPublicKey
		};
	}
};
const SparkWalletUserFromJson = (obj) => {
	return new SparkWalletUser(obj["spark_wallet_user_id"], obj["spark_wallet_user_created_at"], obj["spark_wallet_user_updated_at"], obj["spark_wallet_user_identity_public_key"], "SparkWalletUser");
};
const FRAGMENT = `
fragment SparkWalletUserFragment on SparkWalletUser {
    __typename
    spark_wallet_user_id: id
    spark_wallet_user_created_at: created_at
    spark_wallet_user_updated_at: updated_at
    spark_wallet_user_identity_public_key: identity_public_key
}`;

//#endregion
//#region src/types/sdk-types.ts
function mapSettingsProtoToWalletSettings(proto) {
	return {
		ownerIdentityPublicKey: bytesToHex(proto.ownerIdentityPublicKey),
		privateEnabled: proto.privateEnabled
	};
}
function mapTreeNodeToWalletLeaf(proto) {
	return {
		id: proto.id,
		treeId: proto.treeId,
		value: proto.value,
		parentNodeId: proto.parentNodeId,
		nodeTx: bytesToHex(proto.nodeTx),
		refundTx: bytesToHex(proto.refundTx),
		vout: proto.vout,
		verifyingPublicKey: bytesToHex(proto.verifyingPublicKey),
		ownerIdentityPublicKey: bytesToHex(proto.ownerIdentityPublicKey),
		signingKeyshare: proto.signingKeyshare,
		status: proto.status,
		network: Network[proto.network]
	};
}
let TransferDirection = /* @__PURE__ */ function(TransferDirection) {
	TransferDirection["INCOMING"] = "INCOMING";
	TransferDirection["OUTGOING"] = "OUTGOING";
	return TransferDirection;
}({});
function mapTransferLeafToWalletTransferLeaf(proto) {
	return {
		leaf: proto.leaf ? mapTreeNodeToWalletLeaf(proto.leaf) : void 0,
		secretCipher: bytesToHex(proto.secretCipher),
		signature: bytesToHex(proto.signature),
		intermediateRefundTx: bytesToHex(proto.intermediateRefundTx)
	};
}
function mapTransferToWalletTransfer(proto, identityPublicKey, userRequest) {
	const receiverIdentityPublicKey = bytesToHex(proto.receiverIdentityPublicKey);
	const senderIdentityPublicKey = bytesToHex(proto.senderIdentityPublicKey);
	return {
		id: proto.id,
		senderIdentityPublicKey,
		receiverIdentityPublicKey,
		status: TransferStatus[proto.status],
		totalValue: proto.totalValue,
		expiryTime: proto.expiryTime ? new Date(proto.expiryTime) : void 0,
		leaves: proto.leaves.map(mapTransferLeafToWalletTransferLeaf),
		createdTime: proto.createdTime ? new Date(proto.createdTime) : void 0,
		updatedTime: proto.updatedTime ? new Date(proto.updatedTime) : void 0,
		type: TransferType[proto.type],
		transferDirection: receiverIdentityPublicKey === identityPublicKey ? TransferDirection.INCOMING : TransferDirection.OUTGOING,
		userRequest,
		sparkInvoice: proto.sparkInvoice || void 0
	};
}

//#endregion
export { FRAGMENT$6 as A, getLeavesSwapRequestQuery as C, SparkLeavesSwapRequestStatus as D, TransferFromJson as E, CurrencyAmountFromJson as F, CurrencyUnit$1 as I, ClaimStaticDepositStatus as L, ClaimStaticDepositRequestType as M, ClaimStaticDepositFromJson as N, ExitSpeed as O, getClaimStaticDepositQuery as P, BitcoinNetwork as R, LeavesSwapRequestFromJson as S, Transfer as T, FRAGMENT$3 as _, mapTreeNodeToWalletLeaf as a, LightningReceiveRequestStatus as b, FRAGMENT$1 as c, SparkUserRequestStatus as d, SparkCoopExitRequestStatus as f, LightningSendRequestStatus as g, getLightningSendRequestQuery as h, mapTransferToWalletTransfer as i, getCoopExitFeeQuoteQuery as j, CoopExitFeeQuoteFromJson as k, getUserRequestQuery as l, LightningSendRequestFromJson as m, mapSettingsProtoToWalletSettings as n, SparkWalletUser as o, FRAGMENT$2 as p, mapTransferLeafToWalletTransferLeaf as r, SparkWalletUserToUserRequestsConnectionFromJson as s, TransferDirection as t, SparkUserRequestType as u, LightningReceiveRequestFromJson as v, FRAGMENT$5 as w, FRAGMENT$4 as x, getLightningReceiveRequestQuery as y };
