// @utxos/sdk/dist/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  ApiError: () => ApiError,
  AuthRouteError: () => AuthRouteError,
  CardanoWalletDeveloperControlled: () => CardanoWalletDeveloperControlled,
  NoPRFExtensionError: () => NoPRFExtensionError,
  NotAllowedError: () => NotAllowedError,
  NotAuthenticatedError: () => NotAuthenticatedError,
  SessionExpiredError: () => SessionExpiredError,
  SparkIssuerWalletDeveloperControlled: () => SparkIssuerWalletDeveloperControlled,
  SparkWallet: () => import_spark_sdk5.SparkWallet,
  Sponsorship: () => Sponsorship,
  StorageInsertError: () => StorageInsertError,
  StorageRetrievalError: () => StorageRetrievalError,
  TransferDirection: () => TransferDirection,
  UserAbortedError: () => UserAbortedError,
  WalletDeveloperControlled: () => WalletDeveloperControlled,
  WalletServerCreationError: () => WalletServerCreationError,
  WalletServerRetrievalError: () => WalletServerRetrievalError,
  Web3NonCustodialProvider: () => Web3NonCustodialProvider,
  Web3Sdk: () => Web3Sdk,
  Web3SparkWallet: () => Web3SparkWallet,
  Web3Wallet: () => Web3Wallet,
  bytesToHex: () => bytesToHex2,
  bytesToString: () => bytesToString,
  clientDeriveWallet: () => clientDeriveWallet,
  clientGenerateWallet: () => clientGenerateWallet,
  clientRecovery: () => clientRecovery,
  combineShardsBuildWallet: () => combineShardsBuildWallet,
  convertLegacyToNewFormat: () => convertLegacyToNewFormat,
  createCredential: () => createCredential,
  crypto: () => crypto2,
  decodeBech32mTokenIdentifier: () => import_spark_sdk5.decodeBech32mTokenIdentifier,
  decodeSparkAddress: () => import_spark_sdk5.decodeSparkAddress,
  decryptWithCipher: () => decryptWithCipher,
  decryptWithPrivateKey: () => decryptWithPrivateKey,
  encodeBech32mTokenIdentifier: () => import_spark_sdk5.encodeBech32mTokenIdentifier,
  encryptWithCipher: () => encryptWithCipher,
  encryptWithPublicKey: () => encryptWithPublicKey,
  extractIdentityPublicKey: () => extractIdentityPublicKey,
  generateHash: () => generateHash,
  generateKeyPair: () => generateKeyPair,
  getBitcoinAddressFromPubkey: () => getBitcoinAddressFromPubkey,
  getCardanoAddressFromPubkey: () => getCardanoAddressFromPubkey,
  getCredentialFromCredentialId: () => getCredentialFromCredentialId,
  getCrypto: () => getCrypto,
  getNetworkFromBech32mTokenIdentifier: () => import_spark_sdk5.getNetworkFromBech32mTokenIdentifier,
  getNetworkFromSparkAddress: () => import_spark_sdk5.getNetworkFromSparkAddress,
  getSparkAddressFromPubkey: () => getSparkAddressFromPubkey,
  hashData: () => hashData,
  hexToBytes: () => hexToBytes2,
  isValidMultiChainWalletInfo: () => import_api_contracts.isValidMultiChainWalletInfo,
  isValidSparkAddress: () => import_spark_sdk5.isValidSparkAddress,
  meshUniversalStaticUtxo: () => meshUniversalStaticUtxo,
  openWindow: () => openWindow,
  secretToCryptoKey: () => secretToCryptoKey,
  shamirCombine: () => shamirCombine,
  shamirSplit: () => shamirSplit,
  spiltKeyIntoShards: () => spiltKeyIntoShards,
  stringToBytes: () => stringToBytes,
  validateMultiChainWalletInfo: () => import_api_contracts.validateMultiChainWalletInfo,
  webauthnPublicKeyCredentialToCryptoKey: () => webauthnPublicKeyCredentialToCryptoKey
});
module.exports = __toCommonJS(index_exports);

// src/internal/platform-context.ts
var _adapters = null;
var _initialized = false;
function setAdapters(adapters) {
  if (_initialized) {
    if (process.env.NODE_ENV === "development") {
      console.warn("@utxos/sdk: Re-initializing adapters (development mode)");
    } else {
      console.warn("@utxos/sdk: Adapters already initialized, ignoring");
      return;
    }
  }
  _adapters = adapters;
  _initialized = true;
}
function getAdapters() {
  if (!_adapters) {
    throw new Error(
      '@utxos/sdk: Platform adapters not initialized. This usually means the SDK was imported incorrectly. Import from "@utxos/sdk" directly, not from internal paths.'
    );
  }
  return _adapters;
}
function getCrypto() {
  return getAdapters().crypto;
}
function getStorage() {
  return getAdapters().storage;
}
function getLinking() {
  return getAdapters().linking;
}
function getEncoding() {
  return getAdapters().encoding;
}

// src/platforms/browser/encoding.browser.ts
var encodingAdapter = {
  bytesToBase64(bytes) {
    let binary = "";
    for (let i = 0; i < bytes.length; i++) {
      const byte = bytes[i];
      if (byte !== void 0) {
        binary += String.fromCharCode(byte);
      }
    }
    return btoa(binary);
  },
  base64ToBytes(base64) {
    const normalized = base64.replace(/-/g, "+").replace(/_/g, "/");
    const binary = atob(normalized);
    const bytes = new Uint8Array(binary.length);
    for (let i = 0; i < binary.length; i++) {
      bytes[i] = binary.charCodeAt(i);
    }
    return bytes;
  },
  bytesToHex(bytes) {
    return Array.from(bytes).map((b) => b.toString(16).padStart(2, "0")).join("");
  },
  hexToBytes(hex) {
    const matches = hex.match(/.{1,2}/g);
    if (!matches) return new Uint8Array(0);
    return new Uint8Array(matches.map((byte) => parseInt(byte, 16)));
  },
  bytesToUtf8(bytes) {
    return new TextDecoder().decode(bytes);
  },
  utf8ToBytes(str) {
    return new TextEncoder().encode(str);
  }
};

// src/platforms/browser/storage.browser.ts
var storageAdapter = {
  async getItem(key) {
    try {
      return localStorage.getItem(key);
    } catch (error) {
      console.warn("@utxos/sdk: localStorage.getItem failed:", error);
      return null;
    }
  },
  async setItem(key, value) {
    try {
      localStorage.setItem(key, value);
    } catch (error) {
      console.error("@utxos/sdk: localStorage.setItem failed:", error);
      throw new Error(`Failed to save to storage: ${error}`);
    }
  },
  async removeItem(key) {
    try {
      localStorage.removeItem(key);
    } catch (error) {
      console.warn("@utxos/sdk: localStorage.removeItem failed:", error);
    }
  },
  async clear() {
    try {
      localStorage.clear();
    } catch (error) {
      console.warn("@utxos/sdk: localStorage.clear failed:", error);
    }
  }
};

// src/platforms/browser/linking.browser.ts
function calculatePopupPosition(width, height) {
  const screenWidth = window.innerWidth || document.documentElement.clientWidth || screen.width;
  const screenHeight = window.innerHeight || document.documentElement.clientHeight || screen.height;
  const windowLeft = window.screenX || window.screenLeft || 0;
  const windowTop = window.screenY || window.screenTop || 0;
  const left = (screenWidth - width) / 2 + windowLeft;
  const top = (screenHeight - height) / 2 + windowTop;
  return {
    left: Math.max(0, left),
    top: Math.max(0, top)
  };
}
function buildWindowFeatures(width, height) {
  const windowWidth = window.innerWidth || 0;
  const windowHeight = window.innerHeight || 0;
  const isMobile = windowWidth < 640;
  const isFullScreen = !!(document.fullscreenElement || document.webkitFullscreenElement || document.mozFullScreenElement || document.msFullscreenElement);
  const shouldDisplayFullScreen = isMobile || isFullScreen;
  const finalWidth = shouldDisplayFullScreen ? windowWidth : width;
  const finalHeight = shouldDisplayFullScreen ? windowHeight : height;
  const { left, top } = calculatePopupPosition(finalWidth, finalHeight);
  return `left=${left},top=${top},width=${finalWidth},height=${finalHeight},scrollbars=yes,resizable=yes,status=no,location=no,toolbar=no,menubar=no`;
}
var linkingAdapter = {
  /**
   * Open URL in new browser tab
   */
  async openURL(url) {
    window.open(url, "_blank", "noopener,noreferrer");
  },
  /**
   * Redirect current page to URL (same-tab navigation)
   * Used for auth flow redirects after token storage
   */
  async redirectURL(url) {
    window.location.href = url;
  },
  /**
   * Open OAuth popup window and wait for callback via postMessage
   * Handles popup blocked, user close, and message events
   */
  async openAuthWindow(url, callbackScheme) {
    return new Promise((resolve, reject) => {
      const width = 448;
      const height = 668;
      const features = buildWindowFeatures(width, height);
      const popup = window.open(url, "utxos", features);
      if (!popup) {
        reject(new Error("Popup blocked. Please allow popups for this site."));
        return;
      }
      let pollTimer = null;
      let resolved = false;
      const cleanup = () => {
        if (pollTimer) {
          clearInterval(pollTimer);
          pollTimer = null;
        }
        window.removeEventListener("message", handleMessage);
      };
      const resolveOnce = (result) => {
        if (resolved) return;
        resolved = true;
        cleanup();
        resolve(result);
      };
      const handleMessage = (event) => {
        if (!event.data || typeof event.data !== "object") {
          return;
        }
        const { code, state, error, error_description, target } = event.data;
        if (target === "utxos" || code || error) {
          if (!popup.closed) {
            popup.close();
          }
          resolveOnce({
            code,
            state,
            error,
            errorDescription: error_description,
            // Include full message data for wallet-specific payloads
            data: event.data
          });
        }
      };
      window.addEventListener("message", handleMessage);
      pollTimer = setInterval(() => {
        if (popup.closed) {
          resolveOnce({
            error: "cancelled",
            errorDescription: "User closed the popup"
          });
        }
      }, 500);
    });
  },
  /**
   * Get current page URL
   */
  getCurrentURL() {
    if (typeof window === "undefined") return null;
    return window.location.href;
  },
  /**
   * Parse URL query parameters from current page
   */
  getURLParams() {
    if (typeof window === "undefined") return {};
    const params = new URLSearchParams(window.location.search);
    const result = {};
    params.forEach((value, key) => {
      result[key] = value;
    });
    if (window.location.hash) {
      const hashParams = new URLSearchParams(window.location.hash.slice(1));
      hashParams.forEach((value, key) => {
        result[key] = value;
      });
    }
    return result;
  },
  /**
   * Add listener for URL changes (SPA navigation)
   * Returns cleanup function to remove listener
   */
  addURLListener(callback) {
    if (typeof window === "undefined") {
      return () => {
      };
    }
    const handler = () => callback(window.location.href);
    window.addEventListener("popstate", handler);
    window.addEventListener("hashchange", handler);
    return () => {
      window.removeEventListener("popstate", handler);
      window.removeEventListener("hashchange", handler);
    };
  },
  getUserAgent() {
    if (typeof navigator === "undefined") return null;
    return navigator.userAgent;
  }
};

// src/platforms/browser/crypto.browser.ts
function getHmacAlgorithm(algorithm) {
  const normalized = algorithm.toUpperCase().replace(/SHA-?(\d+)/i, "SHA-$1");
  return {
    name: "HMAC",
    hash: { name: normalized }
  };
}
var cryptoAdapter = {
  /**
   * Generate cryptographically secure random bytes using crypto.getRandomValues
   */
  getRandomBytes(size) {
    const bytes = new Uint8Array(size);
    crypto.getRandomValues(bytes);
    return bytes;
  },
  /**
   * HMAC signing using crypto.subtle
   */
  async hmacSign(algorithm, key, data) {
    const hmacAlgorithm = getHmacAlgorithm(algorithm);
    const cryptoKey = await crypto.subtle.importKey(
      "raw",
      key,
      hmacAlgorithm,
      false,
      ["sign"]
    );
    const signature = await crypto.subtle.sign("HMAC", cryptoKey, data);
    return new Uint8Array(signature);
  },
  /**
   * AES-GCM encryption using crypto.subtle
   */
  async encrypt(data, key, iv) {
    const cryptoKey = await crypto.subtle.importKey(
      "raw",
      key,
      { name: "AES-GCM" },
      false,
      ["encrypt"]
    );
    const ciphertext = await crypto.subtle.encrypt(
      {
        name: "AES-GCM",
        iv
      },
      cryptoKey,
      data
    );
    return new Uint8Array(ciphertext);
  },
  /**
   * AES-GCM decryption using crypto.subtle
   */
  async decrypt(data, key, iv) {
    const cryptoKey = await crypto.subtle.importKey(
      "raw",
      key,
      { name: "AES-GCM" },
      false,
      ["decrypt"]
    );
    const plaintext = await crypto.subtle.decrypt(
      {
        name: "AES-GCM",
        iv
      },
      cryptoKey,
      data
    );
    return new Uint8Array(plaintext);
  },
  /**
   * Generate ECDH key pair using P-256 curve
   * Returns raw public and private key bytes
   */
  async generateKeyPair(_algorithm) {
    const keyPair = await crypto.subtle.generateKey(
      {
        name: "ECDH",
        namedCurve: "P-256"
      },
      true,
      // extractable
      ["deriveBits"]
    );
    const publicKeyBuffer = await crypto.subtle.exportKey("raw", keyPair.publicKey);
    const privateKeyBuffer = await crypto.subtle.exportKey("pkcs8", keyPair.privateKey);
    return {
      publicKey: new Uint8Array(publicKeyBuffer),
      privateKey: new Uint8Array(privateKeyBuffer)
    };
  },
  /**
   * Derive shared secret from ECDH key exchange
   * Uses P-256 curve and returns 256 bits (32 bytes)
   */
  async deriveSharedSecret(privateKey, publicKey) {
    const privateKeyObj = await crypto.subtle.importKey(
      "pkcs8",
      privateKey,
      {
        name: "ECDH",
        namedCurve: "P-256"
      },
      false,
      ["deriveBits"]
    );
    const publicKeyObj = await crypto.subtle.importKey(
      "raw",
      publicKey,
      {
        name: "ECDH",
        namedCurve: "P-256"
      },
      false,
      []
    );
    const sharedSecret = await crypto.subtle.deriveBits(
      {
        name: "ECDH",
        public: publicKeyObj
      },
      privateKeyObj,
      256
      // P-256 produces 256 bits
    );
    return new Uint8Array(sharedSecret);
  },
  /**
   * Import raw key for crypto operations
   * Supports AES-GCM and HMAC algorithms
   */
  async importKey(keyData, algorithm, usages) {
    const normalizedAlgorithm = algorithm.toUpperCase();
    let algorithmParams;
    if (normalizedAlgorithm === "AES-GCM" || normalizedAlgorithm === "AESGCM") {
      algorithmParams = { name: "AES-GCM" };
    } else if (normalizedAlgorithm.startsWith("HMAC")) {
      const hashMatch = normalizedAlgorithm.match(/HMAC[- ]?(SHA[- ]?\d+)?/i);
      const hashName = hashMatch?.[1]?.replace(/[- ]/g, "-") || "SHA-256";
      algorithmParams = {
        name: "HMAC",
        hash: { name: hashName }
      };
    } else if (normalizedAlgorithm.includes("SHA")) {
      algorithmParams = {
        name: "HMAC",
        hash: { name: normalizedAlgorithm.replace(/SHA-?(\d+)/i, "SHA-$1") }
      };
    } else {
      algorithmParams = { name: algorithm };
    }
    return crypto.subtle.importKey(
      "raw",
      keyData,
      algorithmParams,
      false,
      usages
    );
  },
  getSubtleCrypto() {
    return crypto.subtle;
  },
  getRandomValuesInPlace(array) {
    crypto.getRandomValues(array);
    return array;
  }
};

// src/platforms/browser/index.ts
var browserAdapters = {
  crypto: cryptoAdapter,
  storage: storageAdapter,
  linking: linkingAdapter,
  encoding: encodingAdapter
};

// src/chains/bitcoin.ts
var import_bitcoin = require("@meshsdk/bitcoin");
function getBitcoinAddressFromPubkey(pubKeyHash, networkId) {
  return (0, import_bitcoin.resolveAddress)(pubKeyHash, networkId === 1 ? "mainnet" : "testnet").address;
}

// src/chains/cardano.ts
var import_common = require("@meshsdk/common");
var import_core_cst = require("@meshsdk/core-cst");
function getCardanoAddressFromPubkey(pubKeyHash, stakeCredentialHash, networkId) {
  return (0, import_core_cst.serializeAddressObj)(
    (0, import_common.pubKeyAddress)(pubKeyHash, stakeCredentialHash),
    networkId
  );
}

// src/chains/spark/wallet.ts
var import_axios = __toESM(require("axios"), 1);
var Web3SparkWallet = class {
  networkId;
  appUrl;
  projectId;
  address;
  sparkApiUrl;
  publicKeyHex;
  sparkApiKey;
  sparkMainnetStaticDepositAddress;
  sparkRegtestStaticDepositAddress;
  _axiosInstance;
  constructor(options) {
    this.networkId = options.networkId;
    this.appUrl = options.appUrl;
    this.projectId = options.projectId;
    this.address = options.address;
    this.sparkApiUrl = options.sparkApiUrl;
    this.publicKeyHex = options.publicKeyHex;
    this.sparkApiKey = options.sparkApiKey;
    this.sparkMainnetStaticDepositAddress = options.sparkMainnetStaticDepositAddress;
    this.sparkRegtestStaticDepositAddress = options.sparkRegtestStaticDepositAddress;
    this._axiosInstance = import_axios.default.create({
      baseURL: options.sparkApiUrl,
      headers: {
        Accept: "application/json",
        ...options.sparkApiKey && {
          Authorization: `Bearer ${options.sparkApiKey}`
        }
      }
    });
  }
  /** https://docs.xverse.app/sats-connect/spark-methods/spark_getaddress */
  getAddress() {
    return {
      address: this.address,
      network: this.networkId === 0 ? "regtest" : "mainnet",
      publicKey: this.publicKeyHex
    };
  }
  /** https://docs.xverse.app/sats-connect/spark-methods/spark_getbalance */
  async getBalance() {
    if (this.sparkApiKey === void 0) {
      throw new ApiError({
        code: 5,
        info: "Failed to fetch from spark api (no API key)"
      });
    }
    const response = await this._axiosInstance.get(
      `address/${this.address}?network=${this.networkId === 0 ? "REGTEST" : "MAINNET"}`
    );
    const addressSummary = response.data;
    const tokenBalancesMap = /* @__PURE__ */ new Map();
    if (addressSummary.tokens && Array.isArray(addressSummary.tokens)) {
      for (const token of addressSummary.tokens) {
        tokenBalancesMap.set(token.tokenIdentifier, {
          balance: BigInt(token.balance),
          tokenInfo: {
            tokenName: token.name,
            tokenPublicKey: token.issuerPublicKey,
            tokenTicker: token.ticker,
            maxSupply: token.maxSupply === null ? "" : String(token.maxSupply),
            decimals: String(token.decimals)
          }
        });
      }
    }
    return {
      balance: BigInt(addressSummary.balance.btcHardBalanceSats || 0),
      tokenBalances: tokenBalancesMap
    };
  }
  /** https://docs.xverse.app/sats-connect/spark-methods/spark_transfer */
  async transfer({
    receiverSparkAddress,
    amountSats
  }) {
    const res = await openWindow(
      {
        method: "spark-transfer",
        projectId: this.projectId,
        networkId: String(this.networkId),
        receiverSparkAddress,
        amountSats: String(amountSats)
      },
      this.appUrl
    );
    if (res.success === false) {
      throw new ApiError({
        code: 3,
        info: "UserDeclined - User declined the transfer."
      });
    }
    if (res.data.method !== "spark-transfer") {
      throw new ApiError({
        code: 2,
        info: "Recieved the wrong response from wallet popover."
      });
    }
    return { txid: res.data.txid };
  }
  async transferToken({
    receiverSparkAddress,
    tokenIdentifier,
    tokenAmount
  }) {
    const res = await openWindow(
      {
        method: "spark-transfer-token",
        projectId: this.projectId,
        networkId: String(this.networkId),
        receiverSparkAddress,
        tokenAmount: String(tokenAmount),
        tokenIdentifier: String(tokenIdentifier)
      },
      this.appUrl
    );
    if (res.success === false) {
      throw new ApiError({
        code: 3,
        info: "UserDeclined - User declined the transfer."
      });
    }
    if (res.data.method !== "spark-transfer-token") {
      throw new ApiError({
        code: 2,
        info: "Recieved the wrong response from wallet popover."
      });
    }
    return { txid: res.data.txid };
  }
  async claimStaticDeposit({ txId }) {
    const res = await openWindow(
      {
        method: "spark-claim-static-deposit",
        txId,
        projectId: this.projectId,
        networkId: String(this.networkId)
      },
      this.appUrl
    );
    if (res.success === false) {
      throw new ApiError({
        code: 3,
        info: "UserDeclined - User declined the transfer."
      });
    }
    if (res.data.method !== "spark-claim-static-deposit") {
      throw new ApiError({
        code: 2,
        info: "Recieved the wrong response from wallet popover."
      });
    }
    return {
      transferId: res.data.transferId
    };
  }
  async withdrawToBitcoin({
    exitSpeed,
    amountSats,
    deductFeeFromWithdrawalAmount,
    withdrawalAddress
  }) {
    const res = await openWindow(
      {
        method: "spark-withdraw-to-bitcoin",
        exitSpeed,
        withdrawalAddress,
        amountSats: String(amountSats),
        deductFeeFromWithdrawalAmount: deductFeeFromWithdrawalAmount === true ? "true" : "false",
        projectId: this.projectId,
        networkId: String(this.networkId)
      },
      this.appUrl
    );
    if (res.success === false) {
      throw new ApiError({
        code: 3,
        info: "UserDeclined - User declined the transfer."
      });
    }
    if (res.data.method !== "spark-withdraw-to-bitcoin") {
      throw new ApiError({
        code: 2,
        info: "Recieved the wrong response from wallet popover."
      });
    }
    return {
      withdrawalId: res.data.withdrawalId
    };
  }
  async signMessage({ message }) {
    const res = await openWindow(
      {
        method: "spark-sign-message",
        projectId: this.projectId,
        networkId: String(this.networkId),
        message
      },
      this.appUrl
    );
    if (res.success === false) {
      throw new ApiError({
        code: 3,
        info: "UserDeclined - User declined the transfer."
      });
    }
    if (res.data.method !== "spark-sign-message") {
      throw new ApiError({
        code: 2,
        info: "Recieved the wrong response from wallet popover."
      });
    }
    return {
      signature: res.data.signature
    };
  }
};

// src/chains/spark/utils.ts
var import_spark_sdk = require("@buildonspark/spark-sdk");

// node_modules/@scure/base/lib/esm/index.js
function isBytes(a) {
  return a instanceof Uint8Array || ArrayBuffer.isView(a) && a.constructor.name === "Uint8Array";
}
function isArrayOf(isString, arr) {
  if (!Array.isArray(arr))
    return false;
  if (arr.length === 0)
    return true;
  if (isString) {
    return arr.every((item) => typeof item === "string");
  } else {
    return arr.every((item) => Number.isSafeInteger(item));
  }
}
function afn(input) {
  if (typeof input !== "function")
    throw new Error("function expected");
  return true;
}
function astr(label, input) {
  if (typeof input !== "string")
    throw new Error(`${label}: string expected`);
  return true;
}
function anumber(n) {
  if (!Number.isSafeInteger(n))
    throw new Error(`invalid integer: ${n}`);
}
function aArr(input) {
  if (!Array.isArray(input))
    throw new Error("array expected");
}
function astrArr(label, input) {
  if (!isArrayOf(true, input))
    throw new Error(`${label}: array of strings expected`);
}
function anumArr(label, input) {
  if (!isArrayOf(false, input))
    throw new Error(`${label}: array of numbers expected`);
}
// @__NO_SIDE_EFFECTS__
function chain(...args) {
  const id = (a) => a;
  const wrap = (a, b) => (c) => a(b(c));
  const encode = args.map((x) => x.encode).reduceRight(wrap, id);
  const decode = args.map((x) => x.decode).reduce(wrap, id);
  return { encode, decode };
}
// @__NO_SIDE_EFFECTS__
function alphabet(letters) {
  const lettersA = typeof letters === "string" ? letters.split("") : letters;
  const len = lettersA.length;
  astrArr("alphabet", lettersA);
  const indexes = new Map(lettersA.map((l, i) => [l, i]));
  return {
    encode: (digits) => {
      aArr(digits);
      return digits.map((i) => {
        if (!Number.isSafeInteger(i) || i < 0 || i >= len)
          throw new Error(`alphabet.encode: digit index outside alphabet "${i}". Allowed: ${letters}`);
        return lettersA[i];
      });
    },
    decode: (input) => {
      aArr(input);
      return input.map((letter) => {
        astr("alphabet.decode", letter);
        const i = indexes.get(letter);
        if (i === void 0)
          throw new Error(`Unknown letter: "${letter}". Allowed: ${letters}`);
        return i;
      });
    }
  };
}
// @__NO_SIDE_EFFECTS__
function join(separator = "") {
  astr("join", separator);
  return {
    encode: (from) => {
      astrArr("join.decode", from);
      return from.join(separator);
    },
    decode: (to) => {
      astr("join.decode", to);
      return to.split(separator);
    }
  };
}
var gcd = (a, b) => b === 0 ? a : gcd(b, a % b);
var radix2carry = /* @__NO_SIDE_EFFECTS__ */ (from, to) => from + (to - gcd(from, to));
var powers = /* @__PURE__ */ (() => {
  let res = [];
  for (let i = 0; i < 40; i++)
    res.push(2 ** i);
  return res;
})();
function convertRadix2(data, from, to, padding) {
  aArr(data);
  if (from <= 0 || from > 32)
    throw new Error(`convertRadix2: wrong from=${from}`);
  if (to <= 0 || to > 32)
    throw new Error(`convertRadix2: wrong to=${to}`);
  if (/* @__PURE__ */ radix2carry(from, to) > 32) {
    throw new Error(`convertRadix2: carry overflow from=${from} to=${to} carryBits=${/* @__PURE__ */ radix2carry(from, to)}`);
  }
  let carry = 0;
  let pos = 0;
  const max = powers[from];
  const mask = powers[to] - 1;
  const res = [];
  for (const n of data) {
    anumber(n);
    if (n >= max)
      throw new Error(`convertRadix2: invalid data word=${n} from=${from}`);
    carry = carry << from | n;
    if (pos + from > 32)
      throw new Error(`convertRadix2: carry overflow pos=${pos} from=${from}`);
    pos += from;
    for (; pos >= to; pos -= to)
      res.push((carry >> pos - to & mask) >>> 0);
    const pow = powers[pos];
    if (pow === void 0)
      throw new Error("invalid carry");
    carry &= pow - 1;
  }
  carry = carry << to - pos & mask;
  if (!padding && pos >= from)
    throw new Error("Excess padding");
  if (!padding && carry > 0)
    throw new Error(`Non-zero padding: ${carry}`);
  if (padding && pos > 0)
    res.push(carry >>> 0);
  return res;
}
// @__NO_SIDE_EFFECTS__
function radix2(bits, revPadding = false) {
  anumber(bits);
  if (bits <= 0 || bits > 32)
    throw new Error("radix2: bits should be in (0..32]");
  if (/* @__PURE__ */ radix2carry(8, bits) > 32 || /* @__PURE__ */ radix2carry(bits, 8) > 32)
    throw new Error("radix2: carry overflow");
  return {
    encode: (bytes) => {
      if (!isBytes(bytes))
        throw new Error("radix2.encode input should be Uint8Array");
      return convertRadix2(Array.from(bytes), 8, bits, !revPadding);
    },
    decode: (digits) => {
      anumArr("radix2.decode", digits);
      return Uint8Array.from(convertRadix2(digits, bits, 8, revPadding));
    }
  };
}
function unsafeWrapper(fn) {
  afn(fn);
  return function(...args) {
    try {
      return fn.apply(null, args);
    } catch (e) {
    }
  };
}
var BECH_ALPHABET = /* @__PURE__ */ chain(/* @__PURE__ */ alphabet("qpzry9x8gf2tvdw0s3jn54khce6mua7l"), /* @__PURE__ */ join(""));
var POLYMOD_GENERATORS = [996825010, 642813549, 513874426, 1027748829, 705979059];
function bech32Polymod(pre) {
  const b = pre >> 25;
  let chk = (pre & 33554431) << 5;
  for (let i = 0; i < POLYMOD_GENERATORS.length; i++) {
    if ((b >> i & 1) === 1)
      chk ^= POLYMOD_GENERATORS[i];
  }
  return chk;
}
function bechChecksum(prefix, words, encodingConst = 1) {
  const len = prefix.length;
  let chk = 1;
  for (let i = 0; i < len; i++) {
    const c = prefix.charCodeAt(i);
    if (c < 33 || c > 126)
      throw new Error(`Invalid prefix (${prefix})`);
    chk = bech32Polymod(chk) ^ c >> 5;
  }
  chk = bech32Polymod(chk);
  for (let i = 0; i < len; i++)
    chk = bech32Polymod(chk) ^ prefix.charCodeAt(i) & 31;
  for (let v of words)
    chk = bech32Polymod(chk) ^ v;
  for (let i = 0; i < 6; i++)
    chk = bech32Polymod(chk);
  chk ^= encodingConst;
  return BECH_ALPHABET.encode(convertRadix2([chk % powers[30]], 30, 5, false));
}
// @__NO_SIDE_EFFECTS__
function genBech32(encoding) {
  const ENCODING_CONST = encoding === "bech32" ? 1 : 734539939;
  const _words = /* @__PURE__ */ radix2(5);
  const fromWords = _words.decode;
  const toWords = _words.encode;
  const fromWordsUnsafe = unsafeWrapper(fromWords);
  function encode(prefix, words, limit = 90) {
    astr("bech32.encode prefix", prefix);
    if (isBytes(words))
      words = Array.from(words);
    anumArr("bech32.encode", words);
    const plen = prefix.length;
    if (plen === 0)
      throw new TypeError(`Invalid prefix length ${plen}`);
    const actualLength = plen + 7 + words.length;
    if (limit !== false && actualLength > limit)
      throw new TypeError(`Length ${actualLength} exceeds limit ${limit}`);
    const lowered = prefix.toLowerCase();
    const sum = bechChecksum(lowered, words, ENCODING_CONST);
    return `${lowered}1${BECH_ALPHABET.encode(words)}${sum}`;
  }
  function decode(str, limit = 90) {
    astr("bech32.decode input", str);
    const slen = str.length;
    if (slen < 8 || limit !== false && slen > limit)
      throw new TypeError(`invalid string length: ${slen} (${str}). Expected (8..${limit})`);
    const lowered = str.toLowerCase();
    if (str !== lowered && str !== str.toUpperCase())
      throw new Error(`String must be lowercase or uppercase`);
    const sepIndex = lowered.lastIndexOf("1");
    if (sepIndex === 0 || sepIndex === -1)
      throw new Error(`Letter "1" must be present between prefix and data only`);
    const prefix = lowered.slice(0, sepIndex);
    const data = lowered.slice(sepIndex + 1);
    if (data.length < 6)
      throw new Error("Data must be at least 6 characters long");
    const words = BECH_ALPHABET.decode(data).slice(0, -6);
    const sum = bechChecksum(prefix, words, ENCODING_CONST);
    if (!data.endsWith(sum))
      throw new Error(`Invalid checksum in ${str}: expected "${sum}"`);
    return { prefix, words };
  }
  const decodeUnsafe = unsafeWrapper(decode);
  function decodeToBytes(str) {
    const { prefix, words } = decode(str, false);
    return { prefix, words, bytes: fromWords(words) };
  }
  function encodeFromBytes(prefix, bytes) {
    return encode(prefix, toWords(bytes));
  }
  return {
    encode,
    decode,
    encodeFromBytes,
    decodeToBytes,
    decodeUnsafe,
    fromWords,
    fromWordsUnsafe,
    toWords
  };
}
var bech32m = /* @__PURE__ */ genBech32("bech32m");

// node_modules/@bufbuild/protobuf/dist/esm/wire/varint.js
function varint64write(lo, hi, bytes) {
  for (let i = 0; i < 28; i = i + 7) {
    const shift = lo >>> i;
    const hasNext = !(shift >>> 7 == 0 && hi == 0);
    const byte = (hasNext ? shift | 128 : shift) & 255;
    bytes.push(byte);
    if (!hasNext) {
      return;
    }
  }
  const splitBits = lo >>> 28 & 15 | (hi & 7) << 4;
  const hasMoreBits = !(hi >> 3 == 0);
  bytes.push((hasMoreBits ? splitBits | 128 : splitBits) & 255);
  if (!hasMoreBits) {
    return;
  }
  for (let i = 3; i < 31; i = i + 7) {
    const shift = hi >>> i;
    const hasNext = !(shift >>> 7 == 0);
    const byte = (hasNext ? shift | 128 : shift) & 255;
    bytes.push(byte);
    if (!hasNext) {
      return;
    }
  }
  bytes.push(hi >>> 31 & 1);
}
var TWO_PWR_32_DBL = 4294967296;
function int64FromString(dec) {
  const minus = dec[0] === "-";
  if (minus) {
    dec = dec.slice(1);
  }
  const base = 1e6;
  let lowBits = 0;
  let highBits = 0;
  function add1e6digit(begin, end) {
    const digit1e6 = Number(dec.slice(begin, end));
    highBits *= base;
    lowBits = lowBits * base + digit1e6;
    if (lowBits >= TWO_PWR_32_DBL) {
      highBits = highBits + (lowBits / TWO_PWR_32_DBL | 0);
      lowBits = lowBits % TWO_PWR_32_DBL;
    }
  }
  add1e6digit(-24, -18);
  add1e6digit(-18, -12);
  add1e6digit(-12, -6);
  add1e6digit(-6);
  return minus ? negate(lowBits, highBits) : newBits(lowBits, highBits);
}
function int64ToString(lo, hi) {
  let bits = newBits(lo, hi);
  const negative = bits.hi & 2147483648;
  if (negative) {
    bits = negate(bits.lo, bits.hi);
  }
  const result = uInt64ToString(bits.lo, bits.hi);
  return negative ? "-" + result : result;
}
function uInt64ToString(lo, hi) {
  ({ lo, hi } = toUnsigned(lo, hi));
  if (hi <= 2097151) {
    return String(TWO_PWR_32_DBL * hi + lo);
  }
  const low = lo & 16777215;
  const mid = (lo >>> 24 | hi << 8) & 16777215;
  const high = hi >> 16 & 65535;
  let digitA = low + mid * 6777216 + high * 6710656;
  let digitB = mid + high * 8147497;
  let digitC = high * 2;
  const base = 1e7;
  if (digitA >= base) {
    digitB += Math.floor(digitA / base);
    digitA %= base;
  }
  if (digitB >= base) {
    digitC += Math.floor(digitB / base);
    digitB %= base;
  }
  return digitC.toString() + decimalFrom1e7WithLeadingZeros(digitB) + decimalFrom1e7WithLeadingZeros(digitA);
}
function toUnsigned(lo, hi) {
  return { lo: lo >>> 0, hi: hi >>> 0 };
}
function newBits(lo, hi) {
  return { lo: lo | 0, hi: hi | 0 };
}
function negate(lowBits, highBits) {
  highBits = ~highBits;
  if (lowBits) {
    lowBits = ~lowBits + 1;
  } else {
    highBits += 1;
  }
  return newBits(lowBits, highBits);
}
var decimalFrom1e7WithLeadingZeros = (digit1e7) => {
  const partial = String(digit1e7);
  return "0000000".slice(partial.length) + partial;
};
function varint32write(value, bytes) {
  if (value >= 0) {
    while (value > 127) {
      bytes.push(value & 127 | 128);
      value = value >>> 7;
    }
    bytes.push(value);
  } else {
    for (let i = 0; i < 9; i++) {
      bytes.push(value & 127 | 128);
      value = value >> 7;
    }
    bytes.push(1);
  }
}

// node_modules/@bufbuild/protobuf/dist/esm/proto-int64.js
var protoInt64 = /* @__PURE__ */ makeInt64Support();
function makeInt64Support() {
  const dv = new DataView(new ArrayBuffer(8));
  const ok = typeof BigInt === "function" && typeof dv.getBigInt64 === "function" && typeof dv.getBigUint64 === "function" && typeof dv.setBigInt64 === "function" && typeof dv.setBigUint64 === "function" && (!!globalThis.Deno || typeof process != "object" || typeof process.env != "object" || process.env.BUF_BIGINT_DISABLE !== "1");
  if (ok) {
    const MIN = BigInt("-9223372036854775808");
    const MAX = BigInt("9223372036854775807");
    const UMIN = BigInt("0");
    const UMAX = BigInt("18446744073709551615");
    return {
      zero: BigInt(0),
      supported: true,
      parse(value) {
        const bi = typeof value == "bigint" ? value : BigInt(value);
        if (bi > MAX || bi < MIN) {
          throw new Error(`invalid int64: ${value}`);
        }
        return bi;
      },
      uParse(value) {
        const bi = typeof value == "bigint" ? value : BigInt(value);
        if (bi > UMAX || bi < UMIN) {
          throw new Error(`invalid uint64: ${value}`);
        }
        return bi;
      },
      enc(value) {
        dv.setBigInt64(0, this.parse(value), true);
        return {
          lo: dv.getInt32(0, true),
          hi: dv.getInt32(4, true)
        };
      },
      uEnc(value) {
        dv.setBigInt64(0, this.uParse(value), true);
        return {
          lo: dv.getInt32(0, true),
          hi: dv.getInt32(4, true)
        };
      },
      dec(lo, hi) {
        dv.setInt32(0, lo, true);
        dv.setInt32(4, hi, true);
        return dv.getBigInt64(0, true);
      },
      uDec(lo, hi) {
        dv.setInt32(0, lo, true);
        dv.setInt32(4, hi, true);
        return dv.getBigUint64(0, true);
      }
    };
  }
  return {
    zero: "0",
    supported: false,
    parse(value) {
      if (typeof value != "string") {
        value = value.toString();
      }
      assertInt64String(value);
      return value;
    },
    uParse(value) {
      if (typeof value != "string") {
        value = value.toString();
      }
      assertUInt64String(value);
      return value;
    },
    enc(value) {
      if (typeof value != "string") {
        value = value.toString();
      }
      assertInt64String(value);
      return int64FromString(value);
    },
    uEnc(value) {
      if (typeof value != "string") {
        value = value.toString();
      }
      assertUInt64String(value);
      return int64FromString(value);
    },
    dec(lo, hi) {
      return int64ToString(lo, hi);
    },
    uDec(lo, hi) {
      return uInt64ToString(lo, hi);
    }
  };
}
function assertInt64String(value) {
  if (!/^-?[0-9]+$/.test(value)) {
    throw new Error("invalid int64: " + value);
  }
}
function assertUInt64String(value) {
  if (!/^[0-9]+$/.test(value)) {
    throw new Error("invalid uint64: " + value);
  }
}

// node_modules/@bufbuild/protobuf/dist/esm/wire/text-encoding.js
var symbol = /* @__PURE__ */ Symbol.for("@bufbuild/protobuf/text-encoding");
function getTextEncoding() {
  if (globalThis[symbol] == void 0) {
    const te = new globalThis.TextEncoder();
    const td = new globalThis.TextDecoder();
    globalThis[symbol] = {
      encodeUtf8(text) {
        return te.encode(text);
      },
      decodeUtf8(bytes) {
        return td.decode(bytes);
      },
      checkUtf8(text) {
        try {
          encodeURIComponent(text);
          return true;
        } catch (_) {
          return false;
        }
      }
    };
  }
  return globalThis[symbol];
}

// node_modules/@bufbuild/protobuf/dist/esm/wire/binary-encoding.js
var WireType;
(function(WireType2) {
  WireType2[WireType2["Varint"] = 0] = "Varint";
  WireType2[WireType2["Bit64"] = 1] = "Bit64";
  WireType2[WireType2["LengthDelimited"] = 2] = "LengthDelimited";
  WireType2[WireType2["StartGroup"] = 3] = "StartGroup";
  WireType2[WireType2["EndGroup"] = 4] = "EndGroup";
  WireType2[WireType2["Bit32"] = 5] = "Bit32";
})(WireType || (WireType = {}));
var FLOAT32_MAX = 34028234663852886e22;
var FLOAT32_MIN = -34028234663852886e22;
var UINT32_MAX = 4294967295;
var INT32_MAX = 2147483647;
var INT32_MIN = -2147483648;
var BinaryWriter = class {
  constructor(encodeUtf8 = getTextEncoding().encodeUtf8) {
    this.encodeUtf8 = encodeUtf8;
    this.stack = [];
    this.chunks = [];
    this.buf = [];
  }
  /**
   * Return all bytes written and reset this writer.
   */
  finish() {
    if (this.buf.length) {
      this.chunks.push(new Uint8Array(this.buf));
      this.buf = [];
    }
    let len = 0;
    for (let i = 0; i < this.chunks.length; i++)
      len += this.chunks[i].length;
    let bytes = new Uint8Array(len);
    let offset = 0;
    for (let i = 0; i < this.chunks.length; i++) {
      bytes.set(this.chunks[i], offset);
      offset += this.chunks[i].length;
    }
    this.chunks = [];
    return bytes;
  }
  /**
   * Start a new fork for length-delimited data like a message
   * or a packed repeated field.
   *
   * Must be joined later with `join()`.
   */
  fork() {
    this.stack.push({ chunks: this.chunks, buf: this.buf });
    this.chunks = [];
    this.buf = [];
    return this;
  }
  /**
   * Join the last fork. Write its length and bytes, then
   * return to the previous state.
   */
  join() {
    let chunk = this.finish();
    let prev = this.stack.pop();
    if (!prev)
      throw new Error("invalid state, fork stack empty");
    this.chunks = prev.chunks;
    this.buf = prev.buf;
    this.uint32(chunk.byteLength);
    return this.raw(chunk);
  }
  /**
   * Writes a tag (field number and wire type).
   *
   * Equivalent to `uint32( (fieldNo << 3 | type) >>> 0 )`.
   *
   * Generated code should compute the tag ahead of time and call `uint32()`.
   */
  tag(fieldNo, type) {
    return this.uint32((fieldNo << 3 | type) >>> 0);
  }
  /**
   * Write a chunk of raw bytes.
   */
  raw(chunk) {
    if (this.buf.length) {
      this.chunks.push(new Uint8Array(this.buf));
      this.buf = [];
    }
    this.chunks.push(chunk);
    return this;
  }
  /**
   * Write a `uint32` value, an unsigned 32 bit varint.
   */
  uint32(value) {
    assertUInt32(value);
    while (value > 127) {
      this.buf.push(value & 127 | 128);
      value = value >>> 7;
    }
    this.buf.push(value);
    return this;
  }
  /**
   * Write a `int32` value, a signed 32 bit varint.
   */
  int32(value) {
    assertInt32(value);
    varint32write(value, this.buf);
    return this;
  }
  /**
   * Write a `bool` value, a variant.
   */
  bool(value) {
    this.buf.push(value ? 1 : 0);
    return this;
  }
  /**
   * Write a `bytes` value, length-delimited arbitrary data.
   */
  bytes(value) {
    this.uint32(value.byteLength);
    return this.raw(value);
  }
  /**
   * Write a `string` value, length-delimited data converted to UTF-8 text.
   */
  string(value) {
    let chunk = this.encodeUtf8(value);
    this.uint32(chunk.byteLength);
    return this.raw(chunk);
  }
  /**
   * Write a `float` value, 32-bit floating point number.
   */
  float(value) {
    assertFloat32(value);
    let chunk = new Uint8Array(4);
    new DataView(chunk.buffer).setFloat32(0, value, true);
    return this.raw(chunk);
  }
  /**
   * Write a `double` value, a 64-bit floating point number.
   */
  double(value) {
    let chunk = new Uint8Array(8);
    new DataView(chunk.buffer).setFloat64(0, value, true);
    return this.raw(chunk);
  }
  /**
   * Write a `fixed32` value, an unsigned, fixed-length 32-bit integer.
   */
  fixed32(value) {
    assertUInt32(value);
    let chunk = new Uint8Array(4);
    new DataView(chunk.buffer).setUint32(0, value, true);
    return this.raw(chunk);
  }
  /**
   * Write a `sfixed32` value, a signed, fixed-length 32-bit integer.
   */
  sfixed32(value) {
    assertInt32(value);
    let chunk = new Uint8Array(4);
    new DataView(chunk.buffer).setInt32(0, value, true);
    return this.raw(chunk);
  }
  /**
   * Write a `sint32` value, a signed, zigzag-encoded 32-bit varint.
   */
  sint32(value) {
    assertInt32(value);
    value = (value << 1 ^ value >> 31) >>> 0;
    varint32write(value, this.buf);
    return this;
  }
  /**
   * Write a `fixed64` value, a signed, fixed-length 64-bit integer.
   */
  sfixed64(value) {
    let chunk = new Uint8Array(8), view = new DataView(chunk.buffer), tc = protoInt64.enc(value);
    view.setInt32(0, tc.lo, true);
    view.setInt32(4, tc.hi, true);
    return this.raw(chunk);
  }
  /**
   * Write a `fixed64` value, an unsigned, fixed-length 64 bit integer.
   */
  fixed64(value) {
    let chunk = new Uint8Array(8), view = new DataView(chunk.buffer), tc = protoInt64.uEnc(value);
    view.setInt32(0, tc.lo, true);
    view.setInt32(4, tc.hi, true);
    return this.raw(chunk);
  }
  /**
   * Write a `int64` value, a signed 64-bit varint.
   */
  int64(value) {
    let tc = protoInt64.enc(value);
    varint64write(tc.lo, tc.hi, this.buf);
    return this;
  }
  /**
   * Write a `sint64` value, a signed, zig-zag-encoded 64-bit varint.
   */
  sint64(value) {
    const tc = protoInt64.enc(value), sign = tc.hi >> 31, lo = tc.lo << 1 ^ sign, hi = (tc.hi << 1 | tc.lo >>> 31) ^ sign;
    varint64write(lo, hi, this.buf);
    return this;
  }
  /**
   * Write a `uint64` value, an unsigned 64-bit varint.
   */
  uint64(value) {
    const tc = protoInt64.uEnc(value);
    varint64write(tc.lo, tc.hi, this.buf);
    return this;
  }
};
function assertInt32(arg) {
  if (typeof arg == "string") {
    arg = Number(arg);
  } else if (typeof arg != "number") {
    throw new Error("invalid int32: " + typeof arg);
  }
  if (!Number.isInteger(arg) || arg > INT32_MAX || arg < INT32_MIN)
    throw new Error("invalid int32: " + arg);
}
function assertUInt32(arg) {
  if (typeof arg == "string") {
    arg = Number(arg);
  } else if (typeof arg != "number") {
    throw new Error("invalid uint32: " + typeof arg);
  }
  if (!Number.isInteger(arg) || arg > UINT32_MAX || arg < 0)
    throw new Error("invalid uint32: " + arg);
}
function assertFloat32(arg) {
  if (typeof arg == "string") {
    const o = arg;
    arg = Number(arg);
    if (Number.isNaN(arg) && o !== "NaN") {
      throw new Error("invalid float32: " + o);
    }
  } else if (typeof arg != "number") {
    throw new Error("invalid float32: " + typeof arg);
  }
  if (Number.isFinite(arg) && (arg > FLOAT32_MAX || arg < FLOAT32_MIN))
    throw new Error("invalid float32: " + arg);
}

// node_modules/@noble/hashes/esm/utils.js
var hasHexBuiltin = /* @__PURE__ */ (() => (
  // @ts-ignore
  typeof Uint8Array.from([]).toHex === "function" && typeof Uint8Array.fromHex === "function"
))();
var asciis = { _0: 48, _9: 57, A: 65, F: 70, a: 97, f: 102 };
function asciiToBase16(ch) {
  if (ch >= asciis._0 && ch <= asciis._9)
    return ch - asciis._0;
  if (ch >= asciis.A && ch <= asciis.F)
    return ch - (asciis.A - 10);
  if (ch >= asciis.a && ch <= asciis.f)
    return ch - (asciis.a - 10);
  return;
}
function hexToBytes(hex) {
  if (typeof hex !== "string")
    throw new Error("hex string expected, got " + typeof hex);
  if (hasHexBuiltin)
    return Uint8Array.fromHex(hex);
  const hl = hex.length;
  const al = hl / 2;
  if (hl % 2)
    throw new Error("hex string expected, got unpadded hex of length " + hl);
  const array = new Uint8Array(al);
  for (let ai = 0, hi = 0; ai < al; ai++, hi += 2) {
    const n1 = asciiToBase16(hex.charCodeAt(hi));
    const n2 = asciiToBase16(hex.charCodeAt(hi + 1));
    if (n1 === void 0 || n2 === void 0) {
      const char = hex[hi] + hex[hi + 1];
      throw new Error('hex string expected, got non-hex character "' + char + '" at index ' + hi);
    }
    array[ai] = n1 * 16 + n2;
  }
  return array;
}

// src/chains/spark/utils.ts
function extractIdentityPublicKey(sparkAddress) {
  try {
    const network = (0, import_spark_sdk.getNetworkFromSparkAddress)(
      sparkAddress
    );
    const decoded = (0, import_spark_sdk.decodeSparkAddress)(sparkAddress, network);
    return decoded.identityPublicKey || null;
  } catch (error) {
    console.error(
      "[extractIdentityPublicKey] Failed to decode Spark address:",
      {
        address: sparkAddress,
        error: error instanceof Error ? error.message : String(error)
      }
    );
    return null;
  }
}
var NEW_PREFIXES = {
  MAINNET: "spark",
  REGTEST: "sparkrt"
};
function getSparkAddressFromPubkey(identityPublicKey, network) {
  const identityPublicKeyBytes = hexToBytes(identityPublicKey);
  const w = new BinaryWriter();
  w.uint32(10).bytes(identityPublicKeyBytes);
  const serializedPayload = w.finish();
  const words = bech32m.toWords(serializedPayload);
  return bech32m.encode(NEW_PREFIXES[network], words, 1024);
}
function convertLegacyToNewFormat(legacyAddress) {
  if (legacyAddress.startsWith("sparkrt1") || legacyAddress.startsWith("spark1")) {
    console.warn(
      `Address ${legacyAddress} is already in new format, skipping conversion`
    );
    return legacyAddress;
  }
  const network = (0, import_spark_sdk.getNetworkFromSparkAddress)(
    legacyAddress
  );
  const { words } = (0, import_spark_sdk.bech32mDecode)(legacyAddress);
  return bech32m.encode(NEW_PREFIXES[network], words, 1024);
}

// src/functions/crypto/encryption.ts
var IV_LENGTH = 16;
async function encryptWithCipher({
  data,
  key,
  algorithm = "AES-GCM",
  initializationVectorSize = IV_LENGTH
}) {
  const iv = crypto2.getRandomValues(new Uint8Array(initializationVectorSize));
  const encrypted = await crypto2.subtle.encrypt(
    { name: algorithm, iv },
    key,
    new TextEncoder().encode(data)
  );
  const encoding = getEncoding();
  return JSON.stringify({
    iv: encoding.bytesToBase64(new Uint8Array(iv)),
    ciphertext: encoding.bytesToBase64(new Uint8Array(encrypted))
  });
}
async function decryptWithCipher({
  encryptedDataJSON,
  key,
  algorithm = "AES-GCM"
}) {
  const _encryptedData = JSON.parse(encryptedDataJSON);
  const encoding = getEncoding();
  const decodedIv = encoding.base64ToBytes(_encryptedData.iv);
  const decodedEncryptedData = encoding.base64ToBytes(_encryptedData.ciphertext);
  const decrypted = await crypto2.subtle.decrypt(
    { name: algorithm, iv: decodedIv.slice() },
    key,
    decodedEncryptedData.slice()
  );
  return new TextDecoder().decode(decrypted);
}
async function generateKeyPair() {
  const keyPair = await crypto2.subtle.generateKey(
    {
      name: "ECDH",
      namedCurve: "P-256"
    },
    true,
    ["deriveKey"]
  );
  const publicKey = await crypto2.subtle.exportKey("spki", keyPair.publicKey);
  const privateKey = await crypto2.subtle.exportKey("pkcs8", keyPair.privateKey);
  const encoding = getEncoding();
  const key = {
    publicKey: encoding.bytesToBase64(new Uint8Array(publicKey)),
    privateKey: encoding.bytesToBase64(new Uint8Array(privateKey))
  };
  return key;
}
async function encryptWithPublicKey({
  publicKey,
  data
}) {
  const encoding = getEncoding();
  const publicKeyBuffer = encoding.base64ToBytes(publicKey);
  const _publicKey = await crypto2.subtle.importKey(
    "spki",
    publicKeyBuffer.slice(),
    { name: "ECDH", namedCurve: "P-256" },
    false,
    []
  );
  const ephemeralKeyPair = await crypto2.subtle.generateKey(
    { name: "ECDH", namedCurve: "P-256" },
    true,
    ["deriveKey"]
  );
  const sharedSecret = await crypto2.subtle.deriveKey(
    { name: "ECDH", public: _publicKey },
    ephemeralKeyPair.privateKey,
    { name: "AES-GCM", length: 256 },
    false,
    ["encrypt"]
  );
  const iv = crypto2.getRandomValues(new Uint8Array(12));
  const encrypted = await crypto2.subtle.encrypt(
    { name: "AES-GCM", iv },
    sharedSecret,
    new TextEncoder().encode(data)
  );
  const encryptedData = {
    ephemeralPublicKey: await crypto2.subtle.exportKey(
      "spki",
      ephemeralKeyPair.publicKey
    ),
    iv,
    ciphertext: encrypted
  };
  return JSON.stringify({
    ephemeralPublicKey: encoding.bytesToBase64(new Uint8Array(encryptedData.ephemeralPublicKey)),
    iv: encoding.bytesToBase64(new Uint8Array(encryptedData.iv)),
    ciphertext: encoding.bytesToBase64(new Uint8Array(encryptedData.ciphertext))
  });
}
async function decryptWithPrivateKey({
  privateKey,
  encryptedDataJSON
}) {
  const encoding = getEncoding();
  const privateKeyBuffer = encoding.base64ToBytes(privateKey);
  const _encryptedData = JSON.parse(encryptedDataJSON);
  const encryptedData = {
    ephemeralPublicKey: encoding.base64ToBytes(_encryptedData.ephemeralPublicKey),
    iv: encoding.base64ToBytes(_encryptedData.iv),
    ciphertext: encoding.base64ToBytes(_encryptedData.ciphertext)
  };
  const _privateKey = await crypto2.subtle.importKey(
    "pkcs8",
    privateKeyBuffer.slice(),
    { name: "ECDH", namedCurve: "P-256" },
    false,
    ["deriveKey"]
  );
  const ephemeralPublicKey = await crypto2.subtle.importKey(
    "spki",
    encryptedData.ephemeralPublicKey.slice(),
    { name: "ECDH", namedCurve: "P-256" },
    false,
    []
  );
  const sharedSecret = await crypto2.subtle.deriveKey(
    { name: "ECDH", public: ephemeralPublicKey },
    _privateKey,
    { name: "AES-GCM", length: 256 },
    false,
    ["decrypt"]
  );
  const decrypted = await crypto2.subtle.decrypt(
    { name: "AES-GCM", iv: encryptedData.iv.slice() },
    sharedSecret,
    encryptedData.ciphertext.slice()
  );
  return new TextDecoder().decode(decrypted);
}

// src/functions/crypto/hash.ts
function normalizeAlgorithm(algorithm) {
  const upper = algorithm.toUpperCase();
  if (upper.startsWith("SHA") && !upper.includes("-")) {
    return upper.replace(/^SHA(\d+)$/, "SHA-$1");
  }
  return upper;
}
function generateHash({
  size = 64
}) {
  return new Promise((resolve) => {
    const bytes = getCrypto().getRandomBytes(size);
    resolve(getEncoding().bytesToHex(bytes));
  });
}
async function hashData({
  data,
  privateKey = "",
  algorithm = "sha256"
}) {
  const encoding = getEncoding();
  const crypto3 = getCrypto();
  const dataBytes = typeof data === "string" ? encoding.utf8ToBytes(data) : encoding.utf8ToBytes(String(data));
  const keyBytes = encoding.utf8ToBytes(privateKey);
  const normalizedAlgorithm = normalizeAlgorithm(algorithm);
  const signatureBytes = await crypto3.hmacSign(normalizedAlgorithm, keyBytes, dataBytes);
  return encoding.bytesToHex(signatureBytes);
}

// src/functions/crypto/index.ts
var crypto2 = {
  getRandomValues(array) {
    return getCrypto().getRandomValuesInPlace(array);
  },
  get subtle() {
    return getCrypto().getSubtleCrypto();
  }
};

// src/functions/key-shard/combine-shards-build-wallet.ts
var import_wallet = require("@meshsdk/wallet");

// src/functions/key-shard/shamir-secret-sharing.ts
function getRandomBytes(numBytes) {
  return crypto.getRandomValues(new Uint8Array(numBytes));
}
var LOG_TABLE = new Uint8Array([
  0,
  255,
  200,
  8,
  145,
  16,
  208,
  54,
  90,
  62,
  216,
  67,
  153,
  119,
  254,
  24,
  35,
  32,
  7,
  112,
  161,
  108,
  12,
  127,
  98,
  139,
  64,
  70,
  199,
  75,
  224,
  14,
  235,
  22,
  232,
  173,
  207,
  205,
  57,
  83,
  106,
  39,
  53,
  147,
  212,
  78,
  72,
  195,
  43,
  121,
  84,
  40,
  9,
  120,
  15,
  33,
  144,
  135,
  20,
  42,
  169,
  156,
  214,
  116,
  180,
  124,
  222,
  237,
  177,
  134,
  118,
  164,
  152,
  226,
  150,
  143,
  2,
  50,
  28,
  193,
  51,
  238,
  239,
  129,
  253,
  48,
  92,
  19,
  157,
  41,
  23,
  196,
  17,
  68,
  140,
  128,
  243,
  115,
  66,
  30,
  29,
  181,
  240,
  18,
  209,
  91,
  65,
  162,
  215,
  44,
  233,
  213,
  89,
  203,
  80,
  168,
  220,
  252,
  242,
  86,
  114,
  166,
  101,
  47,
  159,
  155,
  61,
  186,
  125,
  194,
  69,
  130,
  167,
  87,
  182,
  163,
  122,
  117,
  79,
  174,
  63,
  55,
  109,
  71,
  97,
  190,
  171,
  211,
  95,
  176,
  88,
  175,
  202,
  94,
  250,
  133,
  228,
  77,
  138,
  5,
  251,
  96,
  183,
  123,
  184,
  38,
  74,
  103,
  198,
  26,
  248,
  105,
  37,
  179,
  219,
  189,
  102,
  221,
  241,
  210,
  223,
  3,
  141,
  52,
  217,
  146,
  13,
  99,
  85,
  170,
  73,
  236,
  188,
  149,
  60,
  132,
  11,
  245,
  230,
  231,
  229,
  172,
  126,
  110,
  185,
  249,
  218,
  142,
  154,
  201,
  36,
  225,
  10,
  21,
  107,
  58,
  160,
  81,
  244,
  234,
  178,
  151,
  158,
  93,
  34,
  136,
  148,
  206,
  25,
  1,
  113,
  76,
  165,
  227,
  197,
  49,
  187,
  204,
  31,
  45,
  59,
  82,
  111,
  246,
  46,
  137,
  247,
  192,
  104,
  27,
  100,
  4,
  6,
  191,
  131,
  56
]);
var EXP_TABLE = new Uint8Array([
  1,
  229,
  76,
  181,
  251,
  159,
  252,
  18,
  3,
  52,
  212,
  196,
  22,
  186,
  31,
  54,
  5,
  92,
  103,
  87,
  58,
  213,
  33,
  90,
  15,
  228,
  169,
  249,
  78,
  100,
  99,
  238,
  17,
  55,
  224,
  16,
  210,
  172,
  165,
  41,
  51,
  89,
  59,
  48,
  109,
  239,
  244,
  123,
  85,
  235,
  77,
  80,
  183,
  42,
  7,
  141,
  255,
  38,
  215,
  240,
  194,
  126,
  9,
  140,
  26,
  106,
  98,
  11,
  93,
  130,
  27,
  143,
  46,
  190,
  166,
  29,
  231,
  157,
  45,
  138,
  114,
  217,
  241,
  39,
  50,
  188,
  119,
  133,
  150,
  112,
  8,
  105,
  86,
  223,
  153,
  148,
  161,
  144,
  24,
  187,
  250,
  122,
  176,
  167,
  248,
  171,
  40,
  214,
  21,
  142,
  203,
  242,
  19,
  230,
  120,
  97,
  63,
  137,
  70,
  13,
  53,
  49,
  136,
  163,
  65,
  128,
  202,
  23,
  95,
  83,
  131,
  254,
  195,
  155,
  69,
  57,
  225,
  245,
  158,
  25,
  94,
  182,
  207,
  75,
  56,
  4,
  185,
  43,
  226,
  193,
  74,
  221,
  72,
  12,
  208,
  125,
  61,
  88,
  222,
  124,
  216,
  20,
  107,
  135,
  71,
  232,
  121,
  132,
  115,
  60,
  189,
  146,
  201,
  35,
  139,
  151,
  149,
  68,
  220,
  173,
  64,
  101,
  134,
  162,
  164,
  204,
  127,
  236,
  192,
  175,
  145,
  253,
  247,
  79,
  129,
  47,
  91,
  234,
  168,
  28,
  2,
  209,
  152,
  113,
  237,
  37,
  227,
  36,
  6,
  104,
  179,
  147,
  44,
  111,
  62,
  108,
  10,
  184,
  206,
  174,
  116,
  177,
  66,
  180,
  30,
  211,
  73,
  233,
  156,
  200,
  198,
  199,
  34,
  110,
  219,
  32,
  191,
  67,
  81,
  82,
  102,
  178,
  118,
  96,
  218,
  197,
  243,
  246,
  170,
  205,
  154,
  160,
  117,
  84,
  14,
  1
]);
function add(a, b) {
  if (!Number.isInteger(a) || a < 0 || a > 255) {
    throw new RangeError("Number is out of Uint8 range");
  }
  if (!Number.isInteger(b) || b < 0 || b > 255) {
    throw new RangeError("Number is out of Uint8 range");
  }
  return a ^ b;
}
function div(a, b) {
  if (!Number.isInteger(a) || a < 0 || a > 255) {
    throw new RangeError("Number is out of Uint8 range");
  }
  if (!Number.isInteger(b) || b < 0 || b > 255) {
    throw new RangeError("Number is out of Uint8 range");
  }
  if (b === 0) {
    throw new Error("cannot divide by zero");
  }
  const logA = LOG_TABLE[a];
  const logB = LOG_TABLE[b];
  const diff = (logA - logB + 255) % 255;
  const result = EXP_TABLE[diff];
  return a === 0 ? 0 : result;
}
function mult(a, b) {
  if (!Number.isInteger(a) || a < 0 || a > 255) {
    throw new RangeError("Number is out of Uint8 range");
  }
  if (!Number.isInteger(b) || b < 0 || b > 255) {
    throw new RangeError("Number is out of Uint8 range");
  }
  const logA = LOG_TABLE[a];
  const logB = LOG_TABLE[b];
  const sum = (logA + logB) % 255;
  const result = EXP_TABLE[sum];
  return a === 0 || b === 0 ? 0 : result;
}
function interpolatePolynomial(xSamples, ySamples, x) {
  if (xSamples.length !== ySamples.length) {
    throw new Error("sample length mistmatch");
  }
  const limit = xSamples.length;
  let basis = 0;
  let result = 0;
  for (let i = 0; i < limit; i++) {
    basis = 1;
    for (let j = 0; j < limit; ++j) {
      if (i === j) {
        continue;
      }
      const num = add(x, xSamples[j]);
      const denom = add(xSamples[i], xSamples[j]);
      const term = div(num, denom);
      basis = mult(basis, term);
    }
    result = add(result, mult(ySamples[i], basis));
  }
  return result;
}
function evaluate(coefficients, x, degree) {
  if (x === 0) {
    throw new Error("cannot evaluate secret polynomial at zero");
  }
  let result = coefficients[degree];
  for (let i = degree - 1; i >= 0; i--) {
    const coefficient = coefficients[i];
    result = add(mult(result, x), coefficient);
  }
  return result;
}
function getRandomByte() {
  return getRandomBytes(1)[0];
}
function getNonZeroRandomByte() {
  while (true) {
    const byte = getRandomByte();
    if (byte > 0) {
      return byte;
    }
  }
}
function newCoefficients(intercept, degree) {
  const coefficients = new Uint8Array(degree + 1);
  coefficients[0] = intercept;
  for (let i = 1; i <= degree; i++) {
    const coefficientTMinus1 = i === degree;
    coefficients[i] = coefficientTMinus1 ? getNonZeroRandomByte() : getRandomByte();
  }
  return coefficients;
}
function newCoordinates() {
  const coordinates = new Uint8Array(255);
  for (let i = 0; i < 255; i++) {
    coordinates[i] = i + 1;
  }
  const randomIndices = getRandomBytes(255);
  for (let i = 0; i < 255; i++) {
    const j = randomIndices[i] % 255;
    const temp = coordinates[i];
    coordinates[i] = coordinates[j];
    coordinates[j] = temp;
  }
  return coordinates;
}
var AssertArgument = {
  instanceOf(object, constructor, message) {
    if (object.constructor !== constructor) {
      throw new TypeError(message);
    }
  },
  inRange(n, start, until, message) {
    if (!(start < until && n >= start && n < until)) {
      throw new RangeError(message);
    }
  },
  greaterThanOrEqualTo(a, b, message) {
    if (a < b) {
      throw new Error(message);
    }
  },
  equalTo(a, b, message) {
    if (a !== b) {
      throw new Error(message);
    }
  }
};
async function shamirSplit(secret, shares, threshold) {
  AssertArgument.instanceOf(secret, Uint8Array, "secret must be a Uint8Array");
  AssertArgument.greaterThanOrEqualTo(
    secret.byteLength,
    1,
    "secret cannot be empty"
  );
  AssertArgument.instanceOf(shares, Number, "shares must be a number");
  AssertArgument.inRange(
    shares,
    2,
    256,
    "shares must be at least 2 and at most 255"
  );
  AssertArgument.instanceOf(threshold, Number, "threshold must be a number");
  AssertArgument.inRange(
    threshold,
    2,
    256,
    "threshold must be at least 2 and at most 255"
  );
  AssertArgument.greaterThanOrEqualTo(
    shares,
    threshold,
    "shares cannot be less than threshold"
  );
  const result = [];
  const secretLength = secret.byteLength;
  const xCoordinates = newCoordinates();
  for (let i = 0; i < shares; i++) {
    const share = new Uint8Array(secretLength + 1);
    share[secretLength] = xCoordinates[i];
    result.push(share);
  }
  const degree = threshold - 1;
  for (let i = 0; i < secretLength; i++) {
    const byte = secret[i];
    const coefficients = newCoefficients(byte, degree);
    for (let j = 0; j < shares; ++j) {
      const x = xCoordinates[j];
      const y = evaluate(coefficients, x, degree);
      result[j][i] = y;
    }
  }
  return result;
}
async function shamirCombine(shares) {
  AssertArgument.instanceOf(shares, Array, "shares must be an Array");
  AssertArgument.inRange(
    shares.length,
    2,
    256,
    "shares must have at least 2 and at most 255 elements"
  );
  const share1 = shares[0];
  AssertArgument.instanceOf(
    share1,
    Uint8Array,
    "each share must be a Uint8Array"
  );
  for (const share of shares) {
    AssertArgument.instanceOf(
      share,
      Uint8Array,
      "each share must be a Uint8Array"
    );
    AssertArgument.greaterThanOrEqualTo(
      share.byteLength,
      2,
      "each share must be at least 2 bytes"
    );
    AssertArgument.equalTo(
      share.byteLength,
      share1.byteLength,
      "all shares must have the same byte length"
    );
  }
  const sharesLength = shares.length;
  const shareLength = share1.byteLength;
  const secretLength = shareLength - 1;
  const secret = new Uint8Array(secretLength);
  const xSamples = new Uint8Array(sharesLength);
  const ySamples = new Uint8Array(sharesLength);
  const samples = /* @__PURE__ */ new Set();
  for (let i = 0; i < sharesLength; i++) {
    const share = shares[i];
    const sample = share[shareLength - 1];
    if (samples.has(sample)) {
      throw new Error(
        "shares must contain unique values but a duplicate was found"
      );
    }
    samples.add(sample);
    xSamples[i] = sample;
  }
  for (let i = 0; i < secretLength; i++) {
    for (let j = 0; j < sharesLength; ++j) {
      ySamples[j] = shares[j][i];
    }
    secret[i] = interpolatePolynomial(xSamples, ySamples, 0);
  }
  return secret;
}

// src/functions/convertors.ts
function stringToBytes(data) {
  return new TextEncoder().encode(data);
}
function bytesToString(data) {
  return new TextDecoder().decode(data);
}
function bytesToHex2(bytes) {
  return Array.from(bytes).map((byte) => byte.toString(16).padStart(2, "0")).join("");
}
function hexToBytes2(hex) {
  const bytes = new Uint8Array(hex.length / 2);
  for (let i = 0; i < bytes.length; i++) {
    bytes[i] = parseInt(hex.substr(i * 2, 2), 16);
  }
  return bytes;
}

// src/functions/key-shard/combine-shards-build-wallet.ts
var import_bitcoin2 = require("@meshsdk/bitcoin");
var import_spark_sdk2 = require("@buildonspark/spark-sdk");
async function combineShardsBuildWallet(networkId, keyShard1, keyShard2, bitcoinProvider, fetcher) {
  const _share1 = hexToBytes2(keyShard1);
  const _share2 = hexToBytes2(keyShard2);
  const reconstructed = await shamirCombine([_share1, _share2]);
  const mnemonic = bytesToString(reconstructed);
  const bitcoinWallet = new import_bitcoin2.EmbeddedWallet({
    network: networkId === 1 ? "Mainnet" : "Testnet",
    key: {
      type: "mnemonic",
      words: mnemonic.split(" ")
    },
    provider: bitcoinProvider
  });
  const cardanoWallet = await import_wallet.MeshCardanoHeadlessWallet.fromMnemonic({
    mnemonic: mnemonic.split(" "),
    networkId,
    walletAddressType: 1,
    fetcher
  });
  const { wallet: sparkWallet } = await import_spark_sdk2.SparkWallet.initialize({
    mnemonicOrSeed: mnemonic,
    options: {
      network: networkId === 1 ? "MAINNET" : "REGTEST"
    }
  });
  return {
    key: mnemonic,
    bitcoinWallet,
    cardanoWallet,
    sparkWallet
  };
}

// src/functions/key-shard/spilt-key-into-shards.ts
async function spiltKeyIntoShards(key) {
  const secret = stringToBytes(key);
  const [share1, share2, share3] = await shamirSplit(secret, 3, 2);
  const keyShare1 = bytesToHex2(share1);
  const keyShare2 = bytesToHex2(share2);
  const keyShare3 = bytesToHex2(share3);
  return [keyShare1, keyShare2, keyShare3];
}

// src/functions/client/derive-wallet.ts
async function clientDeriveWallet(encryptedKeyShard, deviceShardEncryptionKey, custodialShard, networkId, bitcoinProvider, fetcher) {
  const keyShare1 = await decryptWithCipher({
    encryptedDataJSON: encryptedKeyShard,
    key: deviceShardEncryptionKey
  });
  const keyShare2 = custodialShard;
  const { bitcoinWallet, cardanoWallet, sparkWallet, key } = await combineShardsBuildWallet(
    networkId,
    keyShare1,
    keyShare2,
    bitcoinProvider,
    fetcher
  );
  return { bitcoinWallet, cardanoWallet, sparkWallet, key };
}

// src/functions/client/generate-wallet.ts
var import_wallet2 = require("@meshsdk/wallet");
var import_common2 = require("@meshsdk/common");
var import_core_cst2 = require("@meshsdk/core-cst");
var import_bitcoin3 = require("@meshsdk/bitcoin");
var import_spark_sdk3 = require("@buildonspark/spark-sdk");
async function clientGenerateWallet(deviceShardEncryptionKey, recoveryShardEncryptionKey) {
  const mnemonic = await (0, import_common2.generateMnemonic)(256);
  const [keyShare1, keyShare2, keyShare3] = await spiltKeyIntoShards(mnemonic);
  const encryptedDeviceShard = await encryptWithCipher({
    data: keyShare1,
    key: deviceShardEncryptionKey
  });
  const encryptedRecoveryShard = await encryptWithCipher({
    data: keyShare3,
    key: recoveryShardEncryptionKey
  });
  const bitcoinTestnetWallet = new import_bitcoin3.EmbeddedWallet({
    network: "Testnet",
    key: {
      type: "mnemonic",
      words: mnemonic.split(" ")
    }
  });
  const bitcoinMainnetWallet = new import_bitcoin3.EmbeddedWallet({
    network: "Mainnet",
    key: {
      type: "mnemonic",
      words: mnemonic.split(" ")
    }
  });
  const bitcoinTestnetPubKeyHash = bitcoinTestnetWallet.getPublicKey();
  const bitcoinMainnetPubKeyHash = bitcoinMainnetWallet.getPublicKey();
  const cardanoWallet = await import_wallet2.MeshCardanoHeadlessWallet.fromMnemonic({
    mnemonic: mnemonic.split(" "),
    networkId: 1,
    walletAddressType: 1
  });
  const cardanoBaseAddress = await cardanoWallet.getChangeAddressBech32();
  const cardanoKeyHashes = (0, import_core_cst2.deserializeBech32Address)(
    cardanoBaseAddress
  );
  const { wallet: sparkMainnetWallet } = await import_spark_sdk3.SparkWallet.initialize({
    mnemonicOrSeed: mnemonic,
    options: {
      network: "MAINNET"
    }
  });
  const { wallet: sparkRegtestWallet } = await import_spark_sdk3.SparkWallet.initialize({
    mnemonicOrSeed: mnemonic,
    options: {
      network: "REGTEST"
    }
  });
  const [
    sparkMainnetPubKeyHash,
    sparkRegtestPubKeyHash,
    sparkMainnetStaticDepositAddress,
    sparkRegtestStaticDepositAddress
  ] = await Promise.all([
    sparkMainnetWallet.getIdentityPublicKey(),
    sparkRegtestWallet.getIdentityPublicKey(),
    sparkMainnetWallet.getStaticDepositAddress(),
    sparkRegtestWallet.getStaticDepositAddress()
  ]);
  return {
    encryptedDeviceShard,
    authShard: keyShare2,
    encryptedRecoveryShard,
    bitcoinMainnetPubKeyHash,
    bitcoinTestnetPubKeyHash,
    cardanoPubKeyHash: cardanoKeyHashes.pubKeyHash,
    cardanoStakeCredentialHash: cardanoKeyHashes.stakeCredentialHash,
    sparkMainnetPubKeyHash,
    sparkRegtestPubKeyHash,
    sparkMainnetStaticDepositAddress,
    sparkRegtestStaticDepositAddress
  };
}

// src/functions/client/recovery.ts
async function clientRecovery(authShard, recoveryShard, recoveryShardEncryptionKey, newDeviceShardEncryptionKey) {
  try {
    const recoverKeyShare3 = await decryptWithCipher({
      encryptedDataJSON: recoveryShard,
      key: recoveryShardEncryptionKey
    });
    const recoverKeyShare2 = authShard;
    const { key } = await combineShardsBuildWallet(
      0,
      // dont care about network here
      recoverKeyShare2,
      recoverKeyShare3
    );
    const [keyShare1, keyShare2] = await spiltKeyIntoShards(key);
    const encryptedAuthKey = await encryptWithCipher({
      data: keyShare1,
      key: newDeviceShardEncryptionKey
    });
    return {
      deviceShard: encryptedAuthKey,
      authShard: keyShare2,
      fullKey: key
    };
  } catch (e) {
    console.error(e);
    throw new Error("Invalid recovery answer");
  }
}

// src/functions/window/open-window.ts
async function openWindow(params, appUrl = "https://utxos.dev/") {
  const p = new URLSearchParams(params);
  const url = `${appUrl}/client/wallet?${p.toString()}`;
  const result = await getLinking().openAuthWindow(url, "utxos://callback");
  if (result.error === "cancelled") {
    return { success: false, message: "Window was closed by the user" };
  }
  if (result.error) {
    return {
      success: false,
      message: result.errorDescription || result.error
    };
  }
  return result.data;
}

// src/non-custodial/utils.ts
var IV_LENGTH2 = 16;
var staticSalt = new Uint8Array(new Array(32).fill(1)).buffer;
var NoPRFExtensionError = class extends Error {
  constructor(message = "Credential does not support PRF extension") {
    super(message);
    this.name = "NoPRFExtensionError";
  }
};
var UserAbortedError = class extends Error {
  constructor(message = "User aborted passkey flow") {
    super(message);
    this.name = "UserAbortedError";
  }
};
var NotAllowedError = class extends Error {
  constructor(message = "Passkey operation was not allowed") {
    super(message);
    this.name = "NotAllowedError";
  }
};
async function secretToCryptoKey(secret, algorithm = "AES-GCM") {
  const keyMaterial = await crypto.subtle.importKey(
    "raw",
    new TextEncoder().encode(secret),
    { name: "PBKDF2" },
    false,
    ["deriveKey"]
  );
  return await crypto.subtle.deriveKey(
    {
      name: "PBKDF2",
      salt: new Uint8Array(IV_LENGTH2),
      iterations: 1e5,
      hash: "SHA-256"
    },
    keyMaterial,
    { name: algorithm, length: 256 },
    false,
    ["encrypt", "decrypt"]
  );
}
async function webauthnPublicKeyCredentialToCryptoKey(credential, algorithm = "AES-GCM") {
  const extensionResults = credential.getClientExtensionResults();
  const first = extensionResults.prf?.results?.first;
  if (first === void 0) {
    throw new Error("PRF extension not supported or didn't return results");
  }
  return await crypto.subtle.importKey(
    "raw",
    first,
    { name: algorithm, length: 256 },
    false,
    ["encrypt", "decrypt"]
  );
}
async function getCredentialFromCredentialId(credentialId, rp) {
  const opts = {
    publicKey: {
      challenge: new Uint8Array(32).fill(2),
      // Same challenge for consistency
      rpId: rp.id,
      allowCredentials: [
        {
          id: base64urlToUint8Array(credentialId),
          type: "public-key"
        }
      ],
      authenticatorSelection: {
        userVerification: "required",
        residentKey: "required",
        requireResidentKey: true
      },
      extensions: {
        prf: {
          eval: {
            first: staticSalt
          }
        }
      }
    }
  };
  let credential;
  try {
    credential = await navigator.credentials.get(opts);
  } catch (e) {
    if (e instanceof DOMException) {
      if (e.name === "AbortError") {
        return { data: null, error: new UserAbortedError() };
      }
    }
    return { data: null, error: new NotAllowedError() };
  }
  const extensionResults = credential.getClientExtensionResults();
  const first = extensionResults.prf?.results?.first;
  if (first === void 0) {
    throw new Error("PRF extension not supported or didn't return results");
  }
  return { data: { credential }, error: null };
}
async function createCredential(rp) {
  const buf = crypto.getRandomValues(new Uint8Array(32)).buffer;
  const opts = {
    publicKey: {
      challenge: new Uint8Array([1, 2, 3, 4]),
      // Example value
      rp,
      user: {
        id: buf,
        name: "UTXOS Private Key",
        displayName: "UTXOS Private Key"
      },
      pubKeyCredParams: [
        { alg: -8, type: "public-key" },
        // Ed25519
        { alg: -7, type: "public-key" },
        // ES256
        { alg: -257, type: "public-key" }
        // RS256
      ],
      authenticatorSelection: {
        userVerification: "required",
        residentKey: "required",
        requireResidentKey: true
      },
      extensions: {
        prf: {
          eval: {
            first: staticSalt
          }
        },
        credProps: true
      }
    }
  };
  let credential;
  try {
    credential = await navigator.credentials.create(
      opts
    );
  } catch (e) {
    if (e instanceof DOMException) {
      if (e.name === "AbortError") {
        return { data: null, error: new UserAbortedError() };
      }
    }
    return { data: null, error: new NotAllowedError() };
  }
  const extensionResults = credential.getClientExtensionResults();
  const first = extensionResults.prf?.results?.first;
  if (first === void 0) {
    return { data: null, error: new NoPRFExtensionError() };
  }
  return {
    data: {
      credential,
      prfOutput: first
    },
    error: null
  };
}
function base64urlToUint8Array(base64url) {
  const s = base64url.replace(/\s+/g, "");
  const base64 = s.replace(/-/g, "+").replace(/_/g, "/");
  const pad = "=".repeat((4 - base64.length % 4) % 4);
  const binary = atob(base64 + pad);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) bytes[i] = binary.charCodeAt(i);
  return bytes;
}

// src/non-custodial/index.ts
var AUTH_KEY = "mesh-web3-services-auth";
var LOCAL_SHARD_KEY = "mesh-web3-services-local-shard";
var NotAuthenticatedError = class extends Error {
  constructor(message = "User is not authenticated") {
    super(message);
    this.name = "NotAuthenticatedError";
  }
};
var WalletServerRetrievalError = class extends Error {
  constructor(message = "Unable to get custodial shards from the server.") {
    super(message);
    this.name = "WalletServerRetrievalError";
  }
};
var WalletServerCreationError = class extends Error {
  constructor(message = "Unable to create custodial shards on the server.") {
    super(message);
    this.name = "WalletServerCreationError";
  }
};
var SessionExpiredError = class extends Error {
  constructor(message = "User session has expired") {
    super(message);
    this.name = "SessionExpiredError";
  }
};
var AuthRouteError = class extends Error {
  constructor(message = "Unable to finish authentication process.") {
    super(message);
    this.name = "AuthRouteError";
  }
};
var StorageRetrievalError = class extends Error {
  constructor(message = "Unable to retrieve key from storage.") {
    super(message);
    this.name = "StorageRetrievalError";
  }
};
var StorageInsertError = class extends Error {
  constructor(message = "Unable to insert data in storage.") {
    super(message);
    this.name = "StorageInsertError";
  }
};
var Web3NonCustodialProvider = class {
  projectId;
  appOrigin;
  storageLocation;
  googleOauth2ClientId;
  twitterOauth2ClientId;
  discordOauth2ClientId;
  appleOauth2ClientId;
  constructor(params) {
    this.projectId = params.projectId;
    this.appOrigin = params.appOrigin ? params.appOrigin : "https://utxos.dev";
    this.storageLocation = params.storageLocation ? params.storageLocation : "local_storage";
    this.googleOauth2ClientId = params.googleOauth2ClientId;
    this.twitterOauth2ClientId = params.twitterOauth2ClientId;
    this.discordOauth2ClientId = params.discordOauth2ClientId;
    this.appleOauth2ClientId = params.appleOauth2ClientId;
  }
  base64Encode(str) {
    const encoding = getEncoding();
    return encoding.bytesToBase64(encoding.utf8ToBytes(str));
  }
  base64Decode(base64) {
    const encoding = getEncoding();
    const normalized = base64.replace(/-/g, "+").replace(/_/g, "/");
    return encoding.bytesToUtf8(encoding.base64ToBytes(normalized));
  }
  async checkNonCustodialWalletsOnServer() {
    const { data: user, error: userError } = await this.getUser();
    if (userError) {
      return { error: userError, data: null };
    }
    const result = await fetch(this.appOrigin + "/api/wallet", {
      headers: { Authorization: "Bearer " + user.token }
    });
    const wallets = await result.json();
    return { data: wallets, error: null };
  }
  async getWallets() {
    const { data: user, error: userError } = await this.getUser();
    if (userError) {
      return { error: userError, data: null };
    }
    const { data: localShards, error: localShardError } = await this.getFromStorage(
      LOCAL_SHARD_KEY + user.id
    );
    if (localShardError) {
      return { error: localShardError, data: null };
    }
    const ids = localShards.map((item) => item.deviceId);
    const params = new URLSearchParams();
    ids.forEach((id) => params.append("ids", id));
    const res = await fetch(
      this.appOrigin + "/api/device?" + params.toString(),
      { headers: { Authorization: "Bearer " + user.token } }
    );
    if (res.ok === false) {
      return {
        data: null,
        error: new WalletServerRetrievalError(
          "Retrieving wallets from the server failed with status " + res.status
        )
      };
    }
    const walletDevices = await res.json();
    const custodialWallets = walletDevices.map((device) => {
      const localShard = localShards.find(
        (item) => item.deviceId === device.deviceId
      );
      const i = {
        deviceId: device.deviceId,
        walletId: device.walletId,
        authShard: device.authShard,
        userAgent: device.userAgent,
        localShard: localShard.keyShard,
        webauthnCredentialId: device.webauthnCredentialId
      };
      return i;
    });
    return {
      data: custodialWallets,
      error: null
    };
  }
  async createWallet(deviceShardEncryptionKey, recoveryShardEncryptionKey, recoveryQuestion, webauthnCredentialId) {
    const userAgent = getLinking().getUserAgent() ?? "unknown";
    const { data: user, error: userError } = await this.getUser();
    if (userError) {
      return { error: userError, data: null };
    }
    const {
      cardanoPubKeyHash,
      cardanoStakeCredentialHash,
      encryptedDeviceShard,
      encryptedRecoveryShard,
      authShard,
      bitcoinMainnetPubKeyHash,
      bitcoinTestnetPubKeyHash,
      sparkMainnetPubKeyHash,
      sparkRegtestPubKeyHash,
      sparkMainnetStaticDepositAddress,
      sparkRegtestStaticDepositAddress
    } = await clientGenerateWallet(
      deviceShardEncryptionKey,
      recoveryShardEncryptionKey
    );
    const body = {
      userAgent,
      projectId: this.projectId,
      recoveryShard: encryptedRecoveryShard,
      authShard,
      bitcoinMainnetPubKeyHash,
      bitcoinTestnetPubKeyHash,
      cardanoPubKeyHash,
      cardanoStakeCredentialHash,
      sparkMainnetPubKeyHash,
      sparkRegtestPubKeyHash,
      webauthnCredentialId,
      recoveryShardQuestion: recoveryQuestion,
      sparkMainnetStaticDepositAddress,
      sparkRegtestStaticDepositAddress
    };
    const res = await fetch(this.appOrigin + "/api/wallet", {
      method: "POST",
      headers: { Authorization: "Bearer " + user.token },
      body: JSON.stringify(body)
    });
    if (res.ok === false) {
      return {
        error: new WalletServerCreationError(
          "Retrieving wallets from the server failed with status " + res.status
        ),
        data: null
      };
    }
    const result = await res.json();
    await this.pushDevice({
      deviceId: result.deviceId,
      encryptedDeviceShard,
      walletId: result.walletId
    });
    return {
      data: { deviceId: result.deviceId, walletId: result.walletId },
      error: null
    };
  }
  async getUser() {
    const { data } = await this.getFromStorage(AUTH_KEY);
    if (data === null) {
      return { data: null, error: new NotAuthenticatedError() };
    }
    const parts = data.jwt.split(".");
    const bodyUnparsed = parts[1];
    if (bodyUnparsed === void 0) {
      return { data: null, error: new NotAuthenticatedError() };
    }
    const body = JSON.parse(this.base64Decode(bodyUnparsed));
    if (body.exp < Date.now()) {
      return { data: null, error: new SessionExpiredError() };
    }
    return {
      data: {
        id: body.sub,
        scopes: body.scopes,
        provider: body.provider,
        providerId: body.providerId,
        avatarUrl: body.avatarUrl,
        email: body.email,
        username: body.username,
        token: data.jwt
      },
      error: null
    };
  }
  async performRecovery(walletId, recoveryShardEncryptionKey, newDeviceShardEncryptionKey) {
    const { data: user, error: userError } = await this.getUser();
    if (userError) {
      return { error: userError, data: null };
    }
    const res = await fetch(this.appOrigin + "/api/wallet/" + walletId, {
      headers: { Authorization: "Bearer " + user.token }
    });
    if (res.ok === false) {
      return {
        error: new WalletServerRetrievalError(
          "Retrieving wallet " + walletId + "from the server failed with status " + res.status
        ),
        data: null
      };
    }
    const wallet = await res.json();
    const { authShard, deviceShard, fullKey } = await clientRecovery(
      wallet.authShard,
      wallet.recoveryShard,
      recoveryShardEncryptionKey,
      newDeviceShardEncryptionKey
    );
    const userAgent = getLinking().getUserAgent() ?? "unknown";
    const createDeviceBody = {
      walletId,
      authShard,
      userAgent
    };
    const createDeviceRes = await fetch(this.appOrigin + "/api/device", {
      method: "POST",
      headers: { Authorization: "Bearer " + user.token },
      body: JSON.stringify(createDeviceBody)
    });
    if (createDeviceRes.ok === false) {
      return {
        error: new WalletServerCreationError(
          "Retrieving wallets from the server failed with status " + createDeviceRes.status
        ),
        data: null
      };
    }
    const newDeviceWallet = await createDeviceRes.json();
    await this.pushDevice({
      deviceId: newDeviceWallet.deviceId,
      encryptedDeviceShard: deviceShard,
      walletId
    });
    return { error: null, data: { fullKey } };
  }
  signInWithProvider(provider, redirectUrl, callback) {
    if (provider === "google") {
      const googleState = JSON.stringify({
        redirect: redirectUrl,
        provider: "google",
        projectId: this.projectId
      });
      const googleSearchParams = new URLSearchParams({
        client_id: this.googleOauth2ClientId,
        response_type: "code",
        redirect_uri: this.appOrigin + "/api/auth",
        scope: "https://www.googleapis.com/auth/userinfo.email https://www.googleapis.com/auth/userinfo.profile",
        state: this.base64Encode(googleState)
      });
      const googleAuthorizeUrl = "https://accounts.google.com/o/oauth2/v2/auth?" + googleSearchParams.toString();
      callback(googleAuthorizeUrl);
      return;
    } else if (provider === "discord") {
      const discordState = JSON.stringify({
        redirect: redirectUrl,
        provider: "discord",
        projectId: this.projectId
      });
      const discordSearchParams = new URLSearchParams({
        client_id: this.discordOauth2ClientId,
        response_type: "code",
        redirect_uri: this.appOrigin + "/api/auth",
        scope: "identify email",
        state: this.base64Encode(discordState)
      });
      const discordAuthorizeUrl = "https://discord.com/oauth2/authorize?" + discordSearchParams.toString();
      callback(discordAuthorizeUrl);
      return;
    } else if (provider === "twitter") {
      const twitterState = JSON.stringify({
        redirect: redirectUrl,
        provider: "twitter",
        projectId: this.projectId
      });
      const twitterSearchParams = new URLSearchParams({
        response_type: "code",
        client_id: this.twitterOauth2ClientId,
        redirect_uri: this.appOrigin + "/api/auth",
        scope: "users.read+tweet.read+offline.access+users.email",
        state: this.base64Encode(twitterState),
        code_challenge: "challenge",
        code_challenge_method: "plain"
      });
      const twitterAuthorizeUrl = "https://x.com/i/oauth2/authorize?" + twitterSearchParams.toString();
      callback(twitterAuthorizeUrl);
      return;
    } else if (provider === "apple") {
      const appleState = JSON.stringify({
        redirect: redirectUrl,
        provider: "apple",
        projectId: this.projectId
      });
      const appleSearchParams = new URLSearchParams({
        client_id: this.appleOauth2ClientId,
        response_type: "code",
        redirect_uri: this.appOrigin + "/api/auth",
        response_mode: "form_post",
        scope: "name email",
        state: this.base64Encode(appleState)
      });
      const appleAuthorizeUrl = "https://appleid.apple.com/auth/authorize?" + appleSearchParams.toString();
      callback(appleAuthorizeUrl);
      return;
    } else if (provider === "email") {
      throw new Error("Email provider uses OTP flow. Use the email OTP API endpoints instead.");
    }
  }
  /** Always place under /auth/mesh */
  async handleAuthenticationRoute() {
    const linking = getLinking();
    const urlParams = linking.getURLParams();
    console.log("Logging params:", urlParams);
    const token = urlParams["token"] ?? null;
    const redirect = urlParams["redirect"] ?? null;
    console.log(
      "Logging from inside handleAuthenticationRoute:",
      token,
      redirect
    );
    if (token && redirect) {
      await this.putInStorage(AUTH_KEY, { jwt: token });
      await linking.redirectURL(redirect);
      return;
    }
    return {
      error: new AuthRouteError(
        `Either token or redirect are undefined. ?token=${token}, ?redirect=${redirect}`
      )
    };
  }
  /**
   * Sends OTP to email address
   * @param email - The email address to send OTP to
   * @returns Promise that resolves when OTP is sent
   */
  async sendEmailOtp(email) {
    const res = await fetch(this.appOrigin + "/api/auth/email/send-otp", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ email, projectId: this.projectId })
    });
    if (!res.ok) {
      const data = await res.json();
      return { error: new Error(data.error) };
    }
    return { error: null };
  }
  /**
   * Verifies OTP and stores JWT (same pattern as OAuth callback)
   * @param email - The email address used to send OTP
   * @param code - The 6-digit OTP code
   * @returns User data on success, error on failure
   */
  async verifyEmailOtp(email, code) {
    const res = await fetch(this.appOrigin + "/api/auth/email/verify-otp", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ email, code, projectId: this.projectId })
    });
    if (!res.ok) {
      const data = await res.json();
      return { data: null, error: new Error(data.error) };
    }
    const { token } = await res.json();
    await this.putInStorage(AUTH_KEY, { jwt: token });
    return this.getUser();
  }
  async putInStorage(key, data) {
    if (this.storageLocation === "chrome_local") {
      await chrome.storage.local.set({ [key]: data });
    } else if (this.storageLocation === "chrome_sync") {
      await chrome.storage.sync.set({ [key]: data });
    } else if (this.storageLocation === "local_storage") {
      await getStorage().setItem(key, JSON.stringify(data));
    }
  }
  async pushDevice(deviceWallet) {
    const { data: user, error: userError } = await this.getUser();
    if (userError) {
      return { error: userError };
    }
    let { data, error } = await this.getFromStorage(
      LOCAL_SHARD_KEY + user.id
    );
    if (data === null) {
      console.log(
        "We are expecting the error here to be that no local shard wallet objects exists yet: " + error.message
      );
      data = [];
    }
    data.push({
      deviceId: deviceWallet.deviceId,
      keyShard: deviceWallet.encryptedDeviceShard,
      walletId: deviceWallet.walletId
    });
    await this.putInStorage(
      LOCAL_SHARD_KEY + user.id,
      data
    );
    return { error: null };
  }
  async getFromStorage(key) {
    if (this.storageLocation === "chrome_local") {
      const query = await chrome.storage.local.get([key]);
      const data = query[key];
      if (data) {
        return { data, error: null };
      } else {
        return {
          data: null,
          error: new StorageRetrievalError(
            `Unable to retrieve key ${key} from chrome.storage.local.`
          )
        };
      }
    } else if (this.storageLocation === "chrome_sync") {
      const query = await chrome.storage.sync.get([key]);
      const data = query[key];
      if (data) {
        return { data, error: null };
      } else {
        return {
          data: null,
          error: new StorageRetrievalError(
            `Unable to retrieve key ${key} from chrome.storage.sync.`
          )
        };
      }
    } else if (this.storageLocation === "local_storage") {
      const data = await getStorage().getItem(key);
      if (data) {
        return {
          data: JSON.parse(data),
          error: null
        };
      } else {
        return {
          data: null,
          error: new StorageRetrievalError(
            `Unable to retrieve key ${key} from storage.`
          )
        };
      }
    }
    return {
      data: null,
      error: new StorageRetrievalError(
        "Class missing a valid storage location."
      )
    };
  }
};

// src/sdk/index.ts
var import_axios2 = __toESM(require("axios"), 1);

// src/sdk/wallet-developer-controlled/cardano.ts
var import_wallet3 = require("@meshsdk/wallet");
var CardanoWalletDeveloperControlled = class {
  sdk;
  constructor({
    sdk
  }) {
    this.sdk = sdk;
  }
  /**
   * Retrieves all Cardano wallets for the project.
   *
   * @returns Promise resolving to array of wallet information
   *
   * @example
   * ```typescript
   * const wallets = await sdk.wallet.cardano.getWallets();
   * console.log(`Found ${wallets.length} Cardano wallets`);
   * ```
   */
  async getWallets() {
    const { data, status } = await this.sdk.axiosInstance.get(
      `api/project-wallet/${this.sdk.projectId}/cardano`
    );
    if (status === 200) {
      return data;
    }
    throw new Error("Failed to get Cardano wallets");
  }
  /**
   * Retrieves a specific Cardano wallet by ID and initializes a MeshCardanoHeadlessWallet instance.
   *
   * @param walletId - The wallet ID to retrieve
   * @param decryptKey - If true, returns the decrypted mnemonic in wallet info
   * @returns Promise resolving to wallet info and initialized MeshCardanoHeadlessWallet
   *
   * @example
   * ```typescript
   * const { info, wallet } = await sdk.wallet.cardano.getWallet("wallet-id");
   * const address = await wallet.getChangeAddressBech32();
   * console.log("Base address:", address);
   * ```
   */
  async getWallet(walletId, decryptKey = false) {
    if (this.sdk.privateKey === void 0) {
      throw new Error("Private key not found");
    }
    const { data, status } = await this.sdk.axiosInstance.get(
      `api/project-wallet/${this.sdk.projectId}/${walletId}?chain=cardano`
    );
    if (status === 200) {
      const web3Wallet = data;
      const mnemonic = await decryptWithPrivateKey({
        privateKey: this.sdk.privateKey,
        encryptedDataJSON: web3Wallet.key
      });
      if (decryptKey) {
        web3Wallet.key = mnemonic;
      }
      const networkId = this.sdk.network === "mainnet" ? 1 : 0;
      const wallet = await import_wallet3.MeshCardanoHeadlessWallet.fromMnemonic({
        mnemonic: mnemonic.split(" "),
        networkId,
        walletAddressType: 1
      });
      return { info: web3Wallet, wallet };
    }
    throw new Error("Failed to get Cardano wallet");
  }
  /**
   * Gets Cardano wallets filtered by tag.
   *
   * @param tag - The tag to filter by
   * @returns Promise resolving to array of matching wallet information
   *
   * @example
   * ```typescript
   * const wallets = await sdk.wallet.cardano.getWalletsByTag("treasury");
   * ```
   */
  async getWalletsByTag(tag) {
    if (this.sdk.privateKey === void 0) {
      throw new Error("Private key not found");
    }
    const { data, status } = await this.sdk.axiosInstance.get(
      `api/project-wallet/${this.sdk.projectId}/cardano/tag/${tag}`
    );
    if (status === 200) {
      return data;
    }
    throw new Error("Failed to get Cardano wallets by tag");
  }
};

// src/sdk/wallet-developer-controlled/spark-issuer.ts
var SparkIssuerWalletDeveloperControlled = class {
  sdk;
  constructor({ sdk }) {
    this.sdk = sdk;
  }
  /**
   * Gets wallet info by ID.
   *
   * @param walletId The wallet ID to retrieve
   * @returns Promise resolving to wallet info
   *
   * @example
   * ```typescript
   * const walletInfo = await sdk.wallet.sparkIssuer.get("existing-wallet-id");
   * ```
   */
  async getWallet(walletId) {
    const networkParam = this.sdk.network === "mainnet" ? "mainnet" : "regtest";
    const { data, status } = await this.sdk.axiosInstance.get(
      `api/project-wallet/${this.sdk.projectId}/${walletId}?chain=spark&network=${networkParam}`
    );
    if (status === 200) {
      return data;
    }
    throw new Error("Failed to get Spark wallet");
  }
  /**
   * Lists all Spark wallets for the current project.
   * Returns basic wallet information for selection/management purposes.
   *
   * @returns Promise resolving to array of wallet information
   *
   * @example
   * ```typescript
   * const wallets = await sdk.wallet.sparkIssuer.list();
   * console.log(`Found ${wallets.length} Spark wallets:`);
   * wallets.forEach(w => console.log(`- ${w.id}: tags=[${w.tags.join(', ')}]`));
   * ```
   */
  async list() {
    const { data, status } = await this.sdk.axiosInstance.get(
      `api/project-wallet/${this.sdk.projectId}/spark`
    );
    if (status === 200) {
      return data;
    }
    throw new Error("Failed to get Spark wallets");
  }
  /**
   * Gets Spark wallets filtered by tag.
   *
   * @param tag The tag to filter by
   * @returns Promise resolving to array of matching wallet information
   *
   * @example
   * ```typescript
   * const wallets = await sdk.wallet.sparkIssuer.getByTag("tokenization");
   * ```
   */
  async getByTag(tag) {
    const { data, status } = await this.sdk.axiosInstance.get(
      `api/project-wallet/${this.sdk.projectId}/spark/tag/${tag}`
    );
    if (status === 200) {
      return data;
    }
    throw new Error("Failed to get Spark wallets by tag");
  }
};

// src/sdk/wallet-developer-controlled/index.ts
var import_wallet4 = require("@meshsdk/wallet");
var import_common3 = require("@meshsdk/common");
var import_issuer_sdk = require("@buildonspark/issuer-sdk");
var import_core_cst3 = require("@meshsdk/core-cst");
var import_uuid = require("uuid");
var WalletDeveloperControlled = class {
  sdk;
  cardano;
  sparkIssuer;
  constructor({ sdk }) {
    this.sdk = sdk;
    this.cardano = new CardanoWalletDeveloperControlled({ sdk });
    this.sparkIssuer = new SparkIssuerWalletDeveloperControlled({ sdk });
  }
  /**
   * Creates a new developer-controlled wallet with both Spark and Cardano chains using shared mnemonic.
   *
   * @param options - Wallet creation options
   * @param options.tags - Optional tags for the wallet
   * @returns Promise that resolves to wallet info and chain wallet instances
   *
   * @example
   * ```typescript
   * // Create wallet
   * const { info } = await sdk.wallet.createWallet({ tags: ["tokenization"] });
   *
   * // For tokenization, use initWallet then createToken
   * await sdk.tokenization.spark.initWallet({ walletId: info.id });
   * await sdk.tokenization.spark.createToken({
   *   tokenName: "MyToken",
   *   tokenTicker: "MTK",
   *   decimals: 8,
   *   isFreezable: true
   * });
   * ```
   */
  async createWallet(options = {}) {
    const project = await this.sdk.getProject();
    if (!project.publicKey) {
      throw new Error("Project public key not found");
    }
    const networkId = this.sdk.network === "mainnet" ? 1 : 0;
    const walletId = (0, import_uuid.v4)();
    const mnemonic = (await (0, import_common3.generateMnemonic)(256)).split(" ");
    const encryptedKey = await encryptWithPublicKey({
      publicKey: project.publicKey,
      data: mnemonic.join(" ")
    });
    const cardanoWallet = await import_wallet4.MeshCardanoHeadlessWallet.fromMnemonic({
      mnemonic,
      networkId,
      walletAddressType: 1
    });
    const [{ wallet: sparkMainnetWallet }, { wallet: sparkRegtestWallet }] = await Promise.all([
      import_issuer_sdk.IssuerSparkWallet.initialize({
        mnemonicOrSeed: mnemonic.join(" "),
        options: { network: "MAINNET" }
      }),
      import_issuer_sdk.IssuerSparkWallet.initialize({
        mnemonicOrSeed: mnemonic.join(" "),
        options: { network: "REGTEST" }
      })
    ]);
    const baseAddress = await cardanoWallet.getChangeAddressBech32();
    const { pubKeyHash, stakeCredentialHash } = (0, import_core_cst3.deserializeBech32Address)(
      baseAddress
    );
    const [mainnetPublicKey, regtestPublicKey] = await Promise.all([
      sparkMainnetWallet.getIdentityPublicKey(),
      sparkRegtestWallet.getIdentityPublicKey()
    ]);
    const sparkWallet = networkId === 1 ? sparkMainnetWallet : sparkRegtestWallet;
    const walletData = {
      id: walletId,
      projectId: this.sdk.projectId,
      tags: options.tags || [],
      key: encryptedKey,
      chains: {
        cardano: { pubKeyHash, stakeCredentialHash },
        spark: { mainnetPublicKey, regtestPublicKey }
      }
    };
    const { status } = await this.sdk.axiosInstance.post(
      `api/project-wallet`,
      walletData
    );
    if (status === 200) {
      return {
        info: walletData,
        sparkIssuerWallet: sparkWallet,
        cardanoWallet
      };
    }
    throw new Error("Failed to create wallet");
  }
  /**
   * Loads an existing developer-controlled wallet by ID and returns both chain wallet instances.
   *
   * @param walletId - The wallet ID to load
   * @returns Promise that resolves to wallet info and initialized wallet instances
   *
   * @example
   * ```typescript
   * const { info, sparkWallet, cardanoWallet } = await sdk.wallet.initWallet("wallet-id");
   *
   * // Get Spark wallet address
   * const sparkAddress = await sparkWallet.getSparkAddress();
   *
   * // Get Cardano wallet address
   * const address = await cardanoWallet.getChangeAddressBech32();
   * ```
   */
  async initWallet(walletId) {
    if (!this.sdk.privateKey) {
      throw new Error(
        "Private key required to load developer-controlled wallet"
      );
    }
    const walletInfo = await this.getProjectWallet(walletId);
    const effectiveNetworkId = this.sdk.network === "mainnet" ? 1 : 0;
    const sharedMnemonic = await decryptWithPrivateKey({
      privateKey: this.sdk.privateKey,
      encryptedDataJSON: walletInfo.key
    });
    const cardanoWallet = await import_wallet4.MeshCardanoHeadlessWallet.fromMnemonic({
      mnemonic: sharedMnemonic.split(" "),
      networkId: effectiveNetworkId,
      walletAddressType: 1
    });
    const sparkNetwork = effectiveNetworkId === 1 ? "MAINNET" : "REGTEST";
    const { wallet: sparkWallet } = await import_issuer_sdk.IssuerSparkWallet.initialize({
      mnemonicOrSeed: sharedMnemonic,
      options: { network: sparkNetwork }
    });
    return {
      info: walletInfo,
      sparkWallet,
      cardanoWallet
    };
  }
  /**
   * Retrieves a multi-chain wallet for a specific chain.
   *
   * @param walletId - The unique identifier of the wallet
   * @param chain - The chain to load ("spark" or "cardano")
   * @returns Promise that resolves to multi-chain wallet instance
   *
   * @example
   * ```typescript
   * // Load Spark wallet
   * const { sparkIssuerWallet } = await sdk.wallet.getWallet("wallet-id", "spark");
   *
   * // Load Cardano wallet
   * const { cardanoWallet } = await sdk.wallet.getWallet("wallet-id", "cardano");
   * ```
   */
  async getWallet(projectWalletId, chain2) {
    const walletInfo = await this.getProjectWallet(projectWalletId);
    const instance = {
      info: walletInfo
    };
    let mnemonic = null;
    if (this.sdk.privateKey) {
      mnemonic = await decryptWithPrivateKey({
        privateKey: this.sdk.privateKey,
        encryptedDataJSON: walletInfo.key
      });
    }
    const networkId = this.sdk.network === "mainnet" ? 1 : 0;
    if ((chain2 === "cardano" || !chain2) && walletInfo.chains.cardano && mnemonic) {
      const cardanoWallet = await import_wallet4.MeshCardanoHeadlessWallet.fromMnemonic({
        mnemonic: mnemonic.split(" "),
        networkId,
        walletAddressType: 1
      });
      instance.cardanoWallet = cardanoWallet;
    }
    if ((chain2 === "spark" || !chain2) && walletInfo.chains.spark && mnemonic) {
      const sparkNetwork = networkId === 1 ? "MAINNET" : "REGTEST";
      const { wallet: sparkWallet } = await import_issuer_sdk.IssuerSparkWallet.initialize({
        mnemonicOrSeed: mnemonic,
        options: { network: sparkNetwork }
      });
      instance.sparkIssuerWallet = sparkWallet;
    }
    return instance;
  }
  /**
   * Retrieves wallet metadata by ID.
   *
   * @param walletId - The unique identifier of the wallet
   * @returns Promise that resolves to wallet info
   */
  async getProjectWallet(walletId) {
    const { data, status } = await this.sdk.axiosInstance.get(
      `api/project-wallet/${this.sdk.projectId}/${walletId}`
    );
    if (status === 200) {
      return data;
    }
    throw new Error("Project wallet not found");
  }
  /**
   * Retrieves wallets for the project with pagination support.
   *
   * @param options - Pagination options
   * @param options.page - Page number (default: 1)
   * @param options.pageSize - Number of wallets per page (default: 10)
   * @returns Promise that resolves to paginated wallet response
   *
   * @example
   * ```typescript
   * // Get first page of wallets
   * const { data, pagination } = await sdk.wallet.getProjectWallets();
   *
   * // Get specific page
   * const { data, pagination } = await sdk.wallet.getProjectWallets({ page: 2 });
   *
   * // Iterate through all pages
   * let page = 1;
   * let allWallets: MultiChainWalletInfo[] = [];
   * while (true) {
   *   const { data, pagination } = await sdk.wallet.getProjectWallets({ page });
   *   allWallets.push(...data);
   *   if (page >= pagination.totalPages) break;
   *   page++;
   * }
   * ```
   */
  async getProjectWallets(options) {
    const page = options?.page ?? 1;
    const { data, status } = await this.sdk.axiosInstance.get(
      `api/project-wallet/${this.sdk.projectId}?page=${page}`
    );
    if (status === 200) {
      return data;
    }
    throw new Error("Failed to get project wallets");
  }
  /**
   * Retrieves all wallets for the project (fetches all pages).
   *
   * @returns Promise that resolves to array of all wallet info
   *
   * @example
   * ```typescript
   * const wallets = await sdk.wallet.getAllProjectWallets();
   * console.log(`Found ${wallets.length} wallets`);
   * ```
   */
  async getAllProjectWallets() {
    const allWallets = [];
    let page = 1;
    while (true) {
      const { data, pagination } = await this.getProjectWallets({ page });
      allWallets.push(...data);
      if (page >= pagination.totalPages) break;
      page++;
    }
    return allWallets;
  }
};

// src/sdk/sponsorship/index.ts
var import_transaction = require("@meshsdk/transaction");
var CONFIG_DURATION_CONSUME_UTXOS = 1e3 * 60;
var Sponsorship = class {
  sdk;
  constructor({ sdk }) {
    this.sdk = sdk;
  }
  /**
   * Retrieves static information about the sponsorship, including the change address
   * and a predefined UTXO used for sponsorship.
   *
   * @returns An object containing the change address and the static UTXO.
   */
  getStaticInfo(amount = "5") {
    return {
      changeAddress: meshUniversalStaticUtxo[this.sdk.network]["5"].output.address,
      utxo: meshUniversalStaticUtxo[this.sdk.network][amount],
      collateral: meshUniversalStaticUtxo[this.sdk.network]["5"]
    };
  }
  /**
   * Sponsors a transaction by associating it with a specific sponsorship ID and optionally
   * modifying its fee-related parameters.
   *
   * @param params - The parameters for sponsoring the transaction.
   * @param params.sponsorshipId - The unique identifier for the sponsorship.
   * @param params.tx - The transaction in CBOR format to be sponsored.
   * @param params.feeChangeAddress - (Optional) Either `feeChangeAddress` or `feeOutputIndex`. The address to which any change from the fee will be sent, the fee balance will be the first output with this address.
   * @param params.feeOutputIndex - (Optional) Either `feeChangeAddress` or `feeOutputIndex`. The index of the output in the transaction where the fee is specified, the fee balanace will combine with the output with this index.
   * @returns A promise that resolves to the new transaction in CBOR format after sponsorship.
   */
  async sponsorTx({
    sponsorshipId,
    tx
  }) {
    const { data, status } = await this.sdk.axiosInstance.get(
      `api/sponsorship/${sponsorshipId}`
    );
    if (status !== 200) {
      return {
        success: false,
        error: "Invalid sponsorship ID or failed to fetch sponsorship config"
      };
    }
    const sponsorshipConfig = data;
    const signedRebuiltTxHex = await this.sponsorTxAndSign({
      txHex: tx,
      config: sponsorshipConfig
    });
    return { success: true, data: signedRebuiltTxHex };
  }
  /**
   * functions to sponsor a transaction and sign it with the sponsor wallet.
   */
  /**
   * This function looks for UTXOs in the sponsor wallet that can be used to sponsor a transaction.
   * If there are not enough UTXOs, it prepares more UTXOs using the `prepareSponsorUtxosTx` method.
   * It then rebuilds the original transaction by adding the selected UTXO as an input (`rebuildTx`) and signs it with the sponsor wallet.
   */
  async sponsorTxAndSign({
    txHex,
    config
  }) {
    const sponsorWallet = await this.getSponsorWallet(config.projectWalletId);
    const sponsorshipWalletUtxos = await sponsorWallet.getUtxosMesh();
    const sponsorshipWalletAddress = await sponsorWallet.getChangeAddressBech32();
    if (sponsorshipWalletUtxos.length === 0) {
      throw new Error(
        "No UTXOs available in the sponsorship wallet. Please fund the wallet."
      );
    }
    const dbPendingUtxos = await this.dbGetIsPendingUtxo(
      config.projectWalletId
    );
    const pendingUtxoIds = dbPendingUtxos.map((utxo) => {
      return `${utxo.txHash}#${utxo.outputIndex}`;
    });
    const utxosAvailableAsInput = sponsorshipWalletUtxos.filter((utxo) => {
      return utxo.output.amount[0].unit === "lovelace" && utxo.output.amount[0].quantity === (config.utxoAmount * 1e6).toString() && !pendingUtxoIds.includes(
        `${utxo.input.txHash}#${utxo.input.outputIndex}`
      );
    });
    let prepareUtxo = false;
    let sponsorshipTxHash = void 0;
    let sponsorshipIndex = void 0;
    if (utxosAvailableAsInput.length <= config.numUtxosTriggerPrepare) {
      prepareUtxo = true;
    }
    if (prepareUtxo) {
      try {
        sponsorshipTxHash = await this.prepareSponsorUtxosTx({
          config
        });
        sponsorshipIndex = 0;
      } catch (error) {
        console.error("Failed to prepare sponsor UTXOs:", error);
        const { data: resRefreshTxHash, status: refreshStatus } = await this.sdk.axiosInstance.get(
          `api/sponsorship/${config.id}/refreshTxHash`
        );
        if (refreshStatus !== 200 || !resRefreshTxHash.refreshTxHash) {
          throw new Error("Failed to get refresh transaction hash");
        }
        sponsorshipTxHash = resRefreshTxHash.refreshTxHash;
        sponsorshipIndex = Math.floor(Math.random() * config.numUtxosPrepare);
      }
    }
    let selectedUtxo = void 0;
    if (sponsorshipTxHash !== void 0 && sponsorshipIndex !== void 0) {
      selectedUtxo = {
        input: {
          txHash: sponsorshipTxHash,
          outputIndex: sponsorshipIndex
        },
        output: {
          amount: [
            {
              unit: "lovelace",
              quantity: (config.utxoAmount * 1e6).toString()
            }
          ],
          address: sponsorshipWalletAddress
        }
      };
      await this.dbAppendUtxosUsed(config, sponsorshipTxHash, sponsorshipIndex);
    }
    while (selectedUtxo === void 0 && utxosAvailableAsInput.length > 0) {
      const selectedIndex = Math.floor(
        Math.random() * utxosAvailableAsInput.length
      );
      const _selectedUtxo = utxosAvailableAsInput[selectedIndex];
      utxosAvailableAsInput.splice(selectedIndex, 1);
      const isUtxoUsed = await this.dbGetIfUtxoUsed(
        config.projectWalletId,
        _selectedUtxo.input.txHash,
        _selectedUtxo.input.outputIndex
      );
      if (!isUtxoUsed) {
        selectedUtxo = _selectedUtxo;
        if (selectedUtxo) {
          await this.dbAppendUtxosUsed(
            config,
            selectedUtxo.input.txHash,
            selectedUtxo.input.outputIndex
          );
        }
      }
    }
    if (selectedUtxo) {
      let _rebuiltTxHex = void 0;
      try {
        const body = {
          txHex,
          address: sponsorshipWalletAddress,
          utxos: JSON.stringify(sponsorshipWalletUtxos),
          sponsorUtxo: JSON.stringify(selectedUtxo),
          network: this.sdk.network
        };
        const { data, status } = await this.sdk.axiosInstance.post(
          `api/sponsorship/tx-parser`,
          body
        );
        if (status !== 200) {
          throw new Error("Failed to parse Tx on server!");
        }
        const { rebuiltTxHex } = data;
        _rebuiltTxHex = rebuiltTxHex;
      } catch (error) {
        const { data: resRefreshTxHash, status: refreshStatus } = await this.sdk.axiosInstance.get(
          `api/sponsorship/${config.id}/refreshTxHash`
        );
        if (refreshStatus !== 200 || !resRefreshTxHash.refreshTxHash) {
          throw new Error("Failed to get refresh transaction hash");
        }
        const txHash = resRefreshTxHash.refreshTxHash;
        let hasFoundUsableUtxo = false;
        for (let attempt = 0; attempt < config.numUtxosPrepare && !hasFoundUsableUtxo; attempt++) {
          const selectedIndex = Math.floor(
            Math.random() * config.numUtxosPrepare
          );
          try {
            const isUtxoUsed = await this.dbGetIfUtxoUsed(
              config.projectWalletId,
              txHash,
              selectedIndex
            );
            if (!isUtxoUsed) {
              await this.dbAppendUtxosUsed(config, txHash, selectedIndex);
              const newSelectedUtxo = {
                input: {
                  txHash,
                  outputIndex: selectedIndex
                },
                output: {
                  amount: [
                    {
                      unit: "lovelace",
                      quantity: (config.utxoAmount * 1e6).toString()
                    }
                  ],
                  address: sponsorshipWalletAddress
                }
              };
              const body = {
                txHex,
                address: sponsorshipWalletAddress,
                utxos: JSON.stringify(sponsorshipWalletUtxos),
                sponsorUtxo: JSON.stringify(newSelectedUtxo),
                network: this.sdk.network
              };
              const { data, status } = await this.sdk.axiosInstance.post(
                `api/sponsorship/tx-parser`,
                body
              );
              if (status === 200) {
                _rebuiltTxHex = data.rebuiltTxHex;
                hasFoundUsableUtxo = true;
                break;
              }
            }
          } catch (innerError) {
          }
        }
      }
      if (_rebuiltTxHex == void 0) {
        throw new Error("Failed to rebuild transaction with selected UTXO.");
      }
      const signedRebuiltTxHex = await sponsorWallet.signTxReturnFullTx(
        _rebuiltTxHex,
        true
      );
      return signedRebuiltTxHex;
    }
    throw new Error("No available UTXOs to sponsor the transaction.");
  }
  async getSponsorWallet(projectWalletId) {
    const walletResult = await this.sdk.wallet.cardano.getWallet(projectWalletId);
    return walletResult.wallet;
  }
  /**
   * Prepares UTXOs for sponsorship by creating new UTXOs in the sponsor wallet.
   * It uses existing UTXOs that are not the exact sponsor amount and those that have been pending for too long.
   * It creates a transaction that consumes these UTXOs and produces new UTXOs of the specified amount.
   */
  async prepareSponsorUtxosTx({
    config
  }) {
    const wallet = await this.getSponsorWallet(config.projectWalletId);
    const utxos = await wallet.getUtxosMesh();
    const changeAddress = await wallet.getChangeAddressBech32();
    const utxosAsInput = utxos.filter((utxo) => {
      return utxo.output.amount[0].unit === "lovelace" && parseInt(utxo.output.amount[0].quantity) !== config.utxoAmount * 1e6;
    });
    const dbPendingUtxos = await this.dbGetIsPendingUtxo(
      config.projectWalletId
    );
    const pendingUtxosThatAreOld = dbPendingUtxos.filter((utxo) => {
      return utxo.isPending && !utxo.isSpent && new Date(utxo.createdAt).getTime() + CONFIG_DURATION_CONSUME_UTXOS < Date.now();
    });
    const utxosIdsThatAreOld = pendingUtxosThatAreOld.map((utxo) => {
      return `${utxo.txHash}#${utxo.outputIndex}`;
    });
    const utxosNotSpentAfterDuration = utxos.filter((utxo) => {
      return utxo.output.amount[0].unit === "lovelace" && parseInt(utxo.output.amount[0].quantity) === config.utxoAmount * 1e6 && utxosIdsThatAreOld.includes(
        `${utxo.input.txHash}#${utxo.input.outputIndex}`
      );
    });
    const txBuilder = new import_transaction.MeshTxBuilder({
      fetcher: this.sdk.providerFetcher
    });
    txBuilder.changeAddress(changeAddress);
    for (let i = 0; i < utxosAsInput.length; i++) {
      const utxo = utxosAsInput[i];
      txBuilder.txIn(
        utxo.input.txHash,
        utxo.input.outputIndex,
        utxo.output.amount,
        utxo.output.address
      );
    }
    for (let i = 0; i < utxosNotSpentAfterDuration.length; i++) {
      const utxo = utxosNotSpentAfterDuration[i];
      txBuilder.txIn(
        utxo.input.txHash,
        utxo.input.outputIndex,
        utxo.output.amount,
        utxo.output.address
      );
      await this.dbMarkUtxoAsConsumed(
        config,
        utxo.input.txHash,
        utxo.input.outputIndex
      );
    }
    let totalBalance = 0;
    for (const utxo of utxosAsInput) {
      const lovelaceAmount = utxo.output.amount.find(
        (amount) => amount.unit === "lovelace"
      );
      if (lovelaceAmount) {
        totalBalance += parseInt(lovelaceAmount.quantity);
      }
    }
    for (const utxo of utxosNotSpentAfterDuration) {
      const lovelaceAmount = utxo.output.amount.find(
        (amount) => amount.unit === "lovelace"
      );
      if (lovelaceAmount) {
        totalBalance += parseInt(lovelaceAmount.quantity);
      }
    }
    const utxoAmountLovelace = config.utxoAmount * 1e6;
    const maxUtxosWeCanCreate = Math.floor(totalBalance / utxoAmountLovelace);
    const numUtxosToCreate = Math.min(
      config.numUtxosPrepare,
      maxUtxosWeCanCreate
    );
    for (let i = 0; i < numUtxosToCreate; i++) {
      txBuilder.txOut(changeAddress, [
        {
          unit: "lovelace",
          quantity: (config.utxoAmount * 1e6).toString()
        }
      ]);
    }
    const unsignedTx = await txBuilder.complete();
    const signedTx = await wallet.signTxReturnFullTx(unsignedTx);
    const txHash = await this.sdk.providerSubmitter.submitTx(signedTx);
    await this.sdk.axiosInstance.post(
      `api/sponsorship/${config.id}/refreshTxHash`,
      {
        txHash
      }
    );
    return txHash;
  }
  /**
   * DB functions
   */
  async dbGetIsPendingUtxo(projectWalletId) {
    const { data, status } = await this.sdk.axiosInstance.get(
      `api/sponsorship/output/${projectWalletId}/pending`
    );
    if (status === 200) {
      return data;
    }
    throw new Error("Failed to get pending UTXOs");
  }
  async dbGetIfUtxoUsed(projectWalletId, txHash, outputIndex) {
    const { data, status } = await this.sdk.axiosInstance.get(
      `api/sponsorship/output/${projectWalletId}/${txHash}/${outputIndex}`
    );
    if (status === 200) {
      const output = data;
      return output ? output.isSpent : false;
    }
    throw new Error("Failed to check if UTXO is used");
  }
  async dbUpdateUtxos(projectWalletId, txHash, outputIndex, isPending = true, isSpent = false) {
    const { data, status } = await this.sdk.axiosInstance.put(
      `api/sponsorship/output/${projectWalletId}/${txHash}/${outputIndex}`,
      {
        isPending,
        isSpent
      }
    );
    if (status === 200) {
      return data;
    }
    throw new Error("Failed to mark UTXO as used");
  }
  async dbAppendUtxosUsed(config, txHash, outputIndex) {
    return this.dbUpdateUtxos(
      config.projectWalletId,
      txHash,
      outputIndex,
      true,
      false
    );
  }
  async dbMarkUtxoAsConsumed(config, txHash, outputIndex) {
    return this.dbUpdateUtxos(
      config.projectWalletId,
      txHash,
      outputIndex,
      false,
      true
    );
  }
};

// src/sdk/tokenization/spark.ts
var import_issuer_sdk2 = require("@buildonspark/issuer-sdk");
var import_uuid2 = require("uuid");
var import_spark_sdk4 = require("@buildonspark/spark-sdk");
var TokenizationSpark = class {
  sdk;
  wallet = null;
  walletInfo = null;
  walletNetwork = "MAINNET";
  constructor({ sdk }) {
    this.sdk = sdk;
  }
  /**
   * Gets the current wallet ID if one is loaded.
   */
  getWalletId() {
    return this.walletInfo?.id ?? null;
  }
  /**
   * Clears the currently loaded wallet state.
   * Call this before creating a new token if you want to start fresh.
   */
  clearWallet() {
    this.wallet = null;
    this.walletInfo = null;
  }
  /**
   * Internal method to initialize the wallet by wallet ID.
   */
  async initWalletByWalletId(walletId) {
    const networkParam = this.sdk.network === "mainnet" ? "mainnet" : "regtest";
    const { data, status } = await this.sdk.axiosInstance.get(
      `api/project-wallet/${this.sdk.projectId}/${walletId}?chain=spark&network=${networkParam}`
    );
    if (status !== 200) {
      throw new Error("Failed to get Spark wallet");
    }
    const walletProject = data;
    this.walletNetwork = this.sdk.network === "mainnet" ? "MAINNET" : "REGTEST";
    if (this.sdk.privateKey === void 0) {
      throw new Error("Private key not found - required to decrypt wallet");
    }
    const mnemonic = await decryptWithPrivateKey({
      privateKey: this.sdk.privateKey,
      encryptedDataJSON: walletProject.key
    });
    const { wallet } = await import_issuer_sdk2.IssuerSparkWallet.initialize({
      mnemonicOrSeed: mnemonic,
      options: {
        network: this.walletNetwork
      }
    });
    this.wallet = wallet;
    this.walletInfo = walletProject;
  }
  /**
   * Initializes the tokenization by token ID.
   *
   * @param params - { tokenId } - the token ID to load
   * @returns The tokenization policy
   *
   * @example
   * ```typescript
   * // Load existing token by token ID
   * const policy = await sdk.tokenization.spark.initTokenization({ tokenId: "btknrt1..." });
   *
   * // Then perform operations
   * await sdk.tokenization.spark.mintTokens({ amount: BigInt(1000) });
   * ```
   */
  async initTokenization(params) {
    const normalizedTokenId = this.normalizeTokenId(params.tokenId);
    const policy = await this.getTokenizationPolicy(normalizedTokenId);
    await this.initWalletByWalletId(policy.walletId);
    return policy;
  }
  /**
   * Normalizes token ID to hex format.
   * If bech32m encoded (btkn1.., btknrt1.., etc), decodes to hex. Otherwise returns as-is.
   */
  normalizeTokenId(tokenId) {
    const network = this.sdk.network === "mainnet" ? "MAINNET" : "REGTEST";
    try {
      const decoded = (0, import_spark_sdk4.decodeBech32mTokenIdentifier)(
        tokenId,
        network
      );
      return Buffer.from(decoded.tokenIdentifier).toString("hex");
    } catch {
      return tokenId;
    }
  }
  /**
   * Creates a new token on the Spark network.
   * Automatically creates a new wallet if none is loaded.
   *
   * @param params - Token creation parameters
   * @returns Object containing txId, tokenId, and walletId
   *
   * @example
   * ```typescript
   * // Simple one-step token creation (creates wallet automatically)
   * const { tokenId, walletId } = await sdk.tokenization.spark.createToken({
   *   tokenName: "MyToken",
   *   tokenTicker: "MTK",
   *   decimals: 8,
   *   isFreezable: true,
   * });
   *
   * // Or load existing wallet first
   * await sdk.tokenization.spark.initTokenization({ tokenId: "existing-token-id" });
   * const { tokenId } = await sdk.tokenization.spark.createToken({
   *   tokenName: "MyToken",
   *   tokenTicker: "MTK",
   *   decimals: 8,
   *   isFreezable: true,
   * });
   * ```
   */
  async createToken(params) {
    if (!this.wallet || !this.walletInfo) {
      const { info, sparkIssuerWallet } = await this.sdk.wallet.createWallet({
        tags: ["tokenization", "spark"]
      });
      this.walletNetwork = this.sdk.network === "mainnet" ? "MAINNET" : "REGTEST";
      this.walletInfo = info;
      this.wallet = sparkIssuerWallet;
    }
    const txId = await this.wallet.createToken({
      tokenName: params.tokenName,
      tokenTicker: params.tokenTicker,
      decimals: params.decimals,
      maxSupply: params.maxSupply,
      isFreezable: params.isFreezable
    });
    const tokenMetadata = await this.wallet.getIssuerTokenMetadata();
    const tokenIdHex = Buffer.from(tokenMetadata.rawTokenIdentifier).toString(
      "hex"
    );
    const tokenId = (0, import_spark_sdk4.encodeBech32mTokenIdentifier)({
      tokenIdentifier: tokenMetadata.rawTokenIdentifier,
      network: this.walletNetwork
    });
    try {
      await this.sdk.axiosInstance.post("/api/tokenization/tokens", {
        tokenId: tokenIdHex,
        projectId: this.sdk.projectId,
        walletId: this.walletInfo.id,
        chain: "spark",
        network: this.walletNetwork.toLowerCase()
      });
      await this.logTransaction({
        txId,
        tokenId: tokenIdHex,
        walletInfo: this.walletInfo,
        type: "create"
      });
    } catch (saveError) {
      console.warn("Failed to save token to database:", saveError);
    }
    return { txId, tokenId, walletId: this.walletInfo.id };
  }
  /**
   * Mints tokens from the issuer wallet.
   * Requires initTokenization() to be called first.
   *
   * @param params - Mint parameters including amount
   * @returns Transaction ID of the mint operation
   */
  async mintTokens(params) {
    if (!this.wallet || !this.walletInfo) {
      throw new Error("No wallet loaded. Call initTokenization(tokenId) first.");
    }
    const txHash = await this.wallet.mintTokens(params.amount);
    const tokenMetadata = await this.wallet.getIssuerTokenMetadata();
    const tokenId = Buffer.from(tokenMetadata.rawTokenIdentifier).toString(
      "hex"
    );
    await this.logTransaction({
      txId: txHash,
      tokenId,
      walletInfo: this.walletInfo,
      type: "mint",
      amount: params.amount.toString()
    });
    return txHash;
  }
  /**
   * Gets the token balance for an issuer wallet.
   * Requires initTokenization() to be called first.
   *
   * @returns Balance information
   */
  async getTokenBalance() {
    if (!this.wallet) {
      throw new Error("No wallet loaded. Call initTokenization(tokenId) first.");
    }
    const result = await this.wallet.getIssuerTokenBalance();
    return { balance: result.balance.toString() };
  }
  /**
   * Gets metadata for the token created by an issuer wallet.
   * Requires initTokenization() to be called first.
   *
   * @returns Token metadata
   */
  async getTokenMetadata() {
    if (!this.wallet) {
      throw new Error("No wallet loaded. Call initTokenization(tokenId) first.");
    }
    return await this.wallet.getIssuerTokenMetadata();
  }
  /**
   * Transfers tokens from the issuer wallet to another address.
   * Requires initTokenization() to be called first.
   *
   * @param params - Transfer parameters
   * @returns Transaction ID of the transfer
   */
  async transferTokens(params) {
    if (!this.wallet || !this.walletInfo) {
      throw new Error("No wallet loaded. Call initTokenization(tokenId) first.");
    }
    const tokenMetadata = await this.wallet.getIssuerTokenMetadata();
    const tokenId = Buffer.from(tokenMetadata.rawTokenIdentifier).toString(
      "hex"
    );
    const bech32mTokenId = (0, import_spark_sdk4.encodeBech32mTokenIdentifier)({
      tokenIdentifier: tokenMetadata.rawTokenIdentifier,
      network: this.walletNetwork
    });
    const txHash = await this.wallet.transferTokens({
      tokenIdentifier: bech32mTokenId,
      tokenAmount: params.amount,
      receiverSparkAddress: params.toAddress
    });
    const issuerAddress = await this.wallet.getSparkAddress();
    await this.logTransaction({
      txId: txHash,
      tokenId,
      walletInfo: this.walletInfo,
      type: "transfer",
      amount: params.amount.toString(),
      fromAddress: issuerAddress,
      toAddress: params.toAddress
    });
    return txHash;
  }
  /**
   * Burns tokens permanently from circulation.
   * Requires initTokenization() to be called first.
   *
   * @param params - Burn parameters
   * @returns Transaction ID of the burn operation
   */
  async burnTokens(params) {
    if (!this.wallet || !this.walletInfo) {
      throw new Error("No wallet loaded. Call initTokenization(tokenId) first.");
    }
    const txHash = await this.wallet.burnTokens(params.amount);
    const tokenMetadata = await this.wallet.getIssuerTokenMetadata();
    const tokenId = Buffer.from(tokenMetadata.rawTokenIdentifier).toString(
      "hex"
    );
    await this.logTransaction({
      txId: txHash,
      tokenId,
      walletInfo: this.walletInfo,
      type: "burn",
      amount: params.amount.toString()
    });
    return txHash;
  }
  /**
   * Freezes tokens at a specific Spark address for compliance purposes.
   * Requires initTokenization() to be called first.
   *
   * @param params - Freeze parameters
   * @returns Freeze operation results
   */
  async freezeTokens(params) {
    if (!this.wallet || !this.walletInfo) {
      throw new Error("No wallet loaded. Call initTokenization(tokenId) first.");
    }
    const result = await this.wallet.freezeTokens(params.address);
    const tokenMetadata = await this.wallet.getIssuerTokenMetadata();
    const tokenId = Buffer.from(tokenMetadata.rawTokenIdentifier).toString(
      "hex"
    );
    try {
      const publicKeyHash = extractIdentityPublicKey(params.address);
      if (!publicKeyHash) {
        throw new Error(
          `Failed to extract public key hash from Spark address: ${params.address}`
        );
      }
      await this.sdk.axiosInstance.post("/api/tokenization/frozen-addresses", {
        tokenId,
        projectId: this.sdk.projectId,
        projectWalletId: this.walletInfo.id,
        chain: "spark",
        network: this.walletNetwork.toLowerCase(),
        publicKeyHash,
        isFrozen: true,
        freezeReason: params.freezeReason || "Frozen by issuer",
        frozenAt: (/* @__PURE__ */ new Date()).toISOString()
      });
      await this.logTransaction({
        txId: (0, import_uuid2.v4)(),
        tokenId,
        walletInfo: this.walletInfo,
        type: "freeze",
        fromAddress: "Issuer Wallet",
        toAddress: params.address,
        amount: result.impactedTokenAmount.toString()
      });
    } catch (saveError) {
      console.warn("Failed to save freeze operation:", saveError);
    }
    return {
      impactedOutputIds: result.impactedTokenOutputs?.map((o) => o.id) ?? [],
      impactedTokenAmount: result.impactedTokenAmount.toString()
    };
  }
  /**
   * Unfreezes tokens at a specific Spark address.
   * Requires initTokenization() to be called first.
   *
   * @param params - Unfreeze parameters
   * @returns Unfreeze operation results
   */
  async unfreezeTokens(params) {
    if (!this.wallet || !this.walletInfo) {
      throw new Error("No wallet loaded. Call initTokenization(tokenId) first.");
    }
    const result = await this.wallet.unfreezeTokens(params.address);
    const tokenMetadata = await this.wallet.getIssuerTokenMetadata();
    const tokenId = Buffer.from(tokenMetadata.rawTokenIdentifier).toString(
      "hex"
    );
    try {
      const publicKeyHash = extractIdentityPublicKey(params.address);
      if (!publicKeyHash) {
        throw new Error(
          `Failed to extract public key hash from Spark address: ${params.address}`
        );
      }
      await this.sdk.axiosInstance.put("/api/tokenization/frozen-addresses", {
        tokenId,
        publicKeyHash,
        projectId: this.sdk.projectId,
        projectWalletId: this.walletInfo.id
      });
      await this.logTransaction({
        txId: (0, import_uuid2.v4)(),
        tokenId,
        walletInfo: this.walletInfo,
        type: "unfreeze",
        fromAddress: "Issuer Wallet",
        toAddress: params.address,
        amount: result.impactedTokenAmount.toString()
      });
    } catch (saveError) {
      console.warn("Failed to save unfreeze operation:", saveError);
    }
    return {
      impactedOutputIds: result.impactedTokenOutputs?.map((o) => o.id) ?? [],
      impactedTokenAmount: result.impactedTokenAmount.toString()
    };
  }
  /**
   * Lists frozen addresses for a token from the database.
   * Requires initTokenization() to be called first.
   *
   * @param params - Query parameters including pagination
   * @returns List of frozen addresses with pagination info
   */
  async getFrozenAddresses(params) {
    if (!this.wallet) {
      throw new Error("No wallet loaded. Call initTokenization(tokenId) first.");
    }
    const { includeUnfrozen = false, page = 1, limit = 15 } = params || {};
    const tokenMetadata = await this.wallet.getIssuerTokenMetadata();
    const tokenId = Buffer.from(tokenMetadata.rawTokenIdentifier).toString(
      "hex"
    );
    const { data, status } = await this.sdk.axiosInstance.get(
      `api/tokenization/frozen-addresses`,
      {
        params: {
          tokenId,
          projectId: this.sdk.projectId,
          includeUnfrozen,
          page,
          limit
        }
      }
    );
    if (status === 200) {
      return data;
    }
    throw new Error("Failed to get frozen addresses");
  }
  /**
   * Lists token transactions from the database.
   * Requires initTokenization() to be called first.
   *
   * @param params - Query parameters including type filter and pagination
   * @returns List of transactions with pagination info
   */
  async getTransactions(params) {
    if (!this.wallet) {
      throw new Error("No wallet loaded. Call initTokenization(tokenId) first.");
    }
    const { type, page = 1, limit = 50 } = params || {};
    const tokenMetadata = await this.wallet.getIssuerTokenMetadata();
    const tokenId = Buffer.from(tokenMetadata.rawTokenIdentifier).toString(
      "hex"
    );
    const { data, status } = await this.sdk.axiosInstance.get(
      `api/tokenization/transactions`,
      {
        params: {
          tokenId,
          projectId: this.sdk.projectId,
          ...type && { type },
          page,
          limit
        }
      }
    );
    if (status === 200) {
      return data;
    }
    throw new Error("Failed to get transactions");
  }
  /**
   * Lists tokenization policies for the current project from the database.
   *
   * @param params - Optional filter and pagination parameters
   * @returns List of tokenization policies with pagination info
   */
  async getTokenizationPolicies(params) {
    const { tokenId, page = 1, limit = 15 } = params || {};
    const { data, status } = await this.sdk.axiosInstance.get(
      `api/tokenization/policies`,
      {
        params: {
          projectId: this.sdk.projectId,
          ...tokenId && { tokenId },
          page,
          limit
        }
      }
    );
    if (status === 200) {
      return data;
    }
    throw new Error("Failed to get tokenization policies");
  }
  /**
   * Gets a single tokenization policy by token ID.
   *
   * @param tokenId - The token ID to look up
   * @returns The tokenization policy
   */
  async getTokenizationPolicy(tokenId) {
    const { tokens } = await this.getTokenizationPolicies({
      tokenId,
      limit: 1
    });
    const policy = tokens[0];
    if (!policy) {
      throw new Error("Tokenization policy not found");
    }
    return policy;
  }
  /**
   * Internal helper to log token transactions to the database
   */
  async logTransaction(params) {
    try {
      await this.sdk.axiosInstance.post("/api/tokenization/transactions", {
        txId: params.txId,
        tokenId: params.tokenId,
        projectId: this.sdk.projectId,
        projectWalletId: params.walletInfo.id,
        type: params.type,
        chain: "spark",
        network: this.walletNetwork.toLowerCase(),
        amount: params.amount,
        fromAddress: params.fromAddress,
        toAddress: params.toAddress,
        status: params.status || "confirmed"
      });
    } catch (error) {
      console.warn(`Failed to log ${params.type} transaction:`, error);
    }
  }
};

// src/sdk/tokenization/cardano.ts
var import_wallet5 = require("@meshsdk/wallet");
var TokenizationCardano = class {
  sdk;
  wallet = null;
  walletInfo = null;
  constructor({ sdk }) {
    this.sdk = sdk;
  }
  setWallet(wallet, walletInfo) {
    this.wallet = wallet;
    this.walletInfo = walletInfo;
  }
  getWalletId() {
    return this.walletInfo?.id ?? null;
  }
  async initWalletByWalletId(walletId) {
    if (this.sdk.privateKey === void 0) {
      throw new Error("Private key not found - required to decrypt wallet");
    }
    const networkParam = this.sdk.network === "mainnet" ? "mainnet" : "testnet";
    const { data, status } = await this.sdk.axiosInstance.get(
      `api/project-wallet/${this.sdk.projectId}/${walletId}?chain=cardano&network=${networkParam}`
    );
    if (status !== 200) {
      throw new Error("Failed to get Cardano wallet");
    }
    const walletInfo = data;
    const mnemonic = await decryptWithPrivateKey({
      privateKey: this.sdk.privateKey,
      encryptedDataJSON: walletInfo.key
    });
    const networkId = this.sdk.network === "mainnet" ? 1 : 0;
    const wallet = await import_wallet5.MeshCardanoHeadlessWallet.fromMnemonic({
      mnemonic: mnemonic.split(" "),
      networkId,
      walletAddressType: 1
    });
    this.wallet = wallet;
    this.walletInfo = walletInfo;
  }
  async initWallet(tokenId) {
    const policy = await this.getTokenizationPolicy(tokenId);
    await this.initWalletByWalletId(policy.walletId);
    return policy;
  }
  async createToken(params) {
    if (!this.wallet || !this.walletInfo) {
      throw new Error("No wallet loaded. Use createWallet({ enableTokenization: true }) first.");
    }
    throw new Error("Cardano token creation not yet implemented");
  }
  async mintTokens(params) {
    if (!this.wallet || !this.walletInfo) {
      throw new Error("No wallet loaded. Call initWallet(tokenId) first.");
    }
    throw new Error("Cardano token minting not yet implemented");
  }
  async getTokenBalance() {
    if (!this.wallet) {
      throw new Error("No wallet loaded. Call initWallet(tokenId) first.");
    }
    throw new Error("Cardano token balance query not yet implemented");
  }
  async getTokenMetadata() {
    if (!this.wallet) {
      throw new Error("No wallet loaded. Call initWallet(tokenId) first.");
    }
    throw new Error("Cardano token metadata query not yet implemented");
  }
  async transferTokens(params) {
    if (!this.wallet || !this.walletInfo) {
      throw new Error("No wallet loaded. Call initWallet(tokenId) first.");
    }
    throw new Error("Cardano token transfer not yet implemented");
  }
  async burnTokens(params) {
    if (!this.wallet || !this.walletInfo) {
      throw new Error("No wallet loaded. Call initWallet(tokenId) first.");
    }
    throw new Error("Cardano token burning not yet implemented");
  }
  async freezeTokens(params) {
    if (!this.wallet || !this.walletInfo) {
      throw new Error("No wallet loaded. Call initWallet(tokenId) first.");
    }
    throw new Error("Cardano token freezing not yet implemented");
  }
  async unfreezeTokens(params) {
    if (!this.wallet || !this.walletInfo) {
      throw new Error("No wallet loaded. Call initWallet(tokenId) first.");
    }
    throw new Error("Cardano token unfreezing not yet implemented");
  }
  async getFrozenAddresses(params) {
    if (!this.wallet) {
      throw new Error("No wallet loaded. Call initWallet(tokenId) first.");
    }
    const tokenId = "";
    const { includeUnfrozen = false, page = 1, limit = 15 } = params || {};
    const { data, status } = await this.sdk.axiosInstance.get(
      `api/tokenization/frozen-addresses`,
      {
        params: {
          tokenId,
          projectId: this.sdk.projectId,
          includeUnfrozen,
          page,
          limit
        }
      }
    );
    if (status === 200) {
      return data;
    }
    throw new Error("Failed to get frozen addresses");
  }
  async getTransactions(params) {
    if (!this.wallet) {
      throw new Error("No wallet loaded. Call initWallet(tokenId) first.");
    }
    const tokenId = "";
    const { type, page = 1, limit = 50 } = params || {};
    const { data, status } = await this.sdk.axiosInstance.get(
      `api/tokenization/transactions`,
      {
        params: {
          tokenId,
          projectId: this.sdk.projectId,
          ...type && { type },
          page,
          limit
        }
      }
    );
    if (status === 200) {
      return data;
    }
    throw new Error("Failed to get transactions");
  }
  async getTokenizationPolicies(params) {
    const { tokenId, page = 1, limit = 15 } = params || {};
    const { data, status } = await this.sdk.axiosInstance.get(
      `api/tokenization/policies`,
      {
        params: {
          projectId: this.sdk.projectId,
          chain: "cardano",
          ...tokenId && { tokenId },
          page,
          limit
        }
      }
    );
    if (status === 200) {
      return data;
    }
    throw new Error("Failed to get tokenization policies");
  }
  async getTokenizationPolicy(tokenId) {
    const { tokens } = await this.getTokenizationPolicies({ tokenId, limit: 1 });
    const policy = tokens[0];
    if (!policy) {
      throw new Error("Tokenization policy not found");
    }
    return policy;
  }
  async logTransaction(params) {
    try {
      await this.sdk.axiosInstance.post("/api/tokenization/transactions", {
        tokenId: params.tokenId,
        projectId: this.sdk.projectId,
        projectWalletId: params.walletInfo.id,
        type: params.type,
        chain: "cardano",
        network: this.sdk.network,
        txHash: params.txHash,
        amount: params.amount,
        fromAddress: params.fromAddress,
        toAddress: params.toAddress,
        status: params.status || "success",
        metadata: params.metadata
      });
    } catch (error) {
      console.warn(`Failed to log ${params.type} transaction:`, error);
    }
  }
};

// src/sdk/tokenization/index.ts
var Tokenization = class {
  sdk;
  spark;
  cardano;
  constructor({ sdk }) {
    this.sdk = sdk;
    this.spark = new TokenizationSpark({ sdk: this.sdk });
    this.cardano = new TokenizationCardano({ sdk: this.sdk });
  }
};

// src/sdk/index.ts
var meshUniversalStaticUtxo = {
  mainnet: {
    "5": {
      input: {
        outputIndex: 0,
        txHash: "89e9888acc50ec7cc840f3e44ea06d8db1461dd1ba37218e914fcd171d83e4b8"
      },
      output: {
        address: "addr1q8sj3xj6q99m4g9tu9mm2lzzdafy04035eya7hjhpus55r204nlu6dmhgpruq7df228h9gpujt0mtnfcnkcaj3wj457qh5366a",
        amount: [
          {
            unit: "lovelace",
            quantity: "5000000"
          }
        ]
      }
    },
    "99": {
      input: {
        outputIndex: 0,
        txHash: "cb058a1402fcbcd98be15de8c44ea1a44211119050e8028d3c81878158a5d29d"
      },
      output: {
        address: "addr1q8sj3xj6q99m4g9tu9mm2lzzdafy04035eya7hjhpus55r204nlu6dmhgpruq7df228h9gpujt0mtnfcnkcaj3wj457qh5366a",
        amount: [
          {
            unit: "lovelace",
            quantity: "99000000"
          }
        ]
      }
    }
  },
  testnet: {
    "5": {
      input: {
        outputIndex: 0,
        txHash: "5a1edf7da58eff2059030abd456947a96cb2d16b9d8c3822ffff58d167ed8bfc"
      },
      output: {
        address: "addr_test1qrsj3xj6q99m4g9tu9mm2lzzdafy04035eya7hjhpus55r204nlu6dmhgpruq7df228h9gpujt0mtnfcnkcaj3wj457q5zv6kz",
        amount: [
          {
            unit: "lovelace",
            quantity: "5000000"
          }
        ]
      }
    },
    "99": {
      input: {
        outputIndex: 0,
        txHash: "8222b0327a95e8c357016a5df64d93d7cf8a585a07c55327ae618a7e00d58d9e"
      },
      output: {
        address: "addr_test1qrsj3xj6q99m4g9tu9mm2lzzdafy04035eya7hjhpus55r204nlu6dmhgpruq7df228h9gpujt0mtnfcnkcaj3wj457q5zv6kz",
        amount: [
          {
            unit: "lovelace",
            quantity: "99000000"
          }
        ]
      }
    }
  }
};
var Web3Sdk = class {
  axiosInstance;
  appUrl;
  projectId;
  apiKey;
  privateKey;
  providerFetcher;
  providerSubmitter;
  network;
  project;
  wallet;
  sponsorship;
  tokenization;
  constructor({
    appUrl,
    projectId,
    apiKey,
    network,
    privateKey,
    fetcher,
    submitter
  }) {
    this.appUrl = appUrl ? appUrl : "https://utxos.dev/";
    this.projectId = projectId;
    this.apiKey = apiKey;
    this.network = network;
    this.privateKey = privateKey;
    this.providerFetcher = fetcher;
    this.providerSubmitter = submitter;
    this.axiosInstance = import_axios2.default.create({
      baseURL: this.appUrl,
      headers: { "x-api-key": apiKey }
    });
    this.wallet = new WalletDeveloperControlled({
      sdk: this
    });
    this.sponsorship = new Sponsorship({
      sdk: this
    });
    this.tokenization = new Tokenization({
      sdk: this
    });
  }
  async getProject() {
    if (this.project) {
      return this.project;
    }
    const { data, status } = await this.axiosInstance.get(
      `api/project/${this.projectId}`
    );
    if (status === 200) {
      this.project = data;
      return this.project;
    }
    throw new Error("Failed to get project");
  }
};

// src/types/core/multi-chain.ts
var import_api_contracts = require("@utxos/api-contracts");

// src/types/spark/index.ts
var TransferDirection = /* @__PURE__ */ ((TransferDirection2) => {
  TransferDirection2["INCOMING"] = "INCOMING";
  TransferDirection2["OUTGOING"] = "OUTGOING";
  return TransferDirection2;
})(TransferDirection || {});

// src/wallet-user-controlled/web3-wallet.ts
var import_wallet6 = require("@meshsdk/wallet");
var import_bitcoin4 = require("@meshsdk/bitcoin");
var Web3Wallet = class _Web3Wallet {
  projectId;
  appUrl = "https://utxos.dev/";
  user;
  cardano;
  bitcoin;
  spark;
  networkId;
  constructor(options) {
    this.projectId = options.projectId;
    this.appUrl = options.appUrl || "https://utxos.dev/";
    this.user = options.user;
    this.cardano = options.cardano;
    this.bitcoin = options.bitcoin;
    this.spark = options.spark;
    this.networkId = options.networkId;
  }
  /**
   * Initializes a new instance of the Web3Wallet class.
   *
   * @param options - The options to initialize the wallet.
   * @param options.networkId - The network ID (0 for testnet, 1 for mainnet).
   * @param options.fetcher - An optional fetcher for network requests.
   * @param options.submitter - An optional submitter for transaction submissions.
   * @param options.projectId - An optional project ID for analytics or tracking.
   * @param options.appUrl - An optional application URL for the wallet.
   * @param options.directTo - An optional parameter to specify the user-controlled wallet direct-to option.
   * @param options.refreshToken - An optional refresh token for authentication.
   * @param options.keepWindowOpen - An optional flag to keep the wallet window open after operations.
   * @param options.sparkscanApiKey - An optional API key for Sparkscan integration.
   *
   * @returns A promise that resolves to an instance of Web3Wallet.
   */
  static async enable(options) {
    const res = await openWindow(
      {
        method: "enable",
        projectId: options.projectId,
        directTo: options.directTo ? options.directTo : "undefined",
        refreshToken: options.refreshToken ? options.refreshToken : "undefined",
        keepWindowOpen: options.keepWindowOpen ? "true" : "false",
        networkId: String(options.networkId)
      },
      options.appUrl
    );
    if (res.success === false)
      throw new ApiError({
        code: -3,
        info: "Refused - The request was refused due to lack of access - e.g. wallet disconnects."
      });
    if (res.data.method !== "enable") {
      throw new ApiError({
        code: 2,
        info: "Received the wrong response from the iframe."
      });
    }
    const wallet = await _Web3Wallet.initWallet({
      // Not actually deprecated, just want developers to get a danger line when using this function (it's not for public use.)
      networkId: options.networkId,
      fetcher: options.fetcher,
      submitter: options.submitter,
      bitcoinProvider: options.bitcoinProvider,
      projectId: options.projectId,
      appUrl: options.appUrl,
      user: res.data.user,
      keyHashes: {
        cardanoPubKeyHash: res.data.cardanoPubKeyHash,
        cardanoStakeCredentialHash: res.data.cardanoStakeCredentialHash,
        bitcoinMainnetPubKeyHash: res.data.bitcoinMainnetPubKeyHash,
        bitcoinTestnetPubKeyHash: res.data.bitcoinTestnetPubKeyHash,
        sparkMainnetPubKeyHash: res.data.sparkMainnetPubKeyHash,
        sparkRegtestPubKeyHash: res.data.sparkRegtestPubKeyHash
      },
      sparkMainnetStaticDepositAddress: res.data.sparkMainnetStaticDepositAddress,
      sparkRegtestStaticDepositAddress: res.data.sparkRegtestStaticDepositAddress,
      sparkscanApiKey: options.sparkscanApiKey,
      baseUrl: options.baseUrl
    });
    return wallet;
  }
  getUser() {
    return this.user;
  }
  async exportWallet(chain2) {
    const res = await openWindow(
      {
        method: "export-wallet",
        projectId: this.projectId,
        chain: chain2,
        networkId: String(this.networkId)
      },
      this.appUrl
    );
    if (res.success === false)
      throw new ApiError({
        code: 2,
        info: "UserDeclined - User declined to export the wallet."
      });
    if (res.data.method !== "export-wallet") {
      throw new ApiError({
        code: 2,
        info: "Received the wrong response from the iframe."
      });
    }
    return { success: true, data: { method: "export-wallet" } };
  }
  async disable() {
    const res = await openWindow(
      {
        method: "disable",
        projectId: this.projectId
      },
      this.appUrl
    );
    if (res.success === false) {
      throw new ApiError({
        code: 2,
        info: "Received the wrong response from the iframe."
      });
    }
    return { success: true, data: { method: "disable" } };
  }
  async onramp(options) {
    const chain2 = options.chain;
    const method = `${chain2}-onramp`;
    if (!this.projectId) {
      throw new ApiError({
        code: 2,
        info: "Cannot start onramp without a projectId."
      });
    }
    const res = await openWindow(
      {
        method,
        projectId: this.projectId
      },
      this.appUrl
    );
    if (res.success === false)
      throw new ApiError({
        code: 2,
        info: "UserDeclined - User declined to sign the message."
      });
    if (res.data.method !== "cardano-onramp") {
      throw new ApiError({
        code: 2,
        info: "Received the wrong response from the iframe."
      });
    }
    return {
      success: true,
      data: { method }
    };
  }
  static async initBitcoinWallet(options) {
    const bitcoinWallet = new import_bitcoin4.EmbeddedWallet({
      network: options.networkId === 1 ? "Mainnet" : "Testnet",
      key: {
        type: "address",
        address: getBitcoinAddressFromPubkey(
          options.networkId === 1 ? options.keyHashes.bitcoinMainnetPubKeyHash : options.keyHashes.bitcoinTestnetPubKeyHash,
          options.networkId
        )
      },
      provider: options.bitcoinProvider
    });
    bitcoinWallet.signMessage = async (params) => {
      const res = await openWindow(
        {
          method: "bitcoin-sign-message",
          projectId: options.projectId,
          address: params.address,
          message: params.message,
          protocol: params.protocol,
          networkId: String(options.networkId)
        },
        options.appUrl
      );
      if (res.success === false)
        throw new ApiError({
          code: 2,
          info: "UserDeclined - User declined to sign the message."
        });
      if (res.data.method !== "bitcoin-sign-message") {
        throw new ApiError({
          code: 2,
          info: "Received the wrong response from the iframe."
        });
      }
      return {
        signature: res.data.signature,
        messageHash: res.data.messageHash,
        address: res.data.address
      };
    };
    bitcoinWallet.sendTransfer = async (params) => {
      const res = await openWindow(
        {
          method: "bitcoin-send-transfer",
          projectId: options.projectId,
          recipients: JSON.stringify(params.recipients),
          networkId: String(options.networkId)
        },
        options.appUrl
      );
      if (res.success === false)
        throw new ApiError({
          code: 2,
          info: "UserDeclined - User declined to sign the transaction."
        });
      if (res.data.method !== "bitcoin-send-transfer") {
        throw new ApiError({
          code: 2,
          info: "Received the wrong response from the iframe."
        });
      }
      return { txid: res.data.txid };
    };
    bitcoinWallet.signPsbt = async (params) => {
      const res = await openWindow(
        {
          method: "bitcoin-sign-psbt",
          projectId: options.projectId,
          psbt: params.psbt,
          signInputs: JSON.stringify(params.signInputs),
          broadcast: params.broadcast === true ? "true" : "false",
          networkId: String(options.networkId)
        },
        options.appUrl
      );
      if (res.success === false)
        throw new ApiError({
          code: 2,
          info: "UserDeclined - User declined to sign the transaction."
        });
      if (res.data.method !== "bitcoin-sign-psbt") {
        throw new ApiError({
          code: 2,
          info: "Received the wrong response from the iframe."
        });
      }
      return {
        psbt: res.data.psbt,
        txid: res.data.txid
      };
    };
    return bitcoinWallet;
  }
  static async initCardanoWallet(options) {
    const address = getCardanoAddressFromPubkey(
      options.keyHashes.cardanoPubKeyHash,
      options.keyHashes.cardanoStakeCredentialHash,
      options.networkId
    );
    const cardanoWallet = await import_wallet6.MeshCardanoHeadlessWallet.fromCredentialSources({
      paymentCredentialSource: {
        type: "signer",
        signer: {
          getPublicKey: async () => options.keyHashes.cardanoPubKeyHash,
          getPublicKeyHash: async () => options.keyHashes.cardanoPubKeyHash,
          sign: async () => "",
          verify: async () => false
        }
      },
      stakeCredentialSource: options.keyHashes.cardanoStakeCredentialHash ? {
        type: "signer",
        signer: {
          getPublicKey: async () => options.keyHashes.cardanoStakeCredentialHash,
          getPublicKeyHash: async () => options.keyHashes.cardanoStakeCredentialHash,
          sign: async () => "",
          verify: async () => false
        }
      } : void 0,
      networkId: options.networkId,
      walletAddressType: 1
    });
    cardanoWallet.signTx = async (unsignedTx, partialSign = false, returnFullTx = true) => {
      const res = await openWindow(
        {
          method: "cardano-sign-tx",
          projectId: options.projectId,
          unsignedTx,
          partialSign: partialSign === true ? "true" : "false",
          returnFullTx: returnFullTx === true ? "true" : "false",
          networkId: String(options.networkId)
        },
        options.appUrl
      );
      if (res.success === false)
        throw new ApiError({
          code: 2,
          info: "UserDeclined - User declined to sign the message."
        });
      if (res.data.method !== "cardano-sign-tx") {
        throw new ApiError({
          code: 2,
          info: "Received the wrong response from the iframe."
        });
      }
      return res.data.tx;
    };
    cardanoWallet.signTxReturnFullTx = async (unsignedTx, partialSign = false) => {
      const res = await openWindow(
        {
          method: "cardano-sign-tx",
          projectId: options.projectId,
          unsignedTx,
          partialSign: partialSign === true ? "true" : "false",
          returnFullTx: "true",
          networkId: String(options.networkId)
        },
        options.appUrl
      );
      if (res.success === false)
        throw new ApiError({
          code: 2,
          info: "UserDeclined - User declined to sign the message."
        });
      if (res.data.method !== "cardano-sign-tx") {
        throw new ApiError({
          code: 2,
          info: "Received the wrong response from the iframe."
        });
      }
      return res.data.tx;
    };
    cardanoWallet.signData = async (address2, payload) => {
      const res = await openWindow(
        {
          method: "cardano-sign-data",
          projectId: options.projectId,
          payload,
          address: address2,
          networkId: String(options.networkId)
        },
        options.appUrl
      );
      if (res.success === false)
        throw new ApiError({
          code: 3,
          info: "UserDeclined - User declined to sign the data."
        });
      if (res.data.method !== "cardano-sign-data") {
        throw new ApiError({
          code: 2,
          info: "Received the wrong response from the iframe."
        });
      }
      return res.data.dataSignature;
    };
    return cardanoWallet;
  }
  static async initSparkWallet(keyHashes, networkId, projectId, appUrl, sparkMainnetStaticDepositAddress, sparkRegtestStaticDepositAddress, sparkscanApiKey, baseUrl = "https://api.sparkscan.io") {
    const identityPublicKey = networkId === 1 ? keyHashes.sparkMainnetPubKeyHash : keyHashes.sparkRegtestPubKeyHash;
    const sparkOptions = {
      networkId,
      projectId,
      appUrl,
      sparkApiKey: sparkscanApiKey,
      sparkApiUrl: baseUrl,
      sparkMainnetStaticDepositAddress,
      sparkRegtestStaticDepositAddress,
      address: getSparkAddressFromPubkey(
        networkId === 1 ? keyHashes.sparkMainnetPubKeyHash : keyHashes.sparkRegtestPubKeyHash,
        networkId === 1 ? "MAINNET" : "REGTEST"
      ),
      publicKeyHex: identityPublicKey
    };
    return new Web3SparkWallet(sparkOptions);
  }
  /**
   * Initializes a new instance of a Web3 wallet with the specified options.
   * @deprecated - Don't use this function unless you know exactly what you're doing.
   * @param params - The parameters required to initialize the wallet.
   * @param params.networkId - The network ID to connect to. Must be either `0` (testnet) or `1` (mainnet).
   * @param params.address - The wallet address to associate with the wallet instance.
   * @param params.fetcher - (Optional) An implementation of the `IFetcher` interface for fetching data.
   * @param params.submitter - (Optional) An implementation of the `ISubmitter` interface for submitting transactions.
   * @param params.projectId - (Optional) The project ID for analytics or tracking purposes.
   * @param params.appUrl - (Optional) The application URL for associating the wallet with a specific app.
   *
   * @returns A promise that resolves to an initialized instance of `Web3Wallet`.
   */
  static async initWallet({
    networkId,
    fetcher,
    submitter,
    bitcoinProvider,
    projectId,
    appUrl = "https://utxos.dev/",
    user,
    keyHashes,
    baseUrl,
    sparkMainnetStaticDepositAddress,
    sparkRegtestStaticDepositAddress,
    sparkscanApiKey
  }) {
    const bitcoin = await this.initBitcoinWallet({
      projectId,
      appUrl,
      networkId,
      bitcoinProvider,
      keyHashes
    });
    const cardano = await this.initCardanoWallet({
      projectId,
      appUrl,
      networkId,
      fetcher,
      submitter,
      keyHashes
    });
    const spark = await this.initSparkWallet(
      keyHashes,
      networkId,
      projectId,
      appUrl,
      sparkMainnetStaticDepositAddress,
      sparkRegtestStaticDepositAddress,
      sparkscanApiKey,
      baseUrl
    );
    const _options = {
      networkId,
      fetcher,
      submitter,
      appUrl,
      projectId,
      user,
      cardano,
      bitcoin,
      spark
    };
    const wallet = new _Web3Wallet(_options);
    return wallet;
  }
};
var ApiError = class extends Error {
  json;
  constructor(json) {
    super(json.info || "An error occurred");
    this.name = "ApiError";
    this.json = json;
  }
};

// src/index.ts
var import_spark_sdk5 = require("@buildonspark/spark-sdk");
setAdapters(browserAdapters);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ApiError,
  AuthRouteError,
  CardanoWalletDeveloperControlled,
  NoPRFExtensionError,
  NotAllowedError,
  NotAuthenticatedError,
  SessionExpiredError,
  SparkIssuerWalletDeveloperControlled,
  SparkWallet,
  Sponsorship,
  StorageInsertError,
  StorageRetrievalError,
  TransferDirection,
  UserAbortedError,
  WalletDeveloperControlled,
  WalletServerCreationError,
  WalletServerRetrievalError,
  Web3NonCustodialProvider,
  Web3Sdk,
  Web3SparkWallet,
  Web3Wallet,
  bytesToHex,
  bytesToString,
  clientDeriveWallet,
  clientGenerateWallet,
  clientRecovery,
  combineShardsBuildWallet,
  convertLegacyToNewFormat,
  createCredential,
  crypto,
  decodeBech32mTokenIdentifier,
  decodeSparkAddress,
  decryptWithCipher,
  decryptWithPrivateKey,
  encodeBech32mTokenIdentifier,
  encryptWithCipher,
  encryptWithPublicKey,
  extractIdentityPublicKey,
  generateHash,
  generateKeyPair,
  getBitcoinAddressFromPubkey,
  getCardanoAddressFromPubkey,
  getCredentialFromCredentialId,
  getCrypto,
  getNetworkFromBech32mTokenIdentifier,
  getNetworkFromSparkAddress,
  getSparkAddressFromPubkey,
  hashData,
  hexToBytes,
  isValidMultiChainWalletInfo,
  isValidSparkAddress,
  meshUniversalStaticUtxo,
  openWindow,
  secretToCryptoKey,
  shamirCombine,
  shamirSplit,
  spiltKeyIntoShards,
  stringToBytes,
  validateMultiChainWalletInfo,
  webauthnPublicKeyCredentialToCryptoKey
});
/*! Bundled license information:

@scure/base/lib/esm/index.js:
  (*! scure-base - MIT License (c) 2022 Paul Miller (paulmillr.com) *)

@noble/hashes/esm/utils.js:
  (*! noble-hashes - MIT License (c) 2022 Paul Miller (paulmillr.com) *)

@noble/curves/esm/utils.js:
  (*! noble-curves - MIT License (c) 2022 Paul Miller (paulmillr.com) *)
*/
