// @utxos/sdk/dist/index.js
import {
  ApiError,
  AuthRouteError,
  CardanoWalletDeveloperControlled,
  NoPRFExtensionError,
  NotAllowedError,
  NotAuthenticatedError,
  SessionExpiredError,
  SparkIssuerWalletDeveloperControlled,
  SparkWallet,
  Sponsorship,
  StorageInsertError,
  StorageRetrievalError,
  TransferDirection,
  UserAbortedError,
  WalletDeveloperControlled,
  WalletServerCreationError,
  WalletServerRetrievalError,
  Web3NonCustodialProvider,
  Web3Sdk,
  Web3SparkWallet,
  Web3Wallet,
  bytesToHex,
  bytesToString,
  clientDeriveWallet,
  clientGenerateWallet,
  clientRecovery,
  combineShardsBuildWallet,
  convertLegacyToNewFormat,
  createCredential,
  crypto,
  decodeBech32mTokenIdentifier,
  decodeSparkAddress,
  decryptWithCipher,
  decryptWithPrivateKey,
  encodeBech32mTokenIdentifier,
  encryptWithCipher,
  encryptWithPublicKey,
  extractIdentityPublicKey,
  generateHash,
  generateKeyPair,
  getBitcoinAddressFromPubkey,
  getCardanoAddressFromPubkey,
  getCredentialFromCredentialId,
  getCrypto,
  getNetworkFromBech32mTokenIdentifier,
  getNetworkFromSparkAddress,
  getSparkAddressFromPubkey,
  hashData,
  hexToBytes,
  isValidMultiChainWalletInfo,
  isValidSparkAddress,
  meshUniversalStaticUtxo,
  openWindow,
  secretToCryptoKey,
  shamirCombine,
  shamirSplit,
  spiltKeyIntoShards,
  stringToBytes,
  validateMultiChainWalletInfo,
  webauthnPublicKeyCredentialToCryptoKey
} from "./chunk-DIY55XL2.js";
export {
  ApiError,
  AuthRouteError,
  CardanoWalletDeveloperControlled,
  NoPRFExtensionError,
  NotAllowedError,
  NotAuthenticatedError,
  SessionExpiredError,
  SparkIssuerWalletDeveloperControlled,
  SparkWallet,
  Sponsorship,
  StorageInsertError,
  StorageRetrievalError,
  TransferDirection,
  UserAbortedError,
  WalletDeveloperControlled,
  WalletServerCreationError,
  WalletServerRetrievalError,
  Web3NonCustodialProvider,
  Web3Sdk,
  Web3SparkWallet,
  Web3Wallet,
  bytesToHex,
  bytesToString,
  clientDeriveWallet,
  clientGenerateWallet,
  clientRecovery,
  combineShardsBuildWallet,
  convertLegacyToNewFormat,
  createCredential,
  crypto,
  decodeBech32mTokenIdentifier,
  decodeSparkAddress,
  decryptWithCipher,
  decryptWithPrivateKey,
  encodeBech32mTokenIdentifier,
  encryptWithCipher,
  encryptWithPublicKey,
  extractIdentityPublicKey,
  generateHash,
  generateKeyPair,
  getBitcoinAddressFromPubkey,
  getCardanoAddressFromPubkey,
  getCredentialFromCredentialId,
  getCrypto,
  getNetworkFromBech32mTokenIdentifier,
  getNetworkFromSparkAddress,
  getSparkAddressFromPubkey,
  hashData,
  hexToBytes,
  isValidMultiChainWalletInfo,
  isValidSparkAddress,
  meshUniversalStaticUtxo,
  openWindow,
  secretToCryptoKey,
  shamirCombine,
  shamirSplit,
  spiltKeyIntoShards,
  stringToBytes,
  validateMultiChainWalletInfo,
  webauthnPublicKeyCredentialToCryptoKey
};
