// @utxos/sdk/dist/index.native.js
import {
  ApiError,
  AuthRouteError,
  CardanoWalletDeveloperControlled,
  NoPRFExtensionError,
  NotAllowedError,
  NotAuthenticatedError,
  SessionExpiredError,
  SparkIssuerWalletDeveloperControlled,
  Sponsorship,
  StorageInsertError,
  StorageRetrievalError,
  TransferDirection,
  UserAbortedError,
  WalletDeveloperControlled,
  WalletServerCreationError,
  WalletServerRetrievalError,
  Web3NonCustodialProvider,
  Web3Sdk,
  Web3SparkWallet,
  Web3Wallet,
  __require,
  bytesToHex,
  bytesToString,
  clientDeriveWallet,
  clientGenerateWallet,
  clientRecovery,
  combineShardsBuildWallet,
  convertLegacyToNewFormat,
  createCredential,
  crypto,
  decryptWithCipher,
  decryptWithPrivateKey,
  encryptWithCipher,
  encryptWithPublicKey,
  extractIdentityPublicKey,
  generateHash,
  generateKeyPair,
  getBitcoinAddressFromPubkey,
  getCardanoAddressFromPubkey,
  getCredentialFromCredentialId,
  getCrypto,
  getSparkAddressFromPubkey,
  hashData,
  hexToBytes,
  isValidMultiChainWalletInfo,
  meshUniversalStaticUtxo,
  openWindow,
  secretToCryptoKey,
  setAdapters,
  shamirCombine,
  shamirSplit,
  spiltKeyIntoShards,
  stringToBytes,
  validateMultiChainWalletInfo,
  webauthnPublicKeyCredentialToCryptoKey
} from "./chunk-DIY55XL2.js";

// src/platforms/react-native/crypto.native.ts
var QuickCrypto = null;
try {
  QuickCrypto = __require("react-native-quick-crypto");
} catch {
}
function getCrypto2() {
  if (QuickCrypto?.webcrypto) {
    return QuickCrypto.webcrypto;
  }
  if (typeof global !== "undefined" && global.crypto?.subtle) {
    return global.crypto;
  }
  throw new Error(
    "@utxos/sdk: No crypto implementation found. Install react-native-quick-crypto or react-native-get-random-values"
  );
}
function getHmacAlgorithm(algorithm) {
  const normalized = algorithm.toUpperCase().replace(/SHA-?(\d+)/i, "SHA-$1");
  return {
    name: "HMAC",
    hash: { name: normalized }
  };
}
var cryptoAdapter = {
  getRandomBytes(size) {
    const bytes = new Uint8Array(size);
    getCrypto2().getRandomValues(bytes);
    return bytes;
  },
  async hmacSign(algorithm, key, data) {
    const crypto2 = getCrypto2();
    const hmacAlgorithm = getHmacAlgorithm(algorithm);
    const cryptoKey = await crypto2.subtle.importKey(
      "raw",
      key,
      hmacAlgorithm,
      false,
      ["sign"]
    );
    const signature = await crypto2.subtle.sign("HMAC", cryptoKey, data);
    return new Uint8Array(signature);
  },
  async encrypt(data, key, iv) {
    const crypto2 = getCrypto2();
    const cryptoKey = await crypto2.subtle.importKey("raw", key, "AES-GCM", false, ["encrypt"]);
    const encrypted = await crypto2.subtle.encrypt({ name: "AES-GCM", iv }, cryptoKey, data);
    return new Uint8Array(encrypted);
  },
  async decrypt(data, key, iv) {
    const crypto2 = getCrypto2();
    const cryptoKey = await crypto2.subtle.importKey("raw", key, "AES-GCM", false, ["decrypt"]);
    const decrypted = await crypto2.subtle.decrypt({ name: "AES-GCM", iv }, cryptoKey, data);
    return new Uint8Array(decrypted);
  },
  async generateKeyPair(algorithm = "ECDH") {
    const crypto2 = getCrypto2();
    const keyPair = await crypto2.subtle.generateKey(
      { name: algorithm, namedCurve: "P-256" },
      true,
      algorithm === "ECDH" ? ["deriveBits"] : ["sign", "verify"]
    );
    const publicKey = await crypto2.subtle.exportKey("raw", keyPair.publicKey);
    const privateKey = await crypto2.subtle.exportKey("pkcs8", keyPair.privateKey);
    return {
      publicKey: new Uint8Array(publicKey),
      privateKey: new Uint8Array(privateKey)
    };
  },
  async deriveSharedSecret(privateKey, publicKey) {
    const crypto2 = getCrypto2();
    const privKey = await crypto2.subtle.importKey(
      "pkcs8",
      privateKey,
      { name: "ECDH", namedCurve: "P-256" },
      false,
      ["deriveBits"]
    );
    const pubKey = await crypto2.subtle.importKey(
      "raw",
      publicKey,
      { name: "ECDH", namedCurve: "P-256" },
      false,
      []
    );
    const sharedSecret = await crypto2.subtle.deriveBits(
      { name: "ECDH", public: pubKey },
      privKey,
      256
    );
    return new Uint8Array(sharedSecret);
  },
  async importKey(keyData, algorithm, usages) {
    const crypto2 = getCrypto2();
    return crypto2.subtle.importKey("raw", keyData, algorithm, false, usages);
  },
  getSubtleCrypto() {
    return getCrypto2().subtle;
  },
  getRandomValuesInPlace(array) {
    getCrypto2().getRandomValues(array);
    return array;
  }
};

// src/platforms/react-native/storage.native.ts
var AsyncStorage = null;
function getAsyncStorage() {
  if (!AsyncStorage) {
    try {
      AsyncStorage = __require("@react-native-async-storage/async-storage").default;
    } catch {
      throw new Error(
        "@utxos/sdk: @react-native-async-storage/async-storage is required for React Native. Install it with: npm install @react-native-async-storage/async-storage"
      );
    }
  }
  return AsyncStorage;
}
var storageAdapter = {
  async getItem(key) {
    return getAsyncStorage().getItem(key);
  },
  async setItem(key, value) {
    await getAsyncStorage().setItem(key, value);
  },
  async removeItem(key) {
    await getAsyncStorage().removeItem(key);
  },
  async clear() {
    await getAsyncStorage().clear();
  }
};

// src/platforms/react-native/linking.native.ts
var Linking = __require("react-native").Linking;
var InAppBrowser = null;
try {
  InAppBrowser = __require("react-native-inappbrowser-reborn").default;
} catch {
}
var linkingAdapter = {
  async openURL(url) {
    const canOpen = await Linking.canOpenURL(url);
    if (canOpen) {
      await Linking.openURL(url);
    } else {
      throw new Error(`Cannot open URL: ${url}`);
    }
  },
  /**
   * Redirect to URL - in React Native, this is equivalent to openURL
   * since there's no browser-style same-tab navigation
   */
  async redirectURL(url) {
    await this.openURL(url);
  },
  async openAuthWindow(url, callbackScheme) {
    if (InAppBrowser && await InAppBrowser.isAvailable()) {
      try {
        const result = await InAppBrowser.openAuth(url, callbackScheme, {
          showTitle: false,
          enableUrlBarHiding: true,
          enableDefaultShare: false,
          ephemeralWebSession: false
        });
        if (result.type === "success" && result.url) {
          const callbackUrl = new URL(result.url);
          return {
            code: callbackUrl.searchParams.get("code") || void 0,
            state: callbackUrl.searchParams.get("state") || void 0,
            error: callbackUrl.searchParams.get("error") || void 0,
            errorDescription: callbackUrl.searchParams.get("error_description") || void 0
          };
        } else if (result.type === "cancel") {
          return { error: "cancelled", errorDescription: "User cancelled authentication" };
        }
        return { error: "unknown", errorDescription: "Unknown authentication result" };
      } catch (error) {
        return { error: "failed", errorDescription: String(error) };
      }
    }
    return new Promise((resolve) => {
      const subscription = Linking.addEventListener("url", (event) => {
        const callbackUrl = event.url;
        if (callbackUrl.startsWith(callbackScheme)) {
          subscription.remove();
          try {
            const parsedUrl = new URL(callbackUrl);
            resolve({
              code: parsedUrl.searchParams.get("code") || void 0,
              state: parsedUrl.searchParams.get("state") || void 0,
              error: parsedUrl.searchParams.get("error") || void 0,
              errorDescription: parsedUrl.searchParams.get("error_description") || void 0
            });
          } catch {
            resolve({ error: "parse_error", errorDescription: "Failed to parse callback URL" });
          }
        }
      });
      Linking.openURL(url).catch((error) => {
        subscription.remove();
        resolve({ error: "open_failed", errorDescription: String(error) });
      });
    });
  },
  getCurrentURL() {
    return null;
  },
  getURLParams() {
    return {};
  },
  addURLListener(callback) {
    const subscription = Linking.addEventListener("url", (event) => {
      callback(event.url);
    });
    Linking.getInitialURL().then((url) => {
      if (url) callback(url);
    });
    return () => subscription.remove();
  },
  getUserAgent() {
    return "ReactNative";
  }
};

// src/platforms/react-native/encoding.native.ts
var BASE64_CHARS = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
function base64Encode(str) {
  let result = "";
  const len = str.length;
  for (let i = 0; i < len; i += 3) {
    const a = str.charCodeAt(i);
    const b = i + 1 < len ? str.charCodeAt(i + 1) : 0;
    const c = i + 2 < len ? str.charCodeAt(i + 2) : 0;
    result += BASE64_CHARS[a >> 2];
    result += BASE64_CHARS[(a & 3) << 4 | b >> 4];
    result += i + 1 < len ? BASE64_CHARS[(b & 15) << 2 | c >> 6] : "=";
    result += i + 2 < len ? BASE64_CHARS[c & 63] : "=";
  }
  return result;
}
var encodingAdapter = {
  bytesToBase64(bytes) {
    let binary = "";
    for (let i = 0; i < bytes.length; i++) {
      const byte = bytes[i];
      if (byte !== void 0) {
        binary += String.fromCharCode(byte);
      }
    }
    return base64Encode(binary);
  },
  base64ToBytes(base64) {
    const normalized = base64.replace(/-/g, "+").replace(/_/g, "/");
    const cleanBase64 = normalized.replace(/=/g, "");
    const len = cleanBase64.length;
    const byteLen = Math.floor(len * 3 / 4);
    const bytes = new Uint8Array(byteLen);
    let p = 0;
    for (let i = 0; i < len; i += 4) {
      const a = BASE64_CHARS.indexOf(cleanBase64[i] ?? "");
      const b = BASE64_CHARS.indexOf(cleanBase64[i + 1] ?? "");
      const c = i + 2 < len ? BASE64_CHARS.indexOf(cleanBase64[i + 2] ?? "") : 0;
      const d = i + 3 < len ? BASE64_CHARS.indexOf(cleanBase64[i + 3] ?? "") : 0;
      bytes[p++] = a << 2 | b >> 4;
      if (p < byteLen) bytes[p++] = (b & 15) << 4 | c >> 2;
      if (p < byteLen) bytes[p++] = (c & 3) << 6 | d;
    }
    return bytes;
  },
  bytesToHex(bytes) {
    return Array.from(bytes).map((b) => b.toString(16).padStart(2, "0")).join("");
  },
  hexToBytes(hex) {
    const matches = hex.match(/.{1,2}/g);
    if (!matches) return new Uint8Array(0);
    return new Uint8Array(matches.map((byte) => parseInt(byte, 16)));
  },
  bytesToUtf8(bytes) {
    if (typeof TextDecoder !== "undefined") {
      return new TextDecoder().decode(bytes);
    }
    let str = "";
    for (let i = 0; i < bytes.length; i++) {
      const byte = bytes[i];
      if (byte !== void 0) {
        str += String.fromCharCode(byte);
      }
    }
    return decodeURIComponent(escape(str));
  },
  utf8ToBytes(str) {
    if (typeof TextEncoder !== "undefined") {
      return new TextEncoder().encode(str);
    }
    const encoded = unescape(encodeURIComponent(str));
    const bytes = new Uint8Array(encoded.length);
    for (let i = 0; i < encoded.length; i++) {
      bytes[i] = encoded.charCodeAt(i);
    }
    return bytes;
  }
};

// src/platforms/react-native/index.ts
var reactNativeAdapters = {
  crypto: cryptoAdapter,
  storage: storageAdapter,
  linking: linkingAdapter,
  encoding: encodingAdapter
};

// src/index.native.ts
import {
  isValidSparkAddress,
  decodeSparkAddress,
  getNetworkFromSparkAddress,
  encodeBech32mTokenIdentifier,
  decodeBech32mTokenIdentifier,
  getNetworkFromBech32mTokenIdentifier,
  SparkWallet
} from "@buildonspark/spark-sdk";
setAdapters(reactNativeAdapters);
export {
  ApiError,
  AuthRouteError,
  CardanoWalletDeveloperControlled,
  NoPRFExtensionError,
  NotAllowedError,
  NotAuthenticatedError,
  SessionExpiredError,
  SparkIssuerWalletDeveloperControlled,
  SparkWallet,
  Sponsorship,
  StorageInsertError,
  StorageRetrievalError,
  TransferDirection,
  UserAbortedError,
  WalletDeveloperControlled,
  WalletServerCreationError,
  WalletServerRetrievalError,
  Web3NonCustodialProvider,
  Web3Sdk,
  Web3SparkWallet,
  Web3Wallet,
  bytesToHex,
  bytesToString,
  clientDeriveWallet,
  clientGenerateWallet,
  clientRecovery,
  combineShardsBuildWallet,
  convertLegacyToNewFormat,
  createCredential,
  crypto,
  decodeBech32mTokenIdentifier,
  decodeSparkAddress,
  decryptWithCipher,
  decryptWithPrivateKey,
  encodeBech32mTokenIdentifier,
  encryptWithCipher,
  encryptWithPublicKey,
  extractIdentityPublicKey,
  generateHash,
  generateKeyPair,
  getBitcoinAddressFromPubkey,
  getCardanoAddressFromPubkey,
  getCredentialFromCredentialId,
  getCrypto,
  getNetworkFromBech32mTokenIdentifier,
  getNetworkFromSparkAddress,
  getSparkAddressFromPubkey,
  hashData,
  hexToBytes,
  isValidMultiChainWalletInfo,
  isValidSparkAddress,
  meshUniversalStaticUtxo,
  openWindow,
  secretToCryptoKey,
  shamirCombine,
  shamirSplit,
  spiltKeyIntoShards,
  stringToBytes,
  validateMultiChainWalletInfo,
  webauthnPublicKeyCredentialToCryptoKey
};
