// @buildonspark/spark-sdk/dist/graphql/objects/index.js
import {
  BitcoinNetwork_default,
  ClaimStaticDepositRequestType_default,
  ClaimStaticDepositStatus_default,
  CurrencyUnit_default,
  ExitSpeed_default,
  LightningReceiveRequestStatus_default,
  LightningSendRequestStatus_default,
  SparkCoopExitRequestStatus_default,
  SparkLeavesSwapRequestStatus_default,
  SparkUserRequestStatus_default,
  SparkUserRequestType_default,
  SparkWalletUser_default,
  Transfer_default,
  getClaimStaticDepositQuery,
  getCoopExitFeeQuoteQuery,
  getLeavesSwapRequestQuery,
  getLightningReceiveRequestQuery,
  getLightningSendRequestQuery,
  getUserRequestQuery
} from "../../chunk-UYEB2VPG.js";
import "../../chunk-MVRQ5US7.js";
export {
  BitcoinNetwork_default as BitcoinNetwork,
  ClaimStaticDepositRequestType_default as ClaimStaticDepositRequestType,
  ClaimStaticDepositStatus_default as ClaimStaticDepositStatus,
  CurrencyUnit_default as CurrencyUnit,
  ExitSpeed_default as ExitSpeed,
  LightningReceiveRequestStatus_default as LightningReceiveRequestStatus,
  LightningSendRequestStatus_default as LightningSendRequestStatus,
  SparkCoopExitRequestStatus_default as SparkCoopExitRequestStatus,
  SparkLeavesSwapRequestStatus_default as SparkLeavesSwapRequestStatus,
  SparkUserRequestStatus_default as SparkUserRequestStatus,
  SparkUserRequestType_default as SparkUserRequestType,
  SparkWalletUser_default as SparkWalletUser,
  Transfer_default as Transfer,
  getClaimStaticDepositQuery,
  getCoopExitFeeQuoteQuery,
  getLeavesSwapRequestQuery,
  getLightningReceiveRequestQuery,
  getLightningSendRequestQuery,
  getUserRequestQuery
};
