// @buildonspark/spark-sdk/dist/types/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_empty = require('../empty-CblwXbqR.cjs');
const require_proto_spark = require('../proto/spark.cjs');
const require_sdk_types = require('../sdk-types-cbEt9uJw.cjs');

exports.BitcoinNetwork = require_sdk_types.BitcoinNetwork;
exports.ClaimStaticDepositRequestType = require_sdk_types.ClaimStaticDepositRequestType;
exports.ClaimStaticDepositStatus = require_sdk_types.ClaimStaticDepositStatus;
Object.defineProperty(exports, 'CommonProto', {
  enumerable: true,
  get: function () {
    return require_empty.common_exports;
  }
});
exports.CurrencyUnit = require_sdk_types.CurrencyUnit;
exports.ExitSpeed = require_sdk_types.ExitSpeed;
exports.LightningReceiveRequestStatus = require_sdk_types.LightningReceiveRequestStatus;
exports.LightningSendRequestStatus = require_sdk_types.LightningSendRequestStatus;
exports.SparkCoopExitRequestStatus = require_sdk_types.SparkCoopExitRequestStatus;
exports.SparkLeavesSwapRequestStatus = require_sdk_types.SparkLeavesSwapRequestStatus;
Object.defineProperty(exports, 'SparkProto', {
  enumerable: true,
  get: function () {
    return require_proto_spark.spark_exports;
  }
});
exports.SparkUserRequestStatus = require_sdk_types.SparkUserRequestStatus;
exports.SparkUserRequestType = require_sdk_types.SparkUserRequestType;
exports.SparkWalletUser = require_sdk_types.SparkWalletUser;
exports.Transfer = require_sdk_types.Transfer;
exports.TransferDirection = require_sdk_types.TransferDirection;
exports.getClaimStaticDepositQuery = require_sdk_types.getClaimStaticDepositQuery;
exports.getCoopExitFeeQuoteQuery = require_sdk_types.getCoopExitFeeQuoteQuery;
exports.getLeavesSwapRequestQuery = require_sdk_types.getLeavesSwapRequestQuery;
exports.getLightningReceiveRequestQuery = require_sdk_types.getLightningReceiveRequestQuery;
exports.getLightningSendRequestQuery = require_sdk_types.getLightningSendRequestQuery;
exports.getUserRequestQuery = require_sdk_types.getUserRequestQuery;
exports.mapSettingsProtoToWalletSettings = require_sdk_types.mapSettingsProtoToWalletSettings;
exports.mapTransferLeafToWalletTransferLeaf = require_sdk_types.mapTransferLeafToWalletTransferLeaf;
exports.mapTransferToWalletTransfer = require_sdk_types.mapTransferToWalletTransfer;
exports.mapTreeNodeToWalletLeaf = require_sdk_types.mapTreeNodeToWalletLeaf;
