// @buildonspark/spark-sdk/dist/tests/test-utils.js
import {
  BitcoinFaucet,
  ConnectionManagerNodeJS,
  DefaultSparkSigner,
  DepositService,
  SparkError,
  SparkWalletNodeJS,
  UnsafeStatelessSparkSigner,
  WalletConfig,
  WalletConfigService,
  getNetwork,
  getP2TRAddressFromPublicKey
} from "../chunk-RU434ZAE.js";
import "../chunk-ZP6Z6DFX.js";
import "../chunk-EU3I7GFB.js";
import "../chunk-JE3MXMPW.js";
import "../chunk-UYEB2VPG.js";
import "../chunk-MVRQ5US7.js";

// src/tests/test-utils.ts
import { secp256k1 } from "@noble/curves/secp256k1";
import { Address, OutScript, Transaction } from "@scure/btc-signer";

// src/tests/utils/spark-testing-wallet.ts
var SparkWalletTesting = class extends SparkWalletNodeJS {
  async setupBackgroundStream() {
    return;
  }
  async *optimizeLeaves(multiplicity = void 0) {
    return;
  }
  async proxyParentSetupBackgroundStream() {
    return super.setupBackgroundStream();
  }
  getSigner() {
    return this.config.signer;
  }
  async queryPendingTransfers() {
    return await this.transferService.queryPendingTransfers();
  }
  async verifyPendingTransfer(transfer) {
    return await this.transferService.verifyPendingTransfer(transfer);
  }
};

// src/tests/test-utils.ts
import { sha256 } from "@noble/hashes/sha2";
function getTestWalletConfig() {
  const identityPrivateKey = secp256k1.utils.randomPrivateKey();
  return getTestWalletConfigWithIdentityKey(identityPrivateKey);
}
function getTestWalletConfigWithIdentityKey(identityPrivateKey) {
  return {
    ...WalletConfig.LOCAL,
    identityPrivateKey
  };
}
async function createDeposit(wallet, leafId, faucet, amountSats = 100000n) {
  const faucetCoin = await faucet.fund();
  const configService = new WalletConfigService(
    {
      network: "LOCAL"
    },
    wallet.getSigner()
  );
  const connectionManager = new ConnectionManagerNodeJS(configService);
  const depositService = new DepositService(configService, connectionManager);
  const pubKey = await wallet.getSigner().getPublicKeyFromDerivation({
    type: "leaf" /* LEAF */,
    path: leafId
  });
  const depositResp = await depositService.generateDepositAddress({
    signingPubkey: pubKey,
    leafId
  });
  if (!depositResp.depositAddress) {
    throw new SparkError("Deposit address not found", {
      signingPubkey: pubKey,
      leafId
    });
  }
  const depositTx = new Transaction();
  depositTx.addInput(faucetCoin.outpoint);
  const addr = Address(getNetwork(4 /* LOCAL */)).decode(
    depositResp.depositAddress.address
  );
  const script = OutScript.encode(addr);
  depositTx.addOutput({ script, amount: amountSats });
  let treeResp;
  treeResp = await depositService.createTreeRoot({
    keyDerivation: {
      type: "leaf" /* LEAF */,
      path: leafId
    },
    verifyingKey: depositResp.depositAddress.verifyingKey,
    depositTx,
    vout: 0
  });
  const signedDepositTx = await faucet.signFaucetCoin(
    depositTx,
    faucetCoin.txout,
    faucetCoin.key
  );
  await faucet.broadcastTx(signedDepositTx.hex);
  const randomKey = secp256k1.utils.randomPrivateKey();
  const randomPubKey = secp256k1.getPublicKey(randomKey);
  const randomAddress = getP2TRAddressFromPublicKey(
    randomPubKey,
    4 /* LOCAL */
  );
  await faucet.generateToAddress(1, randomAddress);
  await new Promise((resolve) => setTimeout(resolve, 100));
  return treeResp.nodes[0];
}
async function createNewTree(wallet, leafId, faucet, amountSats = 100000n) {
  return await createDeposit(wallet, leafId, faucet, amountSats);
}
var signerTypes = [
  {
    name: "DefaultSparkSigner",
    Signer: DefaultSparkSigner,
    createTree: createNewTree
  },
  {
    name: "UnsafeStatelessSparkSigner",
    Signer: UnsafeStatelessSparkSigner,
    createTree: createNewTree
  }
];
var walletTypes = [...signerTypes];
function createDeterministicKeys(seed) {
  const privateKey = sha256(new TextEncoder().encode(seed));
  const publicKey = secp256k1.getPublicKey(privateKey, true);
  return { privateKey, publicKey };
}
export {
  BitcoinFaucet,
  SparkWalletTesting,
  createDeterministicKeys,
  createNewTree,
  getTestWalletConfig,
  getTestWalletConfigWithIdentityKey,
  signerTypes,
  walletTypes
};
